
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Parameter estimation and optimization for phosphorylation models.">
      
      
        <meta name="author" content="Abhinav Mishra">
      
      
        <link rel="canonical" href="https://bibymaths.github.io/phoskintime/reference/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../Documentation/CHANGELOG/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Reference - phoskintime</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="auto" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="phoskintime" class="md-header__button md-logo" aria-label="phoskintime" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            phoskintime
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Reference
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/bibymaths/phoskintime" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="phoskintime" class="md-nav__button md-logo" aria-label="phoskintime" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    phoskintime
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/bibymaths/phoskintime" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bin.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
    <nav class="md-nav" aria-label="main">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bin.main.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config" class="md-nav__link">
    <span class="md-ellipsis">
      config
    </span>
  </a>
  
    <nav class="md-nav" aria-label="config">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.config.parse_bound_pair" class="md-nav__link">
    <span class="md-ellipsis">
      parse_bound_pair
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.parse_fix_value" class="md-nav__link">
    <span class="md-ellipsis">
      parse_fix_value
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.log_config" class="md-nav__link">
    <span class="md-ellipsis">
      log_config
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.extract_config" class="md-nav__link">
    <span class="md-ellipsis">
      extract_config
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.score_fit" class="md-nav__link">
    <span class="md-ellipsis">
      score_fit
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.constants.get_param_names_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_param_names_ds" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.generate_labels_rand" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.generate_labels_ds" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.location" class="md-nav__link">
    <span class="md-ellipsis">
      location
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_number_of_params_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_number_of_params_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_bounds_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_bounds_rand
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
    <nav class="md-nav" aria-label="logconf">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter" class="md-nav__link">
    <span class="md-ellipsis">
      ColoredFormatter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ColoredFormatter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.format" class="md-nav__link">
    <span class="md-ellipsis">
      format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.remove_ansi" class="md-nav__link">
    <span class="md-ellipsis">
      remove_ansi
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.setup_logger" class="md-nav__link">
    <span class="md-ellipsis">
      setup_logger
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_residuals_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_nsga
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_de" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_de
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo" class="md-nav__link">
    <span class="md-ellipsis">
      minfndiffevo
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfndiffevo">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii" class="md-nav__link">
    <span class="md-ellipsis">
      minfnnsgaii
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfnnsgaii">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_nsga
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_de" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_de
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct.pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      pipeline
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit" class="md-nav__link">
    <span class="md-ellipsis">
      postfit
    </span>
  </a>
  
    <nav class="md-nav" aria-label="postfit">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="md-nav__link">
    <span class="md-ellipsis">
      goodnessoffit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="md-nav__link">
    <span class="md-ellipsis">
      reshape_alpha_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_pca" class="md-nav__link">
    <span class="md-ellipsis">
      perform_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      perform_tsne
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.additional_plots" class="md-nav__link">
    <span class="md-ellipsis">
      additional_plots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="md-nav__link">
    <span class="md-ellipsis">
      create_sankey_from_network
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.important_connections" class="md-nav__link">
    <span class="md-ellipsis">
      important_connections
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT" class="md-nav__link">
    <span class="md-ellipsis">
      KKT
    </span>
  </a>
  
    <nav class="md-nav" aria-label="KKT">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.generate_latex_table" class="md-nav__link">
    <span class="md-ellipsis">
      generate_latex_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_primal_feasibility_results" class="md-nav__link">
    <span class="md-ellipsis">
      print_primal_feasibility_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      print_sensitivity_and_active_constraints
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_constraint_violations" class="md-nav__link">
    <span class="md-ellipsis">
      plot_constraint_violations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      plot_sensitivity_analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.process_excel_results" class="md-nav__link">
    <span class="md-ellipsis">
      process_excel_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.post_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell" class="md-nav__link">
    <span class="md-ellipsis">
      runpowell
    </span>
  </a>
  
    <nav class="md-nav" aria-label="runpowell">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell.run_powell" class="md-nav__link">
    <span class="md-ellipsis">
      run_powell
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers" class="md-nav__link">
    <span class="md-ellipsis">
      helpers
    </span>
  </a>
  
    <nav class="md-nav" aria-label="helpers">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.powerset" class="md-nav__link">
    <span class="md-ellipsis">
      powerset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.state_label" class="md-nav__link">
    <span class="md-ellipsis">
      state_label
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_random_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_random_diagram
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_distributive_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_distributive_diagram
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_successive_model" class="md-nav__link">
    <span class="md-ellipsis">
      create_successive_model
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod" class="md-nav__link">
    <span class="md-ellipsis">
      distmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="distmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.distmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.distmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.distmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod" class="md-nav__link">
    <span class="md-ellipsis">
      randmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="randmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.randmod.prepare_vectorized_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_vectorized_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.randmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.randmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod" class="md-nav__link">
    <span class="md-ellipsis">
      succmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="succmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.succmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.succmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.succmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.testmod" class="md-nav__link">
    <span class="md-ellipsis">
      testmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
    <nav class="md-nav" aria-label="weights">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.weights.early_emphasis" class="md-nav__link">
    <span class="md-ellipsis">
      early_emphasis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.weights.get_weight_options" class="md-nav__link">
    <span class="md-ellipsis">
      get_weight_options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest" class="md-nav__link">
    <span class="md-ellipsis">
      adapest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="adapest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.adaptive_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      adaptive_estimation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.estimate_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_profiles
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core" class="md-nav__link">
    <span class="md-ellipsis">
      core
    </span>
  </a>
  
    <nav class="md-nav" aria-label="core">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.core.process_gene" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.core.process_gene_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene_wrapper
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci" class="md-nav__link">
    <span class="md-ellipsis">
      ci
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ci">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.identifiability.ci.confidence_intervals" class="md-nav__link">
    <span class="md-ellipsis">
      confidence_intervals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="normest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.normest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest" class="md-nav__link">
    <span class="md-ellipsis">
      seqest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="seqest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.fit_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      fit_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.sequential_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      sequential_estimation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle" class="md-nav__link">
    <span class="md-ellipsis">
      toggle
    </span>
  </a>
  
    <nav class="md-nav" aria-label="toggle">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.toggle.estimate_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting" class="md-nav__link">
    <span class="md-ellipsis">
      plotting
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_parallel" class="md-nav__link">
    <span class="md-ellipsis">
      plot_parallel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.pca_components" class="md-nav__link">
    <span class="md-ellipsis">
      pca_components
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      plot_tsne
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_bar" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_bar
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_series" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_series
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      plot_profiles
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_model_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_model_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_A_S" class="md-nav__link">
    <span class="md-ellipsis">
      plot_A_S
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_all" class="md-nav__link">
    <span class="md-ellipsis">
      plot_all
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_clusters" class="md-nav__link">
    <span class="md-ellipsis">
      plot_clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_error_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_error_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_1" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_1
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_5" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_5
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="cleanup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_collecttri" class="md-nav__link">
    <span class="md-ellipsis">
      process_collecttri
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.format_site" class="md-nav__link">
    <span class="md-ellipsis">
      format_site
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss_std" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss_std
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_routlimma" class="md-nav__link">
    <span class="md-ellipsis">
      process_routlimma
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.update_gene_symbols" class="md-nav__link">
    <span class="md-ellipsis">
      update_gene_symbols
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.move_processed_files" class="md-nav__link">
    <span class="md-ellipsis">
      move_processed_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.map" class="md-nav__link">
    <span class="md-ellipsis">
      map
    </span>
  </a>
  
    <nav class="md-nav" aria-label="map">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#processing.map.map_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      map_optimization_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.map.create_cytoscape_table" class="md-nav__link">
    <span class="md-ellipsis">
      create_cytoscape_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.map.generate_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      generate_nodes
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis" class="md-nav__link">
    <span class="md-ellipsis">
      analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_rand" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_ds" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      sensitivity_analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist" class="md-nav__link">
    <span class="md-ellipsis">
      initdist
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initdist">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initdist.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand" class="md-nav__link">
    <span class="md-ellipsis">
      initrand
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initrand">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initrand.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc" class="md-nav__link">
    <span class="md-ellipsis">
      initsucc
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initsucc">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initsucc.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest" class="md-nav__link">
    <span class="md-ellipsis">
      inittest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="inittest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.inittest.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfn">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="md-nav__link">
    <span class="md-ellipsis">
      TFOptimizationMultiObjectiveProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="TFOptimizationMultiObjectiveProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="filter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.load_raw_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_raw_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      filter_mrna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.update_regulations" class="md-nav__link">
    <span class="md-ellipsis">
      update_regulations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_TF" class="md-nav__link">
    <span class="md-ellipsis">
      filter_TF
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.determine_T_use" class="md-nav__link">
    <span class="md-ellipsis">
      determine_T_use
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_mRNA_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_mRNA_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_TF_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_TF_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_no_psite_array" class="md-nav__link">
    <span class="md-ellipsis">
      create_no_psite_array
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.compute_beta_indices" class="md-nav__link">
    <span class="md-ellipsis">
      compute_beta_indices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_initial_guess" class="md-nav__link">
    <span class="md-ellipsis">
      create_initial_guess
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_bounds" class="md-nav__link">
    <span class="md-ellipsis">
      create_bounds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.get_parallel_runner" class="md-nav__link">
    <span class="md-ellipsis">
      get_parallel_runner
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.extract_best_solution" class="md-nav__link">
    <span class="md-ellipsis">
      extract_best_solution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_alpha_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_alpha_mapping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_beta_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_beta_mapping
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper" class="md-nav__link">
    <span class="md-ellipsis">
      helper
    </span>
  </a>
  
    <nav class="md-nav" aria-label="helper">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_alpha_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_beta_barplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap_abs_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_goodness_of_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_time_wise_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfn">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      objective_wrapper
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun.run_optimizer" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimizer
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_alpha_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_alpha_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_beta_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_beta_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_linear_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      build_linear_constraints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="filter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.load_and_filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_filter_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.prepare_data" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.min_max_normalize" class="md-nav__link">
    <span class="md-ellipsis">
      min_max_normalize
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_expression_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_expression_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_tf_protein_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_tf_protein_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.summarize_stats" class="md-nav__link">
    <span class="md-ellipsis">
      summarize_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.get_optimization_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      get_optimization_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.postprocess_results" class="md-nav__link">
    <span class="md-ellipsis">
      postprocess_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display" class="md-nav__link">
    <span class="md-ellipsis">
      display
    </span>
  </a>
  
    <nav class="md-nav" aria-label="display">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utils.display.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.save_result" class="md-nav__link">
    <span class="md-ellipsis">
      save_result
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables" class="md-nav__link">
    <span class="md-ellipsis">
      tables
    </span>
  </a>
  
    <nav class="md-nav" aria-label="tables">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utils.tables.generate_tables" class="md-nav__link">
    <span class="md-ellipsis">
      generate_tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.tables.save_tables" class="md-nav__link">
    <span class="md-ellipsis">
      save_tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.tables.save_master_table" class="md-nav__link">
    <span class="md-ellipsis">
      save_master_table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/CHANGELOG/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Documentation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Documentation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/bin/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Main
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/kinopt/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Kinase Optimization
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/models/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    ODE Models
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/paramest/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Parameter Estimation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/plotting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/sensitivity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sensitivity Analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/steady/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Steady State Calculation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/tfopt/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Transcription Factor Optimization
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Documentation/utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Utils
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bin.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
    <nav class="md-nav" aria-label="main">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bin.main.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config" class="md-nav__link">
    <span class="md-ellipsis">
      config
    </span>
  </a>
  
    <nav class="md-nav" aria-label="config">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.config.parse_bound_pair" class="md-nav__link">
    <span class="md-ellipsis">
      parse_bound_pair
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.parse_fix_value" class="md-nav__link">
    <span class="md-ellipsis">
      parse_fix_value
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.log_config" class="md-nav__link">
    <span class="md-ellipsis">
      log_config
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.extract_config" class="md-nav__link">
    <span class="md-ellipsis">
      extract_config
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.config.score_fit" class="md-nav__link">
    <span class="md-ellipsis">
      score_fit
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.constants.get_param_names_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_param_names_ds" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.generate_labels_rand" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.generate_labels_ds" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.location" class="md-nav__link">
    <span class="md-ellipsis">
      location
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_number_of_params_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_number_of_params_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.constants.get_bounds_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_bounds_rand
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
    <nav class="md-nav" aria-label="logconf">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter" class="md-nav__link">
    <span class="md-ellipsis">
      ColoredFormatter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ColoredFormatter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.format" class="md-nav__link">
    <span class="md-ellipsis">
      format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.remove_ansi" class="md-nav__link">
    <span class="md-ellipsis">
      remove_ansi
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.setup_logger" class="md-nav__link">
    <span class="md-ellipsis">
      setup_logger
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_residuals_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_nsga
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_de" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_de
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo" class="md-nav__link">
    <span class="md-ellipsis">
      minfndiffevo
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfndiffevo">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii" class="md-nav__link">
    <span class="md-ellipsis">
      minfnnsgaii
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfnnsgaii">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_nsga
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_de" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_de
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct.pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      pipeline
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit" class="md-nav__link">
    <span class="md-ellipsis">
      postfit
    </span>
  </a>
  
    <nav class="md-nav" aria-label="postfit">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="md-nav__link">
    <span class="md-ellipsis">
      goodnessoffit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="md-nav__link">
    <span class="md-ellipsis">
      reshape_alpha_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_pca" class="md-nav__link">
    <span class="md-ellipsis">
      perform_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      perform_tsne
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.additional_plots" class="md-nav__link">
    <span class="md-ellipsis">
      additional_plots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="md-nav__link">
    <span class="md-ellipsis">
      create_sankey_from_network
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.important_connections" class="md-nav__link">
    <span class="md-ellipsis">
      important_connections
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT" class="md-nav__link">
    <span class="md-ellipsis">
      KKT
    </span>
  </a>
  
    <nav class="md-nav" aria-label="KKT">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.generate_latex_table" class="md-nav__link">
    <span class="md-ellipsis">
      generate_latex_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_primal_feasibility_results" class="md-nav__link">
    <span class="md-ellipsis">
      print_primal_feasibility_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      print_sensitivity_and_active_constraints
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_constraint_violations" class="md-nav__link">
    <span class="md-ellipsis">
      plot_constraint_violations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      plot_sensitivity_analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.process_excel_results" class="md-nav__link">
    <span class="md-ellipsis">
      process_excel_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.post_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell" class="md-nav__link">
    <span class="md-ellipsis">
      runpowell
    </span>
  </a>
  
    <nav class="md-nav" aria-label="runpowell">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell.run_powell" class="md-nav__link">
    <span class="md-ellipsis">
      run_powell
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers" class="md-nav__link">
    <span class="md-ellipsis">
      helpers
    </span>
  </a>
  
    <nav class="md-nav" aria-label="helpers">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.powerset" class="md-nav__link">
    <span class="md-ellipsis">
      powerset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.state_label" class="md-nav__link">
    <span class="md-ellipsis">
      state_label
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_random_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_random_diagram
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_distributive_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_distributive_diagram
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_successive_model" class="md-nav__link">
    <span class="md-ellipsis">
      create_successive_model
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod" class="md-nav__link">
    <span class="md-ellipsis">
      distmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="distmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.distmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.distmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.distmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod" class="md-nav__link">
    <span class="md-ellipsis">
      randmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="randmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.randmod.prepare_vectorized_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_vectorized_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.randmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.randmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod" class="md-nav__link">
    <span class="md-ellipsis">
      succmod
    </span>
  </a>
  
    <nav class="md-nav" aria-label="succmod">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.succmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.succmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.succmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.testmod" class="md-nav__link">
    <span class="md-ellipsis">
      testmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
    <nav class="md-nav" aria-label="weights">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#models.weights.early_emphasis" class="md-nav__link">
    <span class="md-ellipsis">
      early_emphasis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#models.weights.get_weight_options" class="md-nav__link">
    <span class="md-ellipsis">
      get_weight_options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest" class="md-nav__link">
    <span class="md-ellipsis">
      adapest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="adapest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.adaptive_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      adaptive_estimation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.adapest.estimate_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_profiles
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core" class="md-nav__link">
    <span class="md-ellipsis">
      core
    </span>
  </a>
  
    <nav class="md-nav" aria-label="core">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.core.process_gene" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.core.process_gene_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene_wrapper
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci" class="md-nav__link">
    <span class="md-ellipsis">
      ci
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ci">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.identifiability.ci.confidence_intervals" class="md-nav__link">
    <span class="md-ellipsis">
      confidence_intervals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="normest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.normest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest" class="md-nav__link">
    <span class="md-ellipsis">
      seqest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="seqest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.fit_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      fit_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paramest.seqest.sequential_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      sequential_estimation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle" class="md-nav__link">
    <span class="md-ellipsis">
      toggle
    </span>
  </a>
  
    <nav class="md-nav" aria-label="toggle">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paramest.toggle.estimate_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting" class="md-nav__link">
    <span class="md-ellipsis">
      plotting
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_parallel" class="md-nav__link">
    <span class="md-ellipsis">
      plot_parallel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.pca_components" class="md-nav__link">
    <span class="md-ellipsis">
      pca_components
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      plot_tsne
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_bar" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_bar
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_series" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_series
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      plot_profiles
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_model_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_model_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_A_S" class="md-nav__link">
    <span class="md-ellipsis">
      plot_A_S
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_all" class="md-nav__link">
    <span class="md-ellipsis">
      plot_all
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_clusters" class="md-nav__link">
    <span class="md-ellipsis">
      plot_clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_error_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_error_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_1" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_1
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_5" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_5
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="cleanup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_collecttri" class="md-nav__link">
    <span class="md-ellipsis">
      process_collecttri
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.format_site" class="md-nav__link">
    <span class="md-ellipsis">
      format_site
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss_std" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss_std
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.process_routlimma" class="md-nav__link">
    <span class="md-ellipsis">
      process_routlimma
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.update_gene_symbols" class="md-nav__link">
    <span class="md-ellipsis">
      update_gene_symbols
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.cleanup.move_processed_files" class="md-nav__link">
    <span class="md-ellipsis">
      move_processed_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.map" class="md-nav__link">
    <span class="md-ellipsis">
      map
    </span>
  </a>
  
    <nav class="md-nav" aria-label="map">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#processing.map.map_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      map_optimization_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.map.create_cytoscape_table" class="md-nav__link">
    <span class="md-ellipsis">
      create_cytoscape_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#processing.map.generate_nodes" class="md-nav__link">
    <span class="md-ellipsis">
      generate_nodes
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis" class="md-nav__link">
    <span class="md-ellipsis">
      analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_rand" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_rand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_ds" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_ds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sensitivity.analysis.sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      sensitivity_analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist" class="md-nav__link">
    <span class="md-ellipsis">
      initdist
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initdist">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initdist.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand" class="md-nav__link">
    <span class="md-ellipsis">
      initrand
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initrand">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initrand.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc" class="md-nav__link">
    <span class="md-ellipsis">
      initsucc
    </span>
  </a>
  
    <nav class="md-nav" aria-label="initsucc">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.initsucc.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest" class="md-nav__link">
    <span class="md-ellipsis">
      inittest
    </span>
  </a>
  
    <nav class="md-nav" aria-label="inittest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#steady.inittest.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfn">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="md-nav__link">
    <span class="md-ellipsis">
      TFOptimizationMultiObjectiveProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="TFOptimizationMultiObjectiveProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="filter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.load_raw_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_raw_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      filter_mrna
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.update_regulations" class="md-nav__link">
    <span class="md-ellipsis">
      update_regulations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_TF" class="md-nav__link">
    <span class="md-ellipsis">
      filter_TF
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.determine_T_use" class="md-nav__link">
    <span class="md-ellipsis">
      determine_T_use
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_mRNA_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_mRNA_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_TF_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_TF_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_no_psite_array" class="md-nav__link">
    <span class="md-ellipsis">
      create_no_psite_array
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.compute_beta_indices" class="md-nav__link">
    <span class="md-ellipsis">
      compute_beta_indices
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_initial_guess" class="md-nav__link">
    <span class="md-ellipsis">
      create_initial_guess
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_bounds" class="md-nav__link">
    <span class="md-ellipsis">
      create_bounds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.get_parallel_runner" class="md-nav__link">
    <span class="md-ellipsis">
      get_parallel_runner
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.extract_best_solution" class="md-nav__link">
    <span class="md-ellipsis">
      extract_best_solution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_alpha_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_alpha_mapping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_beta_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_beta_mapping
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper" class="md-nav__link">
    <span class="md-ellipsis">
      helper
    </span>
  </a>
  
    <nav class="md-nav" aria-label="helper">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_alpha_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_beta_barplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap_abs_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_goodness_of_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_time_wise_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="constants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
    <nav class="md-nav" aria-label="plotout">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sheetutils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
    <nav class="md-nav" aria-label="minfn">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      objective_wrapper
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
    <nav class="md-nav" aria-label="optrun">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun.run_optimizer" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimizer
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
    <nav class="md-nav" aria-label="construct">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_alpha_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_alpha_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_beta_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_beta_func
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_linear_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      build_linear_constraints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="filter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.load_and_filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_filter_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.prepare_data" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
    <nav class="md-nav" aria-label="iodata">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.min_max_normalize" class="md-nav__link">
    <span class="md-ellipsis">
      min_max_normalize
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_expression_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_expression_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_tf_protein_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_tf_protein_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.summarize_stats" class="md-nav__link">
    <span class="md-ellipsis">
      summarize_stats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.get_optimization_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      get_optimization_parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.postprocess_results" class="md-nav__link">
    <span class="md-ellipsis">
      postprocess_results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display" class="md-nav__link">
    <span class="md-ellipsis">
      display
    </span>
  </a>
  
    <nav class="md-nav" aria-label="display">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utils.display.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.save_result" class="md-nav__link">
    <span class="md-ellipsis">
      save_result
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.display.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables" class="md-nav__link">
    <span class="md-ellipsis">
      tables
    </span>
  </a>
  
    <nav class="md-nav" aria-label="tables">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utils.tables.generate_tables" class="md-nav__link">
    <span class="md-ellipsis">
      generate_tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.tables.save_tables" class="md-nav__link">
    <span class="md-ellipsis">
      save_tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#utils.tables.save_master_table" class="md-nav__link">
    <span class="md-ellipsis">
      save_master_table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="api-reference">API Reference<a class="headerlink" href="#api-reference" title="Permanent link">&para;</a></h1>


<div class="doc doc-object doc-module">



<h2 id="bin.main" class="doc doc-heading">
            <code>bin.main</code>


<a href="#bin.main" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="bin.main.main" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#bin.main.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to run the phosphorylation modelling process.
It reads the configuration, loads the data, and processes each gene in parallel.
It also handles logging and output organization.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.config" class="doc doc-heading">
            <code>config.config</code>


<a href="#config.config" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.config.parse_bound_pair" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_bound_pair</span><span class="p">(</span><span class="n">val</span><span class="p">)</span></code>

<a href="#config.config.parse_bound_pair" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse a string representing a pair of bounds (lower, upper) into a tuple of floats.
The upper bound can be 'inf' or 'infinity' to represent infinity.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "0,3" or "0,infinity".
Returns:
    tuple: A tuple containing the lower and upper bounds as floats.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.parse_fix_value" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_fix_value</span><span class="p">(</span><span class="n">val</span><span class="p">)</span></code>

<a href="#config.config.parse_fix_value" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse a fixed value or a list of fixed values from a string.
If the input is a single value, it returns that value as a float.
If the input is a comma-separated list, it returns a list of floats.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "1.0" or "1.0,2.0".
Returns:
    float or list: The parsed fixed value(s) as a float or a list of floats.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span></code>

<a href="#config.config.ensure_output_directory" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>:param directory:
:type directory: str</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

<a href="#config.config.parse_args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse command-line arguments for the PhosKinTime script.
This function uses argparse to define and handle the command-line options.
It includes options for setting bounds, fixed parameters, bootstrapping,
profile estimation, and input file paths.
The function returns the parsed arguments as a Namespace object.
The arguments include:
    --A-bound: Bounds for parameter A (default: "0,3")
    --B-bound: Bounds for parameter B (default: "0,3")
    --C-bound: Bounds for parameter C (default: "0,3")
    --D-bound: Bounds for parameter D (default: "0,3")
    --Ssite-bound: Bounds for Ssite (default: "0,3")
    --Dsite-bound: Bounds for Dsite (default: "0,3")
    --fix-A: Fixed value for parameter A
    --fix-B: Fixed value for parameter B
    --fix-C: Fixed value for parameter C
    --fix-D: Fixed value for parameter D
    --fix-Ssite: Fixed value for Ssite
    --fix-Dsite: Fixed value for Dsite
    --fix-t: JSON string mapping time points to fixed param values
    --bootstraps: Number of bootstrapping iterations (default: 0)
    --profile-start: Start time for profile estimation (default: None)
    --profile-end: End time for profile estimation (default: 1)
    --profile-step: Step size for profile estimation (default: 0.5)
    --input-excel: Path to the input Excel file (default: INPUT_EXCEL)
Returns:
    argparse.Namespace: The parsed command-line arguments.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.log_config" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">log_config</span><span class="p">(</span><span class="n">logger</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span></code>

<a href="#config.config.log_config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Log the configuration settings for the PhosKinTime script.
This function logs the parameter bounds, fixed parameters,
bootstrapping iterations, time-specific fixed parameters,
and profile estimation settings.
It uses the provided logger to output the information.
:param logger:
:param bounds:
:param fixed_params:
:param time_fixed:
:param args:
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.extract_config" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_config</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code>

<a href="#config.config.extract_config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract configuration settings from command-line arguments.
This function creates a dictionary containing the parameter bounds,
fixed parameters, bootstrapping iterations, time-specific fixed parameters,
and profile estimation settings. It also sets the maximum number of workers
for parallel processing.
The function returns the configuration dictionary.
:param args:
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.score_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">score_fit</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">prediction</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">ALPHA_WEIGHT</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">BETA_WEIGHT</span><span class="p">,</span> <span class="n">gamma</span><span class="o">=</span><span class="n">GAMMA_WEIGHT</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="n">DELTA_WEIGHT</span><span class="p">,</span> <span class="n">reg_penalty</span><span class="o">=</span><span class="n">MU_REG</span><span class="p">)</span></code>

<a href="#config.config.score_fit" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate the score for the fit of a model to target data.
The score is a weighted combination of various metrics including
mean squared error (MSE), root mean squared error (RMSE),
mean absolute error (MAE), variance, and regularization penalty.
The weights for each metric can be adjusted using the parameters
alpha, beta, gamma, and delta.
The regularization penalty is controlled by the reg_penalty parameter.
The function returns the calculated score.
:param target:
:param prediction:
:param params:
:param alpha:
:param beta:
:param gamma:
:param delta:
:param reg_penalty:
:return:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.constants" class="doc doc-heading">
            <code>config.constants</code>


<a href="#config.constants" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.constants.get_param_names_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_param_names_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#config.constants.get_param_names_rand" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate parameter names for the random model.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        [parameter names for all combinations of dephosphorylation sites].</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_param_names_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_param_names_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#config.constants.get_param_names_ds" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate parameter names for distributive or successive models.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        ['D1', 'D2', ..., 'D<num_psites>'].</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.generate_labels_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_labels_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#config.constants.generate_labels_rand" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the random model.
Returns a list with the base labels "R" and "P", followed by labels for all combinations of phosphorylated sites.
Example for num_psites=2: ["R", "P", "P1", "P2", "P12"]</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.generate_labels_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_labels_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#config.constants.generate_labels_ds" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the distributive or successive models.
Returns a list with the base labels "R" and "P", followed by labels for each individual phosphorylated state.
Example for num_psites=2: ["R", "P", "P1", "P2"]</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.location" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">location</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

<a href="#config.constants.location" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Returns a clickable hyperlink string for supported terminals using ANSI escape sequences.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The file path or URL.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>label</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The display text for the link. Defaults to the path if not provided.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string that, when printed, shows a clickable link in terminals that support ANSI hyperlinks.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_number_of_params_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_number_of_params_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

<a href="#config.constants.get_number_of_params_rand" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate the number of parameters required for the ODE system based on the number of phosphorylation sites.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites (1 to 4).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Total number of parameters.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_bounds_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_bounds_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">ub</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

<a href="#config.constants.get_bounds_rand" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate bounds for the ODE parameters based on the number of phosphorylation sites.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lower</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Lower bound for parameters.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>upper</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Upper bound for parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of bounds as [lower, upper] for each parameter.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.logconf" class="doc doc-heading">
            <code>config.logconf</code>


<a href="#config.logconf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="config.logconf.ColoredFormatter" class="doc doc-heading">
            <code>ColoredFormatter</code>


<a href="#config.logconf.ColoredFormatter" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="logging.Formatter">Formatter</span></code></p>


        <p>Custom formatter to add colors to log messages and elapsed time.
This formatter uses ANSI escape codes to colorize the log messages based on their severity level.
It also includes a right-aligned clock that shows the elapsed time since the logger was initialized.
The elapsed time is displayed in a human-readable format (e.g., "1h 23m 45s").
The formatter is designed to be used with a logger that has a console handler.
The elapsed time is calculated from the time the logger was initialized and is displayed in a right-aligned format.
The formatter also ensures that the log messages are padded to a specified width, which can be adjusted using the <code>width</code> parameter.
The <code>remove_ansi</code> method is used to strip ANSI escape codes from the log message for accurate padding calculation.
The <code>format</code> method is overridden to customize the log message format, including the timestamp, logger name, log level, and message.
The <code>setup_logger</code> function is used to configure the logger with a file handler and a stream handler.
The file handler writes log messages to a specified log file, while the stream handler outputs log messages to the console.
The logger is set to the specified logging level, and the log file is created in the specified directory.
The log file is rotated based on size, and old log files are backed up.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="config.logconf.ColoredFormatter.format" class="doc doc-heading">
            <code class="highlight language-python"><span class="nb">format</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

<a href="#config.logconf.ColoredFormatter.format" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Format the log record with colors and elapsed time.
This method overrides the default format method to customize the log message format.
It includes the timestamp, logger name, log level, and message.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="config.logconf.ColoredFormatter.remove_ansi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">remove_ansi</span><span class="p">(</span><span class="n">s</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#config.logconf.ColoredFormatter.remove_ansi" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Remove ANSI escape codes from a string.</p>


    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="config.logconf.setup_logger" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup_logger</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;phoskintime&#39;</span><span class="p">,</span> <span class="n">log_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">,</span> <span class="n">log_dir</span><span class="o">=</span><span class="n">LOG_DIR</span><span class="p">,</span> <span class="n">rotate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">max_bytes</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">,</span> <span class="n">backup_count</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span></code>

<a href="#config.logconf.setup_logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Setup a logger with colored output and file logging.
This function creates a logger with colored output for console messages
:param name:
:param log_file:
:param level:
:param log_dir:
:param rotate:
:param max_bytes:
:param backup_count:
:return: logger</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.config.constants" class="doc doc-heading">
            <code>kinopt.evol.config.constants</code>


<a href="#kinopt.evol.config.constants" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.config.logconf" class="doc doc-heading">
            <code>kinopt.evol.config.logconf</code>


<a href="#kinopt.evol.config.logconf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.exporter.plotout" class="doc doc-heading">
            <code>kinopt.evol.exporter.plotout</code>


<a href="#kinopt.evol.exporter.plotout" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_residuals_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">)</span></code>

<a href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generates and saves combined residual-related plots for one gene with all psites in the legend.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_data</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary with keys 'psites', 'observed', 'estimated', and 'residuals' containing data for all psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>TIME_POINTS</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span> or <span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time points corresponding to the series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.opt_analyze_nsga" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">opt_analyze_nsga</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">F</span><span class="p">,</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">approx_ideal</span><span class="p">,</span> <span class="n">approx_nadir</span><span class="p">,</span> <span class="n">asf_i</span><span class="p">,</span> <span class="n">pseudo_i</span><span class="p">,</span> <span class="n">n_evals</span><span class="p">,</span> <span class="n">hv</span><span class="p">,</span> <span class="n">hist</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">hist_cv_avg</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">igd</span><span class="p">,</span> <span class="n">best_objectives</span><span class="p">,</span> <span class="n">waterfall_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

<a href="#kinopt.evol.exporter.plotout.opt_analyze_nsga" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generates and saves various plots related to optimization results.
This includes design space plots, objective space plots,
convergence plots, and parameter trend plots.</p>
<p>:param problem:
:param result:
:param F:
:param pairs:
:param approx_ideal:
:param approx_nadir:
:param asf_i:
:param pseudo_i:
:param n_evals:
:param hv:
:param hist:
:param val:
:param hist_cv_avg:
:param k:
:param igd:
:param best_objectives:
:param waterfall_df:
:param convergence_df:
:param alpha_values:
:param beta_values:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.opt_analyze_de" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">opt_analyze_de</span><span class="p">(</span><span class="n">long_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">ordered_optimizer_runs</span><span class="p">,</span> <span class="n">x_values</span><span class="p">,</span> <span class="n">y_values</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span></code>

<a href="#kinopt.evol.exporter.plotout.opt_analyze_de" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generates and saves various plots related to optimization results.
This includes waterfall plots, convergence plots, parameter trend plots,
and parameter scan plots.</p>
<p>:param long_df:
:param convergence_df:
:param ordered_optimizer_runs:
:param x_values:
:param y_values:
:param val:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.exporter.sheetutils" class="doc doc-heading">
            <code>kinopt.evol.exporter.sheetutils</code>


<a href="#kinopt.evol.exporter.sheetutils" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">timepoints</span><span class="p">,</span> <span class="n">OUT_FILE</span><span class="p">)</span></code>

<a href="#kinopt.evol.exporter.sheetutils.output_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>This function is responsible for exporting the results of the optimization process to an Excel file.
It creates multiple sheets in the Excel file, each containing different types of data related to the optimization results.
The function also generates plots for the residuals of each gene.
The data is organized in a structured manner, making it easy to analyze and interpret the results.
The function takes the following parameters:</p>
<p>:param P_initial:
:param P_init_dense:
:param P_estimated:
:param residuals:
:param alpha_values:
:param beta_values:
:param result:
:param timepoints:
:param OUT_FILE:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.objfn.minfndiffevo" class="doc doc-heading">
            <code>kinopt.evol.objfn.minfndiffevo</code>


<a href="#kinopt.evol.objfn.minfndiffevo" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


<a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Custom optimization problem for phosphorylation analysis.</p>
<p>Defines the constraints, bounds, and objective function for optimizing
alpha and beta parameters across gene-psite-kinase relationships.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.P_initial">P_initial</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.P_initial_array">P_initial_array</span></code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.K_index">K_index</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.K_array">K_array</span></code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.gene_psite_counts">gene_psite_counts</span></code></td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.beta_counts">beta_counts</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Initializes the optimization problem with given data and constraints.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>P_initial_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_index</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_psite_counts</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_counts</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_function</span><span class="p">(</span><span class="n">params</span><span class="p">)</span></code>

<a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.objfn.minfnnsgaii" class="doc doc-heading">
            <code>kinopt.evol.objfn.minfnnsgaii</code>


<a href="#kinopt.evol.objfn.minfnnsgaii" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


<a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Multi-objective optimization problem for phosphorylation analysis.</p>
<p>Objectives:
- Minimize sum of squared residuals (main objective).
- Minimize violations of constraints for alpha (secondary objective).
- Minimize violations of constraints for beta (tertiary objective).</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Initializes the multi-objective optimization problem.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>P_initial_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_index</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_psite_counts</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_counts</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_function</span><span class="p">(</span><span class="n">params</span><span class="p">)</span></code>

<a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.opt.optrun" class="doc doc-heading">
            <code>kinopt.evol.opt.optrun</code>


<a href="#kinopt.evol.opt.optrun" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="n">PhosphorylationOptimizationProblem</span><span class="p">)</span></code>

<a href="#kinopt.evol.opt.optrun.run_optimization" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Sets up and runs the multi-objective optimization problem for phosphorylation
using an NSGA2 algorithm and a thread pool for parallelization.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial,</code>
            </td>
            <td>
                  <code>(<span title="P_initial_array">P_initial_array</span>, <span title="K_index">K_index</span>, <span title="K_array">K_array</span>, <span title="gene_psite_counts">gene_psite_counts</span>, <span title="beta_counts">beta_counts</span>)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data structures describing the problem (time-series data, kinases, etc.).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>PhosphorylationOptimizationProblem</code>
            </td>
            <td>
                  <code>class</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The custom problem class to be instantiated.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>result</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The pymoo result object containing the optimized population and history.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>exec_time</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Execution time for the optimization.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.post_optimization_nsga" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_nsga</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]),</span> <span class="n">ref_point</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]))</span></code>

<a href="#kinopt.evol.opt.optrun.post_optimization_nsga" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.
1) Extracts the Pareto front and computes a weighted score to pick a 'best' solution.
2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.
3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>weights</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of length 3 for weighting the objectives.</p>
              </div>
            </td>
            <td>
                  <code><span title="numpy.array">array</span>([1.0, 1.0, 1.0])</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ref_point</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference point for hypervolume computations.</p>
              </div>
            </td>
            <td>
                  <code><span title="numpy.array">array</span>([3, 1, 1])</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hypervolume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.post_optimization_de" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_de</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

<a href="#kinopt.evol.opt.optrun.post_optimization_de" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.
1) Extracts the Pareto front and computes a weighted score to pick a 'best' solution.
2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.
3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hypervolume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.optcon.construct" class="doc doc-heading">
            <code>kinopt.evol.optcon.construct</code>


<a href="#kinopt.evol.optcon.construct" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.optcon.construct.pipeline" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pipeline</span><span class="p">(</span><span class="n">input1_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">input2_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">time_series_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">scaling_method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">split_point</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">segment_points</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">estimate_missing_kinases</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">kinase_to_psites</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">])</span></code>

<a href="#kinopt.evol.optcon.construct.pipeline" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Constructs the pipeline for the optimization process.
This function orchestrates the loading of data, scaling,
and the construction of the P_initial and K_array structures.
It returns the necessary data structures for the optimization process.
The function takes the following parameters:</p>
<p>:param input1_path: Path to the first CSV file (HGNC data)
:param input2_path: Path to the second CSV file (kinase interactions)
:param time_series_columns: List of time series columns to extract
:param scaling_method: Method for scaling the data
:param split_point: Split point for scaling
:param segment_points: Segment points for scaling
:param estimate_missing_kinases: Boolean flag for estimating missing kinases
:param kinase_to_psites: Dictionary mapping kinases to their respective psites</p>
<p>:return: Tuple containing:
    - full_hgnc_df (pd.DataFrame): The scaled data from input1
    - interaction_df (pd.DataFrame): The subset/merged DataFrame from input2
    - observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df
    - P_initial (dict): Initial mapping of gene-psite pairs to kinase relationships and time-series data
    - P_initial_array (np.ndarray): Array containing observed time-series data for gene-psite pairs
    - K_index (dict): Mapping of kinases to their respective psite data
    - K_array (np.ndarray): Array containing time-series data for kinase-psite combinations
    - beta_counts (dict): Mapping of kinase indices to the number of associated psites
    - gene_psite_counts (list): Number of kinases per gene-psite combination
    - n (int): Number of decision variables in the optimization problem</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.utils.iodata" class="doc doc-heading">
            <code>kinopt.evol.utils.iodata</code>


<a href="#kinopt.evol.utils.iodata" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.iodata.format_duration" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Returns a formatted string representing the duration in seconds, minutes, or hours.
The format is:
- seconds: "xx.xx sec"
- minutes: "xx.xx min"
- hours: "xx.xx hr"</p>
<p>:param seconds:
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.iodata.load_and_scale_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Loads the full HGNC data and kinase interaction data, applies scaling to the time-series columns,
and subsets/merges them. The first file is the full HGNC data, and the second file contains kinase interactions.
The function also handles the conversion of kinases from string format to list format.</p>
<p>:param estimate_missing:
:param scaling_method:
:param split_point:
:param seg_points:</p>
<p>:return:
full_hgnc_df (pd.DataFrame): The scaled data from input1
interaction_df (pd.DataFrame): The subset/merged DataFrame from input2
observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">time_series_columns</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">segment_points</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.iodata.apply_scaling" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Applies different scaling methods to the time-series columns of a DataFrame.
Args:
    df (pd.DataFrame): The DataFrame containing the time-series data.
    time_series_columns (list): List of columns to be scaled.
    method (str): The scaling method to apply ('min_max', 'log', 'temporal', 'segmented', 'slope', 'cumulative').
    split_point (int): Column index for temporal scaling.
    segment_points (list): List of column indices for segmented scaling.
:returns:
    pd.DataFrame: The DataFrame with scaled time-series columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.iodata.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.iodata.organize_output_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Organizes output files from the optimization process into separate folders based on protein names.
Each protein's files are moved into a folder named after the protein.
Any remaining files that do not match the protein pattern are moved to a "General" folder.</p>
<p>:param directories: List of directories to organize.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.utils.params" class="doc doc-heading">
            <code>kinopt.evol.utils.params</code>


<a href="#kinopt.evol.utils.params" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.params.extract_parameters" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extracts the optimized alpha and beta values from the optimized parameters.
The function organizes the values into dictionaries for easy access and
interpretation. The alpha values are associated with gene-psite pairs and
their corresponding kinases, while the beta values are associated with
kinase-psite pairs. The function also logs the optimized values for
transparency and debugging purposes.</p>
<p>The function takes the following parameters:
:param P_initial:
:param gene_psite_counts:
:param K_index:
:param optimized_params:</p>
<p>:return:
- alpha_values: Dictionary mapping (gene, psite) to a dictionary of kinases and their alpha values.
- beta_values: Dictionary mapping (kinase, psite) to their corresponding beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span></code>

<a href="#kinopt.evol.utils.params.compute_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Computes various error metrics to evaluate the performance of the optimization process.
The function calculates the Mean Squared Error (MSE), Root Mean Squared Error (RMSE),
Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared value.
These metrics provide insights into the accuracy of the estimated time series compared to the observed data.</p>
<p>The function takes the following parameters:
:param optimized_params:
:param P_initial:
:param P_initial_array:
:param K_index:
:param K_array:
:param gene_psite_counts:
:param beta_counts:
:param n:</p>
<p>:return:
- P_estimated: Estimated time series matrix for all gene-psite combinations.
- residuals: Residuals between observed and estimated values.
- mse: Mean Squared Error.
- rmse: Root Mean Squared Error.
- mae: Mean Absolute Error.
- mape: Mean Absolute Percentage Error.
- r_squared: R-squared value.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.fitanalysis.helpers.postfit" class="doc doc-heading">
            <code>kinopt.fitanalysis.helpers.postfit</code>


<a href="#kinopt.fitanalysis.helpers.postfit" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">goodnessoffit</span><span class="p">(</span><span class="n">estimated</span><span class="p">,</span> <span class="n">observed</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the goodness of fit for estimated and observed values.
It creates scatter plots with confidence intervals and labels for genes outside the 95% CI.
The function also calculates KL divergence and generates a heatmap for optimization progression.</p>
<p>:param estimated:
:param observed:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">reshape_alpha_beta</span><span class="p">(</span><span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to reshape alpha and beta values for plotting.
It renames columns and creates a new 'Parameter' column for each type of value.</p>
<p>:param alpha_values:
:param beta_values:</p>
<p>:returns:
- pd.DataFrame: Reshaped DataFrame containing 'GeneID', 'Value', and 'Parameter' columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.perform_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">perform_pca</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.perform_pca" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Perform PCA analysis on the given DataFrame.
The DataFrame should contain a 'Value' column for PCA analysis.
The function returns a DataFrame with PCA results and additional columns for type and gene/psite information.</p>
<p>:param df: DataFrame containing the data for PCA analysis.
:return: DataFrame with PCA results and additional columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">(</span><span class="n">result_df_sorted</span><span class="p">,</span> <span class="n">y_axis_column</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.plot_pca" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Plot PCA or t-SNE results for each gene/psite.
The function creates scatter plots with different markers for alpha and beta parameters,
and adds labels for each point.
The function also adjusts text labels to avoid overlap using the adjustText library.</p>
<p>:param result_df_sorted: DataFrame containing PCA or t-SNE results.
:param y_axis_column: Column name for the y-axis values in the plot.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.perform_tsne" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">perform_tsne</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.perform_tsne" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Perform t-SNE analysis on the given scaled data.
The function returns a DataFrame with t-SNE results and additional columns for type and gene/psite information.</p>
<p>:param scaled_data:
:param df:</p>
<p>:return:
- pd.DataFrame: DataFrame with t-SNE results and additional columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.additional_plots" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">additional_plots</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">scaled_data</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">residuals_df</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.additional_plots" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to create additional plots including CDF, KDE, Boxplot, and Hierarchical Clustering.</p>
<p>:param df:
:param scaled_data:
:param alpha_values:
:param beta_values:
:param residuals_df:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_sankey_from_network</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a Sankey diagram from the given data and saves it as an HTML file.</p>
<p>This function processes the input data to generate nodes and links for a Sankey diagram.
It assigns colors to nodes and links based on their attributes and values, and uses Plotly
to render the diagram. The resulting diagram is saved as an HTML file in the specified output directory.</p>
<p>:param output_dir: str
    The directory where the Sankey diagram HTML file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the data for the Sankey diagram. It must include the following columns:
    - 'Source': The source node of the link.
    - 'Target': The target node of the link.
    - 'Value': The value of the link, which determines the flow size.
:param title: str
    The title of the Sankey diagram.</p>
<p>The function performs the following steps:
1. Initializes nodes and links for the Sankey diagram.
2. Maps node labels to indices and assigns colors to nodes.
3. Processes the data to create links between nodes, assigning colors based on link values.
4. Builds the Sankey diagram using Plotly.
5. Adds a color bar to explain the flow gradient.
6. Saves the Sankey diagram as an HTML file in the specified output directory.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.important_connections" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">important_connections</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span></code>

<a href="#kinopt.fitanalysis.helpers.postfit.important_connections" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extracts the top N most important connections based on their absolute values
and saves them to a CSV file.</p>
<p>:param output_dir: str
    The directory where the CSV file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the connections with columns 'Source', 'Target', and 'Value'.
:param top_n: int, optional
    The number of top connections to extract (default is 20).</p>
<p>The function sorts the connections by their absolute values in descending order,
selects the top N connections, and saves them to a CSV file named 'top_connections.csv'
in the specified output directory.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.config.constants" class="doc doc-heading">
            <code>kinopt.local.config.constants</code>


<a href="#kinopt.local.config.constants" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

<a href="#kinopt.local.config.constants.parse_args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parses command-line arguments for the optimization script.
This function uses argparse to handle various parameters related to the optimization process.
The parameters include bounds for the optimization, loss function types, estimation of missing kinases,
scaling methods for time-series data, and the optimization method to be used.
The function returns a tuple containing the parsed arguments.</p>
<p>:return: A tuple containing the parsed arguments.
    - lower_bound (float): Lower bound for the optimization.
    - upper_bound (float): Upper bound for the optimization.
    - loss_type (str): Type of loss function to use.
    - estimate_missing (bool): Whether to estimate missing kinase-psite values.
    - scaling_method (str): Method for scaling time-series data.
    - split_point (int): Split point for temporal scaling.
    - segment_points (list of int): Segment points for segmented scaling.
    - method (str): Optimization method to use.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.config.logconf" class="doc doc-heading">
            <code>kinopt.local.config.logconf</code>


<a href="#kinopt.local.config.logconf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.exporter.plotout" class="doc doc-heading">
            <code>kinopt.local.exporter.plotout</code>


<a href="#kinopt.local.exporter.plotout" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.plotout.plot_fits_for_gene" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for a gene.
It generates two plots:
1. A full timepoints plot showing all timepoints.
2. A short timepoints plot showing only the first 7 timepoints.
The plots are saved as PNG files in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.plotout.plot_cumulative_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.
It generates a plot showing the cumulative residuals over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.
It generates a plot showing the autocorrelation values over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.plotout.plot_histogram_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.
It generates a histogram showing the distribution of residuals.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.plotout.plot_qqplot_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.
It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.exporter.sheetutils" class="doc doc-heading">
            <code>kinopt.local.exporter.sheetutils</code>


<a href="#kinopt.local.exporter.sheetutils" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for a gene.
It generates two plots:
1. A full timepoints plot showing all timepoints.
2. A short timepoints plot showing only the first 7 timepoints.
The plots are saved as PNG files in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.
It generates a plot showing the cumulative residuals over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.
It generates a plot showing the autocorrelation values over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.
It generates a histogram showing the distribution of residuals.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.
It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span><span class="p">)</span></code>

<a href="#kinopt.local.exporter.sheetutils.output_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to output the results of the optimization process.
It logs the optimized alpha and beta values, optimization summary,
error metrics, and generates plots for each gene.
It also writes the results to an Excel file with multiple sheets.</p>
<p>The sheets include:
- Alpha Values: Optimized alpha values for each gene and psite.
- Beta Values: Optimized beta values for each kinase and psite.
- Summary: Summary of the optimization process.
- Observed: Observed time-series data for each gene and psite.
- Estimated: Estimated time-series data for each gene and psite.
- Residuals: Residuals for each gene and psite.</p>
<p>:param P_initial:
:param P_init_dense:
:param P_estimated:
:param residuals:
:param alpha_values:
:param beta_values:
:param result:
:param mse:
:param rmse:
:param mae:
:param mape:
:param r_squared:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.objfn.minfn" class="doc doc-heading">
            <code>kinopt.local.objfn.minfn</code>


<a href="#kinopt.local.objfn.minfn" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.opt.optrun" class="doc doc-heading">
            <code>kinopt.local.opt.optrun</code>


<a href="#kinopt.local.opt.optrun" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">obj_fun</span><span class="p">,</span> <span class="n">params_initial</span><span class="p">,</span> <span class="n">opt_method</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">constraints</span><span class="p">)</span></code>

<a href="#kinopt.local.opt.optrun.run_optimization" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run optimization using the specified method.</p>
<p>:param obj_fun:
:param params_initial:
:param opt_method:
:param bounds:
:param constraints:
:return: result, optimized parameters</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.optcon.construct" class="doc doc-heading">
            <code>kinopt.local.optcon.construct</code>


<a href="#kinopt.local.optcon.construct" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.utils.iodata" class="doc doc-heading">
            <code>kinopt.local.utils.iodata</code>


<a href="#kinopt.local.utils.iodata" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.iodata.format_duration" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Formats a duration in seconds into a human-readable string.
- If less than 60 seconds, returns in seconds.
- If less than 3600 seconds, returns in minutes.
- If more than 3600 seconds, returns in hours.</p>
<p>:param seconds:
:return: Formatted string</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.iodata.load_and_scale_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load and scale the data from the specified input files.</p>
<p>:param estimate_missing:
:param scaling_method:
:param split_point:
:param seg_points:
:return: Time series data, interaction data, observed data</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.iodata.apply_scaling" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Apply scaling to the specified columns of a DataFrame based on the given method.
The scaling methods include:
- 'min_max': Min-Max scaling
- 'log': Logarithmic scaling
- 'temporal': Temporal scaling (two segments)
- 'segmented': Segmented scaling (multiple segments)
- 'slope': Slope scaling
- 'cumulative': Cumulative scaling</p>
<p>:param df:
:param cols:
:param method:
:param split_point:
:param seg_points:
:return: df</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.iodata.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.iodata.organize_output_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to organize output files into protein-specific folders.
It moves files matching the pattern 'protein_name_*.{json,svg,png,html,csv,xlsx}'
into a folder named after the protein (e.g., 'ABL2') and moves all other files
into a 'General' folder within the same directory.</p>
<p>:param directories:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.utils.params" class="doc doc-heading">
            <code>kinopt.local.utils.params</code>


<a href="#kinopt.local.utils.params" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">unique_kinases</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.params.extract_parameters" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extracts the alpha and beta parameters from the optimized parameters.</p>
<p>:param P_initial:
:param gene_kinase_counts:
:param total_alpha:
:param unique_kinases:
:param K_index:
:param optimized_params:
:return: Alpha and beta values as dictionaries</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">gene_alpha_starts</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">gene_kinase_idx</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">kinase_beta_starts</span><span class="p">,</span> <span class="n">kinase_beta_counts</span><span class="p">,</span> <span class="n">K_data</span><span class="p">,</span> <span class="n">K_indices</span><span class="p">,</span> <span class="n">K_indptr</span><span class="p">)</span></code>

<a href="#kinopt.local.utils.params.compute_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Computes the estimated series and various metrics based on the optimized parameters.</p>
<p>:param optimized_params:
:param P_init_dense:
:param t_max:
:param gene_alpha_starts:
:param gene_kinase_counts:
:param gene_kinase_idx:
:param total_alpha:
:param kinase_beta_starts:
:param kinase_beta_counts:
:param K_data:
:param K_indices:
:param K_indptr:
:return: Estimated series, residuals, MSE, RMSE, MAE, MAPE, R-squared</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.optimality.KKT" class="doc doc-heading">
            <code>kinopt.optimality.KKT</code>


<a href="#kinopt.optimality.KKT" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.generate_latex_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_latex_table</span><span class="p">(</span><span class="n">summary_dict</span><span class="p">,</span> <span class="n">table_caption</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.generate_latex_table" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to generate a LaTeX table from a summary dictionary.
The table is formatted for use in a LaTeX document.</p>
<p>:param summary_dict: Dictionary containing summary data.
:param table_caption: Caption for the LaTeX table.
:param table: Optional table object to format.
:return: LaTeX table as a string.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.print_primal_feasibility_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_primal_feasibility_results</span><span class="p">(</span><span class="n">primal_summary</span><span class="p">,</span> <span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.print_primal_feasibility_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Logs the primal feasibility summary and violation details.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_sensitivity_and_active_constraints</span><span class="p">(</span><span class="n">sensitivity_summary</span><span class="p">,</span> <span class="n">active_constraints_summary</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Logs the sensitivity summary and active constraints summary.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.plot_constraint_violations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_constraint_violations</span><span class="p">(</span><span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.plot_constraint_violations" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot constraint violations for alpha and beta values.
It creates a stacked bar plot showing the violations for each protein.
The top 5 proteins with the highest violations are highlighted in red.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>alpha_violations</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series containing alpha constraint violations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_violations</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series containing beta constraint violations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.plot_sensitivity_analysis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_sensitivity_analysis</span><span class="p">(</span><span class="n">sensitivity_analysis</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.plot_sensitivity_analysis" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to plot sensitivity analysis results.
It creates a horizontal bar plot showing the mean, max, and min sensitivity for each protein.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sensitivity_analysis</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing sensitivity analysis results.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.process_excel_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_excel_results</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

<a href="#kinopt.optimality.KKT.process_excel_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to process the Excel results file.
It reads the alpha and beta values, estimated and observed values,
validates normalization constraints, computes residuals and gradients,
and generates LaTeX tables for the residuals and sensitivity summaries.
It also performs sensitivity analysis and identifies high sensitivity sites.
The results are returned as a dictionary.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the Excel file containing results.</p>
              </div>
            </td>
            <td>
                  <code><span title="kinopt.evol.config.constants.OUT_FILE">OUT_FILE</span></code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    dict: Dictionary containing the processed results, including alpha and beta values,
          estimated and observed values, constraint violations, residuals summary,
          sensitivity summary, and high sensitivity sites.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.post_optimization_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_results</span><span class="p">()</span></code>

<a href="#kinopt.optimality.KKT.post_optimization_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to process and visualize the results of the optimization.
It reads the results from an Excel file, processes the data,
and generates plots for constraint violations and sensitivity analysis.
It also prints the primal feasibility results and sensitivity summaries.
The results are returned as a dictionary.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing the processed results, including alpha and beta values,
  estimated and observed values, constraint violations, residuals summary,
  sensitivity summary, and high sensitivity sites.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.powell.runpowell" class="doc doc-heading">
            <code>kinopt.powell.runpowell</code>


<a href="#kinopt.powell.runpowell" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.powell.runpowell.run_powell" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_powell</span><span class="p">()</span></code>

<a href="#kinopt.powell.runpowell.run_powell" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run the Powell optimization algorithm using Julia.
It sets the number of threads to half of the available threads
and logs the output in real time.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.diagram.helpers" class="doc doc-heading">
            <code>models.diagram.helpers</code>


<a href="#models.diagram.helpers" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.powerset" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">powerset</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span></code>

<a href="#models.diagram.helpers.powerset" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return the list of all subsets (as frozensets) of the given iterable.
For example, for [1,2] returns:
  [frozenset(), frozenset({1}), frozenset({2}), frozenset({1,2})]</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.state_label" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">state_label</span><span class="p">(</span><span class="n">state</span><span class="p">)</span></code>

<a href="#models.diagram.helpers.state_label" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Convert a set of phosphorylation sites into a node label.
The unphosphorylated state (empty set) is labeled "P".
For nonempty states the label is "P" concatenated with the sorted site numbers.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_random_diagram" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_random_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

<a href="#models.diagram.helpers.create_random_diagram" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a random phosphorylation diagram.
The diagram displays mRNA production/degradation, protein production,
phosphorylation transitions (addition) and dephosphorylation transitions,
as well as degradation edges for each state.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_distributive_diagram" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_distributive_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

<a href="#models.diagram.helpers.create_distributive_diagram" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a distributive phosphorylation diagram.
This model shows single-step phosphorylation and dephosphorylation transitions
for each site, along with production/degradation edges.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_successive_model" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_successive_model</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

<a href="#models.diagram.helpers.create_successive_model" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a successive phosphorylation diagram.
In this model, phosphorylation occurs in a sequential order from P to P1 to P2, etc.,
with corresponding dephosphorylation and degradation transitions.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.distmod" class="doc doc-heading">
            <code>models.distmod</code>


<a href="#models.distmod" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.distmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span></code>

<a href="#models.distmod.ode_core" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>The core ODE system for the distributive phosphorylation model.</p>
<p>The system is defined by the following equations:</p>
<p>dR/dt = A - B * R
dP/dt = C * R - (D + sum(S_rates)) * P + sum(P_sites)
dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</p>
<p>where:</p>
<p>R: the concentration of the mRNA
P: the concentration of the protein
P_sites: the concentration of the phosphorylated sites
A: the rate of production of the mRNA
B: the rate of degradation of the mRNA
C: the rate of production of the protein
D: the rate of degradation of the protein
S_rates: the rates of phosphorylation of each site
D_rates: the rates of dephosphorylation of each site</p>
<p>:param y:
:param A:
:param B:
:param C:
:param D:
:param S_rates:
:param D_rates:
:return: Derivative of y</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.distmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#models.distmod.ode_system" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>The ODE system for the distributive phosphorylation model which calls the core ODE system.</p>
<p>:param y:
:param t:
:param params:
:param num_psites:
:return: ode_core(y, A, B, C, D, S_rates, D_rates)</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.distmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span></code>

<a href="#models.distmod.solve_ode" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Solve the ODE system for the distributive phosphorylation model.</p>
<p>:param params:
:param init_cond:
:param num_psites:
:param t:
:return: solution of the ODE system, solution of phosphorylated sites</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.randmod" class="doc doc-heading">
            <code>models.randmod</code>


<a href="#models.randmod" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.randmod.prepare_vectorized_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_vectorized_arrays</span><span class="p">(</span><span class="n">num_sites</span><span class="p">)</span></code>

<a href="#models.randmod.prepare_vectorized_arrays" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepare vectorized arrays for the Random ODE system.
It creates binary states, phosphorylation targets, and dephosphorylation targets.</p>
<p>Binary states are represented as a 2D array where each row corresponds to a state
and each column corresponds to a phosphorylation site.</p>
<p>Phosphorylation and Dephosphorylation targets are represented as 2D arrays where
each row corresponds to a state and each column corresponds to a phosphorylation site.</p>
<p>Target values indicate the resulting state after phosphorylation or dephosphorylation.</p>
<p>:param num_sites: Number of phosphorylation sites
:return: binary_states, PHOSPHO_TARGET, DEPHOSPHO_TARGET</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.randmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span><span class="p">,</span> <span class="o">*</span><span class="n">params</span><span class="p">)</span></code>

<a href="#models.randmod.ode_system" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>The ODE system for the random phosphorylation model. This function computes the derivatives of the
variables R, P, and X at each time step.</p>
<ul>
<li>R is the concentration of the receptor</li>
<li>P is the concentration of the protein</li>
<li>X is the concentration of the phosphorylated states</li>
<li>S is the phosphorylation rate for each site</li>
<li>D_deg is the degradation rate for each state</li>
<li>A, B, C, D are parameters for the system</li>
<li>num_sites is the number of phosphorylation sites</li>
<li>binary_states is a 2D array representing the binary states of the system</li>
<li>PHOSPHO_TARGET is a 2D array representing the phosphorylation targets</li>
<li>DEPHOSPHO_TARGET is a 2D array representing the dephosphorylation targets</li>
</ul>
<p>:param y:
    y[0] = R
    y[1] = P
    y[2:] = X
:param t:
:param A:
:param B:
:param C:
:param D:
:param num_sites:
:param binary_states:
:param PHOSPHO_TARGET:
:param DEPHOSPHO_TARGET:
:param params:
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.randmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">initial_conditions</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span></code>

<a href="#models.randmod.solve_ode" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Solve the Random ODE system using the provided parameters and initial conditions.
The function integrates the ODE system over the specified time points and returns
the solution.</p>
<p>:param popt:
:param initial_conditions:
:param num_psites:
:param time_points:
:return: solution, solution of phosphorylated states for each site</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.succmod" class="doc doc-heading">
            <code>models.succmod</code>


<a href="#models.succmod" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.succmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span></code>

<a href="#models.succmod.ode_core" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>The core of the ODE system for the successive ODE model.</p>
<p>The system is defined by the following equations:
dR/dt = A - B * R
dP/dt = C * R - D * P - S_rates[0] * P + sum(P_sites)
dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</p>
<p>where:
R: the concentration of the mRNA
P: the concentration of the protein
P_sites: the concentration of the phosphorylated sites
A: the rate of production of the mRNA
B: the rate of degradation of the mRNA
C: the rate of production of the protein
D: the rate of degradation of the protein
S_rates: the rates of phosphorylation of each site
D_rates: the rates of dephosphorylation of each site</p>
<p>:param y:
:param A:
:param B:
:param C:
:param D:
:param S_rates:
:param D_rates:
:return: derivative of y</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.succmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#models.succmod.ode_system" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>The ODE system for the successive ODE model.</p>
<p>:param y:
:param t:
:param params:
:param num_psites:
:return: ode_core(y, A, B, C, D, S_rates, D_rates)</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.succmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span></code>

<a href="#models.succmod.solve_ode" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Solve the ODE system using the given parameters and initial conditions.
The function integrates the ODE system over time and returns the solution.</p>
<p>:param params:
:param init_cond:
:param num_psites:
:param t:
:return: solution, solution of phosphorylated sites</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.testmod" class="doc doc-heading">
            <code>models.testmod</code>


<a href="#models.testmod" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.weights" class="doc doc-heading">
            <code>models.weights</code>


<a href="#models.weights" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.weights.early_emphasis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">early_emphasis</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#models.weights.early_emphasis" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function that calculates custom weights for early time points in a dataset.
The weights are based on the data values and the time differences between points.</p>
<p>The weights are calculated in a way that emphasizes early time points,
while also considering the data values and time intervals.</p>
<p>:param p_data:
:param time_points:
:param num_psites:
:return: flattened array of custom weights</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.weights.get_weight_options" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_weight_options</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">use_regularization</span><span class="p">,</span> <span class="n">reg_len</span><span class="p">,</span> <span class="n">early_weights</span><span class="p">)</span></code>

<a href="#models.weights.get_weight_options" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Function to calculate weights for parameter estimation based on the target data and time points.
The weights are designed to emphasize early time points and account for noise in the data.
The function also includes options for regularization and custom early point emphasis.</p>
<p>The following are the weighting schemes:
- Inverse data: 1 / abs(target)
- Exponential decay: exp(-0.5 * target)
- Log scale: 1 / log(1 + abs(target))
- Time difference: 1 / abs(time_diff)
- Moving average: 1 / abs(target - moving_avg)
- Sigmoid time decay: 1 / (1 + exp(time_indices - 5))
- Exponential early emphasis: exp(-0.5 * time_indices)
- Polynomial decay: 1 / (1 + 0.5 * time_indices)
- MS SNR model: 1 / sqrt(signal)
- MS inverse variance: 1 / (abs(target) ** 0.7)
- Flat region penalty: 1 / abs(grad)
- Steady state decay: exp(-0.1 * time_indices)
- Combined data time: 1 / (abs(target) * (1 + 0.5 * time_indices))
- Inverse sqrt data: 1 / sqrt(abs(target))
- Early emphasis moderate: ones
- Early emphasis steep decay: ones
- Custom early points emphasis: early_weights</p>
<p>:param target:
:param t_target:
:param num_psites:
:param use_regularization:
:param reg_len:
:param early_weights:
:return: dictionary of weights for parameter estimation</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.adapest" class="doc doc-heading">
            <code>paramest.adapest</code>


<a href="#paramest.adapest" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.adapest.prepare_model_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

<a href="#paramest.adapest.prepare_model_func" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepares the model function for curve fitting.
Adaptive Estimation is used to estimate the parameters of the model.</p>
<p>:param num_psites:
:param init_cond:
:param bounds:
:param fixed_params:
:param use_regularization:
:param lambda_reg:
:return: model function, free indices, free bounds, fixed values, total params</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.adapest.adaptive_estimation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">adaptive_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">extra_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

<a href="#paramest.adapest.adaptive_estimation" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Adaptive estimation of parameters using curve fitting.
This function estimates the parameters of the model using curve fitting.
Parameter will be fixed at certain time points.</p>
<p>:param p_data:
:param init_cond:
:param num_psites:
:param time_points:
:param t:
:param bounds:
:param fixed_params:
:param gene:
:param bootstraps:
:param extra_fixed:
:param use_regularization:
:param lambda_reg:
:return: estimated parameters</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.adapest.estimate_profiles" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">estimate_profiles</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">time_fixed_dict</span><span class="p">)</span></code>

<a href="#paramest.adapest.estimate_profiles" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Estimate adaptive profiles for a given gene at specified time points.
This function estimates the parameters of the model using adaptive estimation.
The parameters are estimated at the specified time points and stored in a DataFrame.
The function also plots the profiles.</p>
<p>:param gene:
:param p_data:
:param init_cond:
:param num_psites:
:param time_points:
:param desired_times:
:param bounds:
:param fixed_params:
:param bootstraps:
:param time_fixed_dict:
:return: dataframe, profiles</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.core" class="doc doc-heading">
            <code>paramest.core</code>


<a href="#paramest.core" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.core.process_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">desired_times</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">time_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

<a href="#paramest.core.process_gene" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Process a single gene by estimating its parameters and generating plots.
This function extracts gene-specific data, estimates parameters using the
specified estimation mode, and generates plots for the estimated parameters
and the model fits. It also calculates error metrics and saves the results
to Excel files.</p>
<p>:param gene:
:param measurement_data:
:param time_points:
:param bounds:
:param fixed_params:
:param desired_times:
:param time_fixed:
:param bootstraps:
:param out_dir:
:return:
    - gene: The gene being processed.
    - estimated_params: Estimated parameters for the gene.
    - model_fits: Model fits for the gene.
    - seq_model_fit: Sequential model fit for the gene.
    - errors: Error metrics (MSE, MAE).
    - final_params: Final estimated parameters.
    - profiles: Adaptive profile estimates (if applicable).
    - profiles_df: DataFrame of adaptive profile estimates (if applicable).
    - param_df: DataFrame of estimated parameters.
    - gene_psite_data: Dictionary of gene-specific data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.core.process_gene_wrapper" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_gene_wrapper</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

<a href="#paramest.core.process_gene_wrapper" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Wrapper function to process a gene. This function is a placeholder for
any additional processing or modifications needed before calling the
main processing function.</p>
<p>:param gene:
:param measurement_data:
:param time_points:
:param bounds:
:param fixed_params:
:param desired_times:
:param time_fixed:
:param bootstraps:
:param out_dir:
:return:
    - result: Dictionary containing the results of the gene processing.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.identifiability.ci" class="doc doc-heading">
            <code>paramest.identifiability.ci</code>


<a href="#paramest.identifiability.ci" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.identifiability.ci.confidence_intervals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">confidence_intervals</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">pcov</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">alpha_val</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span></code>

<a href="#paramest.identifiability.ci.confidence_intervals" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Computes the confidence intervals for parameter estimates using a linearization approach.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="popt">popt</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>1D numpy array of best-fit parameter estimates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="pcov">pcov</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Square covariance matrix (numpy array) corresponding to popt.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="target">target</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>1D numpy array of observed data (used to compute degrees of freedom).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="alpha_val">alpha_val</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Significance level (default 0.05 for a 95% confidence interval).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with the following keys:
'beta_hat': Best-fit parameter estimates.
'se_lin': Standard errors (sqrt of diagonal of pcov).
'df_lin': Degrees of freedom (n_obs - n_params).
't_stat': t-statistics for each parameter.
'pval': Two-sided p-values for each parameter.
'qt_lin': t critical value for the given alpha and degrees of freedom.
'lwr_ci': Lower 95% confidence intervals.
'upr_ci': Upper 95% confidence intervals.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.normest" class="doc doc-heading">
            <code>paramest.normest</code>


<a href="#paramest.normest" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.normest.normest" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">normest</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

<a href="#paramest.normest.normest" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Perform normal parameter estimation using all provided time points at once.
Uses the provided bounds (ignores fixed_params so that all parameters are estimated)
and supports bootstrapping if specified.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="p_data">p_data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Measurement data (DataFrame or numpy array). Assumes data starts at column index 2.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="init_cond">init_cond</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial condition for the ODE solver.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="num_psites">num_psites</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="time_points">time_points</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of time points to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bounds">bounds</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of parameter bounds.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="fixed_params">fixed_params</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(Ignored in normest) Provided for interface consistency.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bootstraps">bootstraps</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of bootstrapping iterations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="use_regularization">use_regularization</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flag to apply Tikhonov regularization.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="lambda_reg">lambda_reg</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization strength.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>est_params: List with the full estimated parameter vector.</li>
</ul>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>model_fits: List with the ODE solution and model predictions.</li>
</ul>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>error_vals: List with the squared error (data vs. model prediction).</li>
</ul>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.seqest" class="doc doc-heading">
            <code>paramest.seqest</code>


<a href="#paramest.seqest" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.seqest.prepare_model_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span></code>

<a href="#paramest.seqest.prepare_model_func" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepare the model function for sequential parameter estimation.</p>
<p>This function builds the model function based on the specified
ODE model type and the number of phosphorylation sites. It also
sets up the bounds for the free parameters and handles fixed
parameters. The model function is used for curve fitting to
estimate the parameters of the ODE model.</p>
<p>:param num_psites:
:param init_cond:
:param bounds:
:param fixed_params:
:param use_regularization:
:param lambda_reg:
:return: model_func, free_indices, free_bounds, fixed_values, num_total_params</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.seqest.fit_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fit_parameters</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">p0_free</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">weight_options</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

<a href="#paramest.seqest.fit_parameters" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fit the parameters of the model using curve fitting with different
weighting options. This function iterates over the provided
weight options, performs the curve fitting, and evaluates the
goodness of fit using a scoring function. The best fitting
parameters and their corresponding score are returned.</p>
<p>:param time_points:
:param p_data:
:param model_func:
:param p0_free:
:param bounds:
:param weight_options:
:param use_regularization:
:return: best_fit_params, best_weight_key, scores</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.seqest.sequential_estimation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sequential_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

<a href="#paramest.seqest.sequential_estimation" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Perform sequential parameter estimation for a given gene using
the specified model function. This function iteratively fits the
model to the data at each time point, updating the initial guess
for the parameters based on the previous fit. The estimated
parameters, model fits, and error values are returned.</p>
<p>:param p_data:
:param time_points:
:param init_cond:
:param bounds:
:param fixed_params:
:param num_psites:
:param gene:
:param use_regularization:
:param lambda_reg:
:return:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.toggle" class="doc doc-heading">
            <code>paramest.toggle</code>


<a href="#paramest.toggle" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.toggle.estimate_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">estimate_parameters</span><span class="p">(</span><span class="n">mode</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">)</span></code>

<a href="#paramest.toggle.estimate_parameters" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Toggle between sequential and normal (all timepoints) estimation.</p>
<p>This function allows for the selection of the estimation mode
and handles the parameter estimation process accordingly.</p>
<p>It uses the sequential estimation method for "sequential" mode
and the normal estimation method for "normal" mode.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="mode">mode</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The estimation mode, either "sequential" or "normal".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="gene">gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The gene name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="p_data">p_data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Measurement data (DataFrame or numpy array).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="init_cond">init_cond</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial condition for the ODE solver.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="num_psites">num_psites</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="time_points">time_points</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of time points to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bounds">bounds</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of parameter bounds.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="fixed_params">fixed_params</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of fixed parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bootstraps">bootstraps</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of bootstrapping iterations (only used in normal mode).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>:returns:
    - model_fits: List with the ODE solution and model predictions.
    - estimated_params: List with the full estimated parameter vector.
    - seq_model_fit: Sequential model fit for the gene.
    - errors: Error metrics (MSE, MAE).</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="plotting.plotting" class="doc doc-heading">
            <code>plotting.plotting</code>


<a href="#plotting.plotting" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="plotting.plotting.Plotter" class="doc doc-heading">
            <code>Plotter</code>


<a href="#plotting.plotting.Plotter" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>A class to encapsulate plotting functionalities for ODE model analysis.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.gene">gene</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The gene or experiment name.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.out_dir">out_dir</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where plots will be saved.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.color_palette">color_palette</span></code></td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of color codes used for plotting.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_parallel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_parallel</span><span class="p">(</span><span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_parallel" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a parallel coordinates plot for the given solution.</p>
<p>:param solution: 2D numpy array of shape (sampels, features)
:param labels: list of labels</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.pca_components" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pca_components</span><span class="p">(</span><span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.pca_components" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a scree plot showing the explained variance ratio for PCA components.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param target_variance: The target cumulative explained variance to determine the required number of components.
:return: A tuple containing the number of required components and the explained variance ratio.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">(</span><span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_pca" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the PCA results for the given solution.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param components: Number of PCA components to plot. Defaults to 3.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_tsne" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_tsne</span><span class="p">(</span><span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_tsne" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a t-SNE visualization of the given solution.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param perplexity: Perplexity parameter for t-SNE. Defaults to 30.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_param_bar" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_param_bar</span><span class="p">(</span><span class="n">params_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">s_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_param_bar" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a bar chart of parameter values for the given gene.</p>
<p>This method visualizes the estimated parameter values for a specific gene
and its phosphorylation sites. It uses color coding to distinguish between
different phosphorylation sites and other parameters.</p>
<p>:param params_df: DataFrame containing parameter values.
:param s_df: DataFrame containing phosphorylation site information.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_param_series" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_param_series</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">param_names</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_param_series" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the time series of estimated parameters over the given time points.</p>
<p>This method visualizes the evolution of kinetic rates or parameters
over time for a specific gene.</p>
<p>:param estimated_params: List of estimated parameter values at each time point.
:param param_names: List of parameter names corresponding to the estimated parameters.
:param time_points: 1D numpy array of time points.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_profiles" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_profiles</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_profiles" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the profiles of estimated parameters over time.</p>
<p>:param data: DataFrame containing the estimated parameters and time points.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_model_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_model_fit</span><span class="p">(</span><span class="n">model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">sol</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_model_fit" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the model fit for the given data.</p>
<p>:param model_fit: Estimated model fit values.
:param P_data: Observed data for phosphorylation levels.
:param sol: ODE solution for mRNA and protein levels.
:param num_psites: number of phosphorylation sites.
:param psite_labels: labels for the phosphorylation sites.
:param time_points: time points for the data.
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_A_S" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_A_S</span><span class="p">(</span><span class="n">est_arr</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">time_vals</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_A_S" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the scatter plot of A vs S and the density contour plot.</p>
<p>:param est_arr: Estimated parameters array.
:param num_psites: Number of phosphorylation sites.
:param time_vals: Time values for the data.
:return:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_all" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_all</span><span class="p">(</span><span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_all" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Function that calls parallel, t-SNE, PCA, and model fit plots.
If mode is sequential, it also calls parameter series and A-S plots.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param labels: List of labels for the solution.
:param estimated_params: List of estimated parameter values.
:param time_points: 1D numpy array of time points.
:param P_data: Observed data for phosphorylation levels.
:param seq_model_fit: Estimated model fit values.
:param psite_labels: Labels for the phosphorylation sites.
:param perplexity: Perplexity parameter for t-SNE.
:param components: Number of PCA components to plot.
:param target_variance: The target cumulative explained variance to determine the required number of components.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_clusters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_clusters</span><span class="p">(</span><span class="n">s_values_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">cluster_labels</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_clusters" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the clusters of S values for the given gene.
Expects s_values_df to have columns 'S_value', 'GeneID', and 'Psite'.</p>
<p>:param s_values_df: DataFrame containing S values and gene information.
:param cluster_labels: Cluster labels for each S value.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_heatmap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">param_value_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_heatmap" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Expects param_value_df to have a 'Protein' column.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_error_distribution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_error_distribution</span><span class="p">(</span><span class="n">error_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_error_distribution" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Expects error_df to have a 'MAE' column.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_gof_1" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_gof_1</span><span class="p">(</span><span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_gof_1" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Expects merged_data to contain 'GeneID', 'Psite', and columns 'x1_obs' to 'x14_obs' and 'x1_est' to 'x14_est'.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_gof_5" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_gof_5</span><span class="p">(</span><span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_gof_5" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Uses the row means of observed (x1_obs:x14_obs) and estimated (x1_est:x14_est) values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_kld</span><span class="p">(</span><span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span></code>

<a href="#plotting.plotting.Plotter.plot_kld" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Expects merged_data to have columns 'x1_obs' to 'x14_obs' and 'x1_est' to 'x14_est',
as well as 'GeneID' and 'Psite'.</p>


    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="processing.cleanup" class="doc doc-heading">
            <code>processing.cleanup</code>


<a href="#processing.cleanup" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_collecttri" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_collecttri</span><span class="p">()</span></code>

<a href="#processing.cleanup.process_collecttri" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Processes the CollecTRI file to clean and filter mRNA-TF interactions.
Removes complex interactions, filters by target genes, and saves the result.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.format_site" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_site</span><span class="p">(</span><span class="n">site</span><span class="p">)</span></code>

<a href="#processing.cleanup.format_site" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Formats a phosphorylation site string.</p>
<p>If the input is NaN or an empty string, returns an empty string.
If the input contains an underscore ('_'), splits the string into two parts,
converts the first part to uppercase, and appends the second part unchanged.
Otherwise, converts the entire string to uppercase.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>site</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The phosphorylation site string to format.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The formatted phosphorylation site string.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_msgauss" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_msgauss</span><span class="p">()</span></code>

<a href="#processing.cleanup.process_msgauss" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to generate time series data.</p>
<p>This function performs the following steps:
1. Loads the <code>MS_Gaussian_updated_09032023.csv</code> file.
2. Computes the transformed values as 2^(predict_mean).
3. Pivots the data to create a time series for each (GeneID, Psite) pair.
4. Renames the time point columns to <code>x1</code> to <code>x14</code> and formats the <code>Psite</code> column.
5. Saves the cleaned time series to <code>input1.csv</code>.
6. Filters the data to keep only rows where <code>Psite</code> starts with <code>Y_</code>, <code>S_</code>, <code>T_</code>, or is empty.
7. Saves the filtered time series to <code>input1.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input1.csv</code>: Cleaned and filtered time series data.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_msgauss_std" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_msgauss_std</span><span class="p">()</span></code>

<a href="#processing.cleanup.process_msgauss_std" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to compute transformed means and standard deviations.</p>
<p>This function performs the following steps:
1. Loads the <code>MS_Gaussian_updated_09032023.csv</code> file.
2. Computes transformed values as 2^(predict_mean) and propagates errors using the formula:
   sigma_y = 2^(x) * ln(2) * sigma_x.
3. Pivots the data to create time series for each (GeneID, Psite) pair for both means and standard deviations.
4. Merges the pivoted data for means and standard deviations.
5. Filters the data to keep only rows where <code>Psite</code> starts with <code>Y_</code>, <code>S_</code>, <code>T_</code>, or is empty.
6. Saves the resulting data to <code>input1_wstd.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input1_wstd.csv</code>: Cleaned time series data with transformed means and standard deviations.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data with standard deviations has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_routlimma" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_routlimma</span><span class="p">()</span></code>

<a href="#processing.cleanup.process_routlimma" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Processes the Rout Limma table to generate time series data for mRNA.</p>
<p>This function performs the following steps:
1. Loads the <code>Rout_LimmaTable.csv</code> file.
2. Selects specific columns related to time points and conditions.
3. Renames the selected columns to a standardized format (<code>x1</code> to <code>x9</code>).
4. Converts the values in the renamed columns using the formula <code>2^(value)</code>.
5. Saves the resulting time series data to <code>input3.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input3.csv</code>: Cleaned and transformed time series data for mRNA.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file <code>Rout_LimmaTable.csv</code> is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.update_gene_symbols" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_gene_symbols</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>

<a href="#processing.cleanup.update_gene_symbols" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Updates the GeneID column in a CSV file by mapping GeneIDs to gene/protein symbols.</p>
<p>This function reads a CSV file, queries the MyGeneInfo database to map GeneIDs to their
corresponding gene/protein symbols, and writes the updated DataFrame back to the same file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the CSV file to be updated. The file must contain a 'GeneID' column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the specified file does not exist.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the 'GeneID' column is missing in the file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li>The input file is updated in place with the 'GeneID' column replaced by gene/protein symbols.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A progress bar indicating the mapping process.</li>
<li>A message confirming the update of gene symbols in the file.</li>
</ul>
</details>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.move_processed_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">move_processed_files</span><span class="p">()</span></code>

<a href="#processing.cleanup.move_processed_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Moves or copies processed files to their respective directories.</p>
<p>This function organizes processed files into specific directories for
transcription factor optimization (TFOpt) and kinase optimization (KinOpt).
It creates the target directories if they do not exist and moves or copies
the files based on whether they have already been moved.</p>


<details class="directories" open>
  <summary>Directories</summary>
  <ul>
<li><code>../tfopt/data</code>: Target directory for TFOpt files.</li>
<li><code>../kinopt/data</code>: Target directory for KinOpt files.</li>
</ul>
</details>

<details class="files" open>
  <summary>Files</summary>
  <ul>
<li>KinOpt files: ["input1.csv", "input2.csv"]</li>
<li>TFOpt files: ["input1.csv", "input3.csv", "input4.csv"]</li>
</ul>
</details>

<details class="behavior" open>
  <summary>Behavior</summary>
  <ul>
<li>If a file has already been moved, it is copied to the target directory.</li>
<li>If a file has not been moved, it is moved to the target directory.</li>
<li>If a file does not exist, a message is printed.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>Messages indicating whether files were moved, copied, or not found.</li>
</ul>
</details>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="processing.map" class="doc doc-heading">
            <code>processing.map</code>


<a href="#processing.map" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="processing.map.map_optimization_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">map_optimization_results</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

<a href="#processing.map.map_optimization_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Reads the TF-mRNA optimization results from an Excel file and maps mRNA to each TF.</p>
<p>This function processes the 'Alpha Values' sheet of the provided Excel file to extract
non-zero optimization results, groups mRNA by TF, and merges the results with additional
data from a CSV file containing TF, Psite, and Kinase information. The final DataFrame
is cleaned and formatted for further analysis.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the Excel file containing TF-mRNA optimization results.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the mapped TF, mRNA, Psite, and Kinase information.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.map.create_cytoscape_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_cytoscape_table</span><span class="p">(</span><span class="n">mapping_csv_path</span><span class="p">)</span></code>

<a href="#processing.map.create_cytoscape_table" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a Cytoscape-compatible edge table from a mapping file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mapping_csv_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input CSV file with columns: TF, mRNA, Psite, Kinase</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: Edge table with columns [Source, Target, Interaction]</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.map.generate_nodes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_nodes</span><span class="p">(</span><span class="n">edge_df</span><span class="p">)</span></code>

<a href="#processing.map.generate_nodes" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Infers node types for Cytoscape visualization:
- All nodes default to 'Kinase'
- Nodes that are only targets of 'regulates' are labeled 'mRNA'</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>edge_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Must have columns ['Source', 'Target', 'Interaction']</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame with columns ['Node', 'Type']</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="sensitivity.analysis" class="doc doc-heading">
            <code>sensitivity.analysis</code>


<a href="#sensitivity.analysis" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.define_sensitivity_problem_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">define_sensitivity_problem_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

<a href="#sensitivity.analysis.define_sensitivity_problem_rand" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Problem definition for sensitivity analysis.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.define_sensitivity_problem_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">define_sensitivity_problem_ds</span><span class="p">(</span><span class="n">ub</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#sensitivity.analysis.define_sensitivity_problem_ds" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ub</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Upper bound for the parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Problem definition for sensitivity analysis.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.sensitivity_analysis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sensitivity_analysis</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span></code>

<a href="#sensitivity.analysis.sensitivity_analysis" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Performs sensitivity analysis using the Morris method for a given ODE model.</p>
<p>This function defines the sensitivity problem based on the ODE model type,
generates parameter samples, evaluates the model for each sample, and computes
sensitivity indices. It also generates various plots to visualize the results.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>time_points</code>
            </td>
            <td>
                  <code><span title="list">list</span> or <span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time points for the ODE simulation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>init_cond</code>
            </td>
            <td>
                  <code><span title="list">list</span> or <span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial conditions for the ODE model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene or protein being analyzed.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>None</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function saves sensitivity analysis results and plots to the output directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.initdist" class="doc doc-heading">
            <code>steady.initdist</code>


<a href="#steady.initdist" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initdist.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#steady.initdist.initial_condition" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for distributive phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.initrand" class="doc doc-heading">
            <code>steady.initrand</code>


<a href="#steady.initrand" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initrand.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#steady.initrand.initial_condition" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for random phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.initsucc" class="doc doc-heading">
            <code>steady.initsucc</code>


<a href="#steady.initsucc" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initsucc.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#steady.initsucc.initial_condition" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for successive phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.inittest" class="doc doc-heading">
            <code>steady.inittest</code>


<a href="#steady.inittest" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.inittest.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span></code>

<a href="#steady.inittest.initial_condition" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Compute steady-state conditions for the given model.</p>


<details class="common-parameters" open>
  <summary>Common parameters</summary>
  <p>A, B, C, D = 1, 1, 1, 1
S_rates = ones(num_psites)
D_rates = ones(num_psites)</p>
</details>

<details class="the-parameter-vector-for-the-ode-models-is" open>
  <summary>The parameter vector for the ODE models is</summary>
  <p>[A, B, C, D, S1, ..., S_n, D1, ..., D_n]</p>
</details>        <p>Extra (hardcoded) parameters per model:
  - semi_processive: p_proc = 0.8
  - oscillating:      A_osc = 1.0, omega = 0.5 (steady state computed at t=0)
  - noise:            noise_std = 0.01 (set to 0 for steady state)
  - cooperative:      hill_n = 2, K_half = 0.5</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.config.constants" class="doc doc-heading">
            <code>tfopt.evol.config.constants</code>


<a href="#tfopt.evol.config.constants" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

<a href="#tfopt.evol.config.constants.parse_args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov
- optimizer: Global Evolutionary Optimization method (default: 0).
    Options:
    0: NGSA2
    1: SMSEMOA
    2: AGEMOEA</p>
<p>:returns
- lower_bound: Lower bound for the optimization variables.
- upper_bound: Upper bound for the optimization variables.
- loss_type: Type of loss function to use.
- optimizer: Global Evolutionary Optimization method.
:rtype: tuple
:raises argparse.ArgumentError: If an invalid argument is provided.
:raises SystemExit: If the script is run with invalid arguments.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.config.logconf" class="doc doc-heading">
            <code>tfopt.evol.config.logconf</code>


<a href="#tfopt.evol.config.logconf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.exporter.plotout" class="doc doc-heading">
            <code>tfopt.evol.exporter.plotout</code>


<a href="#tfopt.evol.exporter.plotout" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

<a href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Plot the estimated vs observed expression levels for a set of genes.
This function generates two plots for each gene:
1. A full time series plot showing the observed and estimated expression levels.
2. A plot showing the observed and estimated expression levels for the first 5 time points.
Additionally, it plots the protein signals for the regulators of each gene.
The plots are saved as PNG files in the specified save path.</p>
<p>:param predictions: Predicted expression levels.
:param expression_matrix: Observed expression levels.
:param gene_ids: List of gene identifiers.
:param time_points: Time points for the expression data.
:param regulators: Matrix of regulators for each gene.
:param tf_protein_matrix: Matrix of protein signals for each transcription factor.
:param tf_ids: List of transcription factor identifiers.
:param num_targets: Number of target genes to plot.
:param save_path: Path to save the plots.
:type predictions: np.ndarray
:type expression_matrix: np.ndarray
:type gene_ids: list
:type time_points: np.ndarray
:type regulators: np.ndarray
:type tf_protein_matrix: np.ndarray
:type tf_ids: list
:type num_targets: int
:type save_path: str</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.plotout.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#tfopt.evol.exporter.plotout.compute_predictions" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Compute the predicted expression levels based on the optimization variables.
This function calculates the predicted expression levels for each mRNA based on the optimization variables,
the regulators, and the protein and phosphorylation site tensors.</p>
<p>:param x:
:param regulators:
:param protein_mat:
:param psite_tensor:
:param n_reg:
:param T_use:
:param n_mRNA:
:param beta_start_indices:
:param num_psites:
:return: predictions</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.exporter.sheetutils" class="doc doc-heading">
            <code>tfopt.evol.exporter.sheetutils</code>


<a href="#tfopt.evol.exporter.sheetutils" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_results_to_excel</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

<a href="#tfopt.evol.exporter.sheetutils.save_results_to_excel" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.
This function creates multiple sheets in the Excel file to store different types of data,
including alpha values, beta values, residuals, observed values, estimated values,
and optimization results.
Each sheet is formatted with appropriate column names and data types.
The function also calculates various metrics (MSE, MAE, MAPE, R^2)
to evaluate the performance of the optimization.</p>
<p>:param gene_ids:
:param tf_ids:
:param final_alpha:
:param final_beta:
:param psite_labels_arr:
:param expression_matrix:
:param predictions:
:param objective_value:
:param reg_map:
:param filename:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.objfn.minfn" class="doc doc-heading">
            <code>tfopt.evol.objfn.minfn</code>


<a href="#tfopt.evol.objfn.minfn" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="doc doc-heading">
            <code>TFOptimizationMultiObjectiveProblem</code>


<a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.Problem">Problem</span></code></p>


        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Multi-objective optimization problem for TF optimization.
This class defines a multi-objective optimization problem for the
transcription factor (TF) optimization problem. It inherits from the
<code>Problem</code> class in the pymoo library. The problem is defined with three
objectives: f1 (error), f2 (alpha violation), and f3 (beta violation).</p>
<p>The problem is initialized with the following parameters:
- n_var: Number of decision variables.
- n_mRNA: Number of mRNA genes.
- n_TF: Number of transcription factors.
- n_reg: Number of regulators.
- n_psite_max: Maximum number of phosphorylation sites.
- n_alpha: Number of alpha parameters.
- mRNA_mat: Matrix of measured mRNA values.
- regulators: Matrix of TF regulators for each mRNA.
- protein_mat: Matrix of TF protein time series.
- psite_tensor: Tensor of phosphorylation site signals.
- T_use: Number of time points used.
- beta_start_indices: Integer array giving the starting index (in the β–segment)
                      for each TF.
- num_psites: Integer array with the actual number of phosphorylation sites for each TF.
- no_psite_tf: Boolean array indicating if the TF has no phosphorylation sites.
- xl: Lower bounds for the decision variables.
- xu: Upper bounds for the decision variables.
- kwargs: Additional keyword arguments.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">n_var</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Initialize the multi-objective optimization problem.</p>
<p>:param n_var:
:param n_mRNA:
:param n_TF:
:param n_reg:
:param n_psite_max:
:param n_alpha:
:param mRNA_mat:
:param regulators:
:param protein_mat:
:param psite_tensor:
:param T_use:
:param beta_start_indices:
:param num_psites:
:param no_psite_tf:
:param xl:
:param xu:
:param kwargs:</p>


    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span></code>

<a href="#tfopt.evol.objfn.minfn.objective_" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Computes a loss value using one of several loss functions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mRNA_mat</code>
            </td>
            <td>
                  <code>       </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_mRNA x T_use) measured mRNA values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>     </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_mRNA x n_reg) indices of TF regulators for each mRNA.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>protein_mat</code>
            </td>
            <td>
                  <code>    </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>   </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per mRNA.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_mRNA,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of mRNA and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment)
                 for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>     </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>      </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1,
                 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lam1,</code>
            </td>
            <td>
                  <code><span title="lam2">lam2</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization parameters (used for loss_type 5 and 6).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.opt.optrun" class="doc doc-heading">
            <code>tfopt.evol.opt.optrun</code>


<a href="#tfopt.evol.opt.optrun" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">total_dim</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">)</span></code>

<a href="#tfopt.evol.opt.optrun.run_optimization" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run the optimization using the specified algorithm and problem.
This function sets up the algorithm parameters, initializes the optimizer,
and runs the optimization process.</p>
<p>:param problem:
:param total_dim:
:param optimizer:
:return: result</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.optcon.construct" class="doc doc-heading">
            <code>tfopt.evol.optcon.construct</code>


<a href="#tfopt.evol.optcon.construct" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_fixed_arrays</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

<a href="#tfopt.evol.optcon.construct.build_fixed_arrays" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.
Returns:
  - mRNA_mat: array of shape (n_mRNA, T)
  - regulators: array of shape (n_mRNA, n_reg) with indices into TF_ids.
  - protein_mat: array of shape (n_TF, T)
  - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
  - n_reg: maximum number of regulators per mRNA.
  - n_psite_max: maximum number of PSites among TFs.
  - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
  - num_psites: array of length n_TF with the actual number of PSites for each TF.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.optcon.filter" class="doc doc-heading">
            <code>tfopt.evol.optcon.filter</code>


<a href="#tfopt.evol.optcon.filter" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.load_raw_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_raw_data</span><span class="p">()</span></code>

<a href="#tfopt.evol.optcon.filter.load_raw_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load raw data from files.
This includes mRNA data, TF data, and regulation maps.
The function returns the following:
- mRNA_ids: List of mRNA gene identifiers.
- mRNA_mat: Matrix of mRNA expression data.
- mRNA_time_cols: Time points for mRNA data.
- TF_ids: List of transcription factor identifiers.
- protein_dict: Dictionary mapping TF_ids to their protein data.
- psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.
- psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.
- TF_time_cols: Time points for TF data.
- reg_map: Regulation map, mapping mRNA genes to their regulators.</p>
<p>:return: mRNA_ids, mRNA_mat, mRNA_time_cols, TF_ids, protein_dict, psite_dict, psite_labels_dict, TF_time_cols, reg_map</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.filter_mrna" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_mrna</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

<a href="#tfopt.evol.optcon.filter.filter_mrna" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter mRNA genes to only those with regulators present in the regulation map.
This function returns the filtered mRNA_ids and their corresponding expression matrix.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param mRNA_mat: Matrix of mRNA expression data.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:return: filtered_mRNA_ids, filtered_mRNA_mat</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.update_regulations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_regulations</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">)</span></code>

<a href="#tfopt.evol.optcon.filter.update_regulations" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Update the regulation map to only include relevant transcription factors.
This function modifies the reg_map in place and returns a set of relevant transcription factors.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:param TF_ids: List of transcription factor identifiers.
:return: relevant_TFs: Set of relevant transcription factors.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.filter_TF" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_TF</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">relevant_TFs</span><span class="p">)</span></code>

<a href="#tfopt.evol.optcon.filter.filter_TF" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter transcription factors to only those present in the relevant_TFs set.
This function returns the filtered TF_ids and their corresponding protein and phosphorylation site data.
This is important for ensuring that only relevant transcription factors are included in the analysis.
:param TF_ids: List of transcription factor identifiers.
:param protein_dict: Dictionary mapping TF_ids to their protein data.
:param psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.
:param psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.
:param relevant_TFs: Set of relevant transcription factors.
:return: filtered TF_ids, protein_dict, psite_dict, psite_labels_dict</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.determine_T_use" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">determine_T_use</span><span class="p">(</span><span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_time_cols</span><span class="p">)</span></code>

<a href="#tfopt.evol.optcon.filter.determine_T_use" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Determine the number of time points to use for the analysis.
This function takes the mRNA matrix and TF time columns as input
and returns the minimum number of time points available across both datasets.
This is important for ensuring that the analysis is consistent and
comparable across different datasets.
The function checks the shape of the mRNA matrix and the length of the TF time columns
to determine the minimum number of time points.
If the mRNA matrix has fewer time points than the TF time columns,
it uses the number of time points in the mRNA matrix.
If the TF time columns have fewer time points, it uses that value instead.
This ensures that the analysis is based on the same number of time points
for both mRNA and TF data.</p>
<p>:param mRNA_mat: Matrix of mRNA expression data.
:param TF_time_cols: Time points for TF data.
:return: T_use: Number of time points to use for the analysis.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.utils.iodata" class="doc doc-heading">
            <code>tfopt.evol.utils.iodata</code>


<a href="#tfopt.evol.utils.iodata" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_mRNA_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_mRNA_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.load_mRNA_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load mRNA data from a CSV file.
The file is expected to have a "GeneID" column and time columns for expression data.
The function returns a list of mRNA gene identifiers, a matrix of expression data,
and the time columns.
The mRNA_ids are converted to strings, and the expression data is converted to a numpy array of floats.
The time columns are extracted from the DataFrame, excluding the "GeneID" column.
The function returns:
    - mRNA_ids: List of mRNA gene identifiers (strings).
    - mRNA_mat: Numpy array of expression data (floats).
    - time_cols: List of time columns (excluding "GeneID").</p>
<p>:param filename: str
:return: mRNA_ids, mRNA_mat, time_cols</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_TF_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_TF_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.load_TF_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load TF data from a CSV file.
The file is expected to have a "GeneID" column, a "Psite" column, and time columns for protein data.
The function returns a list of TF gene identifiers, a dictionary mapping TF_ids to their protein data,
a dictionary mapping TF_ids to their phosphorylation site data, a dictionary mapping TF_ids to their phosphorylation site labels,
and the time columns.
The TF_ids are converted to strings, and the protein data is stored in a dictionary with TF_ids as keys.
The phosphorylation site data is stored in a list for each TF_id, and the phosphorylation site labels are also stored in a list.
The time columns are being extracted from the DataFrame, excluding the "GeneID" and "Psite" columns.
The function returns:
    - TF_ids: List of TF gene identifiers (strings).
    - protein_dict: Dictionary mapping TF_ids to their protein data (numpy arrays).
    - psite_dict: Dictionary mapping TF_ids to their phosphorylation site data (lists of numpy arrays).
    - psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels (lists of strings).
    - time_cols: List of time columns (excluding "GeneID" and "Psite").</p>
<p>:param filename: str
:return: TF_ids, protein_dict, psite_dict, psite_labels_dict, time_cols</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.load_regulation" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load regulation data from a CSV file.
The file is expected to have "Source" and "Target" columns.
The function returns a dictionary mapping mRNA gene identifiers to their regulators (TFs).
The mRNA gene identifiers and TF identifiers are converted to strings and stripped of whitespace.
The function returns:
    - reg_map: Dictionary mapping mRNA gene identifiers to their regulators (TFs).</p>
<p>:param filename: str
:return: reg_map</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.organize_output_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Organizes output files from multiple directories into separate folders for each protein.
Files are moved into folders named after the protein identifier extracted from the filename.
Remaining files are moved to a "General" folder.
Args:
    directories (str): List of directories to organize.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.iodata.format_duration" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.
The function converts the duration into hours, minutes, or seconds,
depending on the length of the duration.
Args:
    seconds (float): Duration in seconds.
Returns:
    str: Formatted duration string.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.utils.params" class="doc doc-heading">
            <code>tfopt.evol.utils.params</code>


<a href="#tfopt.evol.utils.params" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_no_psite_array" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_no_psite_array</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.create_no_psite_array" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create an array indicating whether each TF has no phosphorylation sites.
A TF is considered to have no phosphorylation sites if:
1. The number of phosphorylation sites is zero.
2. All labels for the phosphorylation sites are empty strings.
This function is used to determine the initial guess for the beta parameters
in the optimization process.</p>
<p>:param n_TF:
:param num_psites:
:param psite_labels_arr:
:return: array of booleans indicating no phosphorylation sites for each TF</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.compute_beta_indices" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_beta_indices</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.compute_beta_indices" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Compute the starting indices for the beta parameters for each TF.
The beta parameters are stored in a flat array, and this function computes
the starting index for each TF based on the number of phosphorylation sites.
The starting index for the beta parameters of TF i is given by:
beta_start_indices[i] = sum(1 + num_psites[j] for j in range(i))
This function is used to extract the beta parameters from the flat array
during the optimization process.</p>
<p>:param num_psites: array of integers indicating the number of phosphorylation sites for each TF
:param n_TF: number of TFs
:return: beta_start_indices: array of integers indicating the starting index for each TF
         cum: total number of beta parameters</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_initial_guess" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_initial_guess</span><span class="p">(</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.create_initial_guess" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create the initial guess for the optimization variables.
The initial guess is a flat array containing the alpha and beta parameters.
The alpha parameters are initialized to 1.0 / n_reg for each mRNA-regulator pair.
The beta parameters are initialized to 1.0 for the protein and 1.0 / (1 + num_psites[i])
for each phosphorylation site of TF i.
The beta parameters for TFs with no phosphorylation sites are initialized to 1.0.
The initial guess is used as the starting point for the optimization process.
The initial guess is a flat array of length n_alpha + n_beta_total,
where n_alpha is the number of alpha parameters and n_beta_total is the total number of beta parameters.
The alpha parameters are stored in the first n_alpha elements of the array,
and the beta parameters are stored in the remaining elements.
The beta parameters are stored in the order of the TFs, with the protein parameter first
followed by the phosphorylation site parameters.
The beta parameters for TFs with no phosphorylation sites are stored as a single value.
The beta parameters for TFs with phosphorylation sites are stored as a vector of length 1 + num_psites[i].</p>
<p>:param n_mRNA:
:param n_reg:
:param n_TF:
:param num_psites:
:param no_psite_tf:
:return: initial guess array, n_alpha</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_bounds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_bounds</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_beta_total</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.create_bounds" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create the lower and upper bounds for the optimization variables.
The lower bounds are set to 0 for the alpha parameters and lb for the beta parameters.
The upper bounds are set to 1 for the alpha parameters and ub for the beta parameters.
The bounds are used to constrain the optimization process and ensure that the parameters
are within a reasonable range.
The bounds are stored in two separate arrays: xl and xu.
The xl array contains the lower bounds for each parameter,
and the xu array contains the upper bounds.</p>
<p>:param n_alpha: number of alpha parameters
:param n_beta_total: total number of beta parameters
:param lb: lower bound for beta parameters
:param ub: upper bound for beta parameters
:return: xl: lower bounds array
         xu: upper bounds array</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.get_parallel_runner" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_parallel_runner</span><span class="p">()</span></code>

<a href="#tfopt.evol.utils.params.get_parallel_runner" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get a parallel runner for multi-threading.
This function uses the lscpu command to determine the number of threads available
on the system and creates a ThreadPool with that number of threads.
The ThreadPool is used to parallelize the optimization process and speed up the computation.</p>
<p>:return: runner: StarmapParallelization object for parallel execution
            pool: ThreadPool object for managing threads</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.extract_best_solution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_best_solution</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.extract_best_solution" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract the best solution from the optimization results.
This function finds the best solution based on the Pareto front,
which is a set of non-dominated solutions in the objective space.
The best solution is determined by minimizing the weighted sum of the objectives.
The alpha parameters are reshaped into a matrix of shape (n_mRNA, n_reg),
and the beta parameters are extracted based on the starting indices and
the number of phosphorylation sites for each TF.
The beta parameters are stored in a list of arrays, where each array corresponds to a TF
and contains the protein parameter and the phosphorylation site parameters.
The function returns the final alpha and beta parameters, the best objectives,
and the final decision variables.</p>
<p>:param res: optimization results
:param n_alpha: number of alpha parameters
:param n_mRNA: number of mRNAs
:param n_reg: number of regulators
:param n_TF: number of transcription factors
:param num_psites: number of phosphorylation sites for each TF
:param beta_start_indices: starting indices for the beta parameters</p>
<p>:return: final_alpha: final alpha parameters (n_mRNA x n_reg)
            final_beta: final beta parameters (n_TF x (1 + num_psites))
            best_objectives: best objectives (3 objectives)
            final_x: final decision variables (flat array)</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.print_alpha_mapping" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_alpha_mapping</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.print_alpha_mapping" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to mRNAs with their corresponding alpha values.
This function iterates through the mRNA IDs and their corresponding regulators,
and logs the TFs that are present in the final alpha matrix.
The alpha values are printed for each TF that regulates the mRNA.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:param TF_ids: List of transcription factor identifiers.
:param final_alpha: Final alpha parameters (n_mRNA x n_reg).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.print_beta_mapping" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_beta_mapping</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

<a href="#tfopt.evol.utils.params.print_beta_mapping" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to their beta parameters.
This function iterates through the TF IDs and their corresponding beta values,
and logs the beta values for each TF.
The beta values are printed for the protein and each phosphorylation site,
with appropriate labels.</p>
<p>:param TF_ids: List of transcription factor identifiers.
:param final_beta: Final beta parameters (n_TF x (1 + num_psites)).
:param psite_labels_arr: Array of phosphorylation site labels for each TF.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.fitanalysis.helper" class="doc doc-heading">
            <code>tfopt.fitanalysis.helper</code>


<a href="#tfopt.fitanalysis.helper" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="tfopt.fitanalysis.helper.Plotter" class="doc doc-heading">
            <code>Plotter</code>


<a href="#tfopt.fitanalysis.helper.Plotter" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


        <p>A class to plot various analysis results from an Excel file.
The class provides methods to visualize the alpha and beta values,
residuals, observed and estimated values, and other metrics.</p>
<p>The plots include:
- Alpha distribution
- Beta bar plots
- Heatmap of absolute residuals
- Goodness of fit
- Kullback-Leibler divergence
- PCA
- Box plots for alpha and beta values
- CDF for alpha and beta values
- Time-wise residuals</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.__init__" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Initializes the Plotter instance by loading data from the Excel file.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.load_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_data</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.load_data" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Loads data from the specified Excel file.
The data includes residuals, observed values, estimated values,
alpha values, and beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_alpha_distribution</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the distribution of alpha parameter values grouped by transcription factors (TFs)
using a strip plot.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_beta_barplots</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Processes the beta values DataFrame and creates a separate bar plot
for each unique transcription factor (TF).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_heatmap_abs_residuals</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a heatmap of the absolute values of the residuals.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_goodness_of_fit</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Creates a scatter plot comparing observed vs. estimated values,
fits a linear regression model, plots the 95% confidence interval,
and labels points outside the confidence interval.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_kld</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_kld" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the Kullback-Leibler Divergence (KLD) for each mRNA.
The KLD is calculated between the observed and estimated distributions
of the mRNA expression levels.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_pca" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a PCA (Principal Component Analysis) of the observed and estimated values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_boxplot_alpha</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the alpha values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_boxplot_beta</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cdf_alpha</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the alpha values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cdf_beta</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_time_wise_residuals</span><span class="p">()</span></code>

<a href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Plots the residuals over time for each mRNA.</p>


    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.config.constants" class="doc doc-heading">
            <code>tfopt.local.config.constants</code>


<a href="#tfopt.local.config.constants" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

<a href="#tfopt.local.config.constants.parse_args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov</p>
<p>:return: lower_bound, upper_bound, loss_type</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.config.logconf" class="doc doc-heading">
            <code>tfopt.local.config.logconf</code>


<a href="#tfopt.local.config.logconf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.exporter.plotout" class="doc doc-heading">
            <code>tfopt.local.exporter.plotout</code>


<a href="#tfopt.local.exporter.plotout" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

<a href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Plots the estimated vs observed values for a given set of genes and their corresponding TFs.
The function generates two types of plots:
1. A full time series plot showing the observed and estimated values for each gene.
2. A plot showing the first 5 time points for each gene.
Additionally, it generates an interactive plot using Plotly for each gene.</p>
<p>:param predictions: 2D numpy array of predicted values.
:param expression_matrix: 2D numpy array of observed values.
:param gene_ids: List of gene identifiers.
:param time_points: 1D numpy array of time points.
:param regulators: 2D numpy array of regulators for each gene.
:param tf_protein_matrix: 2D numpy array of TF protein signals.
:param tf_ids: List of TF identifiers.
:param num_targets: Number of target genes to plot.
:param save_path: Directory path to save the plots.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.exporter.sheetutils" class="doc doc-heading">
            <code>tfopt.local.exporter.sheetutils</code>


<a href="#tfopt.local.exporter.sheetutils" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_results_to_excel</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

<a href="#tfopt.local.exporter.sheetutils.save_results_to_excel" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.
This function creates multiple sheets in the Excel file, each containing different
aspects of the optimization results, including alpha values, beta values,
residuals, observed values, estimated values, and optimization metrics.
Each sheet is formatted with appropriate column names and data types.
The function also computes various metrics such as Mean Squared Error (MSE),
Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared
score to evaluate the performance of the optimization.</p>
<p>:param gene_ids: List of gene identifiers.
:param tf_ids: List of TF identifiers.
:param final_alpha: 2D numpy array of alpha values.
:param final_beta: 2D numpy array of beta values.
:param psite_labels_arr: List of post-translational modification labels.
:param expression_matrix: 2D numpy array of observed values.
:param predictions: 2D numpy array of predicted values.
:param objective_value: Objective value of the optimization.
:param reg_map: Dictionary mapping genes to their regulators.
:param filename: Path to the output Excel file.
:type filename: str</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.objfn.minfn" class="doc doc-heading">
            <code>tfopt.local.objfn.minfn</code>


<a href="#tfopt.local.objfn.minfn" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">)</span></code>

<a href="#tfopt.local.objfn.minfn.objective_" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Computes a loss value using one of several loss functions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x</code>
            </td>
            <td>
                  <code>                 </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expression_matrix</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>        </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein_matrix</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>      </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>        </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lam1,</code>
            </td>
            <td>
                  <code><span title="lam2">lam2</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization parameters (used for loss_type 5 and 6).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

<a href="#tfopt.local.objfn.minfn.compute_predictions" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Computes the predicted expression matrix based on the decision vector x.
This function uses the regulators, TF protein matrix, and post-translational modification tensor to generate
predictions for each gene at each time point.
Parameters:
  x                  : Decision vector.
  regulators         : (n_genes x n_reg) indices of TF regulators for each gene.
  tf_protein_matrix  : (n_TF x T_use) TF protein time series.
  psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).
  n_reg              : Maximum number of regulators per gene.
  T_use              : Number of time points used.
  n_genes, n_TF     : Number of genes and TF respectively.
  beta_start_indices : Integer array giving the starting index (in the β–segment) for each TF.
  num_psites         : Integer array with the actual number of PSites for each TF.
Returns:
    predictions        : (n_genes x T_use) matrix of predicted gene expression values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.objective_wrapper" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_wrapper</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">)</span></code>

<a href="#tfopt.local.objfn.minfn.objective_wrapper" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Wrapper function for the objective function.
This function is used to call the objective function with the appropriate parameters.
Parameters:
  x                  : Decision vector.
  expression_matrix  : (n_genes x T_use) measured gene expression values.
  regulators         : (n_genes x n_reg) indices of TF regulators for each gene.
  tf_protein_matrix  : (n_TF x T_use) TF protein time series.
  psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).
  n_reg              : Maximum number of regulators per gene.
  T_use              : Number of time points used.
  n_genes, n_TF     : Number of genes and TF respectively.
  beta_start_indices : Integer array giving the starting index (in the β–segment) for each TF.
  num_psites         : Integer array with the actual number of PSites for each TF.
  loss_type          : Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).
Returns:
    loss               : The computed loss (a scalar).</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.opt.optrun" class="doc doc-heading">
            <code>tfopt.local.opt.optrun</code>


<a href="#tfopt.local.opt.optrun" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.opt.optrun.run_optimizer" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimizer</span><span class="p">(</span><span class="n">x0</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">lin_cons</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">)</span></code>

<a href="#tfopt.local.opt.optrun.run_optimizer" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Runs the optimization algorithm to minimize the objective function.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x0</code>
            </td>
            <td>
                  <code>                 </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial guess for the optimization variables.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>bounds</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Bounds for the optimization variables.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lin_cons</code>
            </td>
            <td>
                  <code>           </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Linear constraints for the optimization problem.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expression_matrix</code>
            </td>
            <td>
                  <code>  </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein_matrix</code>
            </td>
            <td>
                  <code>  </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>       </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of loss function to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    result             : Result of the optimization process, including the optimized parameters and objective value.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.optcon.construct" class="doc doc-heading">
            <code>tfopt.local.optcon.construct</code>


<a href="#tfopt.local.optcon.construct" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_fixed_arrays</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

<a href="#tfopt.local.optcon.construct.build_fixed_arrays" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.
Returns:
  - expression_matrix: array of shape (n_genes, T)
  - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.
  - tf_protein_matrix: array of shape (n_TF, T)
  - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
  - n_reg: maximum number of regulators per gene.
  - n_psite_max: maximum number of PSites among TFs.
  - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
  - num_psites: array of length n_TF with the actual number of PSites for each TF.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.constraint_alpha_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">constraint_alpha_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">)</span></code>

<a href="#tfopt.local.optcon.construct.constraint_alpha_func" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>For each gene, the sum of its alpha parameters must equal 1.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.constraint_beta_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">constraint_beta_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

<a href="#tfopt.local.optcon.construct.constraint_beta_func" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>For each TF, the sum of its beta parameters must equal 1.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.build_linear_constraints" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_linear_constraints</span><span class="p">(</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

<a href="#tfopt.local.optcon.construct.build_linear_constraints" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Build linear constraints for the optimization problem.</p>
<p>The constraints are:
1. For each mRNA, the sum of its alpha parameters must equal 1.
2. For each TF, the sum of its beta parameters must equal 1.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.optcon.filter" class="doc doc-heading">
            <code>tfopt.local.optcon.filter</code>


<a href="#tfopt.local.optcon.filter" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.filter.load_and_filter_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_filter_data</span><span class="p">()</span></code>

<a href="#tfopt.local.optcon.filter.load_and_filter_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load and filter data for the optimization problem.
This function loads gene expression data, transcription factor (TF) data,
and regulation data. It filters genes to only include those with at least
one regulator and filters regulators to only include those present in the
TF data. The function returns the filtered gene IDs, expression matrix,
TF IDs, TF protein data, TF phosphorylation site data, TF phosphorylation
site labels, and the regulation map.
It raises a ValueError if no genes with regulators are found.
The function also filters the TF data to only include those present in the
regulation map.</p>
<p>The function returns:
- gene_ids: List of filtered gene IDs.
- expr_matrix: Filtered expression matrix.
- expr_time_cols: Time columns for gene expression data.
- tf_ids: Filtered TF IDs.
- tf_protein: Dictionary mapping TF IDs to their protein data.
- tf_psite_data: Dictionary mapping TF IDs to their phosphorylation site data.
- tf_psite_labels: Dictionary mapping TF IDs to their phosphorylation site labels.
- tf_time_cols: Time columns for TF data.
- reg_map: Regulation map, mapping gene IDs to their regulators.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.filter.prepare_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_data</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">tf_time_cols</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

<a href="#tfopt.local.optcon.filter.prepare_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepares the data for optimization by filtering the expression matrix
to match the number of time points and building fixed arrays.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expr_matrix</code>
            </td>
            <td>
                  <code><span title="np.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene expression matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their protein data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_psite_data</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_psite_labels</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site labels.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_time_cols</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time columns for TF data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reg_map</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    fixed_arrays (tuple): Tuple containing the fixed arrays:
        - expression_matrix: array of shape (n_genes, T)
        - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.
        - tf_protein_matrix: array of shape (n_TF, T)
        - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
        - n_reg: maximum number of regulators per gene.
        - n_psite_max: maximum number of PSites among TFs.
        - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
        - num_psites: array of length n_TF with the actual number of PSites for each TF.
    T_use (int): Number of time points used in the expression matrix.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.utils.iodata" class="doc doc-heading">
            <code>tfopt.local.utils.iodata</code>


<a href="#tfopt.local.utils.iodata" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.min_max_normalize" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">min_max_normalize</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">custom_max</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.min_max_normalize" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Row-wise (per-sample) min-max normalize time-series columns starting with 'x'.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input DataFrame with time-series columns (x1-xN).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>custom_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If given, used as max for all rows.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: Normalized DataFrame with same shape.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_expression_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_expression_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.load_expression_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Loads gene expression (mRNA) data.
Expects a CSV with a 'GeneID' column and time-point columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_tf_protein_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_tf_protein_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.load_tf_protein_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Loads TF protein data along with PSite information.
Expects a CSV with 'GeneID' and 'Psite' columns.
For rows without a valid PSite, the entire row is considered as the protein signal.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.load_regulation" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">



<details class="assumes-the-regulation-file-is-reversed" open>
  <summary>Assumes the regulation file is reversed</summary>
  <ul>
<li>The 'Source' column holds gene (mRNA) identifiers.</li>
<li>The 'Target' column holds TF identifiers.</li>
</ul>
</details>        <p>Returns a mapping from gene (source) to a list of TFs (targets).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.summarize_stats" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">summarize_stats</span><span class="p">(</span><span class="n">input3</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">,</span> <span class="n">input1</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">,</span> <span class="n">input4</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.summarize_stats" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Summarizes statistics for the expression data (input3) and TF protein data (input1).
It also summarizes the data after filtering based on the mapping file (input4).</p>


<details class="the-function-prints-the-following-statistics" open>
  <summary>The function prints the following statistics</summary>
  <ul>
<li>Global min, max, std, var for the full dataset.</li>
<li>Time-wise min, max, std, var for each time point.</li>
<li>Global min, max, std, var for the subset data (filtered by input4).</li>
<li>Time-wise min, max, std, var for the subset data.</li>
</ul>
</details>        <p>Args:
    input3 (str): Path to the expression data CSV file.
    input1 (str): Path to the TF protein data CSV file.
    input4 (str): Path to the mapping file CSV.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.iodata.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.utils.params" class="doc doc-heading">
            <code>tfopt.local.utils.params</code>


<a href="#tfopt.local.utils.params" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.params.get_optimization_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_optimization_parameters</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.params.get_optimization_parameters" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepare the optimization parameters for the optimization problem.
This function initializes the alpha and beta parameters, sets up the bounds,
and constructs the linear constraints for the optimization problem.
It returns the initial guess for the optimization variables, the number of
alpha parameters, the starting indices for beta parameters, the bounds,
the boolean array indicating TFs without phosphorylation sites, and the
number of genes and TFs.</p>
<p>Parameters:
expression_matrix (np.ndarray): Gene expression matrix.
tf_protein_matrix (np.ndarray): TF protein matrix.
n_reg (int): Number of regulators.
T_use (int): Number of time points used in the expression matrix.
psite_labels_arr (list): List of lists containing phosphorylation site labels.
num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.
lb (float): Lower bound for beta parameters.
ub (float): Upper bound for beta parameters.</p>
<p>Returns:
x0 (np.ndarray): Initial guess for the optimization variables.
n_alpha (int): Number of alpha parameters.
beta_start_indices (np.ndarray): Starting indices for beta parameters.
bounds (list): List of tuples specifying the bounds for each optimization variable.
no_psite_tf (np.ndarray): Boolean array indicating TFs without phosphorylation sites.
n_genes (int): Number of genes.
n_TF (int): Number of TFs.
num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.
lin_cons (LinearConstraint): Linear constraints for the optimization problem.
T_use (int): Number of time points used in the expression matrix.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.params.postprocess_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">postprocess_results</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

<a href="#tfopt.local.utils.params.postprocess_results" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Post-process the optimization results to extract the final alpha and beta parameters.
This function reshapes the optimization result into the final alpha and beta matrices,
and builds the mapping of TFs to mRNAs and phosphorylation sites.
It also logs the mappings for debugging purposes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
                  <code><span title="OptimizeResult">OptimizeResult</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The result of the optimization.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_alpha</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Starting indices for beta parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing the number of phosphorylation sites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reg_map</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_labels_arr</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of lists containing phosphorylation site labels.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>final_x</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final optimization result.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>final_alpha</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final alpha parameters reshaped into a matrix.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>final_beta</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final beta parameters reshaped into a matrix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="utils.display" class="doc doc-heading">
            <code>utils.display</code>


<a href="#utils.display" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="utils.display.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span></code>

<a href="#utils.display.ensure_output_directory" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Ensure the output directory exists. If it doesn't, create it.</p>
<p>:param directory: str</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.load_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_data</span><span class="p">(</span><span class="n">excel_file</span><span class="p">,</span> <span class="n">sheet</span><span class="o">=</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span></code>

<a href="#utils.display.load_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load data from an Excel file. The default sheet is "Estimated Values".</p>
<p>:param excel_file: str
:param sheet: str
:return: DataFrame
:rtype: pd.DataFrame</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

<a href="#utils.display.format_duration" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.
The function converts seconds into a string representation in the format:
- "X sec" for seconds
- "X min" for minutes
- "X hr" for hours</p>
<p>:param seconds: float
:return: formatted string</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.save_result" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_result</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">excel_filename</span><span class="p">)</span></code>

<a href="#utils.display.save_result" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save the results to an Excel file with multiple sheets.
Each sheet corresponds to a different gene and contains:
- Sequential Parameter Estimates
- Profiled Estimates (if available)
- Errors summary
The sheet names are prefixed with the gene name, truncated to 25 characters.
Args:
    results (list): List of dictionaries containing results for each gene.
    excel_filename (str): Path to the output Excel file.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

<a href="#utils.display.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

<a href="#utils.display.organize_output_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Organize output files into protein-specific folders and a general folder.
Files matching the pattern "protein_name_*.{json,svg,png,html,csv,xlsx}"
will be moved to a folder named after the protein.
Remaining files will be moved to a "General" folder within the same directory.</p>
<p>:param directories: List of directories to organize.
:type directories: list</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="utils.tables" class="doc doc-heading">
            <code>utils.tables</code>


<a href="#utils.tables" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="utils.tables.generate_tables" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_tables</span><span class="p">(</span><span class="n">xlsx_file_path</span><span class="p">)</span></code>

<a href="#utils.tables.generate_tables" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate hierarchical tables from the XLSX file containing alpha and beta values.
The function reads the alpha and beta values from the specified XLSX file,
processes them to create hierarchical tables, and returns a list of these tables.
Each table is a DataFrame with a MultiIndex for the columns, representing
the alpha and beta values for different kinases and phosphorylation sites.
The tables are structured to facilitate easy comparison and analysis of the
phosphorylation data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.tables.save_tables" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_tables</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">)</span></code>

<a href="#utils.tables.save_tables" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save the generated tables as LaTeX and CSV files.
Each table is saved with a filename based on the protein and phosphorylation site.
The LaTeX files are formatted for easy inclusion in a larger document,
and the CSV files are saved for further analysis.</p>
<p>:param tables: List of tuples containing protein, psite, and the corresponding table.
:param output_dir: Directory where the tables will be saved.
:type output_dir: str</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.tables.save_master_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_master_table</span><span class="p">(</span><span class="n">folder</span><span class="o">=</span><span class="s1">&#39;latex&#39;</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;latex/all_tables.tex&#39;</span><span class="p">)</span></code>

<a href="#utils.tables.save_master_table" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Save a master LaTeX file that includes all individual LaTeX files from the specified folder.
This function generates a LaTeX file that includes all the individual LaTeX files
for each protein and phosphorylation site.</p>
<p>:param folder: Directory containing the individual LaTeX files.
:param output_file: Output LaTeX file name.
:type folder: str
:type output_file: str</p>


    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2025 <a href="https://bibymaths.github.io">Abhinav Mishra</a>.
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.instant", "navigation.sections", "content.code.annotate"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.js"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/auto-render.min.js"></script>
      
        <script src="../js/katex-setup.js"></script>
      
    
  </body>
</html>