
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="..">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Reference - PhosKinTime</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="PhosKinTime" class="md-header__button md-logo" aria-label="PhosKinTime" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PhosKinTime
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Reference
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="PhosKinTime" class="md-nav__button md-logo" aria-label="PhosKinTime" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    PhosKinTime
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bin.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bin.main.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config" class="md-nav__link">
    <span class="md-ellipsis">
      config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.extract_config" class="md-nav__link">
    <span class="md-ellipsis">
      extract_config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.log_config" class="md-nav__link">
    <span class="md-ellipsis">
      log_config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_bound_pair" class="md-nav__link">
    <span class="md-ellipsis">
      parse_bound_pair
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_fix_value" class="md-nav__link">
    <span class="md-ellipsis">
      parse_fix_value
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.score_fit" class="md-nav__link">
    <span class="md-ellipsis">
      score_fit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.generate_labels_ds" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.generate_labels_rand" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_bounds_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_bounds_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_number_of_params_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_number_of_params_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_param_names_ds" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_param_names_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.location" class="md-nav__link">
    <span class="md-ellipsis">
      location
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter" class="md-nav__link">
    <span class="md-ellipsis">
      ColoredFormatter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ColoredFormatter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.format" class="md-nav__link">
    <span class="md-ellipsis">
      format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.remove_ansi" class="md-nav__link">
    <span class="md-ellipsis">
      remove_ansi
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf.setup_logger" class="md-nav__link">
    <span class="md-ellipsis">
      setup_logger
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_de" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_de
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_nsga
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_residuals_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo" class="md-nav__link">
    <span class="md-ellipsis">
      minfndiffevo
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii" class="md-nav__link">
    <span class="md-ellipsis">
      minfnnsgaii
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_de" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_de
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_nsga
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct.pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      pipeline
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit" class="md-nav__link">
    <span class="md-ellipsis">
      postfit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.additional_plots" class="md-nav__link">
    <span class="md-ellipsis">
      additional_plots
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="md-nav__link">
    <span class="md-ellipsis">
      create_sankey_from_network
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="md-nav__link">
    <span class="md-ellipsis">
      goodnessoffit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.important_connections" class="md-nav__link">
    <span class="md-ellipsis">
      important_connections
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_pca" class="md-nav__link">
    <span class="md-ellipsis">
      perform_pca
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      perform_tsne
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="md-nav__link">
    <span class="md-ellipsis">
      reshape_alpha_beta
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT" class="md-nav__link">
    <span class="md-ellipsis">
      KKT
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.generate_latex_table" class="md-nav__link">
    <span class="md-ellipsis">
      generate_latex_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_constraint_violations" class="md-nav__link">
    <span class="md-ellipsis">
      plot_constraint_violations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      plot_sensitivity_analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.post_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_primal_feasibility_results" class="md-nav__link">
    <span class="md-ellipsis">
      print_primal_feasibility_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      print_sensitivity_and_active_constraints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.process_excel_results" class="md-nav__link">
    <span class="md-ellipsis">
      process_excel_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell" class="md-nav__link">
    <span class="md-ellipsis">
      runpowell
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell.run_powell" class="md-nav__link">
    <span class="md-ellipsis">
      run_powell
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers" class="md-nav__link">
    <span class="md-ellipsis">
      helpers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_distributive_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_distributive_diagram
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_random_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_random_diagram
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_successive_model" class="md-nav__link">
    <span class="md-ellipsis">
      create_successive_model
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.powerset" class="md-nav__link">
    <span class="md-ellipsis">
      powerset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.state_label" class="md-nav__link">
    <span class="md-ellipsis">
      state_label
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod" class="md-nav__link">
    <span class="md-ellipsis">
      distmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod" class="md-nav__link">
    <span class="md-ellipsis">
      randmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.prepare_vectorized_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_vectorized_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod" class="md-nav__link">
    <span class="md-ellipsis">
      succmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.testmod" class="md-nav__link">
    <span class="md-ellipsis">
      testmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights.early_emphasis" class="md-nav__link">
    <span class="md-ellipsis">
      early_emphasis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights.get_weight_options" class="md-nav__link">
    <span class="md-ellipsis">
      get_weight_options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest" class="md-nav__link">
    <span class="md-ellipsis">
      adapest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.adaptive_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      adaptive_estimation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.estimate_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_profiles
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core" class="md-nav__link">
    <span class="md-ellipsis">
      core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core.process_gene" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core.process_gene_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene_wrapper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci" class="md-nav__link">
    <span class="md-ellipsis">
      ci
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci.confidence_intervals" class="md-nav__link">
    <span class="md-ellipsis">
      confidence_intervals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest" class="md-nav__link">
    <span class="md-ellipsis">
      seqest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.fit_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      fit_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.sequential_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      sequential_estimation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle" class="md-nav__link">
    <span class="md-ellipsis">
      toggle
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle.estimate_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting" class="md-nav__link">
    <span class="md-ellipsis">
      plotting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.pca_components" class="md-nav__link">
    <span class="md-ellipsis">
      pca_components
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_A_S" class="md-nav__link">
    <span class="md-ellipsis">
      plot_A_S
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_all" class="md-nav__link">
    <span class="md-ellipsis">
      plot_all
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_clusters" class="md-nav__link">
    <span class="md-ellipsis">
      plot_clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_error_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_error_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_1" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_1
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_5" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_5
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_model_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_model_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_parallel" class="md-nav__link">
    <span class="md-ellipsis">
      plot_parallel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_bar" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_bar
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_series" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_series
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      plot_profiles
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      plot_tsne
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.format_site" class="md-nav__link">
    <span class="md-ellipsis">
      format_site
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.move_processed_files" class="md-nav__link">
    <span class="md-ellipsis">
      move_processed_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_collecttri" class="md-nav__link">
    <span class="md-ellipsis">
      process_collecttri
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss_std" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss_std
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_routlimma" class="md-nav__link">
    <span class="md-ellipsis">
      process_routlimma
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.update_gene_symbols" class="md-nav__link">
    <span class="md-ellipsis">
      update_gene_symbols
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.postoptmerge" class="md-nav__link">
    <span class="md-ellipsis">
      postoptmerge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.postoptmerge.map_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      map_optimization_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis" class="md-nav__link">
    <span class="md-ellipsis">
      analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_ds" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_rand" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      sensitivity_analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist" class="md-nav__link">
    <span class="md-ellipsis">
      initdist
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand" class="md-nav__link">
    <span class="md-ellipsis">
      initrand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc" class="md-nav__link">
    <span class="md-ellipsis">
      initsucc
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest" class="md-nav__link">
    <span class="md-ellipsis">
      inittest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="md-nav__link">
    <span class="md-ellipsis">
      TFOptimizationMultiObjectiveProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="TFOptimizationMultiObjectiveProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.determine_T_use" class="md-nav__link">
    <span class="md-ellipsis">
      determine_T_use
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_TF" class="md-nav__link">
    <span class="md-ellipsis">
      filter_TF
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      filter_mrna
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.load_raw_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_raw_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.update_regulations" class="md-nav__link">
    <span class="md-ellipsis">
      update_regulations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_TF_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_TF_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_mRNA_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_mRNA_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.compute_beta_indices" class="md-nav__link">
    <span class="md-ellipsis">
      compute_beta_indices
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_bounds" class="md-nav__link">
    <span class="md-ellipsis">
      create_bounds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_initial_guess" class="md-nav__link">
    <span class="md-ellipsis">
      create_initial_guess
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_no_psite_array" class="md-nav__link">
    <span class="md-ellipsis">
      create_no_psite_array
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.extract_best_solution" class="md-nav__link">
    <span class="md-ellipsis">
      extract_best_solution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.get_parallel_runner" class="md-nav__link">
    <span class="md-ellipsis">
      get_parallel_runner
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_alpha_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_alpha_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_beta_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_beta_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper" class="md-nav__link">
    <span class="md-ellipsis">
      helper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_alpha_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_beta_barplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_goodness_of_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap_abs_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_time_wise_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      objective_wrapper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun.run_optimizer" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimizer
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_linear_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      build_linear_constraints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_alpha_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_alpha_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_beta_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_beta_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.load_and_filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_filter_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.prepare_data" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_expression_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_expression_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_tf_protein_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_tf_protein_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.min_max_normalize" class="md-nav__link">
    <span class="md-ellipsis">
      min_max_normalize
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.summarize_stats" class="md-nav__link">
    <span class="md-ellipsis">
      summarize_stats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.get_optimization_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      get_optimization_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.postprocess_results" class="md-nav__link">
    <span class="md-ellipsis">
      postprocess_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display" class="md-nav__link">
    <span class="md-ellipsis">
      display
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.save_result" class="md-nav__link">
    <span class="md-ellipsis">
      save_result
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables" class="md-nav__link">
    <span class="md-ellipsis">
      tables
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.generate_tables" class="md-nav__link">
    <span class="md-ellipsis">
      generate_tables
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.save_master_table" class="md-nav__link">
    <span class="md-ellipsis">
      save_master_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.save_tables" class="md-nav__link">
    <span class="md-ellipsis">
      save_tables
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bin.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bin.main.main" class="md-nav__link">
    <span class="md-ellipsis">
      main
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config" class="md-nav__link">
    <span class="md-ellipsis">
      config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.extract_config" class="md-nav__link">
    <span class="md-ellipsis">
      extract_config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.log_config" class="md-nav__link">
    <span class="md-ellipsis">
      log_config
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_bound_pair" class="md-nav__link">
    <span class="md-ellipsis">
      parse_bound_pair
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.parse_fix_value" class="md-nav__link">
    <span class="md-ellipsis">
      parse_fix_value
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.config.score_fit" class="md-nav__link">
    <span class="md-ellipsis">
      score_fit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.generate_labels_ds" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.generate_labels_rand" class="md-nav__link">
    <span class="md-ellipsis">
      generate_labels_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_bounds_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_bounds_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_number_of_params_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_number_of_params_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_param_names_ds" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.get_param_names_rand" class="md-nav__link">
    <span class="md-ellipsis">
      get_param_names_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.constants.location" class="md-nav__link">
    <span class="md-ellipsis">
      location
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter" class="md-nav__link">
    <span class="md-ellipsis">
      ColoredFormatter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="ColoredFormatter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.format" class="md-nav__link">
    <span class="md-ellipsis">
      format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#config.logconf.ColoredFormatter.remove_ansi" class="md-nav__link">
    <span class="md-ellipsis">
      remove_ansi
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#config.logconf.setup_logger" class="md-nav__link">
    <span class="md-ellipsis">
      setup_logger
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_de" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_de
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.opt_analyze_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      opt_analyze_nsga
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_residuals_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo" class="md-nav__link">
    <span class="md-ellipsis">
      minfndiffevo
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii" class="md-nav__link">
    <span class="md-ellipsis">
      minfnnsgaii
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="md-nav__link">
    <span class="md-ellipsis">
      PhosphorylationOptimizationProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="PhosphorylationOptimizationProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="md-nav__link">
    <span class="md-ellipsis">
      objective_function
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_de" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_de
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.post_optimization_nsga" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_nsga
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.optcon.construct.pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      pipeline
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.evol.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit" class="md-nav__link">
    <span class="md-ellipsis">
      postfit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.additional_plots" class="md-nav__link">
    <span class="md-ellipsis">
      additional_plots
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="md-nav__link">
    <span class="md-ellipsis">
      create_sankey_from_network
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="md-nav__link">
    <span class="md-ellipsis">
      goodnessoffit
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.important_connections" class="md-nav__link">
    <span class="md-ellipsis">
      important_connections
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_pca" class="md-nav__link">
    <span class="md-ellipsis">
      perform_pca
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.perform_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      perform_tsne
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="md-nav__link">
    <span class="md-ellipsis">
      reshape_alpha_beta
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.plotout.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.output_results" class="md-nav__link">
    <span class="md-ellipsis">
      output_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_autocorrelation_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cumulative_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      plot_fits_for_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_histogram_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_qqplot_residuals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.apply_scaling" class="md-nav__link">
    <span class="md-ellipsis">
      apply_scaling
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.load_and_scale_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_scale_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.compute_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_metrics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.local.utils.params.extract_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      extract_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT" class="md-nav__link">
    <span class="md-ellipsis">
      KKT
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.generate_latex_table" class="md-nav__link">
    <span class="md-ellipsis">
      generate_latex_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_constraint_violations" class="md-nav__link">
    <span class="md-ellipsis">
      plot_constraint_violations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.plot_sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      plot_sensitivity_analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.post_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      post_optimization_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_primal_feasibility_results" class="md-nav__link">
    <span class="md-ellipsis">
      print_primal_feasibility_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      print_sensitivity_and_active_constraints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.optimality.KKT.process_excel_results" class="md-nav__link">
    <span class="md-ellipsis">
      process_excel_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell" class="md-nav__link">
    <span class="md-ellipsis">
      runpowell
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kinopt.powell.runpowell.run_powell" class="md-nav__link">
    <span class="md-ellipsis">
      run_powell
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers" class="md-nav__link">
    <span class="md-ellipsis">
      helpers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_distributive_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_distributive_diagram
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_random_diagram" class="md-nav__link">
    <span class="md-ellipsis">
      create_random_diagram
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.create_successive_model" class="md-nav__link">
    <span class="md-ellipsis">
      create_successive_model
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.powerset" class="md-nav__link">
    <span class="md-ellipsis">
      powerset
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.diagram.helpers.state_label" class="md-nav__link">
    <span class="md-ellipsis">
      state_label
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod" class="md-nav__link">
    <span class="md-ellipsis">
      distmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.distmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod" class="md-nav__link">
    <span class="md-ellipsis">
      randmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.prepare_vectorized_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_vectorized_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.randmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod" class="md-nav__link">
    <span class="md-ellipsis">
      succmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.ode_core" class="md-nav__link">
    <span class="md-ellipsis">
      ode_core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.ode_system" class="md-nav__link">
    <span class="md-ellipsis">
      ode_system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.succmod.solve_ode" class="md-nav__link">
    <span class="md-ellipsis">
      solve_ode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.testmod" class="md-nav__link">
    <span class="md-ellipsis">
      testmod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights" class="md-nav__link">
    <span class="md-ellipsis">
      weights
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights.early_emphasis" class="md-nav__link">
    <span class="md-ellipsis">
      early_emphasis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models.weights.get_weight_options" class="md-nav__link">
    <span class="md-ellipsis">
      get_weight_options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest" class="md-nav__link">
    <span class="md-ellipsis">
      adapest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.adaptive_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      adaptive_estimation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.estimate_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_profiles
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.adapest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core" class="md-nav__link">
    <span class="md-ellipsis">
      core
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core.process_gene" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.core.process_gene_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      process_gene_wrapper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci" class="md-nav__link">
    <span class="md-ellipsis">
      ci
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.identifiability.ci.confidence_intervals" class="md-nav__link">
    <span class="md-ellipsis">
      confidence_intervals
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.normest.normest" class="md-nav__link">
    <span class="md-ellipsis">
      normest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest" class="md-nav__link">
    <span class="md-ellipsis">
      seqest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.fit_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      fit_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.prepare_model_func" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_model_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.seqest.sequential_estimation" class="md-nav__link">
    <span class="md-ellipsis">
      sequential_estimation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle" class="md-nav__link">
    <span class="md-ellipsis">
      toggle
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paramest.toggle.estimate_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      estimate_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting" class="md-nav__link">
    <span class="md-ellipsis">
      plotting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.pca_components" class="md-nav__link">
    <span class="md-ellipsis">
      pca_components
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_A_S" class="md-nav__link">
    <span class="md-ellipsis">
      plot_A_S
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_all" class="md-nav__link">
    <span class="md-ellipsis">
      plot_all
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_clusters" class="md-nav__link">
    <span class="md-ellipsis">
      plot_clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_error_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_error_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_1" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_1
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_gof_5" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gof_5
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_model_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_model_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_parallel" class="md-nav__link">
    <span class="md-ellipsis">
      plot_parallel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_bar" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_bar
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_param_series" class="md-nav__link">
    <span class="md-ellipsis">
      plot_param_series
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_profiles" class="md-nav__link">
    <span class="md-ellipsis">
      plot_profiles
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting.plotting.Plotter.plot_tsne" class="md-nav__link">
    <span class="md-ellipsis">
      plot_tsne
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.format_site" class="md-nav__link">
    <span class="md-ellipsis">
      format_site
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.move_processed_files" class="md-nav__link">
    <span class="md-ellipsis">
      move_processed_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_collecttri" class="md-nav__link">
    <span class="md-ellipsis">
      process_collecttri
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_msgauss_std" class="md-nav__link">
    <span class="md-ellipsis">
      process_msgauss_std
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.process_routlimma" class="md-nav__link">
    <span class="md-ellipsis">
      process_routlimma
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.cleanup.update_gene_symbols" class="md-nav__link">
    <span class="md-ellipsis">
      update_gene_symbols
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.postoptmerge" class="md-nav__link">
    <span class="md-ellipsis">
      postoptmerge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#processing.postoptmerge.map_optimization_results" class="md-nav__link">
    <span class="md-ellipsis">
      map_optimization_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis" class="md-nav__link">
    <span class="md-ellipsis">
      analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_ds" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_ds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.define_sensitivity_problem_rand" class="md-nav__link">
    <span class="md-ellipsis">
      define_sensitivity_problem_rand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sensitivity.analysis.sensitivity_analysis" class="md-nav__link">
    <span class="md-ellipsis">
      sensitivity_analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist" class="md-nav__link">
    <span class="md-ellipsis">
      initdist
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initdist.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand" class="md-nav__link">
    <span class="md-ellipsis">
      initrand
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initrand.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc" class="md-nav__link">
    <span class="md-ellipsis">
      initsucc
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.initsucc.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest" class="md-nav__link">
    <span class="md-ellipsis">
      inittest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steady.inittest.initial_condition" class="md-nav__link">
    <span class="md-ellipsis">
      initial_condition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="md-nav__link">
    <span class="md-ellipsis">
      TFOptimizationMultiObjectiveProblem
    </span>
  </a>
  
    <nav class="md-nav" aria-label="TFOptimizationMultiObjectiveProblem">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.opt.optrun.run_optimization" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.determine_T_use" class="md-nav__link">
    <span class="md-ellipsis">
      determine_T_use
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_TF" class="md-nav__link">
    <span class="md-ellipsis">
      filter_TF
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.filter_mrna" class="md-nav__link">
    <span class="md-ellipsis">
      filter_mrna
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.load_raw_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_raw_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.optcon.filter.update_regulations" class="md-nav__link">
    <span class="md-ellipsis">
      update_regulations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_TF_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_TF_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_mRNA_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_mRNA_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.iodata.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.compute_beta_indices" class="md-nav__link">
    <span class="md-ellipsis">
      compute_beta_indices
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_bounds" class="md-nav__link">
    <span class="md-ellipsis">
      create_bounds
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_initial_guess" class="md-nav__link">
    <span class="md-ellipsis">
      create_initial_guess
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.create_no_psite_array" class="md-nav__link">
    <span class="md-ellipsis">
      create_no_psite_array
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.extract_best_solution" class="md-nav__link">
    <span class="md-ellipsis">
      extract_best_solution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.get_parallel_runner" class="md-nav__link">
    <span class="md-ellipsis">
      get_parallel_runner
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_alpha_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_alpha_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.evol.utils.params.print_beta_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      print_beta_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper" class="md-nav__link">
    <span class="md-ellipsis">
      helper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter" class="md-nav__link">
    <span class="md-ellipsis">
      Plotter
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plotter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="md-nav__link">
    <span class="md-ellipsis">
      plot_alpha_distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_beta_barplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_boxplot_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_alpha
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cdf_beta
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="md-nav__link">
    <span class="md-ellipsis">
      plot_goodness_of_fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_heatmap_abs_residuals
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_kld" class="md-nav__link">
    <span class="md-ellipsis">
      plot_kld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_pca" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pca
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="md-nav__link">
    <span class="md-ellipsis">
      plot_time_wise_residuals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants" class="md-nav__link">
    <span class="md-ellipsis">
      constants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.constants.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      parse_args
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.config.logconf" class="md-nav__link">
    <span class="md-ellipsis">
      logconf
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout" class="md-nav__link">
    <span class="md-ellipsis">
      plotout
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="md-nav__link">
    <span class="md-ellipsis">
      plot_estimated_vs_observed
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils" class="md-nav__link">
    <span class="md-ellipsis">
      sheetutils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.exporter.sheetutils.save_results_to_excel" class="md-nav__link">
    <span class="md-ellipsis">
      save_results_to_excel
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn" class="md-nav__link">
    <span class="md-ellipsis">
      minfn
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.compute_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      compute_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_" class="md-nav__link">
    <span class="md-ellipsis">
      objective_
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.objfn.minfn.objective_wrapper" class="md-nav__link">
    <span class="md-ellipsis">
      objective_wrapper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun" class="md-nav__link">
    <span class="md-ellipsis">
      optrun
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.opt.optrun.run_optimizer" class="md-nav__link">
    <span class="md-ellipsis">
      run_optimizer
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct" class="md-nav__link">
    <span class="md-ellipsis">
      construct
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_fixed_arrays" class="md-nav__link">
    <span class="md-ellipsis">
      build_fixed_arrays
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.build_linear_constraints" class="md-nav__link">
    <span class="md-ellipsis">
      build_linear_constraints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_alpha_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_alpha_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.construct.constraint_beta_func" class="md-nav__link">
    <span class="md-ellipsis">
      constraint_beta_func
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter" class="md-nav__link">
    <span class="md-ellipsis">
      filter
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.load_and_filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_and_filter_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.optcon.filter.prepare_data" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata" class="md-nav__link">
    <span class="md-ellipsis">
      iodata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_expression_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_expression_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_regulation" class="md-nav__link">
    <span class="md-ellipsis">
      load_regulation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.load_tf_protein_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_tf_protein_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.min_max_normalize" class="md-nav__link">
    <span class="md-ellipsis">
      min_max_normalize
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.iodata.summarize_stats" class="md-nav__link">
    <span class="md-ellipsis">
      summarize_stats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params" class="md-nav__link">
    <span class="md-ellipsis">
      params
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.get_optimization_parameters" class="md-nav__link">
    <span class="md-ellipsis">
      get_optimization_parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tfopt.local.utils.params.postprocess_results" class="md-nav__link">
    <span class="md-ellipsis">
      postprocess_results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display" class="md-nav__link">
    <span class="md-ellipsis">
      display
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      create_report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.ensure_output_directory" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_output_directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.format_duration" class="md-nav__link">
    <span class="md-ellipsis">
      format_duration
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.load_data" class="md-nav__link">
    <span class="md-ellipsis">
      load_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.organize_output_files" class="md-nav__link">
    <span class="md-ellipsis">
      organize_output_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.display.save_result" class="md-nav__link">
    <span class="md-ellipsis">
      save_result
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables" class="md-nav__link">
    <span class="md-ellipsis">
      tables
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.generate_tables" class="md-nav__link">
    <span class="md-ellipsis">
      generate_tables
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.save_master_table" class="md-nav__link">
    <span class="md-ellipsis">
      save_master_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utils.tables.save_tables" class="md-nav__link">
    <span class="md-ellipsis">
      save_tables
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="api-reference">API Reference</h1>


<div class="doc doc-object doc-module">



<a id="bin.main"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="bin.main.main" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Main function to run the phosphorylation modelling process.
It reads the configuration, loads the data, and processes each gene in parallel.
It also handles logging and output organization.</p>


            <details class="quote">
              <summary>Source code in <code>bin/main.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to run the phosphorylation modelling process.</span>
<span class="sd">    It reads the configuration, loads the data, and processes each gene in parallel.</span>
<span class="sd">    It also handles logging and output organization.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Set up the logger</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> Phosphorylation Modelling Configuration&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Estimation Mode: </span><span class="si">{</span><span class="n">ESTIMATION_MODE</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">log_config</span><span class="p">(</span><span class="n">logger</span><span class="p">,</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;bounds&#39;</span><span class="p">],</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;fixed_params&#39;</span><span class="p">],</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;time_fixed&#39;</span><span class="p">],</span> <span class="n">args</span><span class="p">)</span>

    <span class="c1"># Make output directory</span>
    <span class="n">ensure_output_directory</span><span class="p">(</span><span class="n">OUT_DIR</span><span class="p">)</span>

    <span class="c1"># Load the data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;input_excel&#39;</span><span class="p">],</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">)</span>

    <span class="c1"># Load protein groups</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="c1"># For testing, only process the first gene</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded Time Series for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> Protein(s)&quot;</span><span class="p">)</span>

    <span class="c1"># Initiate the process pool and run the processing function for each gene</span>
    <span class="k">with</span> <span class="n">ProcessPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;max_workers&#39;</span><span class="p">])</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">executor</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="n">process_gene_wrapper</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span>
            <span class="p">[</span><span class="n">data</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">TIME_POINTS</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;bounds&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;fixed_params&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">desired_times</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;time_fixed&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span>
            <span class="p">[</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;bootstraps&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
        <span class="p">))</span>
    <span class="c1"># Save the results</span>
    <span class="n">save_result</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">excel_filename</span><span class="o">=</span><span class="n">OUT_RESULTS_DIR</span><span class="p">)</span>

    <span class="c1"># Organize output files and create a report</span>
    <span class="n">organize_output_files</span><span class="p">(</span><span class="n">OUT_DIR</span><span class="p">)</span>
    <span class="n">create_report</span><span class="p">(</span><span class="n">OUT_DIR</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Report &amp; Results </span><span class="si">{</span><span class="n">location</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">OUT_DIR</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="config.config"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="config.config.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>:param directory:
:type directory: str</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    :param directory:</span>
<span class="sd">    :type directory: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.extract_config" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_config</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Extract configuration settings from command-line arguments.
This function creates a dictionary containing the parameter bounds,
fixed parameters, bootstrapping iterations, time-specific fixed parameters,
and profile estimation settings. It also sets the maximum number of workers
for parallel processing.
The function returns the configuration dictionary.
:param args:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">extract_config</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract configuration settings from command-line arguments.</span>
<span class="sd">    This function creates a dictionary containing the parameter bounds,</span>
<span class="sd">    fixed parameters, bootstrapping iterations, time-specific fixed parameters,</span>
<span class="sd">    and profile estimation settings. It also sets the maximum number of workers</span>
<span class="sd">    for parallel processing.</span>
<span class="sd">    The function returns the configuration dictionary.</span>
<span class="sd">    :param args:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">A_bound</span><span class="p">,</span>
        <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">B_bound</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">C_bound</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">D_bound</span><span class="p">,</span>
        <span class="s2">&quot;Ssite&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">Ssite_bound</span><span class="p">,</span>
        <span class="s2">&quot;Dsite&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">Dsite_bound</span>
    <span class="p">}</span>
    <span class="n">fixed_params</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_A</span><span class="p">,</span>
        <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_B</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_C</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_D</span><span class="p">,</span>
        <span class="s2">&quot;Ssite&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_Ssite</span><span class="p">,</span>
        <span class="s2">&quot;Dsite&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_Dsite</span>
    <span class="p">}</span>
    <span class="n">time_fixed</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">fix_t</span><span class="p">)</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">fix_t</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">else</span> <span class="p">{}</span>

    <span class="n">config</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;bounds&#39;</span><span class="p">:</span> <span class="n">bounds</span><span class="p">,</span>
        <span class="s1">&#39;fixed_params&#39;</span><span class="p">:</span> <span class="n">fixed_params</span><span class="p">,</span>
        <span class="s1">&#39;time_fixed&#39;</span><span class="p">:</span> <span class="n">time_fixed</span><span class="p">,</span>
        <span class="s1">&#39;bootstraps&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">bootstraps</span><span class="p">,</span>
        <span class="s1">&#39;profile_start&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">profile_start</span><span class="p">,</span>
        <span class="s1">&#39;profile_end&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">profile_end</span><span class="p">,</span>
        <span class="s1">&#39;profile_step&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">profile_step</span><span class="p">,</span>
        <span class="s1">&#39;input_excel&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">input_excel</span><span class="p">,</span>
        <span class="c1"># Adjust as needed for parallel processing</span>
        <span class="s1">&#39;max_workers&#39;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(),</span>  <span class="c1"># Use all CPU cores</span>
        <span class="c1"># &#39;max_workers&#39;: 1,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">config</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.log_config" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">log_config</span><span class="p">(</span><span class="n">logger</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Log the configuration settings for the PhosKinTime script.
This function logs the parameter bounds, fixed parameters,
bootstrapping iterations, time-specific fixed parameters,
and profile estimation settings.
It uses the provided logger to output the information.
:param logger:
:param bounds:
:param fixed_params:
:param time_fixed:
:param args:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">log_config</span><span class="p">(</span><span class="n">logger</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Log the configuration settings for the PhosKinTime script.</span>
<span class="sd">    This function logs the parameter bounds, fixed parameters,</span>
<span class="sd">    bootstrapping iterations, time-specific fixed parameters,</span>
<span class="sd">    and profile estimation settings.</span>
<span class="sd">    It uses the provided logger to output the information.</span>
<span class="sd">    :param logger:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param time_fixed:</span>
<span class="sd">    :param args:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parameter Bounds:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">bounds</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Fixed Parameters:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">fixed_params</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bootstrapping Iterations: </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">bootstraps</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Time-specific Fixed Parameters:&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">time_fixed</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">time_fixed</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   Time </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> min: </span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;   None&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Profile Estimation:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   Start: </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">profile_start</span><span class="si">}</span><span class="s2"> min&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   End:   </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">profile_end</span><span class="si">}</span><span class="s2"> min&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   Step:  </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">profile_step</span><span class="si">}</span><span class="s2"> min&quot;</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">set_printoptions</span><span class="p">(</span><span class="n">suppress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse command-line arguments for the PhosKinTime script.
This function uses argparse to define and handle the command-line options.
It includes options for setting bounds, fixed parameters, bootstrapping,
profile estimation, and input file paths.
The function returns the parsed arguments as a Namespace object.
The arguments include:
    --A-bound: Bounds for parameter A (default: "0,3")
    --B-bound: Bounds for parameter B (default: "0,3")
    --C-bound: Bounds for parameter C (default: "0,3")
    --D-bound: Bounds for parameter D (default: "0,3")
    --Ssite-bound: Bounds for Ssite (default: "0,3")
    --Dsite-bound: Bounds for Dsite (default: "0,3")
    --fix-A: Fixed value for parameter A
    --fix-B: Fixed value for parameter B
    --fix-C: Fixed value for parameter C
    --fix-D: Fixed value for parameter D
    --fix-Ssite: Fixed value for Ssite
    --fix-Dsite: Fixed value for Dsite
    --fix-t: JSON string mapping time points to fixed param values
    --bootstraps: Number of bootstrapping iterations (default: 0)
    --profile-start: Start time for profile estimation (default: None)
    --profile-end: End time for profile estimation (default: 1)
    --profile-step: Step size for profile estimation (default: 0.5)
    --input-excel: Path to the input Excel file (default: INPUT_EXCEL)
Returns:
    argparse.Namespace: The parsed command-line arguments.</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse command-line arguments for the PhosKinTime script.</span>
<span class="sd">    This function uses argparse to define and handle the command-line options.</span>
<span class="sd">    It includes options for setting bounds, fixed parameters, bootstrapping,</span>
<span class="sd">    profile estimation, and input file paths.</span>
<span class="sd">    The function returns the parsed arguments as a Namespace object.</span>
<span class="sd">    The arguments include:</span>
<span class="sd">        --A-bound: Bounds for parameter A (default: &quot;0,3&quot;)</span>
<span class="sd">        --B-bound: Bounds for parameter B (default: &quot;0,3&quot;)</span>
<span class="sd">        --C-bound: Bounds for parameter C (default: &quot;0,3&quot;)</span>
<span class="sd">        --D-bound: Bounds for parameter D (default: &quot;0,3&quot;)</span>
<span class="sd">        --Ssite-bound: Bounds for Ssite (default: &quot;0,3&quot;)</span>
<span class="sd">        --Dsite-bound: Bounds for Dsite (default: &quot;0,3&quot;)</span>
<span class="sd">        --fix-A: Fixed value for parameter A</span>
<span class="sd">        --fix-B: Fixed value for parameter B</span>
<span class="sd">        --fix-C: Fixed value for parameter C</span>
<span class="sd">        --fix-D: Fixed value for parameter D</span>
<span class="sd">        --fix-Ssite: Fixed value for Ssite</span>
<span class="sd">        --fix-Dsite: Fixed value for Dsite</span>
<span class="sd">        --fix-t: JSON string mapping time points to fixed param values</span>
<span class="sd">        --bootstraps: Number of bootstrapping iterations (default: 0)</span>
<span class="sd">        --profile-start: Start time for profile estimation (default: None)</span>
<span class="sd">        --profile-end: End time for profile estimation (default: 1)</span>
<span class="sd">        --profile-step: Step size for profile estimation (default: 0.5)</span>
<span class="sd">        --input-excel: Path to the input Excel file (default: INPUT_EXCEL)</span>
<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: The parsed command-line arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;PhosKinTime - ODE Parameter Estimation of Phosphorylation Events in Temporal Space&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--A-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--B-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--C-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--D-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--Ssite-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--Dsite-bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_bound_pair</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,20&quot;</span><span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-A&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-B&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-C&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-D&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-Ssite&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_fix_value</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-Dsite&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">parse_fix_value</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--fix-t&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;{ &#39;</span>
                                                     <span class="s1">&#39;</span><span class="se">\&quot;</span><span class="s1">0</span><span class="se">\&quot;</span><span class="s1">: {</span><span class="se">\&quot;</span><span class="s1">A</span><span class="se">\&quot;</span><span class="s1">: 0.85, </span><span class="se">\&quot;</span><span class="s1">S</span><span class="se">\&quot;</span><span class="s1">: 0.1},  &#39;</span>
                                                     <span class="s1">&#39;</span><span class="se">\&quot;</span><span class="s1">60</span><span class="se">\&quot;</span><span class="s1">: {</span><span class="se">\&quot;</span><span class="s1">A</span><span class="se">\&quot;</span><span class="s1">:0.85, </span><span class="se">\&quot;</span><span class="s1">S</span><span class="se">\&quot;</span><span class="s1">: 0.2},  &#39;</span>
                                                     <span class="s1">&#39;</span><span class="se">\&quot;</span><span class="s1">inf</span><span class="se">\&quot;</span><span class="s1">: {</span><span class="se">\&quot;</span><span class="s1">A</span><span class="se">\&quot;</span><span class="s1">:0.85, </span><span class="se">\&quot;</span><span class="s1">S</span><span class="se">\&quot;</span><span class="s1">: 0.4} &#39;</span>
                                                     <span class="s1">&#39;}&#39;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;JSON string mapping time points to fixed param values, e.g. &#39;{</span><span class="se">\&quot;</span><span class="s2">60</span><span class="se">\&quot;</span><span class="s2">: {</span><span class="se">\&quot;</span><span class="s2">A</span><span class="se">\&quot;</span><span class="s2">: 1.3}}&#39;&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bootstraps&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--profile-start&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--profile-end&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--profile-step&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--input-excel&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                        <span class="n">default</span><span class="o">=</span><span class="n">INPUT_EXCEL</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the input Excel file&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.parse_bound_pair" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_bound_pair</span><span class="p">(</span><span class="n">val</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse a string representing a pair of bounds (lower, upper) into a tuple of floats.
The upper bound can be 'inf' or 'infinity' to represent infinity.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "0,3" or "0,infinity".
Returns:
    tuple: A tuple containing the lower and upper bounds as floats.</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_bound_pair</span><span class="p">(</span><span class="n">val</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse a string representing a pair of bounds (lower, upper) into a tuple of floats.</span>
<span class="sd">    The upper bound can be &#39;inf&#39; or &#39;infinity&#39; to represent infinity.</span>
<span class="sd">    Raises ValueError if the input is not in the correct format.</span>
<span class="sd">    Args:</span>
<span class="sd">        val (str): The string to parse, e.g., &quot;0,3&quot; or &quot;0,infinity&quot;.</span>
<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing the lower and upper bounds as floats.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="n">val</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Bounds must be provided as &#39;lower,upper&#39;&quot;</span><span class="p">)</span>
        <span class="n">lower</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">upper_str</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">upper_str</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;inf&quot;</span><span class="p">,</span> <span class="s2">&quot;infinity&quot;</span><span class="p">]:</span>
            <span class="n">upper</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">upper</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">lower</span><span class="p">,</span> <span class="n">upper</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentTypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid bound pair &#39;</span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.parse_fix_value" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_fix_value</span><span class="p">(</span><span class="n">val</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse a fixed value or a list of fixed values from a string.
If the input is a single value, it returns that value as a float.
If the input is a comma-separated list, it returns a list of floats.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "1.0" or "1.0,2.0".
Returns:
    float or list: The parsed fixed value(s) as a float or a list of floats.</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_fix_value</span><span class="p">(</span><span class="n">val</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse a fixed value or a list of fixed values from a string.</span>
<span class="sd">    If the input is a single value, it returns that value as a float.</span>
<span class="sd">    If the input is a comma-separated list, it returns a list of floats.</span>
<span class="sd">    Raises ValueError if the input is not in the correct format.</span>
<span class="sd">    Args:</span>
<span class="sd">        val (str): The string to parse, e.g., &quot;1.0&quot; or &quot;1.0,2.0&quot;.</span>
<span class="sd">    Returns:</span>
<span class="sd">        float or list: The parsed fixed value(s) as a float or a list of floats.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">val</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s1">&#39;,&#39;</span> <span class="ow">in</span> <span class="n">val</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">val</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)]</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentTypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid fixed value list &#39;</span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentTypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid fixed value &#39;</span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.config.score_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">score_fit</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">prediction</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">ALPHA_WEIGHT</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">BETA_WEIGHT</span><span class="p">,</span> <span class="n">gamma</span><span class="o">=</span><span class="n">GAMMA_WEIGHT</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="n">DELTA_WEIGHT</span><span class="p">,</span> <span class="n">reg_penalty</span><span class="o">=</span><span class="n">MU_REG</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculate the score for the fit of a model to target data.
The score is a weighted combination of various metrics including
mean squared error (MSE), root mean squared error (RMSE),
mean absolute error (MAE), variance, and regularization penalty.
The weights for each metric can be adjusted using the parameters
alpha, beta, gamma, and delta.
The regularization penalty is controlled by the reg_penalty parameter.
The function returns the calculated score.
:param target:
:param prediction:
:param params:
:param alpha:
:param beta:
:param gamma:
:param delta:
:param reg_penalty:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>config/config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">score_fit</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">prediction</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span>
              <span class="n">alpha</span><span class="o">=</span><span class="n">ALPHA_WEIGHT</span><span class="p">,</span>
              <span class="n">beta</span><span class="o">=</span><span class="n">BETA_WEIGHT</span><span class="p">,</span>
              <span class="n">gamma</span><span class="o">=</span><span class="n">GAMMA_WEIGHT</span><span class="p">,</span>
              <span class="n">delta</span><span class="o">=</span><span class="n">DELTA_WEIGHT</span><span class="p">,</span>
              <span class="n">reg_penalty</span><span class="o">=</span><span class="n">MU_REG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the score for the fit of a model to target data.</span>
<span class="sd">    The score is a weighted combination of various metrics including</span>
<span class="sd">    mean squared error (MSE), root mean squared error (RMSE),</span>
<span class="sd">    mean absolute error (MAE), variance, and regularization penalty.</span>
<span class="sd">    The weights for each metric can be adjusted using the parameters</span>
<span class="sd">    alpha, beta, gamma, and delta.</span>
<span class="sd">    The regularization penalty is controlled by the reg_penalty parameter.</span>
<span class="sd">    The function returns the calculated score.</span>
<span class="sd">    :param target:</span>
<span class="sd">    :param prediction:</span>
<span class="sd">    :param params:</span>
<span class="sd">    :param alpha:</span>
<span class="sd">    :param beta:</span>
<span class="sd">    :param gamma:</span>
<span class="sd">    :param delta:</span>
<span class="sd">    :param reg_penalty:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">residual</span> <span class="o">=</span> <span class="n">target</span> <span class="o">-</span> <span class="n">prediction</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residual</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">residual</span> <span class="o">**</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residual</span><span class="p">))</span>
    <span class="n">variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">residual</span><span class="p">)</span>
    <span class="n">l2_norm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>

    <span class="n">score</span> <span class="o">=</span> <span class="n">delta</span> <span class="o">*</span> <span class="n">mse</span> <span class="o">+</span> <span class="n">alpha</span> <span class="o">*</span> <span class="n">rmse</span> <span class="o">+</span> <span class="n">beta</span> <span class="o">*</span> <span class="n">mae</span> <span class="o">+</span> <span class="n">gamma</span> <span class="o">*</span> <span class="n">variance</span> <span class="o">+</span> <span class="n">reg_penalty</span> <span class="o">*</span> <span class="n">l2_norm</span>
    <span class="k">return</span> <span class="n">score</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="config.constants"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="config.constants.generate_labels_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_labels_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the distributive or successive models.
Returns a list with the base labels "R" and "P", followed by labels for each individual phosphorylated state.
Example for num_psites=2: ["R", "P", "P1", "P2"]</p>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_labels_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates labels for the states based on the number of phosphorylation sites for the distributive or successive models.</span>
<span class="sd">    Returns a list with the base labels &quot;R&quot; and &quot;P&quot;, followed by labels for each individual phosphorylated state.</span>
<span class="sd">    Example for num_psites=2: [&quot;R&quot;, &quot;P&quot;, &quot;P1&quot;, &quot;P2&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="s2">&quot;R&quot;</span><span class="p">,</span> <span class="s2">&quot;P&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.generate_labels_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_labels_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the random model.
Returns a list with the base labels "R" and "P", followed by labels for all combinations of phosphorylated sites.
Example for num_psites=2: ["R", "P", "P1", "P2", "P12"]</p>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_labels_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates labels for the states based on the number of phosphorylation sites for the random model.</span>
<span class="sd">    Returns a list with the base labels &quot;R&quot; and &quot;P&quot;, followed by labels for all combinations of phosphorylated sites.</span>
<span class="sd">    Example for num_psites=2: [&quot;R&quot;, &quot;P&quot;, &quot;P1&quot;, &quot;P2&quot;, &quot;P12&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;R&quot;</span><span class="p">,</span> <span class="s2">&quot;P&quot;</span><span class="p">]</span>
    <span class="n">subsets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">comb</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">k</span><span class="p">):</span>
            <span class="n">subsets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;P&quot;</span> <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">comb</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">labels</span> <span class="o">+</span> <span class="n">subsets</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.get_bounds_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_bounds_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">ub</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generate bounds for the ODE parameters based on the number of phosphorylation sites.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lower</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Lower bound for parameters.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>upper</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Upper bound for parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of bounds as [lower, upper] for each parameter.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_bounds_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">ub</span><span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate bounds for the ODE parameters based on the number of phosphorylation sites.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites.</span>
<span class="sd">        lower (float): Lower bound for parameters.</span>
<span class="sd">        upper (float): Upper bound for parameters.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of bounds as [lower, upper] for each parameter.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="p">[[</span><span class="n">lower</span><span class="p">,</span> <span class="n">ub</span><span class="p">]]</span> <span class="o">*</span> <span class="mi">4</span>
    <span class="n">bounds</span> <span class="o">+=</span> <span class="p">[[</span><span class="n">lower</span><span class="p">,</span> <span class="n">ub</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span><span class="p">):</span>
            <span class="n">bounds</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lower</span><span class="p">,</span> <span class="n">ub</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">bounds</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.get_number_of_params_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_number_of_params_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculate the number of parameters required for the ODE system based on the number of phosphorylation sites.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites (1 to 4).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Total number of parameters.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_number_of_params_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the number of parameters required for the ODE system based on the number of phosphorylation sites.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites (1 to 4).</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Total number of parameters.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">base_params</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">phosphorylation_params</span> <span class="o">=</span> <span class="n">num_psites</span>
    <span class="n">dephosphorylation_params</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">comb</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">total_params</span> <span class="o">=</span> <span class="n">base_params</span> <span class="o">+</span> <span class="n">phosphorylation_params</span> <span class="o">+</span> <span class="n">dephosphorylation_params</span>
    <span class="k">return</span> <span class="n">total_params</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.get_param_names_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_param_names_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generate parameter names for distributive or successive models.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        ['D1', 'D2', ..., 'D<num_psites>'].</p>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_param_names_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate parameter names for distributive or successive models.</span>
<span class="sd">    Format: [&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;] +</span>
<span class="sd">            [&#39;S1&#39;, &#39;S2&#39;, ..., &#39;S&lt;num_psites&gt;&#39;] +</span>
<span class="sd">            [&#39;D1&#39;, &#39;D2&#39;, ..., &#39;D&lt;num_psites&gt;&#39;].</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.get_param_names_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_param_names_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generate parameter names for the random model.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        [parameter names for all combinations of dephosphorylation sites].</p>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_param_names_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate parameter names for the random model.</span>
<span class="sd">    Format: [&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;] +</span>
<span class="sd">            [&#39;S1&#39;, &#39;S2&#39;, ..., &#39;S&lt;num_psites&gt;&#39;] +</span>
<span class="sd">            [parameter names for all combinations of dephosphorylation sites].</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">param_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">]</span>
    <span class="n">param_names</span> <span class="o">+=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;S</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span><span class="p">):</span>
            <span class="n">param_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;D</span><span class="si">{</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="w"> </span><span class="n">combo</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">param_names</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="config.constants.location" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">location</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Returns a clickable hyperlink string for supported terminals using ANSI escape sequences.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The file path or URL.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>label</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The display text for the link. Defaults to the path if not provided.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string that, when printed, shows a clickable link in terminals that support ANSI hyperlinks.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>config/helpers/__init__.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">location</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a clickable hyperlink string for supported terminals using ANSI escape sequences.</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str): The file path or URL.</span>
<span class="sd">        label (str, optional): The display text for the link. Defaults to the path if not provided.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A string that, when printed, shows a clickable link in terminals that support ANSI hyperlinks.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">label</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">label</span> <span class="o">=</span> <span class="n">path</span>
    <span class="c1"># Ensure the path is a URL (for file paths, prepend file://)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">path</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;http://&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">path</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;https://&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">path</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;file://&quot;</span><span class="p">)):</span>
        <span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;file://</span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="c1"># ANSI escape sequence format: ESC ] 8 ; ; &lt;URL&gt; ESC \ &lt;label&gt; ESC ] 8 ; ; ESC \</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\033</span><span class="s2">]8;;</span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="se">\033\\</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="se">\033</span><span class="s2">]8;;</span><span class="se">\033\\</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="config.logconf"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="config.logconf.ColoredFormatter" class="doc doc-heading">
            <code>ColoredFormatter</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="logging.Formatter">Formatter</span></code></p>


        <p>Custom formatter to add colors to log messages and elapsed time.
This formatter uses ANSI escape codes to colorize the log messages based on their severity level.
It also includes a right-aligned clock that shows the elapsed time since the logger was initialized.
The elapsed time is displayed in a human-readable format (e.g., "1h 23m 45s").
The formatter is designed to be used with a logger that has a console handler.
The elapsed time is calculated from the time the logger was initialized and is displayed in a right-aligned format.
The formatter also ensures that the log messages are padded to a specified width, which can be adjusted using the <code>width</code> parameter.
The <code>remove_ansi</code> method is used to strip ANSI escape codes from the log message for accurate padding calculation.
The <code>format</code> method is overridden to customize the log message format, including the timestamp, logger name, log level, and message.
The <code>setup_logger</code> function is used to configure the logger with a file handler and a stream handler.
The file handler writes log messages to a specified log file, while the stream handler outputs log messages to the console.
The logger is set to the specified logging level, and the log file is created in the specified directory.
The log file is rotated based on size, and old log files are backed up.</p>







              <details class="quote">
                <summary>Source code in <code>config/logconf.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">ColoredFormatter</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">Formatter</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Custom formatter to add colors to log messages and elapsed time.</span>
<span class="sd">    This formatter uses ANSI escape codes to colorize the log messages based on their severity level.</span>
<span class="sd">    It also includes a right-aligned clock that shows the elapsed time since the logger was initialized.</span>
<span class="sd">    The elapsed time is displayed in a human-readable format (e.g., &quot;1h 23m 45s&quot;).</span>
<span class="sd">    The formatter is designed to be used with a logger that has a console handler.</span>
<span class="sd">    The elapsed time is calculated from the time the logger was initialized and is displayed in a right-aligned format.</span>
<span class="sd">    The formatter also ensures that the log messages are padded to a specified width, which can be adjusted using the `width` parameter.</span>
<span class="sd">    The `remove_ansi` method is used to strip ANSI escape codes from the log message for accurate padding calculation.</span>
<span class="sd">    The `format` method is overridden to customize the log message format, including the timestamp, logger name, log level, and message.</span>
<span class="sd">    The `setup_logger` function is used to configure the logger with a file handler and a stream handler.</span>
<span class="sd">    The file handler writes log messages to a specified log file, while the stream handler outputs log messages to the console.</span>
<span class="sd">    The logger is set to the specified logging level, and the log file is created in the specified directory.</span>
<span class="sd">    The log file is rotated based on size, and old log files are backed up.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">datefmt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">120</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">fmt</span><span class="p">,</span> <span class="n">datefmt</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">start_time</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">width</span> <span class="o">=</span> <span class="n">width</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">record</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Format the log record with colors and elapsed time.</span>
<span class="sd">        This method overrides the default format method to customize the log message format.</span>
<span class="sd">        It includes the timestamp, logger name, log level, and message.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">elapsed</span> <span class="o">=</span> <span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start_time</span><span class="p">)</span><span class="o">.</span><span class="n">total_seconds</span><span class="p">()</span>
        <span class="n">elapsed_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ELAPSED&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">format_duration</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Compose colored parts</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">LOG_COLORS</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">levelname</span><span class="p">,</span> <span class="n">LOG_COLORS</span><span class="p">[</span><span class="s2">&quot;INFO&quot;</span><span class="p">])</span>
        <span class="n">time_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;DEBUG&#39;</span><span class="p">]</span><span class="si">}{</span><span class="bp">self</span><span class="o">.</span><span class="n">formatTime</span><span class="p">(</span><span class="n">record</span><span class="p">)</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">name_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;WARNING&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">level_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">color</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">levelname</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">msg_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">color</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">getMessage</span><span class="p">()</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">raw_msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">time_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">name_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">level_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">msg_str</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">no_ansi_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">remove_ansi</span><span class="p">(</span><span class="n">raw_msg</span><span class="p">))</span>
        <span class="n">padding</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">width</span> <span class="o">-</span> <span class="n">no_ansi_len</span><span class="p">)</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">raw_msg</span><span class="si">}{</span><span class="s1">&#39; &#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">padding</span><span class="si">}{</span><span class="n">elapsed_str</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">remove_ansi</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove ANSI escape codes from a string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ansi_escape</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\x1B\[[0-?]*[ -/]*[@-~]&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ansi_escape</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.logconf.ColoredFormatter.format" class="doc doc-heading">
            <code class="highlight language-python"><span class="nb">format</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Format the log record with colors and elapsed time.
This method overrides the default format method to customize the log message format.
It includes the timestamp, logger name, log level, and message.</p>


            <details class="quote">
              <summary>Source code in <code>config/logconf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">record</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format the log record with colors and elapsed time.</span>
<span class="sd">    This method overrides the default format method to customize the log message format.</span>
<span class="sd">    It includes the timestamp, logger name, log level, and message.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">elapsed</span> <span class="o">=</span> <span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start_time</span><span class="p">)</span><span class="o">.</span><span class="n">total_seconds</span><span class="p">()</span>
    <span class="n">elapsed_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ELAPSED&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">format_duration</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="c1"># Compose colored parts</span>
    <span class="n">color</span> <span class="o">=</span> <span class="n">LOG_COLORS</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">levelname</span><span class="p">,</span> <span class="n">LOG_COLORS</span><span class="p">[</span><span class="s2">&quot;INFO&quot;</span><span class="p">])</span>
    <span class="n">time_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;DEBUG&#39;</span><span class="p">]</span><span class="si">}{</span><span class="bp">self</span><span class="o">.</span><span class="n">formatTime</span><span class="p">(</span><span class="n">record</span><span class="p">)</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">name_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;WARNING&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">level_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">color</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">levelname</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">msg_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">color</span><span class="si">}{</span><span class="n">record</span><span class="o">.</span><span class="n">getMessage</span><span class="p">()</span><span class="si">}{</span><span class="n">LOG_COLORS</span><span class="p">[</span><span class="s1">&#39;ENDC&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">raw_msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">time_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">name_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">level_str</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">msg_str</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">no_ansi_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">remove_ansi</span><span class="p">(</span><span class="n">raw_msg</span><span class="p">))</span>
    <span class="n">padding</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">width</span> <span class="o">-</span> <span class="n">no_ansi_len</span><span class="p">)</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">raw_msg</span><span class="si">}{</span><span class="s1">&#39; &#39;</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">padding</span><span class="si">}{</span><span class="n">elapsed_str</span><span class="si">}</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.logconf.ColoredFormatter.remove_ansi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">remove_ansi</span><span class="p">(</span><span class="n">s</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Remove ANSI escape codes from a string.</p>


            <details class="quote">
              <summary>Source code in <code>config/logconf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">remove_ansi</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove ANSI escape codes from a string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ansi_escape</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\x1B\[[0-?]*[ -/]*[@-~]&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ansi_escape</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="config.logconf.setup_logger" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup_logger</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;phoskintime&#39;</span><span class="p">,</span> <span class="n">log_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">,</span> <span class="n">log_dir</span><span class="o">=</span><span class="n">LOG_DIR</span><span class="p">,</span> <span class="n">rotate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">max_bytes</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">,</span> <span class="n">backup_count</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Setup a logger with colored output and file logging.
This function creates a logger with colored output for console messages
:param name:
:param log_file:
:param level:
:param log_dir:
:param rotate:
:param max_bytes:
:param backup_count:
:return: logger</p>


            <details class="quote">
              <summary>Source code in <code>config/logconf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">setup_logger</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;phoskintime&quot;</span><span class="p">,</span>
    <span class="n">log_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">,</span>
    <span class="n">log_dir</span><span class="o">=</span><span class="n">LOG_DIR</span><span class="p">,</span>
    <span class="n">rotate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">max_bytes</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span><span class="p">,</span>
    <span class="n">backup_count</span><span class="o">=</span><span class="mi">5</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Setup a logger with colored output and file logging.</span>
<span class="sd">    This function creates a logger with colored output for console messages</span>
<span class="sd">    :param name:</span>
<span class="sd">    :param log_file:</span>
<span class="sd">    :param level:</span>
<span class="sd">    :param log_dir:</span>
<span class="sd">    :param rotate:</span>
<span class="sd">    :param max_bytes:</span>
<span class="sd">    :param backup_count:</span>
<span class="sd">    :return: logger</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">log_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">log_dir</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">log_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">timestamp</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">log_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">log_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">timestamp</span><span class="si">}</span><span class="s2">.log&quot;</span><span class="p">)</span>

    <span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">level</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">logger</span><span class="o">.</span><span class="n">hasHandlers</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">handlers</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>

    <span class="c1"># File Handler</span>
    <span class="k">if</span> <span class="n">rotate</span><span class="p">:</span>
        <span class="n">file_handler</span> <span class="o">=</span> <span class="n">RotatingFileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">maxBytes</span><span class="o">=</span><span class="n">max_bytes</span><span class="p">,</span> <span class="n">backupCount</span><span class="o">=</span><span class="n">backup_count</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">file_handler</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">)</span>

    <span class="n">file_format</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">Formatter</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> - </span><span class="si">%(name)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">file_handler</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">file_format</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">file_handler</span><span class="p">)</span>

    <span class="c1"># Stream Handler (Console)</span>
    <span class="n">stream_handler</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">()</span>
    <span class="n">stream_format</span> <span class="o">=</span> <span class="n">ColoredFormatter</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> - </span><span class="si">%(name)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">stream_handler</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">stream_format</span><span class="p">)</span>
    <span class="n">stream_handler</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">stream_handler</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">logger</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.config.constants"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.config.logconf"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.exporter.plotout"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.exporter.plotout.opt_analyze_de" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">opt_analyze_de</span><span class="p">(</span><span class="n">long_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">ordered_optimizer_runs</span><span class="p">,</span> <span class="n">x_values</span><span class="p">,</span> <span class="n">y_values</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generates and saves various plots related to optimization results.
This includes waterfall plots, convergence plots, parameter trend plots,
and parameter scan plots.</p>
<p>:param long_df:
:param convergence_df:
:param ordered_optimizer_runs:
:param x_values:
:param y_values:
:param val:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">opt_analyze_de</span><span class="p">(</span><span class="n">long_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">ordered_optimizer_runs</span><span class="p">,</span>
                   <span class="n">x_values</span><span class="p">,</span> <span class="n">y_values</span><span class="p">,</span> <span class="n">val</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates and saves various plots related to optimization results.</span>
<span class="sd">    This includes waterfall plots, convergence plots, parameter trend plots,</span>
<span class="sd">    and parameter scan plots.</span>

<span class="sd">    :param long_df:</span>
<span class="sd">    :param convergence_df:</span>
<span class="sd">    :param ordered_optimizer_runs:</span>
<span class="sd">    :param x_values:</span>
<span class="sd">    :param y_values:</span>
<span class="sd">    :param val:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Waterfall plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">])),</span>
        <span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Objective Value&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Customize the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimizer Runs&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/waterfall.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="c1"># Waterfall plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_values</span><span class="p">,</span> <span class="n">y_values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="c1"># Plot the points</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">x_values</span><span class="p">,</span>
        <span class="n">y_values</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Objective Value&quot;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimizer Runs&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/waterfall_2.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Covergence plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span>
        <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Iteration&quot;</span><span class="p">],</span>
        <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">]),</span>  <span class="c1"># Changes in Value</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;coral&quot;</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Error&quot;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Iteration&quot;</span><span class="p">],</span> <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Error&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Iteration&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/convergence_2.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="p">)),</span> <span class="n">val</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Iteration&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/convergence.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">style</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s1">&#39;teal&#39;</span><span class="p">,</span> <span class="s2">&quot;marker&quot;</span><span class="p">:</span> <span class="s2">&quot;o&quot;</span><span class="p">},</span>
             <span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s1">&#39;indigo&#39;</span><span class="p">,</span> <span class="s2">&quot;marker&quot;</span><span class="p">:</span> <span class="s2">&quot;o&quot;</span><span class="p">}}</span>
    <span class="k">for</span> <span class="n">param_type</span><span class="p">,</span> <span class="n">props</span> <span class="ow">in</span> <span class="n">style</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">long_df</span><span class="p">[</span><span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">param_type</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">subset</span><span class="p">[</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">],</span>
            <span class="n">subset</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
            <span class="n">label</span><span class="o">=</span><span class="n">param_type</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">props</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">],</span>
            <span class="n">marker</span><span class="o">=</span><span class="n">props</span><span class="p">[</span><span class="s2">&quot;marker&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/parameter_trend.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="c1"># Use a hexbin plot to visualize distributions of parameter values</span>
    <span class="c1"># across the objective function</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">hb</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hexbin</span><span class="p">(</span>
        <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">],</span>
        <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
        <span class="n">gridsize</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span>
        <span class="n">mincnt</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">hb</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/parameter_scan.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="c1"># Plot the distributional plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust width to accommodate many parameters</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">,</span>
        <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">,</span>  <span class="c1"># This shows the distribution with respect to objective values</span>
        <span class="n">data</span><span class="o">=</span><span class="n">long_df</span><span class="p">,</span>
        <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span>
        <span class="n">density_norm</span><span class="o">=</span><span class="s1">&#39;width&#39;</span><span class="p">,</span>
        <span class="n">cut</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>  <span class="c1"># Remove x-axis ticks</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>  <span class="c1"># Remove the x-axis label</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/parameter_scatter.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.exporter.plotout.opt_analyze_nsga" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">opt_analyze_nsga</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">F</span><span class="p">,</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">approx_ideal</span><span class="p">,</span> <span class="n">approx_nadir</span><span class="p">,</span> <span class="n">asf_i</span><span class="p">,</span> <span class="n">pseudo_i</span><span class="p">,</span> <span class="n">n_evals</span><span class="p">,</span> <span class="n">hv</span><span class="p">,</span> <span class="n">hist</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">hist_cv_avg</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">igd</span><span class="p">,</span> <span class="n">best_objectives</span><span class="p">,</span> <span class="n">waterfall_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generates and saves various plots related to optimization results.
This includes design space plots, objective space plots,
convergence plots, and parameter trend plots.</p>
<p>:param problem:
:param result:
:param F:
:param pairs:
:param approx_ideal:
:param approx_nadir:
:param asf_i:
:param pseudo_i:
:param n_evals:
:param hv:
:param hist:
:param val:
:param hist_cv_avg:
:param k:
:param igd:
:param best_objectives:
:param waterfall_df:
:param convergence_df:
:param alpha_values:
:param beta_values:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">opt_analyze_nsga</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">F</span><span class="p">,</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">approx_ideal</span><span class="p">,</span>
                <span class="n">approx_nadir</span><span class="p">,</span> <span class="n">asf_i</span><span class="p">,</span> <span class="n">pseudo_i</span><span class="p">,</span> <span class="n">n_evals</span><span class="p">,</span>
                <span class="n">hv</span><span class="p">,</span> <span class="n">hist</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">hist_cv_avg</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">igd</span><span class="p">,</span> <span class="n">best_objectives</span><span class="p">,</span>
                <span class="n">waterfall_df</span><span class="p">,</span> <span class="n">convergence_df</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span>
                <span class="n">beta_values</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates and saves various plots related to optimization results.</span>
<span class="sd">    This includes design space plots, objective space plots,</span>
<span class="sd">    convergence plots, and parameter trend plots.</span>

<span class="sd">    :param problem:</span>
<span class="sd">    :param result:</span>
<span class="sd">    :param F:</span>
<span class="sd">    :param pairs:</span>
<span class="sd">    :param approx_ideal:</span>
<span class="sd">    :param approx_nadir:</span>
<span class="sd">    :param asf_i:</span>
<span class="sd">    :param pseudo_i:</span>
<span class="sd">    :param n_evals:</span>
<span class="sd">    :param hv:</span>
<span class="sd">    :param hist:</span>
<span class="sd">    :param val:</span>
<span class="sd">    :param hist_cv_avg:</span>
<span class="sd">    :param k:</span>
<span class="sd">    :param igd:</span>
<span class="sd">    :param best_objectives:</span>
<span class="sd">    :param waterfall_df:</span>
<span class="sd">    :param convergence_df:</span>
<span class="sd">    :param alpha_values:</span>
<span class="sd">    :param beta_values:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">xl</span><span class="p">,</span> <span class="n">xu</span> <span class="o">=</span> <span class="n">problem</span><span class="o">.</span><span class="n">bounds</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">result</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="n">xl</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">xu</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">xl</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">xu</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Design Space&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/design_space.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pairs</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="n">x</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="n">y</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Objective Space (F[</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">] vs F[</span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2">])&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;F[</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;F[</span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/objective_space_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">axes</span><span class="p">(</span><span class="n">projection</span><span class="o">=</span><span class="s2">&quot;3d&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Solutions&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">approx_ideal</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">approx_ideal</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">approx_ideal</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Ideal Point (Approx)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">approx_nadir</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">approx_nadir</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">approx_nadir</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Nadir Point (Approx)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Objective Space (3D)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;F[0]&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;F[1]&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s2">&quot;F[2]&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/objective_space_3d.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;ASF (Alpha Constraints vs Error)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/asf_plot1.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;ASF (Alpha vs Beta Constraints)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/asf_plot2.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">asf_i</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;ASF (Beta Constraints vs Error)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/asf_plot3.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Pseudo Weights (Alpha Constraints vs Error)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/pseudo_weights_plot1.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Pseudo Weights (Alpha vs Beta Constraints)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/pseudo_weights_plot2.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">facecolors</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">F</span><span class="p">[</span><span class="n">pseudo_i</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Pseudo Weights (Beta Constraints vs Error)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/pseudo_weights_plot2.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">hv</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Avg. CV of Pop&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">hv</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Convergence&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Function Evaluations&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Hypervolume&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/hypervolume_plot.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">hist_cv_avg</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Avg. CV of Pop&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">hist_cv_avg</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">n_evals</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;All Feasible&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Convergence&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Function Evaluations&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Constraint Violation&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/convergence_plot.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">igd</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;IGD+&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">n_evals</span><span class="p">,</span> <span class="n">igd</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;p&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">10</span> <span class="o">**</span> <span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10^-2&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;IGD+ Convergence&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Function Evaluations&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;IGD+&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/igd_convergence.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plot</span> <span class="o">=</span> <span class="n">Radar</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[</span><span class="n">approx_ideal</span><span class="p">,</span> <span class="n">approx_nadir</span><span class="p">],</span> <span class="n">normalize_each_objective</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">tight_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plot</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">best_objectives</span><span class="p">)</span>
    <span class="n">plot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">plot</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/radar_plot.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

    <span class="c1"># Get min and max values for each objective across all generations</span>
    <span class="n">all_f</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">algo</span><span class="o">.</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">algo</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">])</span>  <span class="c1"># Combine all generations</span>
    <span class="n">min_f</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">all_f</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">max_f</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">all_f</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Set up the figure and axis for 3D plotting</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s2">&quot;3d&quot;</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">update_frame</span><span class="p">(</span><span class="n">frame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Update the frame for the animation.</span>
<span class="sd">        Args:</span>
<span class="sd">            frame (int): The current frame number.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
        <span class="n">gen_data</span> <span class="o">=</span> <span class="n">hist</span><span class="p">[</span><span class="n">frame</span><span class="p">]</span><span class="o">.</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">)</span>  <span class="c1"># Extract objective values for the current generation</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">gen_data</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">gen_data</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">gen_data</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generation </span><span class="si">{</span><span class="n">frame</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;F[0]&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;F[1]&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s2">&quot;F[2]&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="n">min_f</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">max_f</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="n">min_f</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">max_f</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_zlim</span><span class="p">([</span><span class="n">min_f</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">max_f</span><span class="p">[</span><span class="mi">2</span><span class="p">]])</span>
    <span class="c1"># Create the animation</span>
    <span class="n">anim</span> <span class="o">=</span> <span class="n">FuncAnimation</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">update_frame</span><span class="p">,</span> <span class="n">frames</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">hist</span><span class="p">),</span> <span class="n">repeat</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Save as GIF</span>
    <span class="n">anim</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/optimization_run.gif&quot;</span><span class="p">,</span> <span class="n">writer</span><span class="o">=</span><span class="n">PillowWriter</span><span class="p">(</span><span class="n">fps</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

    <span class="n">ordered_optimizer_runs</span> <span class="o">=</span> <span class="n">waterfall_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Generate a waterfall plot for the convergence data</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">])),</span>
        <span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Objective Value&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Customize the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimizer Runs&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/waterfall.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Generate a waterfall plot for the convergence data</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="n">objective_values</span> <span class="o">=</span> <span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># Dynamically determine the threshold based on the range of objective values</span>
    <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="o">*</span> <span class="p">(</span><span class="n">objective_values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">objective_values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>

    <span class="c1"># Determine indices to plot</span>
    <span class="n">indices_to_plot</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Always plot the first point</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">objective_values</span><span class="p">)):</span>
        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">objective_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">objective_values</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">:</span>  <span class="c1"># Significant change</span>
            <span class="n">indices_to_plot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Plot sparsely for small changes</span>
            <span class="n">indices_to_plot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

            <span class="c1"># Plot only the selected indices</span>
    <span class="n">x_values</span> <span class="o">=</span> <span class="n">indices_to_plot</span>
    <span class="n">y_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">objective_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indices_to_plot</span><span class="p">]</span>

    <span class="c1"># Plot the line connecting points</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_values</span><span class="p">,</span> <span class="n">y_values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>

    <span class="c1"># Plot the points</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">x_values</span><span class="p">,</span>
        <span class="n">y_values</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Objective Value&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Customize the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimizer Runs&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/waterfall_2.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Covergence plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span>
        <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Iteration&quot;</span><span class="p">],</span>
        <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">]),</span>  <span class="c1"># Changes in Value</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;coral&quot;</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Error&quot;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Iteration&quot;</span><span class="p">],</span> <span class="n">convergence_df</span><span class="p">[</span><span class="s2">&quot;Value&quot;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Error&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Iteration&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{OUT_DIR}</span><span class="s1">/convergence_2.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="p">)),</span> <span class="n">val</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Iteration&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/convergence.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Combine alpha and beta labels with Greek symbols</span>
    <span class="n">param_labels</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Add alpha labels</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">param_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Add beta labels</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">param_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Melt the DataFrame to make it long-form for easy plotting</span>
    <span class="n">long_df</span> <span class="o">=</span> <span class="n">waterfall_df</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Individual&quot;</span><span class="p">,</span> <span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
                                <span class="n">value_vars</span><span class="o">=</span><span class="n">param_labels</span><span class="p">,</span>
                                <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span>
                                <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">)</span>

    <span class="c1"># Add a column to classify parameters as &#39;&#39; or &#39;&#39;</span>
    <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Parameter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Other&quot;</span><span class="p">))</span>

    <span class="c1"># Sort the DataFrame by &quot;Parameter Value&quot;</span>
    <span class="n">long_df</span> <span class="o">=</span> <span class="n">long_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">style</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s1">&#39;teal&#39;</span><span class="p">,</span> <span class="s2">&quot;marker&quot;</span><span class="p">:</span> <span class="s2">&quot;o&quot;</span><span class="p">},</span>
             <span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s1">&#39;indigo&#39;</span><span class="p">,</span> <span class="s2">&quot;marker&quot;</span><span class="p">:</span> <span class="s2">&quot;o&quot;</span><span class="p">}}</span>
    <span class="k">for</span> <span class="n">param_type</span><span class="p">,</span> <span class="n">props</span> <span class="ow">in</span> <span class="n">style</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">long_df</span><span class="p">[</span><span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">param_type</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">subset</span><span class="p">[</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">],</span>
            <span class="n">subset</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
            <span class="n">label</span><span class="o">=</span><span class="n">param_type</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">props</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">],</span>
            <span class="n">marker</span><span class="o">=</span><span class="n">props</span><span class="p">[</span><span class="s2">&quot;marker&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/parameter_trend.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Use a hexbin plot to visualize distributions of parameter values across the objective function</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="n">hb</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hexbin</span><span class="p">(</span>
        <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">],</span>
        <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
        <span class="n">gridsize</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span>
        <span class="n">mincnt</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">hb</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/parameter_scan.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Plot the distributional plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust width to accommodate many parameters</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">,</span>
        <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">,</span>  <span class="c1"># This shows the distribution with respect to objective values</span>
        <span class="n">data</span><span class="o">=</span><span class="n">long_df</span><span class="p">,</span>
        <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span>
        <span class="n">density_norm</span><span class="o">=</span><span class="s1">&#39;width&#39;</span><span class="p">,</span>
        <span class="n">cut</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>  <span class="c1"># Remove x-axis ticks</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>  <span class="c1"># Remove the x-axis label</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Optimized Values&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/parameter_scatter.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_residuals_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generates and saves combined residual-related plots for one gene with all psites in the legend.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_data</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary with keys 'psites', 'observed', 'estimated', and 'residuals' containing data for all psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>TIME_POINTS</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span> or <span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time points corresponding to the series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_residuals_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates and saves combined residual-related plots for one gene with all psites in the legend.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene (str): Gene identifier.</span>
<span class="sd">        gene_data (dict): Dictionary with keys &#39;psites&#39;, &#39;observed&#39;, &#39;estimated&#39;, and &#39;residuals&#39; containing data for all psites.</span>
<span class="sd">        TIME_POINTS (np.ndarray or list): Time points corresponding to the series.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get colors from Dark2 palette</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Dark2&quot;</span><span class="p">)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set1&quot;)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set2&quot;)</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Full timepoints plot</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">TIME_POINTS</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">TIME_POINTS</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation Level (FC)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">TIME_POINTS</span><span class="p">[</span><span class="mi">9</span><span class="p">:])</span>

    <span class="c1"># First 7 timepoints plot</span>
    <span class="n">short_timepoints</span> <span class="o">=</span> <span class="n">TIME_POINTS</span><span class="p">[:</span><span class="mi">7</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="c1"># axs[1].set_title(f&quot;{gene}&quot;)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_fit_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># 2. Cumulative Sum of Residuals</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
            <span class="n">TIME_POINTS</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]),</span>
            <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span>
        <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cumulative Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/cumulative_residuals_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># 3. Autocorrelation of Residuals</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plot_acf</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">lags</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">TIME_POINTS</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Lags&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Autocorrelation&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/autocorrelation_residuals_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># 4. Histogram of Residuals</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/histogram_residuals_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># 5. QQ Plot of Residuals</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">qqplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">line</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/qqplot_residuals_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.exporter.sheetutils"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">timepoints</span><span class="p">,</span> <span class="n">OUT_FILE</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>This function is responsible for exporting the results of the optimization process to an Excel file.
It creates multiple sheets in the Excel file, each containing different types of data related to the optimization results.
The function also generates plots for the residuals of each gene.
The data is organized in a structured manner, making it easy to analyze and interpret the results.
The function takes the following parameters:</p>
<p>:param P_initial:
:param P_init_dense:
:param P_estimated:
:param residuals:
:param alpha_values:
:param beta_values:
:param result:
:param timepoints:
:param OUT_FILE:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/exporter/sheetutils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span>
                   <span class="n">result</span><span class="p">,</span> <span class="n">timepoints</span><span class="p">,</span> <span class="n">OUT_FILE</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is responsible for exporting the results of the optimization process to an Excel file.</span>
<span class="sd">    It creates multiple sheets in the Excel file, each containing different types of data related to the optimization results.</span>
<span class="sd">    The function also generates plots for the residuals of each gene.</span>
<span class="sd">    The data is organized in a structured manner, making it easy to analyze and interpret the results.</span>
<span class="sd">    The function takes the following parameters:</span>

<span class="sd">    :param P_initial:</span>
<span class="sd">    :param P_init_dense:</span>
<span class="sd">    :param P_estimated:</span>
<span class="sd">    :param residuals:</span>
<span class="sd">    :param alpha_values:</span>
<span class="sd">    :param beta_values:</span>
<span class="sd">    :param result:</span>
<span class="sd">    :param timepoints:</span>
<span class="sd">    :param OUT_FILE:</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build a genes_data dictionary from computed metrics.</span>
    <span class="n">genes_data</span> <span class="o">=</span> <span class="n">build_genes_data</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">)</span>

    <span class="c1"># For each gene, call the plotting functions.</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span> <span class="ow">in</span> <span class="n">genes_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plot_residuals_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">)</span>

        <span class="c1"># Write results to Excel.</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="n">OUT_FILE</span>
    <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;openpyxl&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
        <span class="c1"># Sheet 1: Alpha Values</span>
        <span class="c1"># Create a DataFrame for alphas</span>
        <span class="n">alpha_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">alpha_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;Protein&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">:</span> <span class="n">value</span><span class="p">})</span>
        <span class="n">alpha_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">alpha_data</span><span class="p">)</span>

        <span class="c1"># Write alpha values to a separate sheet</span>
        <span class="n">alpha_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Alpha Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Sheet 2: Beta Values</span>
        <span class="c1"># Create a DataFrame for betas</span>
        <span class="c1"># Collect beta data for each kinase and psite combination</span>
        <span class="n">beta_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">beta_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">,</span>
                              <span class="s1">&#39;Beta&#39;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="o">.</span><span class="n">item</span><span class="p">())</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;item&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)})</span>

        <span class="c1"># Create DataFrame from the collected data</span>
        <span class="n">beta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">beta_data</span><span class="p">)</span>

        <span class="c1"># Write beta values to a separate sheet</span>
        <span class="n">beta_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Beta Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Sheet 4: Residuals with Gene and Psite Labels</span>
        <span class="n">residuals_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">residual_row</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">}</span>
            <span class="n">residual_row</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">time</span><span class="p">:</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">]</span> <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">time</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">timepoints</span><span class="p">)})</span>
            <span class="n">residuals_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">residual_row</span><span class="p">)</span>
        <span class="n">residuals_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">residuals_data</span><span class="p">)</span>
        <span class="n">residuals_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Residuals&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Sheet 5: Estimated Values with Gene and Psite Labels</span>
        <span class="n">estimated_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">estimated_row</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">}</span>
            <span class="n">estimated_row</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">time</span><span class="p">:</span> <span class="n">P_estimated</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">t</span><span class="p">]</span> <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">time</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">timepoints</span><span class="p">)})</span>
            <span class="n">estimated_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">estimated_row</span><span class="p">)</span>
        <span class="n">estimated_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">estimated_data</span><span class="p">)</span>
        <span class="n">estimated_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Estimated Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Optimization results saved for ODE modelling.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.objfn.minfndiffevo"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Custom optimization problem for phosphorylation analysis.</p>
<p>Defines the constraints, bounds, and objective function for optimizing
alpha and beta parameters across gene-psite-kinase relationships.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.P_initial">P_initial</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.P_initial_array">P_initial_array</span></code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.K_index">K_index</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.K_array">K_array</span></code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.gene_psite_counts">gene_psite_counts</span></code></td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.beta_counts">beta_counts</span></code></td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>kinopt/evol/objfn/minfndiffevo.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">PhosphorylationOptimizationProblem</span><span class="p">(</span><span class="n">ElementwiseProblem</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Custom optimization problem for phosphorylation analysis.</span>

<span class="sd">    Defines the constraints, bounds, and objective function for optimizing</span>
<span class="sd">    alpha and beta parameters across gene-psite-kinase relationships.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        P_initial (dict): Mapping of gene-psite pairs to kinase relationships and time-series data.</span>
<span class="sd">        P_initial_array (np.ndarray): Observed time-series data for gene-psite pairs.</span>
<span class="sd">        K_index (dict): Mapping of kinases to their respective psite data.</span>
<span class="sd">        K_array (np.ndarray): Array containing time-series data for kinase-psite combinations.</span>
<span class="sd">        gene_psite_counts (list): Number of kinases per gene-psite combination.</span>
<span class="sd">        beta_counts (dict): Mapping of kinase indices to the number of associated psites.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the optimization problem with given data and constraints.</span>

<span class="sd">        Args:</span>
<span class="sd">            P_initial (dict): Mapping of gene-psite pairs to kinase relationships and time-series data.</span>
<span class="sd">            P_initial_array (np.ndarray): Observed time-series data for gene-psite pairs.</span>
<span class="sd">            K_index (dict): Mapping of kinases to their respective psite data.</span>
<span class="sd">            K_array (np.ndarray): Array containing time-series data for kinase-psite combinations.</span>
<span class="sd">            gene_psite_counts (list): Number of kinases per gene-psite combination.</span>
<span class="sd">            beta_counts (dict): Mapping of kinase indices to the number of associated psites.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span> <span class="o">=</span> <span class="n">P_initial</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">=</span> <span class="n">P_initial_array</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span> <span class="o">=</span> <span class="n">K_index</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">K_array</span> <span class="o">=</span> <span class="n">K_array</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span> <span class="o">=</span> <span class="n">gene_psite_counts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span> <span class="o">=</span> <span class="n">beta_counts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Define the problem with pymoo</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">n_var</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">,</span>
            <span class="n">n_obj</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>  <span class="c1"># Single objective (sum of squared residuals)</span>
            <span class="n">n_ieq_constr</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_counts</span><span class="p">),</span>  <span class="c1"># Constraints for alpha and beta</span>
            <span class="n">xl</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">0</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">lb</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">]),</span>
            <span class="n">xu</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">ub</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">])</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluates the objective function and constraints for the given decision variables.</span>

<span class="sd">        Args:</span>
<span class="sd">            x (np.ndarray): Decision variable vector.</span>
<span class="sd">            out (dict): Dictionary to store objective function value and constraint values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Calculate the residuals for the objective function</span>
        <span class="n">error</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">objective_function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="c1"># Initialize an empty list for constraints</span>
        <span class="n">constraints</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Constraints for alphas (sum to 1 for each gene-psite-kinase group)</span>
        <span class="n">alpha_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
            <span class="n">constraints</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Constraints for betas (sum to 1 for each kinase across its psites)</span>
        <span class="n">beta_start</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>
        <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">psites</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">num_psites</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">psites</span><span class="p">)</span>
            <span class="n">constraints</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">num_psites</span>

        <span class="c1"># Ensure constraints list matches the expected length (self.n_ieq_constr)</span>
        <span class="c1"># Pad constraints with zeros if fewer constraints are defined</span>
        <span class="n">constraints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">constraints</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">constraints</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_ieq_constr</span><span class="p">:</span>
            <span class="n">constraints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">constraints</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_ieq_constr</span> <span class="o">-</span> <span class="n">constraints</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])])</span>

        <span class="c1"># Output the objective and constraints, reshaping G to have shape (1, n_ieq_constr)</span>
        <span class="n">out</span><span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">error</span>
        <span class="n">out</span><span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">constraints</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_ieq_constr</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">objective_function</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes the loss value for the given parameters using the selected loss type.</span>

<span class="sd">        Args:</span>
<span class="sd">            params (np.ndarray): Decision variables vector.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Computed loss value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
        <span class="n">alpha_start</span><span class="p">,</span> <span class="n">beta_start</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>

        <span class="c1"># Extract alphas for each gene-psite-kinase combination</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
            <span class="n">alpha</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span>
            <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Extract betas for each kinase-psite combination</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Calculate predicted matrix using alpha and beta values</span>
        <span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">P_i_t_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">kinases</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
            <span class="n">gene_psite_prediction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">t_max</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

            <span class="c1"># Sum contributions of each kinase for the gene-psite</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinases</span><span class="p">):</span>
                <span class="n">kinase_psites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">kinase</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">kinase_psites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Sum contributions across all psites of the kinase</span>
                <span class="k">for</span> <span class="n">k_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">k_psite</span><span class="p">,</span> <span class="n">k_time_series</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_psites</span><span class="p">):</span>
                    <span class="n">kinase_betas</span> <span class="o">=</span> <span class="n">beta</span><span class="p">[</span><span class="n">k_idx</span><span class="p">]</span>
                    <span class="n">gene_psite_prediction</span> <span class="o">+=</span> <span class="n">alpha</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">kinase_betas</span> <span class="o">*</span> <span class="n">k_time_series</span>

            <span class="n">P_i_t_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">gene_psite_prediction</span>

        <span class="c1"># Calculate residuals and sum of squared errors</span>
        <span class="n">residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">P_i_t_matrix</span>

        <span class="c1"># Select the loss function based on global loss_type</span>
        <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span><span class="p">:</span>
            <span class="c1"># MSE</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># MSE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span><span class="p">:</span>
            <span class="c1"># Autocorrelation Loss</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># Autocorrelation Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span><span class="p">:</span>
            <span class="c1"># Huber Loss</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
                <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
                <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="p">))</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># Huber Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
                <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
                <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span><span class="p">:</span>
            <span class="c1"># MAPE</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># MAPE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initializes the optimization problem with given data and constraints.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>P_initial_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_index</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_psite_counts</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_counts</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/objfn/minfndiffevo.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the optimization problem with given data and constraints.</span>

<span class="sd">    Args:</span>
<span class="sd">        P_initial (dict): Mapping of gene-psite pairs to kinase relationships and time-series data.</span>
<span class="sd">        P_initial_array (np.ndarray): Observed time-series data for gene-psite pairs.</span>
<span class="sd">        K_index (dict): Mapping of kinases to their respective psite data.</span>
<span class="sd">        K_array (np.ndarray): Array containing time-series data for kinase-psite combinations.</span>
<span class="sd">        gene_psite_counts (list): Number of kinases per gene-psite combination.</span>
<span class="sd">        beta_counts (dict): Mapping of kinase indices to the number of associated psites.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span> <span class="o">=</span> <span class="n">P_initial</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">=</span> <span class="n">P_initial_array</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span> <span class="o">=</span> <span class="n">K_index</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">K_array</span> <span class="o">=</span> <span class="n">K_array</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span> <span class="o">=</span> <span class="n">gene_psite_counts</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span> <span class="o">=</span> <span class="n">beta_counts</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="c1"># Define the problem with pymoo</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
        <span class="n">n_var</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">,</span>
        <span class="n">n_obj</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>  <span class="c1"># Single objective (sum of squared residuals)</span>
        <span class="n">n_ieq_constr</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_counts</span><span class="p">),</span>  <span class="c1"># Constraints for alpha and beta</span>
        <span class="n">xl</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">0</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">lb</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">]),</span>
        <span class="n">xu</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">ub</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">])</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_function</span><span class="p">(</span><span class="n">params</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/objfn/minfndiffevo.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">objective_function</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the loss value for the given parameters using the selected loss type.</span>

<span class="sd">    Args:</span>
<span class="sd">        params (np.ndarray): Decision variables vector.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: Computed loss value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
    <span class="n">alpha_start</span><span class="p">,</span> <span class="n">beta_start</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>

    <span class="c1"># Extract alphas for each gene-psite-kinase combination</span>
    <span class="n">alpha</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
        <span class="n">alpha</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span>
        <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

    <span class="c1"># Extract betas for each kinase-psite combination</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
        <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">count</span>

    <span class="c1"># Calculate predicted matrix using alpha and beta values</span>
    <span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span><span class="o">.</span><span class="n">shape</span>
    <span class="n">P_i_t_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="n">kinases</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
        <span class="n">gene_psite_prediction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">t_max</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

        <span class="c1"># Sum contributions of each kinase for the gene-psite</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinases</span><span class="p">):</span>
            <span class="n">kinase_psites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">kinase</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">kinase_psites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Sum contributions across all psites of the kinase</span>
            <span class="k">for</span> <span class="n">k_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">k_psite</span><span class="p">,</span> <span class="n">k_time_series</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_psites</span><span class="p">):</span>
                <span class="n">kinase_betas</span> <span class="o">=</span> <span class="n">beta</span><span class="p">[</span><span class="n">k_idx</span><span class="p">]</span>
                <span class="n">gene_psite_prediction</span> <span class="o">+=</span> <span class="n">alpha</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">kinase_betas</span> <span class="o">*</span> <span class="n">k_time_series</span>

        <span class="n">P_i_t_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">gene_psite_prediction</span>

    <span class="c1"># Calculate residuals and sum of squared errors</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">P_i_t_matrix</span>

    <span class="c1"># Select the loss function based on global loss_type</span>
    <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span><span class="p">:</span>
        <span class="c1"># MSE</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># MSE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span><span class="p">:</span>
        <span class="c1"># Autocorrelation Loss</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># Autocorrelation Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span><span class="p">:</span>
        <span class="c1"># Huber Loss</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
            <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
            <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="p">))</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># Huber Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
            <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
            <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span><span class="p">:</span>
        <span class="c1"># MAPE</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># MAPE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.objfn.minfnnsgaii"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Multi-objective optimization problem for phosphorylation analysis.</p>
<p>Objectives:
- Minimize sum of squared residuals (main objective).
- Minimize violations of constraints for alpha (secondary objective).
- Minimize violations of constraints for beta (tertiary objective).</p>







              <details class="quote">
                <summary>Source code in <code>kinopt/evol/objfn/minfnnsgaii.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">PhosphorylationOptimizationProblem</span><span class="p">(</span><span class="n">ElementwiseProblem</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Multi-objective optimization problem for phosphorylation analysis.</span>

<span class="sd">    Objectives:</span>
<span class="sd">    - Minimize sum of squared residuals (main objective).</span>
<span class="sd">    - Minimize violations of constraints for alpha (secondary objective).</span>
<span class="sd">    - Minimize violations of constraints for beta (tertiary objective).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the multi-objective optimization problem.</span>

<span class="sd">        Args:</span>
<span class="sd">            P_initial (dict): Mapping of gene-psite pairs to kinase relationships and time-series data.</span>
<span class="sd">            P_initial_array (np.ndarray): Observed time-series data for gene-psite pairs.</span>
<span class="sd">            K_index (dict): Mapping of kinases to their respective psite data.</span>
<span class="sd">            K_array (np.ndarray): Array containing time-series data for kinase-psite combinations.</span>
<span class="sd">            gene_psite_counts (list): Number of kinases per gene-psite combination.</span>
<span class="sd">            beta_counts (dict): Mapping of kinase indices to the number of associated psites.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span> <span class="o">=</span> <span class="n">P_initial</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">=</span> <span class="n">P_initial_array</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span> <span class="o">=</span> <span class="n">K_index</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">K_array</span> <span class="o">=</span> <span class="n">K_array</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span> <span class="o">=</span> <span class="n">gene_psite_counts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span> <span class="o">=</span> <span class="n">beta_counts</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Define the problem with pymoo</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="n">n_var</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">,</span>
            <span class="c1"># Multi-objective: [residuals, alpha constraints, beta constraints]</span>
            <span class="n">n_obj</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
            <span class="c1"># Constraints are part of the objectives</span>
            <span class="n">n_ieq_constr</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">xl</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">0</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">lb</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">]),</span>
            <span class="n">xu</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">ub</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">])</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluates the objectives for the given decision variables.</span>

<span class="sd">        Args:</span>
<span class="sd">            x (np.ndarray): Decision variable vector.</span>
<span class="sd">            out (dict): Dictionary to store objective function values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Primary objective: sum of squared residuals</span>
        <span class="n">error</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">objective_function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="c1"># Secondary objective: alpha constraint violations</span>
        <span class="n">alpha_violations</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">alpha_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
            <span class="n">alpha_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span>
            <span class="n">alpha_violations</span> <span class="o">+=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">alpha_sum</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
            <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Tertiary objective: beta constraint violations</span>
        <span class="n">beta_violations</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">beta_start</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>
        <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">psites</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">num_psites</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">psites</span><span class="p">)</span>
            <span class="n">beta_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">])</span>
            <span class="n">beta_violations</span> <span class="o">+=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">beta_sum</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">num_psites</span>

        <span class="c1"># Set objectives</span>
        <span class="n">out</span><span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">error</span><span class="p">,</span> <span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">objective_function</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes the loss value for the given parameters using the selected loss type.</span>

<span class="sd">        Args:</span>
<span class="sd">            params (np.ndarray): Decision variables vector.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Computed loss value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
        <span class="n">alpha_start</span><span class="p">,</span> <span class="n">beta_start</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>

        <span class="c1"># Extract alphas for each gene-psite-kinase combination</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
            <span class="n">alpha</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span>
            <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Extract betas for each kinase-psite combination</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">count</span>

        <span class="c1"># Calculate predicted matrix using alpha and beta values</span>
        <span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">P_i_t_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">kinases</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
            <span class="n">gene_psite_prediction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">t_max</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

            <span class="c1"># Sum contributions of each kinase for the gene-psite</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinases</span><span class="p">):</span>
                <span class="n">kinase_psites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">kinase</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">kinase_psites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Sum contributions across all psites of the kinase</span>
                <span class="k">for</span> <span class="n">k_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">k_psite</span><span class="p">,</span> <span class="n">k_time_series</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_psites</span><span class="p">):</span>
                    <span class="n">kinase_betas</span> <span class="o">=</span> <span class="n">beta</span><span class="p">[</span><span class="n">k_idx</span><span class="p">]</span>
                    <span class="n">gene_psite_prediction</span> <span class="o">+=</span> <span class="n">alpha</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">kinase_betas</span> <span class="o">*</span> <span class="n">k_time_series</span>

            <span class="n">P_i_t_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">gene_psite_prediction</span>

        <span class="c1"># Calculate residuals and sum of squared errors</span>
        <span class="n">residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">P_i_t_matrix</span>

        <span class="c1"># Select the loss function based on global loss_type</span>
        <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span><span class="p">:</span>
            <span class="c1"># MSE</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># MSE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span><span class="p">:</span>
            <span class="c1"># Autocorrelation Loss</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># Autocorrelation Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span><span class="p">:</span>
            <span class="c1"># Huber Loss</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
                <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
                <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="p">))</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># Huber Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
                <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
                <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span><span class="p">:</span>
            <span class="c1"># MAPE</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
            <span class="c1"># MAPE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initializes the multi-objective optimization problem.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>P_initial_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_index</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>K_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_psite_counts</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_counts</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/objfn/minfnnsgaii.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the multi-objective optimization problem.</span>

<span class="sd">    Args:</span>
<span class="sd">        P_initial (dict): Mapping of gene-psite pairs to kinase relationships and time-series data.</span>
<span class="sd">        P_initial_array (np.ndarray): Observed time-series data for gene-psite pairs.</span>
<span class="sd">        K_index (dict): Mapping of kinases to their respective psite data.</span>
<span class="sd">        K_array (np.ndarray): Array containing time-series data for kinase-psite combinations.</span>
<span class="sd">        gene_psite_counts (list): Number of kinases per gene-psite combination.</span>
<span class="sd">        beta_counts (dict): Mapping of kinase indices to the number of associated psites.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span> <span class="o">=</span> <span class="n">P_initial</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">=</span> <span class="n">P_initial_array</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span> <span class="o">=</span> <span class="n">K_index</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">K_array</span> <span class="o">=</span> <span class="n">K_array</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span> <span class="o">=</span> <span class="n">gene_psite_counts</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span> <span class="o">=</span> <span class="n">beta_counts</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="c1"># Define the problem with pymoo</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
        <span class="n">n_var</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">,</span>
        <span class="c1"># Multi-objective: [residuals, alpha constraints, beta constraints]</span>
        <span class="n">n_obj</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
        <span class="c1"># Constraints are part of the objectives</span>
        <span class="n">n_ieq_constr</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">xl</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">0</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">lb</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">]),</span>
        <span class="n">xu</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([(</span><span class="mi">1</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span><span class="p">,</span> <span class="p">(</span><span class="n">ub</span><span class="p">,)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_beta</span><span class="p">])</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_function</span><span class="p">(</span><span class="n">params</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/objfn/minfnnsgaii.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">objective_function</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the loss value for the given parameters using the selected loss type.</span>

<span class="sd">    Args:</span>
<span class="sd">        params (np.ndarray): Decision variables vector.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: Computed loss value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
    <span class="n">alpha_start</span><span class="p">,</span> <span class="n">beta_start</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_alpha</span>

    <span class="c1"># Extract alphas for each gene-psite-kinase combination</span>
    <span class="n">alpha</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_psite_counts</span><span class="p">:</span>
        <span class="n">alpha</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">])</span>
        <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

    <span class="c1"># Extract betas for each kinase-psite combination</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
        <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">count</span>

    <span class="c1"># Calculate predicted matrix using alpha and beta values</span>
    <span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span><span class="o">.</span><span class="n">shape</span>
    <span class="n">P_i_t_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">i_max</span><span class="p">,</span> <span class="n">t_max</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="n">kinases</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
        <span class="n">gene_psite_prediction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">t_max</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

        <span class="c1"># Sum contributions of each kinase for the gene-psite</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinases</span><span class="p">):</span>
            <span class="n">kinase_psites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">K_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">kinase</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">kinase_psites</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Sum contributions across all psites of the kinase</span>
            <span class="k">for</span> <span class="n">k_idx</span><span class="p">,</span> <span class="p">(</span><span class="n">k_psite</span><span class="p">,</span> <span class="n">k_time_series</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_psites</span><span class="p">):</span>
                <span class="n">kinase_betas</span> <span class="o">=</span> <span class="n">beta</span><span class="p">[</span><span class="n">k_idx</span><span class="p">]</span>
                <span class="n">gene_psite_prediction</span> <span class="o">+=</span> <span class="n">alpha</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">kinase_betas</span> <span class="o">*</span> <span class="n">k_time_series</span>

        <span class="n">P_i_t_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">gene_psite_prediction</span>

    <span class="c1"># Calculate residuals and sum of squared errors</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">P_i_t_matrix</span>

    <span class="c1"># Select the loss function based on global loss_type</span>
    <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span><span class="p">:</span>
        <span class="c1"># MSE</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;base&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># MSE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span><span class="p">:</span>
        <span class="c1"># Autocorrelation Loss</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;autocorrelation&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># Autocorrelation Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">residuals</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_max</span><span class="p">)])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span><span class="p">:</span>
        <span class="c1"># Huber Loss</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
            <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
            <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="p">))</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;huber&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># Huber Loss + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">1.0</span><span class="p">,</span>  <span class="c1"># Delta (adjust as necessary)</span>
            <span class="mf">0.5</span> <span class="o">*</span> <span class="n">residuals</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span>
            <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span><span class="p">:</span>
        <span class="c1"># MAPE</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="s2">&quot;mape&quot;</span> <span class="ow">and</span> <span class="n">include_regularization</span><span class="p">:</span>
        <span class="c1"># MAPE + L1 penalty (absolute values) + L2 penalty (squared values)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">P_initial_array</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">params</span><span class="p">))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.opt.optrun"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.opt.optrun.post_optimization_de" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_de</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.
1) Extracts the Pareto front and computes a weighted score to pick a 'best' solution.
2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.
3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hypervolume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">post_optimization_de</span><span class="p">(</span>
    <span class="n">result</span><span class="p">,</span>
    <span class="n">alpha_values</span><span class="p">,</span>
    <span class="n">beta_values</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Post-processes the result of a multi-objective optimization run.</span>
<span class="sd">    1) Extracts the Pareto front and computes a weighted score to pick a &#39;best&#39; solution.</span>
<span class="sd">    2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.</span>
<span class="sd">    3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</span>

<span class="sd">    Args:</span>
<span class="sd">        result: The final result object from the optimizer (e.g., a pymoo result).</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary with keys:</span>
<span class="sd">            &#39;best_solution&#39;: The best individual from the weighted scoring.</span>
<span class="sd">            &#39;best_objectives&#39;: Its corresponding objective vector.</span>
<span class="sd">            &#39;optimized_params&#39;: The individual&#39;s decision variables (X).</span>
<span class="sd">            &#39;scores&#39;: Weighted scores for each solution in the Pareto front.</span>
<span class="sd">            &#39;best_index&#39;: The index of the best solution according to weighted score.</span>
<span class="sd">            &#39;hist_hv&#39;: The hypervolume per generation.</span>
<span class="sd">            &#39;hist_igd&#39;: The IGD+ per generation.</span>
<span class="sd">            &#39;convergence_df&#39;: The DataFrame with iteration vs. best objective value</span>
<span class="sd">                for each iteration in the result history.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># # Display key results</span>
    <span class="c1"># print(&quot;Optimization Results:\n&quot;)</span>
    <span class="c1"># print(&quot;Objective Value (F):&quot;, result.F)  # Best objective value</span>
    <span class="c1"># print(&quot;Best Solution Variables (X):&quot;, result.X)  # Best solution variables</span>
    <span class="c1"># print(&quot;Constraint Violations (G):&quot;, result.G)  # Constraint violations</span>
    <span class="c1"># print(&quot;Constraint Violation Summary (CV):&quot;, result.CV)  # Summary of constraint violations</span>
    <span class="c1"># # Additional attributes</span>
    <span class="c1"># print(&quot;\nAdditional Information:&quot;)</span>
    <span class="c1"># print(&quot;Algorithm:&quot;, result.algorithm)</span>
    <span class="c1"># print(&quot;Archive:&quot;, result.archive)</span>
    <span class="c1"># print(&quot;Feasibility (feas):&quot;, result.feas)</span>
    <span class="c1"># # Display details about the population with more attributes for each individual</span>
    <span class="c1"># print(&quot;\nFirst 5 Population Details:&quot;)</span>
    <span class="c1"># for i, individual in enumerate(result.pop[:5]):  # Displaying first 5 individuals for brevity</span>
    <span class="c1">#     print(f&quot;\nIndividual {i + 1}:&quot;)</span>
    <span class="c1">#     print(&quot;  Decision Variables (X):&quot;, individual.X)</span>
    <span class="c1">#     print(&quot;  Objective Value (F):&quot;, individual.F)</span>
    <span class="c1">#     print(&quot;  Constraint Violation (CV):&quot;, individual.CV if hasattr(individual, &#39;CV&#39;) else &quot;N/A&quot;)</span>
    <span class="c1">#     print(&quot;  Feasibility (feas):&quot;, individual.feas if hasattr(individual, &#39;feas&#39;) else &quot;N/A&quot;)</span>
    <span class="c1"># Combined alpha and beta labels with Greek symbols</span>
    <span class="n">param_labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Add alpha () labels</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">param_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># Add beta () labels</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">param_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># Initialize an empty list to store individual data</span>
    <span class="n">pops</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Loop through the first 5 individuals and extract data</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">individual</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">pop</span><span class="p">):</span>  <span class="c1"># Displaying first 5 individuals for brevity</span>
        <span class="n">row</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Individual&quot;</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">:</span> <span class="n">individual</span><span class="o">.</span><span class="n">F</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>  <span class="c1"># Add individual info</span>
        <span class="n">row</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">param_labels</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">individual</span><span class="o">.</span><span class="n">X</span><span class="p">)})</span>  <span class="c1"># Add decision variables</span>
        <span class="n">pops</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="c1"># Create a DataFrame from the collected data</span>
    <span class="n">waterfall_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span>
    <span class="n">waterfall_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/parameter_scan.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Visualize the convergence</span>
    <span class="n">val</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span><span class="o">.</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">history</span><span class="p">]</span>
    <span class="c1"># Flatten the list or extract the first element of each value</span>
    <span class="n">flattened_val</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">))</span> <span class="k">else</span> <span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">val</span><span class="p">]</span>
    <span class="c1"># Correctly construct the DataFrame</span>
    <span class="n">convergence_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s2">&quot;Iteration&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">flattened_val</span><span class="p">)),</span>
        <span class="s2">&quot;Value&quot;</span><span class="p">:</span> <span class="n">flattened_val</span>
    <span class="p">})</span>
    <span class="c1"># Save the DataFrame to a CSV file</span>
    <span class="n">convergence_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/convergence.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ordered_optimizer_runs</span> <span class="o">=</span> <span class="n">waterfall_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">objective_values</span> <span class="o">=</span> <span class="n">ordered_optimizer_runs</span><span class="p">[</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="c1"># Dynamically determine the threshold based on the range of objective values</span>
    <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="o">*</span> <span class="p">(</span><span class="n">objective_values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">objective_values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="c1"># Determine indices to plot</span>
    <span class="n">indices_to_plot</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Always plot the first point</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">objective_values</span><span class="p">)):</span>
        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">objective_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">objective_values</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">:</span>  <span class="c1"># Significant change</span>
            <span class="n">indices_to_plot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Plot sparsely for small changes</span>
            <span class="n">indices_to_plot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="c1"># Plot only the selected indices</span>
    <span class="n">x_values</span> <span class="o">=</span> <span class="n">indices_to_plot</span>
    <span class="n">y_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">objective_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indices_to_plot</span><span class="p">]</span>
    <span class="c1"># Melt the DataFrame to make it long-form for easy plotting</span>
    <span class="n">long_df</span> <span class="o">=</span> <span class="n">waterfall_df</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Individual&quot;</span><span class="p">,</span> <span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">],</span>
                                <span class="n">value_vars</span><span class="o">=</span><span class="n">param_labels</span><span class="p">,</span>
                                <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;Parameter&quot;</span><span class="p">,</span>
                                <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;Parameter Value&quot;</span><span class="p">)</span>
    <span class="c1"># Add a column to classify parameters as &#39;&#39; or &#39;&#39;</span>
    <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">long_df</span><span class="p">[</span><span class="s2">&quot;Parameter&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Other&quot;</span><span class="p">))</span>
    <span class="c1"># Sort the DataFrame by &quot;Parameter Value&quot;</span>
    <span class="n">long_df</span> <span class="o">=</span> <span class="n">long_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">ordered_optimizer_runs</span><span class="p">,</span>
        <span class="n">convergence_df</span><span class="p">,</span>
        <span class="n">long_df</span><span class="p">,</span>
        <span class="n">x_values</span><span class="p">,</span>
        <span class="n">y_values</span><span class="p">,</span>
        <span class="n">val</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.opt.optrun.post_optimization_nsga" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_nsga</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]),</span> <span class="n">ref_point</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]))</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.
1) Extracts the Pareto front and computes a weighted score to pick a 'best' solution.
2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.
3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>weights</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of length 3 for weighting the objectives.</p>
              </div>
            </td>
            <td>
                  <code><span title="numpy.array">array</span>([1.0, 1.0, 1.0])</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ref_point</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference point for hypervolume computations.</p>
              </div>
            </td>
            <td>
                  <code><span title="numpy.array">array</span>([3, 1, 1])</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hypervolume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">post_optimization_nsga</span><span class="p">(</span>
    <span class="n">result</span><span class="p">,</span>
    <span class="n">weights</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]),</span>
    <span class="n">ref_point</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Post-processes the result of a multi-objective optimization run.</span>
<span class="sd">    1) Extracts the Pareto front and computes a weighted score to pick a &#39;best&#39; solution.</span>
<span class="sd">    2) Gathers metrics like Hypervolume (HV) and IGD+ over the optimization history.</span>
<span class="sd">    3) Logs feasibility generation info, saves waterfall and convergence data to CSV.</span>

<span class="sd">    Args:</span>
<span class="sd">        result: The final result object from the optimizer (e.g., a pymoo result).</span>
<span class="sd">        weights (np.ndarray): Array of length 3 for weighting the objectives.</span>
<span class="sd">        ref_point (np.ndarray): Reference point for hypervolume computations.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary with keys:</span>
<span class="sd">            &#39;best_solution&#39;: The best individual from the weighted scoring.</span>
<span class="sd">            &#39;best_objectives&#39;: Its corresponding objective vector.</span>
<span class="sd">            &#39;optimized_params&#39;: The individual&#39;s decision variables (X).</span>
<span class="sd">            &#39;scores&#39;: Weighted scores for each solution in the Pareto front.</span>
<span class="sd">            &#39;best_index&#39;: The index of the best solution according to weighted score.</span>
<span class="sd">            &#39;hist_hv&#39;: The hypervolume per generation.</span>
<span class="sd">            &#39;hist_igd&#39;: The IGD+ per generation.</span>
<span class="sd">            &#39;convergence_df&#39;: The DataFrame with iteration vs. best objective value</span>
<span class="sd">                for each iteration in the result history.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 1) Extract the Pareto front from the final population</span>
    <span class="n">pareto_front</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">ind</span><span class="o">.</span><span class="n">F</span> <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">pop</span><span class="p">])</span>

    <span class="c1"># Weighted scoring for picking a single &#39;best&#39; solution from Pareto</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">weights</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">weights</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">])</span>
    <span class="n">best_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="n">best_solution</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">pop</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>
    <span class="n">best_objectives</span> <span class="o">=</span> <span class="n">pareto_front</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>
    <span class="n">optimized_params</span> <span class="o">=</span> <span class="n">best_solution</span><span class="o">.</span><span class="n">X</span>

    <span class="c1"># Additional references from the result</span>
    <span class="n">F</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">F</span>  <span class="c1"># The entire final objective set</span>
    <span class="n">pairs</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">),(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)]</span>
    <span class="n">approx_ideal</span> <span class="o">=</span> <span class="n">F</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">approx_nadir</span> <span class="o">=</span> <span class="n">F</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Decomposition objects</span>
    <span class="n">decomp</span> <span class="o">=</span> <span class="n">ASF</span><span class="p">()</span>
    <span class="n">asf_i</span> <span class="o">=</span> <span class="n">decomp</span><span class="o">.</span><span class="n">do</span><span class="p">(</span><span class="n">F</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="n">weights</span><span class="p">)</span><span class="o">.</span><span class="n">argmin</span><span class="p">()</span>

    <span class="n">hist</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">history</span>  <span class="c1"># the full history of the optimization</span>

    <span class="n">n_evals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">hist_F</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">hist_cv</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">hist_cv_avg</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># 2) Gather feasibility, objective space data over each generation</span>
    <span class="k">for</span> <span class="n">algo</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">:</span>
        <span class="n">n_evals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">algo</span><span class="o">.</span><span class="n">evaluator</span><span class="o">.</span><span class="n">n_eval</span><span class="p">)</span>

        <span class="n">opt</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">opt</span>
        <span class="n">hist_cv</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CV&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">hist_cv_avg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">algo</span><span class="o">.</span><span class="n">pop</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CV&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">())</span>

        <span class="n">feas</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;feasible&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">hist_F</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">)[</span><span class="n">feas</span><span class="p">])</span>

    <span class="c1"># Identify when we first got a feasible solution</span>
    <span class="n">k</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">hist_cv</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">0.0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;At least one feasible solution in Generation </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> after </span><span class="si">{</span><span class="n">n_evals</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="si">}</span><span class="s2"> evaluations.&quot;</span><span class="p">)</span>

    <span class="c1"># Identify when the whole population became feasible</span>
    <span class="n">vals</span> <span class="o">=</span> <span class="n">hist_cv_avg</span>
    <span class="n">k</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mf">0.0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Whole population feasible in Generation </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> after </span><span class="si">{</span><span class="n">n_evals</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="si">}</span><span class="s2"> evaluations.&quot;</span><span class="p">)</span>

    <span class="c1"># 3) Metrics: Hypervolume &amp; IGD+ across generations</span>
    <span class="n">metric_hv</span> <span class="o">=</span> <span class="n">Hypervolume</span><span class="p">(</span>
        <span class="n">ref_point</span><span class="o">=</span><span class="n">ref_point</span><span class="p">,</span>
        <span class="n">norm_ref_point</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">zero_to_one</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">ideal</span><span class="o">=</span><span class="n">approx_ideal</span><span class="p">,</span>
        <span class="n">nadir</span><span class="o">=</span><span class="n">approx_nadir</span>
    <span class="p">)</span>
    <span class="n">hist_hv</span> <span class="o">=</span> <span class="p">[</span><span class="n">metric_hv</span><span class="o">.</span><span class="n">do</span><span class="p">(</span><span class="n">_F</span><span class="p">)</span> <span class="k">for</span> <span class="n">_F</span> <span class="ow">in</span> <span class="n">hist_F</span><span class="p">]</span>

    <span class="n">metric_igd</span> <span class="o">=</span> <span class="n">IGDPlus</span><span class="p">(</span><span class="n">F</span><span class="p">)</span>
    <span class="n">hist_igd</span> <span class="o">=</span> <span class="p">[</span><span class="n">metric_igd</span><span class="o">.</span><span class="n">do</span><span class="p">(</span><span class="n">_F</span><span class="p">)</span> <span class="k">for</span> <span class="n">_F</span> <span class="ow">in</span> <span class="n">hist_F</span><span class="p">]</span>

    <span class="c1"># 4) Waterfall or &#39;pops&#39; data (currently empty in snippet)</span>
    <span class="n">pops</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># You could populate &#39;pops&#39; with relevant info from each generation if desired</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">individual</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">pop</span><span class="p">):</span>  <span class="c1"># Displaying first 5 individuals for brevity</span>
        <span class="n">row</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Individual&quot;</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Objective Value (F)&quot;</span><span class="p">:</span> <span class="n">individual</span><span class="o">.</span><span class="n">F</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>  <span class="c1"># Add individual info</span>
        <span class="n">row</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">j</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">individual</span><span class="o">.</span><span class="n">X</span><span class="p">)})</span>  <span class="c1"># Add decision variables</span>
        <span class="n">pops</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>


    <span class="n">waterfall_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span>
    <span class="n">waterfall_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/parameter_scan.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># 5) Convergence data (best objective value each generation)</span>
    <span class="n">val</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span><span class="o">.</span><span class="n">opt</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">]</span>  <span class="c1"># each iteration&#39;s best F</span>
    <span class="c1"># Flatten if it is list or array</span>
    <span class="n">flattened_val</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">))</span> <span class="k">else</span> <span class="n">v</span>
        <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">val</span>
    <span class="p">]</span>
    <span class="n">convergence_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s2">&quot;Iteration&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">flattened_val</span><span class="p">)),</span>
        <span class="s2">&quot;Value&quot;</span><span class="p">:</span> <span class="n">flattened_val</span>
    <span class="p">})</span>
    <span class="n">convergence_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/convergence.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Display the selected solution and objectives</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;--- Best Solution ---&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Objective Values (F): </span><span class="si">{</span><span class="n">best_objectives</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">F</span><span class="p">,</span>
        <span class="n">pairs</span><span class="p">,</span>
        <span class="n">n_evals</span><span class="p">,</span>
        <span class="n">hist_cv</span><span class="p">,</span>
        <span class="n">hist_cv_avg</span><span class="p">,</span>
        <span class="n">k</span><span class="p">,</span>
        <span class="n">metric_igd</span><span class="p">,</span>
        <span class="n">metric_hv</span><span class="p">,</span>
        <span class="n">best_solution</span><span class="p">,</span>
        <span class="n">best_objectives</span><span class="p">,</span>
        <span class="n">optimized_params</span><span class="p">,</span>
        <span class="n">approx_nadir</span><span class="p">,</span>
        <span class="n">approx_ideal</span><span class="p">,</span>
        <span class="n">scores</span><span class="p">,</span>
        <span class="n">best_index</span><span class="p">,</span>
        <span class="n">hist</span><span class="p">,</span>
        <span class="n">hist_hv</span><span class="p">,</span>
        <span class="n">hist_igd</span><span class="p">,</span>
        <span class="n">convergence_df</span><span class="p">,</span>
        <span class="n">waterfall_df</span><span class="p">,</span>
        <span class="n">asf_i</span><span class="p">,</span>
        <span class="n">PseudoWeights</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span><span class="o">.</span><span class="n">do</span><span class="p">(</span><span class="n">F</span><span class="p">),</span>
        <span class="n">pairs</span><span class="p">,</span>
        <span class="n">val</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="n">PhosphorylationOptimizationProblem</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Sets up and runs the multi-objective optimization problem for phosphorylation
using an NSGA2 algorithm and a thread pool for parallelization.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>P_initial,</code>
            </td>
            <td>
                  <code>(<span title="P_initial_array">P_initial_array</span>, <span title="K_index">K_index</span>, <span title="K_array">K_array</span>, <span title="gene_psite_counts">gene_psite_counts</span>, <span title="beta_counts">beta_counts</span>)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data structures describing the problem (time-series data, kinases, etc.).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>PhosphorylationOptimizationProblem</code>
            </td>
            <td>
                  <code>class</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The custom problem class to be instantiated.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>result</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The pymoo result object containing the optimized population and history.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>exec_time</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Execution time for the optimization.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_optimization</span><span class="p">(</span>
    <span class="n">P_initial</span><span class="p">,</span>
    <span class="n">P_initial_array</span><span class="p">,</span>
    <span class="n">K_index</span><span class="p">,</span>
    <span class="n">K_array</span><span class="p">,</span>
    <span class="n">gene_psite_counts</span><span class="p">,</span>
    <span class="n">beta_counts</span><span class="p">,</span>
    <span class="n">PhosphorylationOptimizationProblem</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sets up and runs the multi-objective optimization problem for phosphorylation</span>
<span class="sd">    using an NSGA2 algorithm and a thread pool for parallelization.</span>

<span class="sd">    Args:</span>
<span class="sd">        P_initial, P_initial_array, K_index, K_array, gene_psite_counts, beta_counts:</span>
<span class="sd">            Data structures describing the problem (time-series data, kinases, etc.).</span>
<span class="sd">        PhosphorylationOptimizationProblem (class):</span>
<span class="sd">            The custom problem class to be instantiated.</span>

<span class="sd">    Returns:</span>
<span class="sd">        result: The pymoo result object containing the optimized population and history.</span>
<span class="sd">        exec_time: Execution time for the optimization.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 1) Determine the number of threads via lscpu</span>
    <span class="n">n_threads_cmd</span> <span class="o">=</span> <span class="s2">&quot;lscpu -p | grep -v &#39;^#&#39; | wc -l&quot;</span>
    <span class="n">n_threads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">n_threads_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of threads: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># 2) Create a thread pool and a parallelization runner</span>
    <span class="n">pool</span> <span class="o">=</span> <span class="n">ThreadPool</span><span class="p">(</span><span class="n">n_threads</span><span class="p">)</span>
    <span class="n">runner</span> <span class="o">=</span> <span class="n">StarmapParallelization</span><span class="p">(</span><span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">)</span>

    <span class="c1"># 3) Instantiate the problem</span>
    <span class="n">problem</span> <span class="o">=</span> <span class="n">PhosphorylationOptimizationProblem</span><span class="p">(</span>
        <span class="n">P_initial</span><span class="o">=</span><span class="n">P_initial</span><span class="p">,</span>
        <span class="n">P_initial_array</span><span class="o">=</span><span class="n">P_initial_array</span><span class="p">,</span>
        <span class="n">K_index</span><span class="o">=</span><span class="n">K_index</span><span class="p">,</span>
        <span class="n">K_array</span><span class="o">=</span><span class="n">K_array</span><span class="p">,</span>
        <span class="n">gene_psite_counts</span><span class="o">=</span><span class="n">gene_psite_counts</span><span class="p">,</span>
        <span class="n">beta_counts</span><span class="o">=</span><span class="n">beta_counts</span><span class="p">,</span>
        <span class="n">elementwise_runner</span><span class="o">=</span><span class="n">runner</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">METHOD</span> <span class="o">==</span> <span class="s2">&quot;DE&quot;</span><span class="p">:</span>
        <span class="c1"># 4) Set up the algorithm and termination criteria</span>
        <span class="c1"># for single-objective optimization</span>
        <span class="n">algorithm</span> <span class="o">=</span> <span class="n">DE</span><span class="p">(</span>
            <span class="n">pop_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
            <span class="n">sampling</span><span class="o">=</span><span class="n">LHS</span><span class="p">(),</span>
            <span class="n">variant</span><span class="o">=</span><span class="s2">&quot;DE/rand/1/bin&quot;</span><span class="p">,</span>
            <span class="n">CR</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
            <span class="n">dither</span><span class="o">=</span><span class="s2">&quot;vector&quot;</span><span class="p">,</span>
            <span class="n">jitter</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        <span class="n">termination</span> <span class="o">=</span> <span class="n">DefaultSingleObjectiveTermination</span><span class="p">(</span>
            <span class="c1"># xtol=1e-8,</span>
            <span class="c1"># cvtol=1e-6,</span>
            <span class="c1"># ftol=1e-6,</span>
            <span class="c1"># period=20,</span>
            <span class="c1"># n_max_gen=1000,</span>
            <span class="c1"># n_max_evals=100000</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># 4) Set up the algorithm and termination criteria</span>
        <span class="c1"># for multi-objective optimization</span>
        <span class="n">algorithm</span> <span class="o">=</span> <span class="n">NSGA2</span><span class="p">(</span>
            <span class="n">pop_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
            <span class="n">crossover</span><span class="o">=</span><span class="n">TwoPointCrossover</span><span class="p">(),</span>
            <span class="n">eliminate_duplicates</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">termination</span> <span class="o">=</span> <span class="n">DefaultMultiObjectiveTermination</span><span class="p">(</span>
            <span class="c1"># xtol=1e-8,</span>
            <span class="c1"># cvtol=1e-6,</span>
            <span class="c1"># ftol=0.0025,</span>
            <span class="c1"># n_max_gen=1000,</span>
            <span class="c1"># n_max_evals=100000</span>
        <span class="p">)</span>

    <span class="c1"># 5) Run the optimization</span>
    <span class="c1"># buf = io.StringIO()</span>
    <span class="c1"># with contextlib.redirect_stdout(buf):</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span>
        <span class="n">problem</span><span class="p">,</span>
        <span class="n">algorithm</span><span class="p">,</span>
        <span class="n">termination</span><span class="o">=</span><span class="n">termination</span><span class="p">,</span>
        <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">save_history</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>

    <span class="c1"># # Log the captured pymoo progress</span>
    <span class="c1"># pymoo_progress = buf.getvalue()</span>
    <span class="c1"># if pymoo_progress.strip():  # only log if there&#39;s actual text</span>
    <span class="c1">#     logger.info(&quot;--- Progress Output ---\n&quot; + pymoo_progress)</span>

    <span class="c1"># 6) Grab execution time and close the pool</span>
    <span class="c1"># Convert execution time to minutes and hours</span>
    <span class="n">exec_time_seconds</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">exec_time</span>
    <span class="n">exec_time_minutes</span> <span class="o">=</span> <span class="n">exec_time_seconds</span> <span class="o">/</span> <span class="mi">60</span>
    <span class="n">exec_time_hours</span> <span class="o">=</span> <span class="n">exec_time_seconds</span> <span class="o">/</span> <span class="mi">3600</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Execution Time: </span><span class="si">{</span><span class="n">exec_time_seconds</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> seconds |  &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">exec_time_minutes</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> minutes |  &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">exec_time_hours</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hours&quot;</span><span class="p">)</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">problem</span><span class="p">,</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.optcon.construct"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.optcon.construct.pipeline" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pipeline</span><span class="p">(</span><span class="n">input1_path</span><span class="p">,</span> <span class="n">input2_path</span><span class="p">,</span> <span class="n">time_series_columns</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">segment_points</span><span class="p">,</span> <span class="n">estimate_missing_kinases</span><span class="p">,</span> <span class="n">kinase_to_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Constructs the pipeline for the optimization process.
This function orchestrates the loading of data, scaling,
and the construction of the P_initial and K_array structures.
It returns the necessary data structures for the optimization process.
The function takes the following parameters:</p>
<p>:param input1_path: Path to the first CSV file (HGNC data)
:param input2_path: Path to the second CSV file (kinase interactions)
:param time_series_columns: List of time series columns to extract
:param scaling_method: Method for scaling the data
:param split_point: Split point for scaling
:param segment_points: Segment points for scaling
:param estimate_missing_kinases: Boolean flag for estimating missing kinases
:param kinase_to_psites: Dictionary mapping kinases to their respective psites</p>
<p>:return: Tuple containing:
    - full_hgnc_df (pd.DataFrame): The scaled data from input1
    - interaction_df (pd.DataFrame): The subset/merged DataFrame from input2
    - observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df
    - P_initial (dict): Initial mapping of gene-psite pairs to kinase relationships and time-series data
    - P_initial_array (np.ndarray): Array containing observed time-series data for gene-psite pairs
    - K_index (dict): Mapping of kinases to their respective psite data
    - K_array (np.ndarray): Array containing time-series data for kinase-psite combinations
    - beta_counts (dict): Mapping of kinase indices to the number of associated psites
    - gene_psite_counts (list): Number of kinases per gene-psite combination
    - n (int): Number of decision variables in the optimization problem</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">pipeline</span><span class="p">(</span>
    <span class="n">input1_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">input2_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">time_series_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">scaling_method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">split_point</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">segment_points</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
    <span class="n">estimate_missing_kinases</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">kinase_to_psites</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructs the pipeline for the optimization process.</span>
<span class="sd">    This function orchestrates the loading of data, scaling,</span>
<span class="sd">    and the construction of the P_initial and K_array structures.</span>
<span class="sd">    It returns the necessary data structures for the optimization process.</span>
<span class="sd">    The function takes the following parameters:</span>

<span class="sd">    :param input1_path: Path to the first CSV file (HGNC data)</span>
<span class="sd">    :param input2_path: Path to the second CSV file (kinase interactions)</span>
<span class="sd">    :param time_series_columns: List of time series columns to extract</span>
<span class="sd">    :param scaling_method: Method for scaling the data</span>
<span class="sd">    :param split_point: Split point for scaling</span>
<span class="sd">    :param segment_points: Segment points for scaling</span>
<span class="sd">    :param estimate_missing_kinases: Boolean flag for estimating missing kinases</span>
<span class="sd">    :param kinase_to_psites: Dictionary mapping kinases to their respective psites</span>

<span class="sd">    :return: Tuple containing:</span>
<span class="sd">        - full_hgnc_df (pd.DataFrame): The scaled data from input1</span>
<span class="sd">        - interaction_df (pd.DataFrame): The subset/merged DataFrame from input2</span>
<span class="sd">        - observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df</span>
<span class="sd">        - P_initial (dict): Initial mapping of gene-psite pairs to kinase relationships and time-series data</span>
<span class="sd">        - P_initial_array (np.ndarray): Array containing observed time-series data for gene-psite pairs</span>
<span class="sd">        - K_index (dict): Mapping of kinases to their respective psite data</span>
<span class="sd">        - K_array (np.ndarray): Array containing time-series data for kinase-psite combinations</span>
<span class="sd">        - beta_counts (dict): Mapping of kinase indices to the number of associated psites</span>
<span class="sd">        - gene_psite_counts (list): Number of kinases per gene-psite combination</span>
<span class="sd">        - n (int): Number of decision variables in the optimization problem</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 1) Load and scale</span>
    <span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">interaction_df</span><span class="p">,</span> <span class="n">observed</span> <span class="o">=</span> <span class="n">_load_and_scale_data</span><span class="p">(</span>
        <span class="n">input1_path</span><span class="o">=</span><span class="n">input1_path</span><span class="p">,</span>
        <span class="n">input2_path</span><span class="o">=</span><span class="n">input2_path</span><span class="p">,</span>
        <span class="n">time_series_columns</span><span class="o">=</span><span class="n">time_series_columns</span><span class="p">,</span>
        <span class="n">scaling_method</span><span class="o">=</span><span class="n">scaling_method</span><span class="p">,</span>
        <span class="n">split_point</span><span class="o">=</span><span class="n">split_point</span><span class="p">,</span>
        <span class="n">segment_points</span><span class="o">=</span><span class="n">segment_points</span><span class="p">,</span>
        <span class="n">estimate_missing_kinases</span><span class="o">=</span><span class="n">estimate_missing_kinases</span>
    <span class="p">)</span>

    <span class="c1"># 2) Build P_initial</span>
    <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span> <span class="o">=</span> <span class="n">_build_p_initial</span><span class="p">(</span>
        <span class="n">interaction_df</span><span class="p">,</span> <span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">time_series_columns</span>
    <span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">P_initial_array</span><span class="o">.</span><span class="n">size</span>

    <span class="c1"># 3) Build K_array</span>
    <span class="n">K_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">beta_counts</span> <span class="o">=</span> <span class="n">_build_k_array</span><span class="p">(</span>
        <span class="n">interaction_df</span><span class="o">=</span><span class="n">interaction_df</span><span class="p">,</span>
        <span class="n">full_hgnc_df</span><span class="o">=</span><span class="n">full_hgnc_df</span><span class="p">,</span>
        <span class="n">time</span><span class="o">=</span><span class="n">time_series_columns</span><span class="p">,</span>
        <span class="n">estimate_missing_kinases</span><span class="o">=</span><span class="n">estimate_missing_kinases</span><span class="p">,</span>
        <span class="n">kinase_to_psites</span><span class="o">=</span><span class="n">kinase_to_psites</span>
    <span class="p">)</span>

    <span class="c1"># 4) gene_psite_counts for alpha parameters</span>
    <span class="n">gene_psite_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Kinases&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">P_initial</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">full_hgnc_df</span><span class="p">,</span>         <span class="c1"># pd.DataFrame</span>
        <span class="n">interaction_df</span><span class="p">,</span>       <span class="c1"># pd.DataFrame</span>
        <span class="n">observed</span><span class="p">,</span>             <span class="c1"># pd.DataFrame</span>
        <span class="n">P_initial</span><span class="p">,</span>            <span class="c1"># dict</span>
        <span class="n">P_initial_array</span><span class="p">,</span>      <span class="c1"># np.ndarray</span>
        <span class="n">K_array</span><span class="p">,</span>              <span class="c1"># np.ndarray</span>
        <span class="n">K_index</span><span class="p">,</span>              <span class="c1"># dict</span>
        <span class="n">beta_counts</span><span class="p">,</span>          <span class="c1"># dict</span>
        <span class="n">gene_psite_counts</span><span class="p">,</span>    <span class="c1"># list</span>
        <span class="n">n</span>                     <span class="c1"># int</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.utils.iodata"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">time_series_columns</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">segment_points</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Applies different scaling methods to the time-series columns of a DataFrame.
Args:
    df (pd.DataFrame): The DataFrame containing the time-series data.
    time_series_columns (list): List of columns to be scaled.
    method (str): The scaling method to apply ('min_max', 'log', 'temporal', 'segmented', 'slope', 'cumulative').
    split_point (int): Column index for temporal scaling.
    segment_points (list): List of column indices for segmented scaling.
:returns:
    pd.DataFrame: The DataFrame with scaled time-series columns.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">time_series_columns</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">segment_points</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Applies different scaling methods to the time-series columns of a DataFrame.</span>
<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): The DataFrame containing the time-series data.</span>
<span class="sd">        time_series_columns (list): List of columns to be scaled.</span>
<span class="sd">        method (str): The scaling method to apply (&#39;min_max&#39;, &#39;log&#39;, &#39;temporal&#39;, &#39;segmented&#39;, &#39;slope&#39;, &#39;cumulative&#39;).</span>
<span class="sd">        split_point (int): Column index for temporal scaling.</span>
<span class="sd">        segment_points (list): List of column indices for segmented scaling.</span>
<span class="sd">    :returns:</span>
<span class="sd">        pd.DataFrame: The DataFrame with scaled time-series columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;min_max&#39;</span><span class="p">:</span>
        <span class="c1"># Min-Max Scaling (row-wise)</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span>
            <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
            <span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span>
        <span class="p">)</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;log&#39;</span><span class="p">:</span>
        <span class="c1"># Log Scaling</span>
        <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;temporal&#39;</span><span class="p">:</span>
        <span class="c1"># Temporal Scaling (first split_point columns vs the rest)</span>
        <span class="n">first_part_columns</span> <span class="o">=</span> <span class="n">time_series_columns</span><span class="p">[:</span><span class="n">split_point</span><span class="p">]</span>
        <span class="n">second_part_columns</span> <span class="o">=</span> <span class="n">time_series_columns</span><span class="p">[</span><span class="n">split_point</span><span class="p">:]</span>

        <span class="n">scaler_first</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">scaler_second</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>

        <span class="n">df</span><span class="p">[</span><span class="n">first_part_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler_first</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">first_part_columns</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="n">second_part_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler_second</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">second_part_columns</span><span class="p">])</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;segmented&#39;</span><span class="p">:</span>
        <span class="c1"># Segmented Scaling (multiple segments)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">segment_points</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;segment_points must be provided for segmented scaling.&quot;</span><span class="p">)</span>
        <span class="n">segment_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">time_series_columns</span><span class="p">[</span><span class="n">segment_points</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span><span class="n">segment_points</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>
                           <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">segment_points</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">segment</span> <span class="ow">in</span> <span class="n">segment_columns</span><span class="p">:</span>
            <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
            <span class="n">df</span><span class="p">[</span><span class="n">segment</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">segment</span><span class="p">])</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;slope&#39;</span><span class="p">:</span>
        <span class="c1"># Slope Scaling (rate of change)</span>
        <span class="n">diffs_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">diffs_df</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;cumulative&#39;</span><span class="p">:</span>
        <span class="c1"># Cumulative Scaling (cumulative sum)</span>
        <span class="n">cum_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="n">time_series_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">cum_df</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;None&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Invalid scaling method. Choose from &#39;min_max&#39;, &#39;log&#39;, &#39;temporal&#39;, &#39;segmented&#39;, &#39;slope&#39;, &#39;cumulative&#39;.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;report.html&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a single global report HTML file from all gene folders inside the results directory.</span>

<span class="sd">    For each gene folder (e.g. &quot;ABL2&quot;), the report will include:</span>
<span class="sd">      - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.</span>
<span class="sd">      - Each plot is confined to a fixed size of 900px by 900px.</span>
<span class="sd">      - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</span>

<span class="sd">    Args:</span>
<span class="sd">        results_dir (str): Path to the root results directory.</span>
<span class="sd">        output_file (str): Name of the generated global report file (placed inside results_dir).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Gather gene folders (skip &quot;General&quot; and &quot;logs&quot;)</span>
    <span class="n">gene_folders</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">results_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="ow">and</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;General&quot;</span><span class="p">,</span> <span class="s2">&quot;logs&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Build HTML content with updated CSS for spacing.</span>
    <span class="n">html_parts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;meta charset=&#39;UTF-8&#39;&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;title&gt;Estimation Report&lt;/title&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;body { font-family: Arial, sans-serif; margin: 20px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h1 { color: #333; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h2 { color: #555; font-size: 1.8em; border-bottom: 1px solid #ccc; padding-bottom: 5px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h3 { color: #666; font-size: 1.4em; margin-top: 10px; margin-bottom: 10px; }&quot;</span><span class="p">,</span>
        <span class="c1"># /* CSS grid for plots: two per row, fixed size 500px x 500px, extra space between rows */</span>
        <span class="s2">&quot;.plot-container {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  display: grid;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  grid-template-columns: repeat(2, 500px);&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  column-gap: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  row-gap: 40px;&quot;</span><span class="p">,</span> <span class="c1"># /* extra vertical gap */</span>
        <span class="s2">&quot;  justify-content: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;.plot-item {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;img, iframe {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  object-fit: contain;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: none;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="c1"># /* Data tables: full width, one per row */</span>
        <span class="s2">&quot;.data-table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 50%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border-collapse: collapse;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-top: 10px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;th, td {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: 1px solid #ccc;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  padding: 8px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  text-align: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;body&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;h1&gt;Kinase Optimization Report&lt;/h1&gt;&quot;</span>
    <span class="p">]</span>

    <span class="c1"># For each gene folder, create a section in the report.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_folders</span><span class="p">):</span>
        <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;Protein Group: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>

        <span class="c1"># Create grid container for fixed-size plots.</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;div class=&quot;plot-container&quot;&gt;&#39;</span><span class="p">)</span>
        <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                    <span class="n">rel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;plot-item&quot;&gt;&lt;h3&gt;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;img src=&quot;</span><span class="si">{</span><span class="n">rel_path</span><span class="si">}</span><span class="s1">&quot; alt=&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&quot;&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;/div&gt;&#39;</span><span class="p">)</span>  <span class="c1"># End of plot container</span>

        <span class="c1"># Data tables: display XLSX or CSV files from the gene folder, one per row.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.xlsx&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                    <span class="n">table_html</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">border</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;</span><span class="si">{</span><span class="n">table_html</span><span class="si">}</span><span class="s1">&lt;/div&gt;&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;p&gt;Error reading </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&lt;/p&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>

    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span>

    <span class="c1"># Write the report into the results directory.</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">html_parts</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Returns a formatted string representing the duration in seconds, minutes, or hours.
The format is:
- seconds: "xx.xx sec"
- minutes: "xx.xx min"
- hours: "xx.xx hr"</p>
<p>:param seconds:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a formatted string representing the duration in seconds, minutes, or hours.</span>
<span class="sd">    The format is:</span>
<span class="sd">    - seconds: &quot;xx.xx sec&quot;</span>
<span class="sd">    - minutes: &quot;xx.xx min&quot;</span>
<span class="sd">    - hours: &quot;xx.xx hr&quot;</span>

<span class="sd">    :param seconds:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">60</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> sec&quot;</span>
    <span class="k">elif</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">3600</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> min&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hr&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Loads the full HGNC data and kinase interaction data, applies scaling to the time-series columns,
and subsets/merges them. The first file is the full HGNC data, and the second file contains kinase interactions.
The function also handles the conversion of kinases from string format to list format.</p>
<p>:param estimate_missing:
:param scaling_method:
:param split_point:
:param seg_points:</p>
<p>:return:
full_hgnc_df (pd.DataFrame): The scaled data from input1
interaction_df (pd.DataFrame): The subset/merged DataFrame from input2
observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads the full HGNC data and kinase interaction data, applies scaling to the time-series columns,</span>
<span class="sd">    and subsets/merges them. The first file is the full HGNC data, and the second file contains kinase interactions.</span>
<span class="sd">    The function also handles the conversion of kinases from string format to list format.</span>

<span class="sd">    :param estimate_missing:</span>
<span class="sd">    :param scaling_method:</span>
<span class="sd">    :param split_point:</span>
<span class="sd">    :param seg_points:</span>

<span class="sd">    :return:</span>
<span class="sd">    full_hgnc_df (pd.DataFrame): The scaled data from input1</span>
<span class="sd">    interaction_df (pd.DataFrame): The subset/merged DataFrame from input2</span>
<span class="sd">    observed (pd.DataFrame): Subset of full_hgnc_df merged with interaction_df</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">INPUT1</span><span class="p">)</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="p">[</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span>
    <span class="n">time_series_cols</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)]</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">apply_scaling</span><span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">time_series_cols</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span>
    <span class="n">interaction_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">INPUT2</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">estimate_missing</span><span class="p">:</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interaction_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;max&quot;</span><span class="p">,</span> <span class="s2">&quot;min&quot;</span><span class="p">])</span>
        <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">interaction_df</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="nb">all</span><span class="p">(</span><span class="n">kinase</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span> <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)))]</span>
        <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)])</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interaction_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;max&quot;</span><span class="p">,</span> <span class="s2">&quot;min&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">interaction_df</span><span class="p">,</span> <span class="n">observed</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Organizes output files from the optimization process into separate folders based on protein names.
Each protein's files are moved into a folder named after the protein.
Any remaining files that do not match the protein pattern are moved to a "General" folder.</p>
<p>:param directories: List of directories to organize.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Organizes output files from the optimization process into separate folders based on protein names.</span>
<span class="sd">    Each protein&#39;s files are moved into a folder named after the protein.</span>
<span class="sd">    Any remaining files that do not match the protein pattern are moved to a &quot;General&quot; folder.</span>

<span class="sd">    :param directories: List of directories to organize.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Za-z0-9]+)_.*\.(json|svg|png|html|csv|xlsx)$&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">directory</span> <span class="ow">in</span> <span class="n">directories</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">directory</span><span class="si">}</span><span class="s2">&#39; is not a valid directory. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Move files matching the protein pattern.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">protein_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">protein_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">protein</span><span class="p">)</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>

        <span class="c1"># After protein files have been moved, move remaining files to a &quot;General&quot; folder.</span>
        <span class="n">general_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="s2">&quot;General&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.evol.utils.params"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Computes various error metrics to evaluate the performance of the optimization process.
The function calculates the Mean Squared Error (MSE), Root Mean Squared Error (RMSE),
Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared value.
These metrics provide insights into the accuracy of the estimated time series compared to the observed data.</p>
<p>The function takes the following parameters:
:param optimized_params:
:param P_initial:
:param P_initial_array:
:param K_index:
:param K_array:
:param gene_psite_counts:
:param beta_counts:
:param n:</p>
<p>:return:
- P_estimated: Estimated time series matrix for all gene-psite combinations.
- residuals: Residuals between observed and estimated values.
- mse: Mean Squared Error.
- rmse: Root Mean Squared Error.
- mae: Mean Absolute Error.
- mape: Mean Absolute Percentage Error.
- r_squared: R-squared value.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">P_initial_array</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span>
                    <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes various error metrics to evaluate the performance of the optimization process.</span>
<span class="sd">    The function calculates the Mean Squared Error (MSE), Root Mean Squared Error (RMSE),</span>
<span class="sd">    Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared value.</span>
<span class="sd">    These metrics provide insights into the accuracy of the estimated time series compared to the observed data.</span>

<span class="sd">    The function takes the following parameters:</span>
<span class="sd">    :param optimized_params:</span>
<span class="sd">    :param P_initial:</span>
<span class="sd">    :param P_initial_array:</span>
<span class="sd">    :param K_index:</span>
<span class="sd">    :param K_array:</span>
<span class="sd">    :param gene_psite_counts:</span>
<span class="sd">    :param beta_counts:</span>
<span class="sd">    :param n:</span>

<span class="sd">    :return:</span>
<span class="sd">    - P_estimated: Estimated time series matrix for all gene-psite combinations.</span>
<span class="sd">    - residuals: Residuals between observed and estimated values.</span>
<span class="sd">    - mse: Mean Squared Error.</span>
<span class="sd">    - rmse: Root Mean Squared Error.</span>
<span class="sd">    - mae: Mean Absolute Error.</span>
<span class="sd">    - mape: Mean Absolute Percentage Error.</span>
<span class="sd">    - r_squared: R-squared value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">P_estimated</span> <span class="o">=</span> <span class="n">estimated_series</span><span class="p">(</span>
        <span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_initial</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">K_array</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">beta_counts</span>
    <span class="p">)</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">P_estimated</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span>
    <span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">))</span>
    <span class="n">mape</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="n">P_initial_array</span><span class="p">))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">r_squared</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">residuals</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">P_initial_array</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">P_initial_array</span><span class="p">))</span> <span class="o">**</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;--- Error Metrics ---&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean Squared Error (MSE): </span><span class="si">{</span><span class="n">mse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Root Mean Squared Error (RMSE): </span><span class="si">{</span><span class="n">rmse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean Absolute Error (MAE): </span><span class="si">{</span><span class="n">mae</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean Absolute Percentage Error (MAPE): </span><span class="si">{</span><span class="n">mape</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;R-squared (R^2): </span><span class="si">{</span><span class="n">r_squared</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.evol.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Extracts the optimized alpha and beta values from the optimized parameters.
The function organizes the values into dictionaries for easy access and
interpretation. The alpha values are associated with gene-psite pairs and
their corresponding kinases, while the beta values are associated with
kinase-psite pairs. The function also logs the optimized values for
transparency and debugging purposes.</p>
<p>The function takes the following parameters:
:param P_initial:
:param gene_psite_counts:
:param K_index:
:param optimized_params:</p>
<p>:return:
- alpha_values: Dictionary mapping (gene, psite) to a dictionary of kinases and their alpha values.
- beta_values: Dictionary mapping (kinase, psite) to their corresponding beta values.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_psite_counts</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts the optimized alpha and beta values from the optimized parameters.</span>
<span class="sd">    The function organizes the values into dictionaries for easy access and</span>
<span class="sd">    interpretation. The alpha values are associated with gene-psite pairs and</span>
<span class="sd">    their corresponding kinases, while the beta values are associated with</span>
<span class="sd">    kinase-psite pairs. The function also logs the optimized values for</span>
<span class="sd">    transparency and debugging purposes.</span>

<span class="sd">    The function takes the following parameters:</span>
<span class="sd">    :param P_initial:</span>
<span class="sd">    :param gene_psite_counts:</span>
<span class="sd">    :param K_index:</span>
<span class="sd">    :param optimized_params:</span>

<span class="sd">    :return:</span>
<span class="sd">    - alpha_values: Dictionary mapping (gene, psite) to a dictionary of kinases and their alpha values.</span>
<span class="sd">    - beta_values: Dictionary mapping (kinase, psite) to their corresponding beta values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">alpha_start</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">):</span>
        <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">kinases</span> <span class="o">=</span> <span class="n">P_initial</span><span class="p">[(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">)][</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
        <span class="n">alpha_values</span><span class="p">[(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">kinases</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]))</span>
        <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>

    <span class="c1"># Extract betas for kinase-psite</span>
    <span class="n">beta_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">beta_start</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">gene_psite_counts</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">kinase_psites</span> <span class="ow">in</span> <span class="n">K_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">psite</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">kinase_psites</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># Each psite in beta_counts has a count of 1</span>
            <span class="n">beta_values</span><span class="p">[(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">)]</span> <span class="o">=</span> <span class="n">optimized_params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">:</span><span class="n">beta_start</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="n">count</span>
            <span class="c1"># Display optimized values</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Optimized Alpha Values:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Protein </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">, Psite </span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Kinase </span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Optimized Beta Values:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Kinase </span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">, Psite </span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.fitanalysis.helpers.postfit"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.additional_plots" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">additional_plots</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">scaled_data</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">residuals_df</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to create additional plots including CDF, KDE, Boxplot, and Hierarchical Clustering.</p>
<p>:param df:
:param scaled_data:
:param alpha_values:
:param beta_values:
:param residuals_df:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">additional_plots</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">scaled_data</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">residuals_df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to create additional plots including CDF, KDE, Boxplot, and Hierarchical Clustering.</span>

<span class="sd">    :param df:</span>
<span class="sd">    :param scaled_data:</span>
<span class="sd">    :param alpha_values:</span>
<span class="sd">    :param beta_values:</span>
<span class="sd">    :param residuals_df:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)]</span>
    <span class="n">beta_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)]</span>

    <span class="c1"># KDE Plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">(</span><span class="n">alpha_df</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">(</span><span class="n">beta_df</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/distribution_parameters.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Box Plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Type&#39;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Value&#39;</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">alpha_df</span><span class="p">,</span> <span class="n">beta_df</span><span class="p">])</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/boxplot_parameters.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># CDF Plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="n">alpha_df</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="n">beta_df</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;CDF&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/cdf_parameters.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Prepare data</span>
    <span class="n">alpha_data</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">:</span> <span class="s1">&#39;Value&#39;</span><span class="p">})</span>
    <span class="n">alpha_data</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;$\alpha$&#39;</span>
    <span class="n">beta_data</span> <span class="o">=</span> <span class="n">beta_values</span><span class="p">[[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="s1">&#39;Beta&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="s1">&#39;Beta&#39;</span><span class="p">:</span> <span class="s1">&#39;Value&#39;</span><span class="p">})</span>
    <span class="n">beta_data</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;$\beta$&#39;</span>
    <span class="n">combined_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">alpha_data</span><span class="p">,</span> <span class="n">beta_data</span><span class="p">])</span>
    <span class="n">sorted_data</span> <span class="o">=</span> <span class="n">combined_data</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>

    <span class="c1"># Plot the violin plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">sorted_data</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Value&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;Parameter&#39;</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Estimated Values&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/violin_parameters.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Rename the columns of time points in the residuals DataFrame to 1, 2, ..., 14</span>
    <span class="c1"># Create a mapping of known column names to their replacements</span>
    <span class="n">column_mapping</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)}</span>
    <span class="c1"># Rename the columns directly using the mapping</span>
    <span class="n">residuals_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">column_mapping</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Define consistent time columns</span>
    <span class="n">time_columns</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)]</span>

    <span class="c1"># Heatmap data for optimization progression</span>
    <span class="n">heatmap_data</span> <span class="o">=</span> <span class="n">residuals_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)[</span><span class="n">time_columns</span><span class="p">]</span>

    <span class="c1"># Optimization progression heatmap</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
        <span class="n">heatmap_data</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span>
        <span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;Magnitude&#39;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Group&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/time_residuals.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Variance across time points</span>
    <span class="n">time_point_variance</span> <span class="o">=</span> <span class="n">residuals_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)[</span><span class="n">time_columns</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Identify the top 5 highest variance time points</span>
    <span class="n">top_5_variances</span> <span class="o">=</span> <span class="n">time_point_variance</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

    <span class="c1"># Plot the variance across time points with top 5 highlighted</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">bars</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">time_point_variance</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">time_point_variance</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightblue&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">time_point</span> <span class="ow">in</span> <span class="n">top_5_variances</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="n">bars</span><span class="p">[</span><span class="n">time_point_variance</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">time_point</span><span class="p">)]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;coral&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$\mathrm</span><span class="si">{Var}</span><span class="s2">(\text</span><span class="si">{residuals}</span><span class="s2">)$&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

    <span class="c1"># Annotate the top 5 points</span>
    <span class="k">for</span> <span class="n">time_point</span><span class="p">,</span> <span class="n">variance</span> <span class="ow">in</span> <span class="n">top_5_variances</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">time_point_variance</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">time_point</span><span class="p">),</span> <span class="n">variance</span> <span class="o">+</span> <span class="mf">0.01</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">variance</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                 <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/variance_residuals.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Error trends over time</span>
    <span class="c1"># Mean absolute residuals across all Genes for each time point</span>
    <span class="n">mean_absolute_error</span> <span class="o">=</span> <span class="n">residuals_df</span><span class="p">[</span><span class="n">time_columns</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Identify top 3 lowest and top 3 highest MAE time points</span>
    <span class="n">top_3_highest_mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="n">top_3_lowest_mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="o">.</span><span class="n">nsmallest</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mean_absolute_error</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">mean_absolute_error</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightblue&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>

    <span class="c1"># Highlight top 3 highest with upward triangles and lowest with downward triangles</span>
    <span class="k">for</span> <span class="n">time_point</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">top_3_highest_mae</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">mean_absolute_error</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">time_point</span><span class="p">),</span> <span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">time_point</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">top_3_lowest_mae</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">mean_absolute_error</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">time_point</span><span class="p">),</span> <span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;MAE&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/error_trends.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Residual profiles across Genes</span>
    <span class="c1"># Sum residuals across time points for each gene</span>
    <span class="n">residual_profiles</span> <span class="o">=</span> <span class="n">residuals_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)[</span><span class="n">time_columns</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Sort and identify top 5 and bottom 5 Genes with highest and lowest cumulative residuals</span>
    <span class="n">sorted_residual_profiles</span> <span class="o">=</span> <span class="n">residual_profiles</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">top_5_residuals</span> <span class="o">=</span> <span class="n">sorted_residual_profiles</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    <span class="n">bottom_5_residuals</span> <span class="o">=</span> <span class="n">sorted_residual_profiles</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Assign unique colors to the top 5 and bottom 5 bars</span>
    <span class="n">top_colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="s1">&#39;purple&#39;</span><span class="p">]</span>
    <span class="n">bottom_colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cyan&#39;</span><span class="p">,</span> <span class="s1">&#39;magenta&#39;</span><span class="p">,</span> <span class="s1">&#39;yellow&#39;</span><span class="p">,</span> <span class="s1">&#39;brown&#39;</span><span class="p">,</span> <span class="s1">&#39;pink&#39;</span><span class="p">]</span>
    <span class="n">default_color</span> <span class="o">=</span> <span class="s1">&#39;teal&#39;</span>

    <span class="c1"># Create bars with distinct colors for the top 5 and bottom 5</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_residual_profiles</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">top_5_residuals</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">color_index</span> <span class="o">=</span> <span class="n">top_5_residuals</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">top_colors</span><span class="p">[</span><span class="n">color_index</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                    <span class="n">label</span><span class="o">=</span><span class="n">gene</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">bottom_5_residuals</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">color_index</span> <span class="o">=</span> <span class="n">bottom_5_residuals</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">bottom_colors</span><span class="p">[</span><span class="n">color_index</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                    <span class="n">label</span><span class="o">=</span><span class="n">gene</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">default_color</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

    <span class="c1"># Add a single legend for the top 5 and bottom 5 Genes</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Protein&quot;</span><span class="p">)</span>

    <span class="c1"># Remove x-axis labels and ticks for a cleaner look</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([],</span> <span class="p">[])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cumulative Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/residual_profiles.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_sankey_from_network</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a Sankey diagram from the given data and saves it as an HTML file.</p>
<p>This function processes the input data to generate nodes and links for a Sankey diagram.
It assigns colors to nodes and links based on their attributes and values, and uses Plotly
to render the diagram. The resulting diagram is saved as an HTML file in the specified output directory.</p>
<p>:param output_dir: str
    The directory where the Sankey diagram HTML file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the data for the Sankey diagram. It must include the following columns:
    - 'Source': The source node of the link.
    - 'Target': The target node of the link.
    - 'Value': The value of the link, which determines the flow size.
:param title: str
    The title of the Sankey diagram.</p>
<p>The function performs the following steps:
1. Initializes nodes and links for the Sankey diagram.
2. Maps node labels to indices and assigns colors to nodes.
3. Processes the data to create links between nodes, assigning colors based on link values.
4. Builds the Sankey diagram using Plotly.
5. Adds a color bar to explain the flow gradient.
6. Saves the Sankey diagram as an HTML file in the specified output directory.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_sankey_from_network</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a Sankey diagram from the given data and saves it as an HTML file.</span>

<span class="sd">    This function processes the input data to generate nodes and links for a Sankey diagram.</span>
<span class="sd">    It assigns colors to nodes and links based on their attributes and values, and uses Plotly</span>
<span class="sd">    to render the diagram. The resulting diagram is saved as an HTML file in the specified output directory.</span>

<span class="sd">    :param output_dir: str</span>
<span class="sd">        The directory where the Sankey diagram HTML file will be saved.</span>
<span class="sd">    :param data: pd.DataFrame</span>
<span class="sd">        A DataFrame containing the data for the Sankey diagram. It must include the following columns:</span>
<span class="sd">        - &#39;Source&#39;: The source node of the link.</span>
<span class="sd">        - &#39;Target&#39;: The target node of the link.</span>
<span class="sd">        - &#39;Value&#39;: The value of the link, which determines the flow size.</span>
<span class="sd">    :param title: str</span>
<span class="sd">        The title of the Sankey diagram.</span>

<span class="sd">    The function performs the following steps:</span>
<span class="sd">    1. Initializes nodes and links for the Sankey diagram.</span>
<span class="sd">    2. Maps node labels to indices and assigns colors to nodes.</span>
<span class="sd">    3. Processes the data to create links between nodes, assigning colors based on link values.</span>
<span class="sd">    4. Builds the Sankey diagram using Plotly.</span>
<span class="sd">    5. Adds a color bar to explain the flow gradient.</span>
<span class="sd">    6. Saves the Sankey diagram as an HTML file in the specified output directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initialize nodes and links for Sankey diagram</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">links</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Create a mapping for node indices</span>
    <span class="n">node_indices</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Define color scale</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab20</span>  <span class="c1"># Choose a colormap</span>
    <span class="n">norm</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">vmax</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>

    <span class="c1"># Process the data to extract nodes and edges</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span>
        <span class="n">value</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">])</span> <span class="o">*</span> <span class="mi">100</span>  <span class="c1"># Use absolute value for Sankey flow size</span>

        <span class="c1"># Add source and target nodes if not already added</span>
        <span class="k">if</span> <span class="n">source</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">node_indices</span><span class="p">:</span>
            <span class="n">nodes</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;green&quot;</span> <span class="k">if</span> <span class="s2">&quot;Kinase&quot;</span> <span class="ow">in</span> <span class="n">source</span> <span class="k">else</span> <span class="s2">&quot;red&quot;</span><span class="p">})</span>
            <span class="n">node_indices</span><span class="p">[</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="n">index</span>
            <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">target</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">node_indices</span><span class="p">:</span>
            <span class="n">nodes</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;gray&quot;</span><span class="p">})</span>
            <span class="n">node_indices</span><span class="p">[</span><span class="n">target</span><span class="p">]</span> <span class="o">=</span> <span class="n">index</span>
            <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># Generate color for the link based on value</span>
        <span class="n">flow_color</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">rgb2hex</span><span class="p">(</span><span class="n">cmap</span><span class="p">(</span><span class="n">norm</span><span class="p">(</span><span class="n">value</span><span class="p">)))</span>

        <span class="c1"># Add a link between the source and target</span>
        <span class="n">links</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">node_indices</span><span class="p">[</span><span class="n">source</span><span class="p">],</span>
            <span class="s2">&quot;target&quot;</span><span class="p">:</span> <span class="n">node_indices</span><span class="p">[</span><span class="n">target</span><span class="p">],</span>
            <span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="n">value</span><span class="p">,</span>
            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">flow_color</span>
        <span class="p">})</span>

    <span class="c1"># Build the Sankey diagram</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Sankey</span><span class="p">(</span>
        <span class="n">node</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
            <span class="n">pad</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
            <span class="n">thickness</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
            <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="n">node</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">node</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">]</span>
        <span class="p">),</span>
        <span class="n">link</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
            <span class="n">source</span><span class="o">=</span><span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="p">],</span>
            <span class="n">target</span><span class="o">=</span><span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="p">],</span>
            <span class="n">value</span><span class="o">=</span><span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;color&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="p">]</span>  <span class="c1"># Set link colors</span>
        <span class="p">)</span>
    <span class="p">))</span>
    <span class="c1"># Add a color bar to explain the flow gradient</span>
    <span class="n">colorbar</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">],</span>
        <span class="n">y</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">],</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers&#39;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
            <span class="n">colorscale</span><span class="o">=</span><span class="s1">&#39;Viridis&#39;</span><span class="p">,</span>
            <span class="n">cmin</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
            <span class="n">cmax</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span>
            <span class="n">colorbar</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Edge Value&quot;</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">hoverinfo</span><span class="o">=</span><span class="s1">&#39;none&#39;</span>
    <span class="p">)</span>
    <span class="c1"># Add the color bar scatter plot</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">colorbar</span><span class="p">)</span>
    <span class="c1"># Update layout</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">xaxis_visible</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">yaxis_visible</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">update_xaxes</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">update_yaxes</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">title_text</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s2">/sankey.html&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">goodnessoffit</span><span class="p">(</span><span class="n">estimated</span><span class="p">,</span> <span class="n">observed</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the goodness of fit for estimated and observed values.
It creates scatter plots with confidence intervals and labels for genes outside the 95% CI.
The function also calculates KL divergence and generates a heatmap for optimization progression.</p>
<p>:param estimated:
:param observed:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">goodnessoffit</span><span class="p">(</span><span class="n">estimated</span><span class="p">,</span> <span class="n">observed</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the goodness of fit for estimated and observed values.</span>
<span class="sd">    It creates scatter plots with confidence intervals and labels for genes outside the 95% CI.</span>
<span class="sd">    The function also calculates KL divergence and generates a heatmap for optimization progression.</span>

<span class="sd">    :param estimated:</span>
<span class="sd">    :param observed:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">merged_data</span> <span class="o">=</span> <span class="n">estimated</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                                     <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_est&#39;</span><span class="p">,</span> <span class="s1">&#39;_obs&#39;</span><span class="p">))</span>
    <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">tab20</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)))</span>

    <span class="n">ci_color_95</span> <span class="o">=</span> <span class="s1">&#39;red&#39;</span>
    <span class="n">ci_color_99</span> <span class="o">=</span> <span class="s1">&#39;gray&#39;</span>
    <span class="n">diagonal_color</span> <span class="o">=</span> <span class="s1">&#39;gray&#39;</span>

    <span class="c1"># Generate a palette with as many distinct colors as genes</span>
    <span class="n">unique_genes_list</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

    <span class="c1"># Generate a categorical color palette with fully distinct colors (tab20 has 20 distinct colors)</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes_list</span><span class="p">))</span>

    <span class="c1"># Map each gene to a unique color from the categorical palette</span>
    <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">palette</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes_list</span><span class="p">)}</span>

    <span class="c1"># Calculate overall mean and standard deviation</span>
    <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

    <span class="c1"># Calculate CI offsets for 95% and 99%</span>
    <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
    <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

    <span class="c1"># Goodness of Fit: Time-Series</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="n">gene_handles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)):</span>
        <span class="c1"># Sort values for plotting</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
        <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="c1"># Plot observed and estimated values</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>

        <span class="c1"># Add label for each gene once, matching the point color</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">handle</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">gene_handles</span><span class="p">]:</span>
            <span class="n">handle</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>

    <span class="c1"># Add diagonal line through the origin</span>
    <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
                  <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span>
                  <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">diagonal_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add lines parallel to the diagonal for 95% and 99% CI</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Expand the axes limits slightly to include all points without clipping</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>

    <span class="c1"># Add labels and legends</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Add first legend for CI lines and diagonal</span>
    <span class="n">ci_legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">ci_legend</span><span class="p">)</span>

    <span class="c1"># Add second legend for gene labels</span>
    <span class="n">gene_legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">gene_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Proteins&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">gene_legend</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_1.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Goodness of Fit: Time-Series</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="n">gene_handles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)):</span>
        <span class="c1"># Sort values for plotting</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
        <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="c1"># Plot observed and estimated values</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>

        <span class="c1"># Add label for each gene once, matching the point color</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">handle</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">gene_handles</span><span class="p">]:</span>
            <span class="n">handle</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>

    <span class="c1"># Add diagonal line through the origin</span>
    <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
                  <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span>
                  <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">diagonal_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add lines parallel to the diagonal for 95% and 99% CI</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add labels and legends</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Add first legend for CI lines and diagonal</span>
    <span class="n">ci_legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">ci_legend</span><span class="p">)</span>

    <span class="c1"># Add second legend for gene labels</span>
    <span class="n">gene_legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">gene_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Proteins&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">gene_legend</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_2.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">text_annotations</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Collect text objects for adjustment</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)):</span>
        <span class="c1"># Sort values for plotting</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
        <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="c1"># Plot observed and estimated values</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>

        <span class="c1"># Add label only for genes outside the 95% CI</span>
        <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                         <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                         <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Adjust text positions to avoid overlap</span>
    <span class="n">adjust_text</span><span class="p">(</span><span class="n">text_annotations</span><span class="p">,</span> <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>

    <span class="c1"># Add diagonal line through the origin</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">diagonal_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add lines parallel to the diagonal for 95% and 99% CI</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add labels and grid</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_3.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Plot Observed vs. Estimated for x1 to x14 values</span>
    <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">text_annotations</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Collect text objects for adjustment</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                              <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)):</span>
        <span class="c1"># Sort values for plotting</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
        <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="c1"># Plot observed and estimated values</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>

        <span class="c1"># Add label only for genes outside the 95% CI</span>
        <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                         <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                         <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Adjust text positions to avoid overlap</span>
    <span class="n">adjust_text</span><span class="p">(</span><span class="n">text_annotations</span><span class="p">,</span> <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>

    <span class="c1"># Add diagonal line through the origin</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">diagonal_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add lines parallel to the diagonal for 95% and 99% CI</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_95</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">ci_color_99</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
             <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Expand the axes limits slightly to include all points without clipping</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span>
                                                                      <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>

    <span class="c1"># Add labels and grid</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_4.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Goodness of Fit: Overall</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Define the color palette for distinct colors</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">merged_data</span><span class="p">))</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())}</span>

    <span class="c1"># Calculate 95% and 99% CI bounds</span>
    <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
    <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

    <span class="n">lower_bound_95</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_95</span>
    <span class="n">upper_bound_95</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_95</span>
    <span class="n">lower_bound_99</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_99</span>
    <span class="n">upper_bound_99</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_99</span>

    <span class="c1"># Plot points</span>
    <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">],</span>
                                     <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">],</span>
                                     <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                     <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]):</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>

        <span class="c1"># Add labels only for points outside the 95% CI, ensuring each gene is labeled once</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">upper_bound_95</span><span class="p">(</span><span class="n">obs</span><span class="p">)</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">lower_bound_95</span><span class="p">(</span><span class="n">obs</span><span class="p">)):</span>
            <span class="n">label</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span>
                     <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
            <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Add straight grey line through origin diagonal</span>
    <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()),</span>
              <span class="nb">max</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add dashed 95% CI lines in red parallel to the diagonal</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">upper_bound_95</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">lower_bound_95</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add dashed 99% CI lines in grey parallel to the diagonal</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">upper_bound_99</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">lower_bound_99</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add labels and grid</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_5.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Goodness of Fit: Overall with sorting by values directly in the loop</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Define the color palette for distinct colors</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">merged_data</span><span class="p">))</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())}</span>

    <span class="c1"># Calculate 95% and 99% CI bounds</span>
    <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
    <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

    <span class="n">lower_bound_95</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_95</span>
    <span class="n">upper_bound_95</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_95</span>
    <span class="n">lower_bound_99</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_99</span>
    <span class="n">upper_bound_99</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_99</span>

    <span class="c1"># Sort observed and estimated values while iterating</span>
    <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">sorted_indices</span><span class="p">:</span>
        <span class="n">obs</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">est</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="c1"># psite = merged_data[&#39;Psite&#39;].iloc[idx]</span>

        <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>

        <span class="c1"># Add labels only for points outside the 95% CI, ensuring each gene is labeled once</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">upper_bound_95</span><span class="p">(</span><span class="n">obs</span><span class="p">)</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">lower_bound_95</span><span class="p">(</span><span class="n">obs</span><span class="p">)):</span>
            <span class="n">label</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span>
                     <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
            <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Add straight grey line through origin diagonal</span>
    <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># Add dashed 95% CI lines in red parallel to the diagonal</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">upper_bound_95</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">lower_bound_95</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add dashed 99% CI lines in grey parallel to the diagonal</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">upper_bound_99</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">lower_bound_99</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Expand the axes limits slightly to include all points without clipping</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span>
                <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>

    <span class="c1"># Add labels and grid</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/gof_6.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># KL Divergence</span>
    <span class="n">normalized_obs</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                                                                <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">normalized_est</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
                                                                <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">kl_divergence</span> <span class="o">=</span> <span class="n">normalized_obs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">entropy</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">normalized_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Sort before plotting</span>
    <span class="n">kl_divergence_df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">kl_divergence_df</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kl_divergence</span><span class="o">.</span><span class="n">values</span>
    <span class="n">kl_divergence_by_gene</span> <span class="o">=</span> <span class="n">kl_divergence_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;GeneID&#39;</span><span class="p">)[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Ensure distinct values for index (GeneID) on x-axis</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">kl_divergence_by_gene</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">kl_divergence_by_gene</span><span class="o">.</span><span class="n">values</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$\bar</span><span class="si">{x}</span><span class="s2">$&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>  <span class="c1"># Rotate and align x-axis labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Entropy&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/kld.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.important_connections" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">important_connections</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Extracts the top N most important connections based on their absolute values
and saves them to a CSV file.</p>
<p>:param output_dir: str
    The directory where the CSV file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the connections with columns 'Source', 'Target', and 'Value'.
:param top_n: int, optional
    The number of top connections to extract (default is 20).</p>
<p>The function sorts the connections by their absolute values in descending order,
selects the top N connections, and saves them to a CSV file named 'top_connections.csv'
in the specified output directory.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">important_connections</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts the top N most important connections based on their absolute values</span>
<span class="sd">    and saves them to a CSV file.</span>

<span class="sd">    :param output_dir: str</span>
<span class="sd">        The directory where the CSV file will be saved.</span>
<span class="sd">    :param data: pd.DataFrame</span>
<span class="sd">        A DataFrame containing the connections with columns &#39;Source&#39;, &#39;Target&#39;, and &#39;Value&#39;.</span>
<span class="sd">    :param top_n: int, optional</span>
<span class="sd">        The number of top connections to extract (default is 20).</span>

<span class="sd">    The function sorts the connections by their absolute values in descending order,</span>
<span class="sd">    selects the top N connections, and saves them to a CSV file named &#39;top_connections.csv&#39;</span>
<span class="sd">    in the specified output directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sorted_edges</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Value&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="nb">abs</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span>
    <span class="n">important_connections</span> <span class="o">=</span> <span class="n">sorted_edges</span><span class="p">[[</span><span class="s2">&quot;Source&quot;</span><span class="p">,</span> <span class="s2">&quot;Target&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">]]</span>
    <span class="n">important_connections</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s2">/top_connections.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.perform_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">perform_pca</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Perform PCA analysis on the given DataFrame.
The DataFrame should contain a 'Value' column for PCA analysis.
The function returns a DataFrame with PCA results and additional columns for type and gene/psite information.</p>
<p>:param df: DataFrame containing the data for PCA analysis.
:return: DataFrame with PCA results and additional columns.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">perform_pca</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform PCA analysis on the given DataFrame.</span>
<span class="sd">    The DataFrame should contain a &#39;Value&#39; column for PCA analysis.</span>
<span class="sd">    The function returns a DataFrame with PCA results and additional columns for type and gene/psite information.</span>

<span class="sd">    :param df: DataFrame containing the data for PCA analysis.</span>
<span class="sd">    :return: DataFrame with PCA results and additional columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">numeric_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;Value&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
    <span class="n">scaled_data</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">numeric_df</span><span class="p">)</span>

    <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">pca_result</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">)</span>

    <span class="n">pca_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pca_result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PCA&#39;</span><span class="p">])</span>
    <span class="n">result_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">[[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]],</span> <span class="n">pca_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Alpha&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
    <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;Gene_Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;_(.*?)_|_(.*)&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">combine_first</span><span class="p">(</span>
        <span class="n">result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;_(.*?)_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">result_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene_Psite&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.perform_tsne" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">perform_tsne</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Perform t-SNE analysis on the given scaled data.
The function returns a DataFrame with t-SNE results and additional columns for type and gene/psite information.</p>
<p>:param scaled_data:
:param df:</p>
<p>:return:
- pd.DataFrame: DataFrame with t-SNE results and additional columns.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">perform_tsne</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">,</span> <span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform t-SNE analysis on the given scaled data.</span>
<span class="sd">    The function returns a DataFrame with t-SNE results and additional columns for type and gene/psite information.</span>

<span class="sd">    :param scaled_data:</span>
<span class="sd">    :param df:</span>

<span class="sd">    :return:</span>
<span class="sd">    - pd.DataFrame: DataFrame with t-SNE results and additional columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tsne</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
    <span class="n">tsne_result</span> <span class="o">=</span> <span class="n">tsne</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">)</span>

    <span class="n">tsne_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tsne_result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;tSNE&#39;</span><span class="p">])</span>
    <span class="n">tsne_result_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">[[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]],</span> <span class="n">tsne_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">tsne_result_df</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tsne_result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Alpha&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
    <span class="n">tsne_result_df</span><span class="p">[</span><span class="s1">&#39;Gene_Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tsne_result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;_(.*?)_|_(.*)&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">combine_first</span><span class="p">(</span>
        <span class="n">tsne_result_df</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;_(.*?)_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">tsne_result_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gene_Psite&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">(</span><span class="n">result_df_sorted</span><span class="p">,</span> <span class="n">y_axis_column</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Plot PCA or t-SNE results for each gene/psite.
The function creates scatter plots with different markers for alpha and beta parameters,
and adds labels for each point.
The function also adjusts text labels to avoid overlap using the adjustText library.</p>
<p>:param result_df_sorted: DataFrame containing PCA or t-SNE results.
:param y_axis_column: Column name for the y-axis values in the plot.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_pca</span><span class="p">(</span><span class="n">result_df_sorted</span><span class="p">,</span> <span class="n">y_axis_column</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot PCA or t-SNE results for each gene/psite.</span>
<span class="sd">    The function creates scatter plots with different markers for alpha and beta parameters,</span>
<span class="sd">    and adds labels for each point.</span>
<span class="sd">    The function also adjusts text labels to avoid overlap using the adjustText library.</span>

<span class="sd">    :param result_df_sorted: DataFrame containing PCA or t-SNE results.</span>
<span class="sd">    :param y_axis_column: Column name for the y-axis values in the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">gene_psite</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">result_df_sorted</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Gene_Psite&#39;</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">alpha_marker</span> <span class="o">=</span> <span class="s1">&#39;^&#39;</span>
        <span class="n">beta_marker</span> <span class="o">=</span> <span class="s1">&#39;v&#39;</span>

        <span class="n">texts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">param_type</span><span class="p">,</span> <span class="n">marker</span> <span class="ow">in</span> <span class="p">[(</span><span class="s1">&#39;Alpha&#39;</span><span class="p">,</span> <span class="n">alpha_marker</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;Beta&#39;</span><span class="p">,</span> <span class="n">beta_marker</span><span class="p">)]:</span>
            <span class="n">subset</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="n">group</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">param_type</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">subset</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">subset</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">],</span> <span class="n">subset</span><span class="p">[</span><span class="n">y_axis_column</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">param_type</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">subset</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">param_type</span> <span class="o">==</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">:</span>
                        <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">])</span>
                    <span class="n">texts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="n">y_axis_column</span><span class="p">],</span> <span class="n">label</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">))</span>

        <span class="n">adjustText</span><span class="o">.</span><span class="n">adjust_text</span><span class="p">(</span>
            <span class="n">texts</span><span class="p">,</span> <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">shrinkA</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">shrinkB</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene_psite</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="n">y_axis_column</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Type&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">y_axis_column</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">gene_psite</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">reshape_alpha_beta</span><span class="p">(</span><span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to reshape alpha and beta values for plotting.
It renames columns and creates a new 'Parameter' column for each type of value.</p>
<p>:param alpha_values:
:param beta_values:</p>
<p>:returns:
- pd.DataFrame: Reshaped DataFrame containing 'GeneID', 'Value', and 'Parameter' columns.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/fitanalysis/helpers/postfit.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">reshape_alpha_beta</span><span class="p">(</span><span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to reshape alpha and beta values for plotting.</span>
<span class="sd">    It renames columns and creates a new &#39;Parameter&#39; column for each type of value.</span>

<span class="sd">    :param alpha_values:</span>
<span class="sd">    :param beta_values:</span>

<span class="sd">    :returns:</span>
<span class="sd">    - pd.DataFrame: Reshaped DataFrame containing &#39;GeneID&#39;, &#39;Value&#39;, and &#39;Parameter&#39; columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="n">alpha_values_reshaped</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">:</span> <span class="s1">&#39;Value&#39;</span><span class="p">})</span>
    <span class="n">alpha_values_reshaped</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> \
                                         <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span>

    <span class="n">beta_values_reshaped</span> <span class="o">=</span> <span class="n">beta_values</span><span class="p">[[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">,</span> <span class="s1">&#39;Beta&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Beta&#39;</span><span class="p">:</span> <span class="s1">&#39;Value&#39;</span><span class="p">})</span>
    <span class="n">beta_values_reshaped</span><span class="p">[</span><span class="s1">&#39;Parameter&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">alpha_values_reshaped</span><span class="p">,</span> <span class="n">beta_values_reshaped</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.config.constants"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parses command-line arguments for the optimization script.
This function uses argparse to handle various parameters related to the optimization process.
The parameters include bounds for the optimization, loss function types, estimation of missing kinases,
scaling methods for time-series data, and the optimization method to be used.
The function returns a tuple containing the parsed arguments.</p>
<p>:return: A tuple containing the parsed arguments.
    - lower_bound (float): Lower bound for the optimization.
    - upper_bound (float): Upper bound for the optimization.
    - loss_type (str): Type of loss function to use.
    - estimate_missing (bool): Whether to estimate missing kinase-psite values.
    - scaling_method (str): Method for scaling time-series data.
    - split_point (int): Split point for temporal scaling.
    - segment_points (list of int): Segment points for segmented scaling.
    - method (str): Optimization method to use.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/config/constants.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses command-line arguments for the optimization script.</span>
<span class="sd">    This function uses argparse to handle various parameters related to the optimization process.</span>
<span class="sd">    The parameters include bounds for the optimization, loss function types, estimation of missing kinases,</span>
<span class="sd">    scaling methods for time-series data, and the optimization method to be used.</span>
<span class="sd">    The function returns a tuple containing the parsed arguments.</span>

<span class="sd">    :return: A tuple containing the parsed arguments.</span>
<span class="sd">        - lower_bound (float): Lower bound for the optimization.</span>
<span class="sd">        - upper_bound (float): Upper bound for the optimization.</span>
<span class="sd">        - loss_type (str): Type of loss function to use.</span>
<span class="sd">        - estimate_missing (bool): Whether to estimate missing kinase-psite values.</span>
<span class="sd">        - scaling_method (str): Method for scaling time-series data.</span>
<span class="sd">        - split_point (int): Split point for temporal scaling.</span>
<span class="sd">        - segment_points (list of int): Segment points for segmented scaling.</span>
<span class="sd">        - method (str): Optimization method to use.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;PhosKinTime - SLSQP/TRUST-CONSTR Kinase Phosphorylation Optimization Problem prior to ODE Modelling.&quot;</span>
    <span class="p">)</span>

    <span class="c1"># lower_bound and upper_bound are the bounds for the optimization.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--lower_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Lower Beta bound.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--upper_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Upper Beta bound.&quot;</span><span class="p">)</span>

    <span class="c1"># loss_type is the type of loss function to use.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--loss_type&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;base&quot;</span><span class="p">,</span> <span class="s2">&quot;weighted&quot;</span><span class="p">,</span> <span class="s2">&quot;softl1&quot;</span><span class="p">,</span> <span class="s2">&quot;cauchy&quot;</span><span class="p">,</span> <span class="s2">&quot;arctan&quot;</span><span class="p">],</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;weighted&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Loss function to use.&quot;</span><span class="p">)</span>

    <span class="c1"># estimate_missing_kinases indicates whether to estimate missing kinase-psite values.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--estimate_missing_kinases&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;yes&quot;</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;yes&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Estimate missing kinase-psite values?&quot;</span><span class="p">)</span>

    <span class="c1"># scaling_method is the method for scaling time-series data.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--scaling_method&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;min_max&quot;</span><span class="p">,</span> <span class="s2">&quot;log&quot;</span><span class="p">,</span> <span class="s2">&quot;temporal&quot;</span><span class="p">,</span> <span class="s2">&quot;segmented&quot;</span><span class="p">,</span> <span class="s2">&quot;slope&quot;</span><span class="p">,</span> <span class="s2">&quot;cumulative&quot;</span><span class="p">,</span> <span class="s2">&quot;none&quot;</span><span class="p">],</span>
                        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Scaling method for time-series data.&quot;</span><span class="p">)</span>

    <span class="c1"># split_point is the split point for temporal scaling.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--split_point&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Split point for temporal scaling.&quot;</span><span class="p">)</span>

    <span class="c1"># segment_points is a list of segment points for segmented scaling.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--segment_points&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;0,3,6,9,14&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Comma-separated segment points for segmented scaling.&quot;</span><span class="p">)</span>

    <span class="c1"># method is the optimization method to use.</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--method&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;slsqp&quot;</span><span class="p">,</span> <span class="s2">&quot;trust-constr&quot;</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;slsqp&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Optimization method.&quot;</span><span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">estimate_missing</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">estimate_missing_kinases</span> <span class="o">==</span> <span class="s2">&quot;yes&quot;</span>
    <span class="n">seg_points</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">segment_points</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)))</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">scaling_method</span> <span class="o">==</span> <span class="s2">&quot;segmented&quot;</span> <span class="k">else</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">lower_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">upper_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">loss_type</span><span class="p">,</span> <span class="n">estimate_missing</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">scaling_method</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.config.logconf"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.exporter.plotout"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.
It generates a plot showing the autocorrelation values over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the autocorrelation of residuals for each psite of a gene.</span>
<span class="sd">    It generates a plot showing the autocorrelation values over time.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plot_acf</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">lags</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="n">alpha</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">(),</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Lags&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Autocorrelation&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_autocorrelation_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.plotout.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.
It generates a plot showing the cumulative residuals over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the cumulative residuals for each psite of a gene.</span>
<span class="sd">    It generates a plot showing the cumulative residuals over time.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">)</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]),</span>
                 <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                 <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cumulative Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_cumulative_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.plotout.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for a gene.
It generates two plots:
1. A full timepoints plot showing all timepoints.
2. A short timepoints plot showing only the first 7 timepoints.
The plots are saved as PNG files in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the observed and estimated phosphorylation levels for a gene.</span>
<span class="sd">    It generates two plots:</span>
<span class="sd">    1. A full timepoints plot showing all timepoints.</span>
<span class="sd">    2. A short timepoints plot showing only the first 7 timepoints.</span>
<span class="sd">    The plots are saved as PNG files in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get colors from Dark2 palette</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Dark2&quot;</span><span class="p">)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set1&quot;)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set2&quot;)</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Full timepoints plot</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation Level (FC)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">[</span><span class="mi">9</span><span class="p">:])</span>

    <span class="c1"># First 7 timepoints plot</span>
    <span class="n">short_timepoints</span> <span class="o">=</span> <span class="n">real_timepoints</span><span class="p">[:</span><span class="mi">7</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="c1"># axs[1].set_title(f&quot;{gene}&quot;)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_fit_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.plotout.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.
It generates a histogram showing the distribution of residuals.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot histograms of residuals for each psite of a gene.</span>
<span class="sd">    It generates a histogram showing the distribution of residuals.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">)</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                     <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_histogram_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span> 
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.plotout.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.
It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot QQ plots of residuals for each psite of a gene.</span>
<span class="sd">    It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">qqplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">line</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_qqplot_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.exporter.sheetutils"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to output the results of the optimization process.
It logs the optimized alpha and beta values, optimization summary,
error metrics, and generates plots for each gene.
It also writes the results to an Excel file with multiple sheets.</p>
<p>The sheets include:
- Alpha Values: Optimized alpha values for each gene and psite.
- Beta Values: Optimized beta values for each kinase and psite.
- Summary: Summary of the optimization process.
- Observed: Observed time-series data for each gene and psite.
- Estimated: Estimated time-series data for each gene and psite.
- Residuals: Residuals for each gene and psite.</p>
<p>:param P_initial:
:param P_init_dense:
:param P_estimated:
:param residuals:
:param alpha_values:
:param beta_values:
:param result:
:param mse:
:param rmse:
:param mae:
:param mape:
:param r_squared:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/sheetutils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">output_results</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span><span class="p">,</span>
                   <span class="n">result</span><span class="p">,</span> <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to output the results of the optimization process.</span>
<span class="sd">    It logs the optimized alpha and beta values, optimization summary,</span>
<span class="sd">    error metrics, and generates plots for each gene.</span>
<span class="sd">    It also writes the results to an Excel file with multiple sheets.</span>

<span class="sd">    The sheets include:</span>
<span class="sd">    - Alpha Values: Optimized alpha values for each gene and psite.</span>
<span class="sd">    - Beta Values: Optimized beta values for each kinase and psite.</span>
<span class="sd">    - Summary: Summary of the optimization process.</span>
<span class="sd">    - Observed: Observed time-series data for each gene and psite.</span>
<span class="sd">    - Estimated: Estimated time-series data for each gene and psite.</span>
<span class="sd">    - Residuals: Residuals for each gene and psite.</span>

<span class="sd">    :param P_initial:</span>
<span class="sd">    :param P_init_dense:</span>
<span class="sd">    :param P_estimated:</span>
<span class="sd">    :param residuals:</span>
<span class="sd">    :param alpha_values:</span>
<span class="sd">    :param beta_values:</span>
<span class="sd">    :param result:</span>
<span class="sd">    :param mse:</span>
<span class="sd">    :param rmse:</span>
<span class="sd">    :param mae:</span>
<span class="sd">    :param mape:</span>
<span class="sd">    :param r_squared:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Optimized Alpha values:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Protein </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">, Psite </span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Kinase </span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Optimized Beta values:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Kinase </span><span class="si">{</span><span class="n">kinase</span><span class="si">}</span><span class="s2">, Psite </span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;--- Optimization Summary ---&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Success: </span><span class="si">{</span><span class="s1">&#39;Success&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;Failure&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Message: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">message</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Iterations: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">nit</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Function Evaluations: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">nfev</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SSE: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">fun</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;--- Error Metrics ---&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MSE: </span><span class="si">{</span><span class="n">mse</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;RMSE: </span><span class="si">{</span><span class="n">rmse</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MAE: </span><span class="si">{</span><span class="n">mae</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MAPE: </span><span class="si">{</span><span class="n">mape</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;R-squared: </span><span class="si">{</span><span class="n">r_squared</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Build a genes_data dictionary from computed metrics.</span>
    <span class="n">genes_data</span> <span class="o">=</span> <span class="n">build_genes_data</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">P_estimated</span><span class="p">,</span> <span class="n">residuals</span><span class="p">)</span>

    <span class="c1"># For each gene, call the plotting functions.</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">genes_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">TIME_POINTS</span><span class="p">)</span>
        <span class="n">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">TIME_POINTS</span><span class="p">)</span>
        <span class="n">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">TIME_POINTS</span><span class="p">)</span>
        <span class="n">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">TIME_POINTS</span><span class="p">)</span>
        <span class="n">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">TIME_POINTS</span><span class="p">)</span>

    <span class="c1"># Write results to Excel.</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="n">OUT_FILE</span>
    <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;openpyxl&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
        <span class="n">alpha_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">kinases</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">kinase</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kinases</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">alpha_list</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;Alpha&#39;</span><span class="p">:</span> <span class="n">value</span><span class="p">})</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">alpha_list</span><span class="p">)</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Alpha Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">beta_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">beta_list</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">:</span> <span class="n">psite</span><span class="p">,</span> <span class="s1">&#39;Beta&#39;</span><span class="p">:</span> <span class="n">value</span><span class="p">})</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">beta_list</span><span class="p">)</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Beta Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">summary</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;Metric&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Success&quot;</span><span class="p">,</span> <span class="s2">&quot;Message&quot;</span><span class="p">,</span> <span class="s2">&quot;Iterations&quot;</span><span class="p">,</span> <span class="s2">&quot;Function Evaluations&quot;</span><span class="p">,</span> <span class="s2">&quot;SSE&quot;</span><span class="p">,</span> <span class="s2">&quot;MSE&quot;</span><span class="p">,</span> <span class="s2">&quot;RMSE&quot;</span><span class="p">,</span> <span class="s2">&quot;MAE&quot;</span><span class="p">,</span> <span class="s2">&quot;MAPE&quot;</span><span class="p">,</span>
                       <span class="s2">&quot;R-squared&quot;</span><span class="p">],</span>
            <span class="s1">&#39;Value&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="s2">&quot;Success&quot;</span> <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span> <span class="k">else</span> <span class="s2">&quot;Failure&quot;</span><span class="p">),</span> <span class="n">result</span><span class="o">.</span><span class="n">message</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">nit</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">nfev</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">fun</span><span class="p">,</span>
                      <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span><span class="p">]</span>
        <span class="p">}</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">summary</span><span class="p">)</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Summary&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">timepoints</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">P_init_dense</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>

        <span class="c1"># For Observed sheet: use &quot;GeneID&quot; as the column name.</span>
        <span class="n">observed_keys_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span>
        <span class="n">observed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">P_init_dense</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">timepoints</span><span class="p">)</span>
        <span class="n">observed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">observed_keys_df</span><span class="p">,</span> <span class="n">observed_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">observed_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Observed&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># For Estimated sheet: use &quot;gene&quot; as the column name.</span>
        <span class="n">estimated_keys_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span>
        <span class="n">estimated_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">P_estimated</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">timepoints</span><span class="p">)</span>
        <span class="n">estimated_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">estimated_keys_df</span><span class="p">,</span> <span class="n">estimated_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">estimated_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Estimated&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># For Residuals sheet: use &quot;gene&quot; as the column name.</span>
        <span class="n">residuals_keys_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span>
        <span class="n">residuals_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">timepoints</span><span class="p">)</span>
        <span class="n">residuals_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">residuals_keys_df</span><span class="p">,</span> <span class="n">residuals_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">residuals_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Residuals&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Optimization results saved for ODE modelling.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.
It generates a plot showing the autocorrelation values over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_autocorrelation_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the autocorrelation of residuals for each psite of a gene.</span>
<span class="sd">    It generates a plot showing the autocorrelation values over time.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plot_acf</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">lags</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="n">alpha</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">(),</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Lags&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Autocorrelation&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_autocorrelation_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.
It generates a plot showing the cumulative residuals over time.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cumulative_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the cumulative residuals for each psite of a gene.</span>
<span class="sd">    It generates a plot showing the cumulative residuals over time.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">)</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]),</span>
                 <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                 <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cumulative Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_cumulative_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for a gene.
It generates two plots:
1. A full timepoints plot showing all timepoints.
2. A short timepoints plot showing only the first 7 timepoints.
The plots are saved as PNG files in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_fits_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the observed and estimated phosphorylation levels for a gene.</span>
<span class="sd">    It generates two plots:</span>
<span class="sd">    1. A full timepoints plot showing all timepoints.</span>
<span class="sd">    2. A short timepoints plot showing only the first 7 timepoints.</span>
<span class="sd">    The plots are saved as PNG files in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the observed and estimated phosphorylation levels for the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get colors from Dark2 palette</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Dark2&quot;</span><span class="p">)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set1&quot;)</span>
    <span class="c1"># cmap = mpl.cm.get_cmap(&quot;Set2&quot;)</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Full timepoints plot</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation Level (FC)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">real_timepoints</span><span class="p">[</span><span class="mi">9</span><span class="p">:])</span>

    <span class="c1"># First 7 timepoints plot</span>
    <span class="n">short_timepoints</span> <span class="o">=</span> <span class="n">real_timepoints</span><span class="p">[:</span><span class="mi">7</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;observed&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;estimated&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">][:</span><span class="mi">7</span><span class="p">],</span>
                    <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="c1"># axs[1].set_title(f&quot;{gene}&quot;)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">short_timepoints</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_fit_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.
It generates a histogram showing the distribution of residuals.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_histogram_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot histograms of residuals for each psite of a gene.</span>
<span class="sd">    It generates a histogram showing the distribution of residuals.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">)</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">%</span> <span class="mi">20</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]))]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                     <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Residuals&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_histogram_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span> 
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.
It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.
The plot is saved as a PNG file in the specified output directory.</p>
<p>Parameters:
gene (str): The name of the gene.
gene_data (dict): A dictionary containing the residuals for each psite of the gene.
real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_qqplot_residuals</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">gene_data</span><span class="p">,</span> <span class="n">real_timepoints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot QQ plots of residuals for each psite of a gene.</span>
<span class="sd">    It generates a QQ plot showing the quantiles of the residuals against the quantiles of a normal distribution.</span>
<span class="sd">    The plot is saved as a PNG file in the specified output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene (str): The name of the gene.</span>
<span class="sd">    gene_data (dict): A dictionary containing the residuals for each psite of the gene.</span>
<span class="sd">    real_timepoints (list): A list of timepoints corresponding to the observed and estimated data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;psites&quot;</span><span class="p">]):</span>
        <span class="n">qqplot</span><span class="p">(</span><span class="n">gene_data</span><span class="p">[</span><span class="s2">&quot;residuals&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">line</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">())</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_qqplot_residuals_.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.objfn.minfn"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.opt.optrun"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">obj_fun</span><span class="p">,</span> <span class="n">params_initial</span><span class="p">,</span> <span class="n">opt_method</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">constraints</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Run optimization using the specified method.</p>
<p>:param obj_fun:
:param params_initial:
:param opt_method:
:param bounds:
:param constraints:
:return: result, optimized parameters</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_optimization</span><span class="p">(</span><span class="n">obj_fun</span><span class="p">,</span> <span class="n">params_initial</span><span class="p">,</span> <span class="n">opt_method</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">constraints</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run optimization using the specified method.</span>

<span class="sd">    :param obj_fun:</span>
<span class="sd">    :param params_initial:</span>
<span class="sd">    :param opt_method:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param constraints:</span>
<span class="sd">    :return: result, optimized parameters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span><span class="n">obj_fun</span><span class="p">,</span> <span class="n">params_initial</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">opt_method</span><span class="p">,</span>
                      <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span> <span class="n">constraints</span><span class="o">=</span><span class="n">constraints</span><span class="p">,</span>
                      <span class="n">options</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;maxiter&#39;</span><span class="p">:</span> <span class="mi">20000</span><span class="p">,</span> <span class="s1">&#39;verbose&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}</span> <span class="k">if</span> <span class="n">opt_method</span> <span class="o">==</span> <span class="s2">&quot;trust-constr&quot;</span> <span class="k">else</span> <span class="p">{</span><span class="s1">&#39;maxiter&#39;</span><span class="p">:</span> <span class="mi">20000</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">result</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.optcon.construct"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.utils.iodata"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Apply scaling to the specified columns of a DataFrame based on the given method.
The scaling methods include:
- 'min_max': Min-Max scaling
- 'log': Logarithmic scaling
- 'temporal': Temporal scaling (two segments)
- 'segmented': Segmented scaling (multiple segments)
- 'slope': Slope scaling
- 'cumulative': Cumulative scaling</p>
<p>:param df:
:param cols:
:param method:
:param split_point:
:param seg_points:
:return: df</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">apply_scaling</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Apply scaling to the specified columns of a DataFrame based on the given method.</span>
<span class="sd">    The scaling methods include:</span>
<span class="sd">    - &#39;min_max&#39;: Min-Max scaling</span>
<span class="sd">    - &#39;log&#39;: Logarithmic scaling</span>
<span class="sd">    - &#39;temporal&#39;: Temporal scaling (two segments)</span>
<span class="sd">    - &#39;segmented&#39;: Segmented scaling (multiple segments)</span>
<span class="sd">    - &#39;slope&#39;: Slope scaling</span>
<span class="sd">    - &#39;cumulative&#39;: Cumulative scaling</span>

<span class="sd">    :param df:</span>
<span class="sd">    :param cols:</span>
<span class="sd">    :param method:</span>
<span class="sd">    :param split_point:</span>
<span class="sd">    :param seg_points:</span>
<span class="sd">    :return: df</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;min_max&#39;</span><span class="p">:</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">r</span><span class="p">:</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;log&#39;</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;temporal&#39;</span><span class="p">:</span>
        <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[:</span><span class="n">split_point</span><span class="p">],</span> <span class="n">cols</span><span class="p">[</span><span class="n">split_point</span><span class="p">:]</span>
        <span class="n">scaler1</span><span class="p">,</span> <span class="n">scaler2</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">(),</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="n">first</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler1</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">first</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="n">second</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler2</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">second</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;segmented&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">seg_points</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Segment points must be provided.&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seg</span> <span class="ow">in</span> <span class="p">[</span><span class="n">cols</span><span class="p">[</span><span class="n">seg_points</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span><span class="n">seg_points</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seg_points</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)]:</span>
            <span class="n">df</span><span class="p">[</span><span class="n">seg</span><span class="p">]</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">seg</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;slope&#39;</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;cumulative&#39;</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;report.html&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a single global report HTML file from all gene folders inside the results directory.</span>

<span class="sd">    For each gene folder (e.g. &quot;ABL2&quot;), the report will include:</span>
<span class="sd">      - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.</span>
<span class="sd">      - Each plot is confined to a fixed size of 900px by 900px.</span>
<span class="sd">      - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</span>

<span class="sd">    Args:</span>
<span class="sd">        results_dir (str): Path to the root results directory.</span>
<span class="sd">        output_file (str): Name of the generated global report file (placed inside results_dir).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Gather gene folders (skip &quot;General&quot; and &quot;logs&quot;)</span>
    <span class="n">gene_folders</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">results_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="ow">and</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;General&quot;</span><span class="p">,</span> <span class="s2">&quot;logs&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Build HTML content with updated CSS for spacing.</span>
    <span class="n">html_parts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;meta charset=&#39;UTF-8&#39;&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;title&gt;Estimation Report&lt;/title&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;body { font-family: Arial, sans-serif; margin: 20px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h1 { color: #333; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h2 { color: #555; font-size: 1.8em; border-bottom: 1px solid #ccc; padding-bottom: 5px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h3 { color: #666; font-size: 1.4em; margin-top: 10px; margin-bottom: 10px; }&quot;</span><span class="p">,</span>
        <span class="c1"># /* CSS grid for plots: two per row, fixed size 500px x 500px, extra space between rows */</span>
        <span class="s2">&quot;.plot-container {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  display: grid;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  grid-template-columns: repeat(2, 500px);&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  column-gap: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  row-gap: 40px;&quot;</span><span class="p">,</span> <span class="c1"># /* extra vertical gap */</span>
        <span class="s2">&quot;  justify-content: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;.plot-item {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;img, iframe {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  object-fit: contain;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: none;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="c1"># /* Data tables: full width, one per row */</span>
        <span class="s2">&quot;.data-table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 50%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border-collapse: collapse;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-top: 10px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;th, td {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: 1px solid #ccc;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  padding: 8px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  text-align: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;body&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;h1&gt;Kinase Optimization Report&lt;/h1&gt;&quot;</span>
    <span class="p">]</span>

    <span class="c1"># For each gene folder, create a section in the report.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_folders</span><span class="p">):</span>
        <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;Protein Group: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>

        <span class="c1"># Create grid container for fixed-size plots.</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;div class=&quot;plot-container&quot;&gt;&#39;</span><span class="p">)</span>
        <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                    <span class="n">rel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;plot-item&quot;&gt;&lt;h3&gt;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;img src=&quot;</span><span class="si">{</span><span class="n">rel_path</span><span class="si">}</span><span class="s1">&quot; alt=&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&quot;&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;/div&gt;&#39;</span><span class="p">)</span>  <span class="c1"># End of plot container</span>

        <span class="c1"># Data tables: display XLSX or CSV files from the gene folder, one per row.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.xlsx&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                    <span class="n">table_html</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">border</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;</span><span class="si">{</span><span class="n">table_html</span><span class="si">}</span><span class="s1">&lt;/div&gt;&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;p&gt;Error reading </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&lt;/p&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>

    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span>

    <span class="c1"># Write the report into the results directory.</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">html_parts</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Formats a duration in seconds into a human-readable string.
- If less than 60 seconds, returns in seconds.
- If less than 3600 seconds, returns in minutes.
- If more than 3600 seconds, returns in hours.</p>
<p>:param seconds:
:return: Formatted string</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Formats a duration in seconds into a human-readable string.</span>
<span class="sd">    - If less than 60 seconds, returns in seconds.</span>
<span class="sd">    - If less than 3600 seconds, returns in minutes.</span>
<span class="sd">    - If more than 3600 seconds, returns in hours.</span>

<span class="sd">    :param seconds:</span>
<span class="sd">    :return: Formatted string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">60</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> sec&quot;</span>
    <span class="k">elif</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">3600</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> min&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hr&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load and scale the data from the specified input files.</p>
<p>:param estimate_missing:
:param scaling_method:
:param split_point:
:param seg_points:
:return: Time series data, interaction data, observed data</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_and_scale_data</span><span class="p">(</span><span class="n">estimate_missing</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load and scale the data from the specified input files.</span>

<span class="sd">    :param estimate_missing:</span>
<span class="sd">    :param scaling_method:</span>
<span class="sd">    :param split_point:</span>
<span class="sd">    :param seg_points:</span>
<span class="sd">    :return: Time series data, interaction data, observed data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">INPUT1</span><span class="p">)</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="p">[</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">notna</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span>
    <span class="n">time_series_cols</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">15</span><span class="p">)]</span>
    <span class="n">full_hgnc_df</span> <span class="o">=</span> <span class="n">apply_scaling</span><span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">time_series_cols</span><span class="p">,</span> <span class="n">scaling_method</span><span class="p">,</span> <span class="n">split_point</span><span class="p">,</span> <span class="n">seg_points</span><span class="p">)</span>
    <span class="n">interaction_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">INPUT2</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">estimate_missing</span><span class="p">:</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interaction_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span>
        <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">interaction_df</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="nb">all</span><span class="p">(</span><span class="n">kinase</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">full_hgnc_df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span> <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)))]</span>
        <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interaction_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">k</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)])</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="n">full_hgnc_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interaction_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">full_hgnc_df</span><span class="p">,</span> <span class="n">interaction_df</span><span class="p">,</span> <span class="n">observed</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to organize output files into protein-specific folders.
It moves files matching the pattern 'protein_name_*.{json,svg,png,html,csv,xlsx}'
into a folder named after the protein (e.g., 'ABL2') and moves all other files
into a 'General' folder within the same directory.</p>
<p>:param directories:</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to organize output files into protein-specific folders.</span>
<span class="sd">    It moves files matching the pattern &#39;protein_name_*.{json,svg,png,html,csv,xlsx}&#39;</span>
<span class="sd">    into a folder named after the protein (e.g., &#39;ABL2&#39;) and moves all other files</span>
<span class="sd">    into a &#39;General&#39; folder within the same directory.</span>

<span class="sd">    :param directories:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Za-z0-9]+)_.*\.(json|svg|png|html|csv|xlsx)$&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">directory</span> <span class="ow">in</span> <span class="n">directories</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">directory</span><span class="si">}</span><span class="s2">&#39; is not a valid directory. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Move files matching the protein pattern.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">protein_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">protein_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">protein</span><span class="p">)</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>

        <span class="c1"># After protein files have been moved, move remaining files to a &quot;General&quot; folder.</span>
        <span class="n">general_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="s2">&quot;General&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.local.utils.params"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">gene_alpha_starts</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">gene_kinase_idx</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">kinase_beta_starts</span><span class="p">,</span> <span class="n">kinase_beta_counts</span><span class="p">,</span> <span class="n">K_data</span><span class="p">,</span> <span class="n">K_indices</span><span class="p">,</span> <span class="n">K_indptr</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Computes the estimated series and various metrics based on the optimized parameters.</p>
<p>:param optimized_params:
:param P_init_dense:
:param t_max:
:param gene_alpha_starts:
:param gene_kinase_counts:
:param gene_kinase_idx:
:param total_alpha:
:param kinase_beta_starts:
:param kinase_beta_counts:
:param K_data:
:param K_indices:
:param K_indptr:
:return: Estimated series, residuals, MSE, RMSE, MAE, MAPE, R-squared</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_metrics</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">gene_alpha_starts</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span>
                    <span class="n">gene_kinase_idx</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">kinase_beta_starts</span><span class="p">,</span> <span class="n">kinase_beta_counts</span><span class="p">,</span>
                    <span class="n">K_data</span><span class="p">,</span> <span class="n">K_indices</span><span class="p">,</span> <span class="n">K_indptr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the estimated series and various metrics based on the optimized parameters.</span>

<span class="sd">    :param optimized_params:</span>
<span class="sd">    :param P_init_dense:</span>
<span class="sd">    :param t_max:</span>
<span class="sd">    :param gene_alpha_starts:</span>
<span class="sd">    :param gene_kinase_counts:</span>
<span class="sd">    :param gene_kinase_idx:</span>
<span class="sd">    :param total_alpha:</span>
<span class="sd">    :param kinase_beta_starts:</span>
<span class="sd">    :param kinase_beta_counts:</span>
<span class="sd">    :param K_data:</span>
<span class="sd">    :param K_indices:</span>
<span class="sd">    :param K_indptr:</span>
<span class="sd">    :return: Estimated series, residuals, MSE, RMSE, MAE, MAPE, R-squared</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">P_est</span> <span class="o">=</span> <span class="n">estimated_series</span><span class="p">(</span><span class="n">optimized_params</span><span class="p">,</span> <span class="n">t_max</span><span class="p">,</span> <span class="n">P_init_dense</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                 <span class="n">gene_alpha_starts</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">gene_kinase_idx</span><span class="p">,</span>
                                 <span class="n">total_alpha</span><span class="p">,</span> <span class="n">kinase_beta_starts</span><span class="p">,</span> <span class="n">kinase_beta_counts</span><span class="p">,</span>
                                 <span class="n">K_data</span><span class="p">,</span> <span class="n">K_indices</span><span class="p">,</span> <span class="n">K_indptr</span><span class="p">)</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="n">P_init_dense</span> <span class="o">-</span> <span class="n">P_est</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">residuals</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">P_init_dense</span><span class="o">.</span><span class="n">size</span>
    <span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span><span class="p">))</span>
    <span class="n">mape</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">residuals</span> <span class="o">/</span> <span class="p">(</span><span class="n">P_init_dense</span> <span class="o">+</span> <span class="mf">1e-12</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">r_squared</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">residuals</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">P_init_dense</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">P_init_dense</span><span class="p">))</span><span class="o">**</span><span class="mi">2</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">P_est</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">mse</span><span class="p">,</span> <span class="n">rmse</span><span class="p">,</span> <span class="n">mae</span><span class="p">,</span> <span class="n">mape</span><span class="p">,</span> <span class="n">r_squared</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.local.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">unique_kinases</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Extracts the alpha and beta parameters from the optimized parameters.</p>
<p>:param P_initial:
:param gene_kinase_counts:
:param total_alpha:
:param unique_kinases:
:param K_index:
:param optimized_params:
:return: Alpha and beta values as dictionaries</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/local/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">extract_parameters</span><span class="p">(</span><span class="n">P_initial</span><span class="p">,</span> <span class="n">gene_kinase_counts</span><span class="p">,</span> <span class="n">total_alpha</span><span class="p">,</span> <span class="n">unique_kinases</span><span class="p">,</span> <span class="n">K_index</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts the alpha and beta parameters from the optimized parameters.</span>

<span class="sd">    :param P_initial:</span>
<span class="sd">    :param gene_kinase_counts:</span>
<span class="sd">    :param total_alpha:</span>
<span class="sd">    :param unique_kinases:</span>
<span class="sd">    :param K_index:</span>
<span class="sd">    :param optimized_params:</span>
<span class="sd">    :return: Alpha and beta values as dictionaries</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">alpha_start</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">P_initial</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">gene_kinase_counts</span><span class="p">):</span>
        <span class="n">kinases</span> <span class="o">=</span> <span class="n">P_initial</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s1">&#39;Kinases&#39;</span><span class="p">]</span>
        <span class="n">alpha_values</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">kinases</span><span class="p">,</span> <span class="n">optimized_params</span><span class="p">[</span><span class="n">alpha_start</span><span class="p">:</span><span class="n">alpha_start</span><span class="o">+</span><span class="n">count</span><span class="p">]))</span>
        <span class="n">alpha_start</span> <span class="o">+=</span> <span class="n">count</span>
    <span class="n">beta_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">beta_start</span> <span class="o">=</span> <span class="n">total_alpha</span>
    <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">unique_kinases</span><span class="p">:</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">psite</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="n">K_index</span><span class="p">[</span><span class="n">kinase</span><span class="p">]:</span>
            <span class="n">beta_values</span><span class="p">[(</span><span class="n">kinase</span><span class="p">,</span> <span class="n">psite</span><span class="p">)]</span> <span class="o">=</span> <span class="n">optimized_params</span><span class="p">[</span><span class="n">beta_start</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">alpha_values</span><span class="p">,</span> <span class="n">beta_values</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.optimality.KKT"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.generate_latex_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_latex_table</span><span class="p">(</span><span class="n">summary_dict</span><span class="p">,</span> <span class="n">table_caption</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to generate a LaTeX table from a summary dictionary.
The table is formatted for use in a LaTeX document.</p>
<p>:param summary_dict: Dictionary containing summary data.
:param table_caption: Caption for the LaTeX table.
:param table: Optional table object to format.
:return: LaTeX table as a string.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_latex_table</span><span class="p">(</span><span class="n">summary_dict</span><span class="p">,</span> <span class="n">table_caption</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to generate a LaTeX table from a summary dictionary.</span>
<span class="sd">    The table is formatted for use in a LaTeX document.</span>

<span class="sd">    :param summary_dict: Dictionary containing summary data.</span>
<span class="sd">    :param table_caption: Caption for the LaTeX table.</span>
<span class="sd">    :param table: Optional table object to format.</span>
<span class="sd">    :return: LaTeX table as a string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">latex_table</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n\\</span><span class="s2">begin</span><span class="si">{table}</span><span class="s2">[H]</span><span class="se">\n\\</span><span class="s2">centering</span><span class="se">\n\\</span><span class="s2">begin</span><span class="si">{tabular}</span><span class="s2">{|l|c|}</span><span class="se">\\</span><span class="s2">hline</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">latex_table</span> <span class="o">+=</span> <span class="s2">&quot;Metric &amp; Value </span><span class="se">\\\\</span><span class="s2"> </span><span class="se">\\</span><span class="s2">hline</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">summary_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">latex_table</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> &amp; </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2"> </span><span class="se">\\\\</span><span class="s2"> </span><span class="se">\\</span><span class="s2">hline</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">latex_table</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">end</span><span class="si">{tabular}</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">latex_table</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">caption</span><span class="se">{{</span><span class="si">{</span><span class="n">table_caption</span><span class="si">}</span><span class="se">}}\n\\</span><span class="s2">end</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">return</span> <span class="n">latex_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.plot_constraint_violations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_constraint_violations</span><span class="p">(</span><span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot constraint violations for alpha and beta values.
It creates a stacked bar plot showing the violations for each protein.
The top 5 proteins with the highest violations are highlighted in red.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>alpha_violations</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series containing alpha constraint violations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_violations</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series containing beta constraint violations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_constraint_violations</span><span class="p">(</span><span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot constraint violations for alpha and beta values.</span>
<span class="sd">    It creates a stacked bar plot showing the violations for each protein.</span>
<span class="sd">    The top 5 proteins with the highest violations are highlighted in red.</span>

<span class="sd">    Args:</span>
<span class="sd">        alpha_violations (pd.Series): Series containing alpha constraint violations.</span>
<span class="sd">        beta_violations (pd.Series): Series containing beta constraint violations.</span>
<span class="sd">        out_dir (str): Directory to save the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Group and combine violations</span>
    <span class="n">alpha_violations_abs</span> <span class="o">=</span> <span class="n">alpha_violations</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">beta_violations_abs</span> <span class="o">=</span> <span class="n">beta_violations</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">alpha_violations_abs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s2">&quot;Alpha Violations&quot;</span><span class="p">:</span> <span class="n">alpha_violations_abs</span><span class="p">,</span>
        <span class="s2">&quot;Beta Violations&quot;</span><span class="p">:</span> <span class="n">beta_violations_abs</span>
    <span class="p">})</span>
    <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Total Violations&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Total Violations&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">top_proteins</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">bar_alpha</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">combined</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">combined</span><span class="p">[</span><span class="s2">&quot;Alpha Violations&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$\alpha$&#39;</span><span class="p">)</span>
    <span class="n">bar_beta</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">combined</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">combined</span><span class="p">[</span><span class="s2">&quot;Beta Violations&quot;</span><span class="p">],</span> <span class="n">bottom</span><span class="o">=</span><span class="n">combined</span><span class="p">[</span><span class="s2">&quot;Alpha Violations&quot;</span><span class="p">],</span>
                       <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightgreen&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$\beta$&#39;</span><span class="p">)</span>

    <span class="c1"># Highlight top violations in red</span>
    <span class="k">for</span> <span class="n">bar</span> <span class="ow">in</span> <span class="n">bar_alpha</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">bar</span><span class="o">.</span><span class="n">get_x</span><span class="p">()</span> <span class="ow">in</span> <span class="n">top_proteins</span><span class="p">:</span>
            <span class="n">bar</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">bar</span> <span class="ow">in</span> <span class="n">bar_beta</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">bar</span><span class="o">.</span><span class="n">get_x</span><span class="p">()</span> <span class="ow">in</span> <span class="n">top_proteins</span><span class="p">:</span>
            <span class="n">bar</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Proteins&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Constraint Violations&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;constraint_violations.png&quot;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.plot_sensitivity_analysis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_sensitivity_analysis</span><span class="p">(</span><span class="n">sensitivity_analysis</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to plot sensitivity analysis results.
It creates a horizontal bar plot showing the mean, max, and min sensitivity for each protein.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sensitivity_analysis</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing sensitivity analysis results.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_sensitivity_analysis</span><span class="p">(</span><span class="n">sensitivity_analysis</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot sensitivity analysis results.</span>
<span class="sd">    It creates a horizontal bar plot showing the mean, max, and min sensitivity for each protein.</span>

<span class="sd">    Args:</span>
<span class="sd">        sensitivity_analysis (pd.DataFrame): DataFrame containing sensitivity analysis results.</span>
<span class="sd">        out_dir (str): Directory to save the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">sensitivity_analysis</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;GeneID&quot;</span><span class="p">)[[</span><span class="s2">&quot;Sensitivity Mean&quot;</span><span class="p">,</span> <span class="s2">&quot;Max Sensitivity&quot;</span><span class="p">,</span> <span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">summary</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Sensitivity Mean&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">bar_min</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">summary</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightgreen&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Min&#39;</span><span class="p">)</span>
    <span class="n">bar_mean</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">summary</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Sensitivity Mean&quot;</span><span class="p">],</span>
                        <span class="n">left</span><span class="o">=</span><span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Mean&#39;</span><span class="p">)</span>
    <span class="n">bar_max</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">summary</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Max Sensitivity&quot;</span><span class="p">],</span>
                       <span class="n">left</span><span class="o">=</span><span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;Sensitivity Mean&quot;</span><span class="p">],</span>
                       <span class="n">color</span><span class="o">=</span><span class="s1">&#39;coral&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Max&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Sensitivity&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Proteins&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;sensitivity.png&quot;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.post_optimization_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">post_optimization_results</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to process and visualize the results of the optimization.
It reads the results from an Excel file, processes the data,
and generates plots for constraint violations and sensitivity analysis.
It also prints the primal feasibility results and sensitivity summaries.
The results are returned as a dictionary.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing the processed results, including alpha and beta values,
  estimated and observed values, constraint violations, residuals summary,
  sensitivity summary, and high sensitivity sites.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">post_optimization_results</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to process and visualize the results of the optimization.</span>
<span class="sd">    It reads the results from an Excel file, processes the data,</span>
<span class="sd">    and generates plots for constraint violations and sensitivity analysis.</span>
<span class="sd">    It also prints the primal feasibility results and sensitivity summaries.</span>
<span class="sd">    The results are returned as a dictionary.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing the processed results, including alpha and beta values,</span>
<span class="sd">              estimated and observed values, constraint violations, residuals summary,</span>
<span class="sd">              sensitivity summary, and high sensitivity sites.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">process_excel_results</span><span class="p">()</span>
    <span class="c1"># Plot violation and sensitivity figures</span>
    <span class="n">plot_constraint_violations</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;alpha_constraint_violations&quot;</span><span class="p">],</span> <span class="n">results</span><span class="p">[</span><span class="s2">&quot;beta_constraint_violations&quot;</span><span class="p">],</span> <span class="n">OUT_DIR</span><span class="p">)</span>
    <span class="n">plot_sensitivity_analysis</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;sensitivity_analysis&quot;</span><span class="p">],</span> <span class="n">OUT_DIR</span><span class="p">)</span>

    <span class="c1"># For demonstration, call the printing functions (you may pass in appropriate summaries)</span>
    <span class="n">print_primal_feasibility_results</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;alpha_constraint_violations&quot;</span><span class="p">],</span>
                                     <span class="n">results</span><span class="p">[</span><span class="s2">&quot;alpha_constraint_violations&quot;</span><span class="p">],</span>
                                     <span class="n">results</span><span class="p">[</span><span class="s2">&quot;beta_constraint_violations&quot;</span><span class="p">])</span>
    <span class="n">print_sensitivity_and_active_constraints</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;sensitivity_summary&quot;</span><span class="p">],</span>
                                             <span class="n">results</span><span class="p">[</span><span class="s2">&quot;sensitivity_summary&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.print_primal_feasibility_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_primal_feasibility_results</span><span class="p">(</span><span class="n">primal_summary</span><span class="p">,</span> <span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Logs the primal feasibility summary and violation details.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">print_primal_feasibility_results</span><span class="p">(</span><span class="n">primal_summary</span><span class="p">,</span> <span class="n">alpha_violations</span><span class="p">,</span> <span class="n">beta_violations</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Logs the primal feasibility summary and violation details.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">logger_obj</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger_obj</span> <span class="o">=</span> <span class="n">logger</span>
    <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Primal Feasibility Summary:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">primal_summary</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Alpha Violations:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">alpha_violations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Beta Violations:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">beta_violations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_sensitivity_and_active_constraints</span><span class="p">(</span><span class="n">sensitivity_summary</span><span class="p">,</span> <span class="n">active_constraints_summary</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Logs the sensitivity summary and active constraints summary.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">print_sensitivity_and_active_constraints</span><span class="p">(</span><span class="n">sensitivity_summary</span><span class="p">,</span> <span class="n">active_constraints_summary</span><span class="p">,</span> <span class="n">logger_obj</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Logs the sensitivity summary and active constraints summary.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">logger_obj</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger_obj</span> <span class="o">=</span> <span class="n">logger</span>
    <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sensitivity Summary:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">sensitivity_summary</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Active Constraints Summary:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">active_constraints_summary</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger_obj</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="kinopt.optimality.KKT.process_excel_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_excel_results</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to process the Excel results file.
It reads the alpha and beta values, estimated and observed values,
validates normalization constraints, computes residuals and gradients,
and generates LaTeX tables for the residuals and sensitivity summaries.
It also performs sensitivity analysis and identifies high sensitivity sites.
The results are returned as a dictionary.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the Excel file containing results.</p>
              </div>
            </td>
            <td>
                  <code><span title="kinopt.evol.config.constants.OUT_FILE">OUT_FILE</span></code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    dict: Dictionary containing the processed results, including alpha and beta values,
          estimated and observed values, constraint violations, residuals summary,
          sensitivity summary, and high sensitivity sites.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/optimality/KKT.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_excel_results</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to process the Excel results file.</span>
<span class="sd">    It reads the alpha and beta values, estimated and observed values,</span>
<span class="sd">    validates normalization constraints, computes residuals and gradients,</span>
<span class="sd">    and generates LaTeX tables for the residuals and sensitivity summaries.</span>
<span class="sd">    It also performs sensitivity analysis and identifies high sensitivity sites.</span>
<span class="sd">    The results are returned as a dictionary.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path (str): Path to the Excel file containing results.</span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing the processed results, including alpha and beta values,</span>
<span class="sd">              estimated and observed values, constraint violations, residuals summary,</span>
<span class="sd">              sensitivity summary, and high sensitivity sites.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alpha_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Alpha Values&#39;</span><span class="p">)</span>
    <span class="n">beta_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Beta Values&#39;</span><span class="p">)</span>
    <span class="n">estimated_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">)</span>
    <span class="n">observed_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">)</span>

    <span class="c1"># Validate normalization constraints</span>
    <span class="n">alpha_sum</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">])[</span><span class="s1">&#39;Alpha&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">alpha_violations</span> <span class="o">=</span> <span class="n">alpha_sum</span><span class="p">[</span><span class="n">alpha_sum</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">beta_sum</span> <span class="o">=</span> <span class="n">beta_values</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Kinase&#39;</span><span class="p">])[</span><span class="s1">&#39;Beta&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">beta_violations</span> <span class="o">=</span> <span class="n">beta_sum</span><span class="p">[</span><span class="n">beta_sum</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Compute residuals and gradients</span>
    <span class="n">observed_matrix</span> <span class="o">=</span> <span class="n">observed_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">estimated_matrix</span> <span class="o">=</span> <span class="n">estimated_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="n">observed_matrix</span> <span class="o">-</span> <span class="n">estimated_matrix</span>
    <span class="n">gradients</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">gradient</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">residuals_summary</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Max Residual&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">residuals</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Min Residual&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">residuals</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Mean Residual&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">residuals</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Max Gradient&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">gradients</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Min Gradient&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">gradients</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Mean Gradient&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">gradients</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">sensitivity_summary</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Max Sensitivity&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
        <span class="s2">&quot;Mean Sensitivity&quot;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">latex_res_table</span> <span class="o">=</span> <span class="n">generate_latex_table</span><span class="p">(</span><span class="n">residuals_summary</span><span class="p">,</span> <span class="s2">&quot;Residual Summary&quot;</span><span class="p">)</span>
    <span class="n">latex_sens_table</span> <span class="o">=</span> <span class="n">generate_latex_table</span><span class="p">(</span><span class="n">sensitivity_summary</span><span class="p">,</span> <span class="s2">&quot;Sensitivity Summary&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">latex_res_table</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">latex_sens_table</span><span class="p">)</span>

    <span class="n">sensitivity_analysis</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span> <span class="n">observed_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
        <span class="s2">&quot;Psite&quot;</span><span class="p">:</span> <span class="n">observed_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="s2">&quot;Sensitivity Mean&quot;</span><span class="p">:</span> <span class="n">observed_matrix</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
        <span class="s2">&quot;Max Sensitivity&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
        <span class="s2">&quot;Min Sensitivity&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">observed_matrix</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="p">})</span>

    <span class="c1"># Threshold for high sensitivity sites</span>
    <span class="n">high_thresh</span> <span class="o">=</span> <span class="mf">0.75</span>
    <span class="n">high_sites_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">observed_matrix</span> <span class="o">&gt;=</span> <span class="n">high_thresh</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">high_sites</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">observed_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">observed_values</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">high_sites_idx</span>
    <span class="p">]</span>

    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;alpha_values&quot;</span><span class="p">:</span> <span class="n">alpha_values</span><span class="p">,</span>
        <span class="s2">&quot;beta_values&quot;</span><span class="p">:</span> <span class="n">beta_values</span><span class="p">,</span>
        <span class="s2">&quot;estimated_values&quot;</span><span class="p">:</span> <span class="n">estimated_values</span><span class="p">,</span>
        <span class="s2">&quot;observed_values&quot;</span><span class="p">:</span> <span class="n">observed_values</span><span class="p">,</span>
        <span class="s2">&quot;alpha_constraint_violations&quot;</span><span class="p">:</span> <span class="n">alpha_violations</span><span class="p">,</span>
        <span class="s2">&quot;beta_constraint_violations&quot;</span><span class="p">:</span> <span class="n">beta_violations</span><span class="p">,</span>
        <span class="s2">&quot;residuals_summary&quot;</span><span class="p">:</span> <span class="n">residuals_summary</span><span class="p">,</span>
        <span class="s2">&quot;sensitivity_summary&quot;</span><span class="p">:</span> <span class="n">sensitivity_summary</span><span class="p">,</span>
        <span class="s2">&quot;sensitivity_analysis&quot;</span><span class="p">:</span> <span class="n">sensitivity_analysis</span><span class="p">,</span>
        <span class="s2">&quot;high_sensitivity_sites&quot;</span><span class="p">:</span> <span class="n">high_sites</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="kinopt.powell.runpowell"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="kinopt.powell.runpowell.run_powell" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_powell</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Run the Powell optimization algorithm using Julia.
It sets the number of threads to half of the available threads
and logs the output in real time.</p>


            <details class="quote">
              <summary>Source code in <code>kinopt/powell/runpowell.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_powell</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run the Powell optimization algorithm using Julia.</span>
<span class="sd">    It sets the number of threads to half of the available threads</span>
<span class="sd">    and logs the output in real time.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Run the command to get the number of threads (as a string)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="s2">&quot;lscpu -p | grep -v &#39;^#&#39; | wc -l&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">num_threads_str</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="c1"># Convert to integer, compute half, and round it</span>
    <span class="n">num_threads_int</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">num_threads_str</span><span class="p">)</span>
    <span class="n">half_threads</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">num_threads_int</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Set up the environment variable (convert back to string)</span>
    <span class="n">env</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">env</span><span class="p">[</span><span class="s2">&quot;JULIA_NUM_THREADS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">half_threads</span><span class="p">)</span>

    <span class="c1"># Start the Julia script and log output in real time</span>
    <span class="n">process</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;julia&quot;</span><span class="p">,</span> <span class="s2">&quot;kinopt/powell/powell.jl&quot;</span><span class="p">],</span>
        <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
        <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span>
        <span class="n">env</span><span class="o">=</span><span class="n">env</span><span class="p">,</span>
        <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">bufsize</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>

    <span class="c1"># Read each line</span>
    <span class="k">with</span> <span class="n">process</span><span class="o">.</span><span class="n">stdout</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">iter</span><span class="p">(</span><span class="n">process</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">readline</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>

    <span class="n">process</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.diagram.helpers"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="models.diagram.helpers.create_distributive_diagram" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_distributive_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create a distributive phosphorylation diagram.
This model shows single-step phosphorylation and dephosphorylation transitions
for each site, along with production/degradation edges.</p>


            <details class="quote">
              <summary>Source code in <code>models/diagram/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_distributive_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a distributive phosphorylation diagram.</span>
<span class="sd">    This model shows single-step phosphorylation and dephosphorylation transitions</span>
<span class="sd">    for each site, along with production/degradation edges.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dot</span> <span class="o">=</span> <span class="n">Digraph</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s1">&#39;neato&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">rankdir</span><span class="o">=</span><span class="s1">&#39;LR&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;Distributive&quot;</span><span class="p">,</span> <span class="n">labelloc</span><span class="o">=</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s2">&quot;15&quot;</span><span class="p">,</span>
             <span class="n">fontname</span><span class="o">=</span><span class="s2">&quot;Helvetica&quot;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;graph&#39;</span><span class="p">,</span> <span class="n">bgcolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="s1">&#39;300&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;node&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;ellipse&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled,rounded&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;12&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;edge&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;10&#39;</span><span class="p">)</span>

    <span class="c1"># --- mRNA (R) and its null state ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;lightcoral&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Protein (P) and its null state ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_P&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;NULL_P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Phosphorylation sites ---</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">p_node</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="n">null_p_node</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;NULL_P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">p_node</span><span class="p">,</span> <span class="n">p_node</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;turquoise3&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">null_p_node</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">p_node</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;S</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">p_node</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">p_node</span><span class="p">,</span> <span class="n">null_p_node</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;D</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>

    <span class="n">dot</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">cleanup</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.diagram.helpers.create_random_diagram" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_random_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create a random phosphorylation diagram.
The diagram displays mRNA production/degradation, protein production,
phosphorylation transitions (addition) and dephosphorylation transitions,
as well as degradation edges for each state.</p>


            <details class="quote">
              <summary>Source code in <code>models/diagram/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_random_diagram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a random phosphorylation diagram.</span>
<span class="sd">    The diagram displays mRNA production/degradation, protein production,</span>
<span class="sd">    phosphorylation transitions (addition) and dephosphorylation transitions,</span>
<span class="sd">    as well as degradation edges for each state.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dot</span> <span class="o">=</span> <span class="n">Digraph</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s1">&#39;neato&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">rankdir</span><span class="o">=</span><span class="s1">&#39;LR&#39;</span><span class="p">)</span>
    <span class="c1"># Global graph attributes: title and overall styling.</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;Random&quot;</span><span class="p">,</span> <span class="n">labelloc</span><span class="o">=</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s2">&quot;15&quot;</span><span class="p">,</span>
             <span class="n">fontname</span><span class="o">=</span><span class="s2">&quot;Helvetica&quot;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;graph&#39;</span><span class="p">,</span> <span class="n">bgcolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="s1">&#39;300&#39;</span><span class="p">)</span>

    <span class="c1"># Global node and edge attributes for a modern, clean look.</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;node&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;ellipse&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled,rounded&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span>
             <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;12&#39;</span><span class="p">,</span> <span class="n">fixedsize</span><span class="o">=</span><span class="s1">&#39;false&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;edge&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;10&#39;</span><span class="p">)</span>

    <span class="c1"># --- mRNA (R) Production and Degradation ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;lightcoral&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Protein (P) and its phosphorylation states ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_P&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
    <span class="n">all_states</span> <span class="o">=</span> <span class="n">powerset</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span>

    <span class="c1"># Create nodes for each phosphorylation state with custom colors.</span>
    <span class="n">state_nodes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">all_states</span><span class="p">:</span>
        <span class="n">label</span> <span class="o">=</span> <span class="n">state_label</span><span class="p">(</span><span class="n">state</span><span class="p">)</span>
        <span class="n">state_nodes</span><span class="p">[</span><span class="n">state</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span>
        <span class="k">if</span> <span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;P&quot;</span><span class="p">:</span>
            <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;turquoise3&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="c1"># Create corresponding NULL node for degradation of this state.</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_&#39;</span> <span class="o">+</span> <span class="n">label</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>

    <span class="c1"># --- Production of protein P ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">state_nodes</span><span class="p">[</span><span class="nb">frozenset</span><span class="p">()],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Phosphorylation transitions (Addition) ---</span>
    <span class="n">all_sites_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sites</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">all_states</span><span class="p">:</span>
        <span class="n">current_label</span> <span class="o">=</span> <span class="n">state_nodes</span><span class="p">[</span><span class="n">state</span><span class="p">]</span>
        <span class="n">missing_sites</span> <span class="o">=</span> <span class="n">all_sites_set</span> <span class="o">-</span> <span class="n">state</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">missing_sites</span><span class="p">:</span>
            <span class="n">new_state</span> <span class="o">=</span> <span class="n">state</span><span class="o">.</span><span class="n">union</span><span class="p">({</span><span class="n">site</span><span class="p">})</span>
            <span class="n">new_label</span> <span class="o">=</span> <span class="n">state_nodes</span><span class="p">[</span><span class="n">new_state</span><span class="p">]</span>
            <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">current_label</span><span class="p">,</span> <span class="n">new_label</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;S&#39;</span> <span class="o">+</span> <span class="n">site</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span>
                     <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>

    <span class="c1"># --- Dephosphorylation transitions (Removal) ---</span>
    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">all_states</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">state</span><span class="p">:</span>  <span class="c1"># Only states with at least one phosphorylation can dephosphorylate.</span>
            <span class="n">current_label</span> <span class="o">=</span> <span class="n">state_nodes</span><span class="p">[</span><span class="n">state</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">state</span><span class="p">:</span>
                <span class="n">new_state</span> <span class="o">=</span> <span class="n">state</span><span class="o">.</span><span class="n">difference</span><span class="p">({</span><span class="n">site</span><span class="p">})</span>
                <span class="n">new_label</span> <span class="o">=</span> <span class="n">state_nodes</span><span class="p">[</span><span class="n">new_state</span><span class="p">]</span>
                <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">current_label</span><span class="p">,</span> <span class="n">new_label</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span>
                         <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>

    <span class="c1"># --- Degradation transitions for all protein states ---</span>
    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">all_states</span><span class="p">:</span>
        <span class="n">label</span> <span class="o">=</span> <span class="n">state_nodes</span><span class="p">[</span><span class="n">state</span><span class="p">]</span>
        <span class="n">deg_label</span> <span class="o">=</span> <span class="s1">&#39;D&#39;</span> <span class="k">if</span> <span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;P&quot;</span> <span class="k">else</span> <span class="s1">&#39;D&#39;</span> <span class="o">+</span> <span class="n">label</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="s1">&#39;NULL_&#39;</span> <span class="o">+</span> <span class="n">label</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">deg_label</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span>
                 <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>

    <span class="n">dot</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">cleanup</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.diagram.helpers.create_successive_model" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_successive_model</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create a successive phosphorylation diagram.
In this model, phosphorylation occurs in a sequential order from P to P1 to P2, etc.,
with corresponding dephosphorylation and degradation transitions.</p>


            <details class="quote">
              <summary>Source code in <code>models/diagram/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_successive_model</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a successive phosphorylation diagram.</span>
<span class="sd">    In this model, phosphorylation occurs in a sequential order from P to P1 to P2, etc.,</span>
<span class="sd">    with corresponding dephosphorylation and degradation transitions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dot</span> <span class="o">=</span> <span class="n">Digraph</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s1">&#39;neato&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">rankdir</span><span class="o">=</span><span class="s1">&#39;LR&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;Successive&quot;</span><span class="p">,</span> <span class="n">labelloc</span><span class="o">=</span><span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s2">&quot;15&quot;</span><span class="p">,</span>
             <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;graph&#39;</span><span class="p">,</span> <span class="n">bgcolor</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="s1">&#39;300&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;node&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;ellipse&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled,rounded&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;12&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">attr</span><span class="p">(</span><span class="s1">&#39;edge&#39;</span><span class="p">,</span> <span class="n">fontname</span><span class="o">=</span><span class="s1">&#39;Helvetica&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;10&#39;</span><span class="p">)</span>

    <span class="c1"># --- mRNA (R) Production and Degradation ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;lightcoral&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;NULL_R&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;forestgreen&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Protein (P) and its null state ---</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;NULL_P&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;dodgerblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>
    <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;NULL_P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;2&#39;</span><span class="p">)</span>

    <span class="c1"># --- Successive Phosphorylation States ---</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">state</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="n">null_state</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;NULL_P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">state</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;filled&#39;</span><span class="p">,</span> <span class="n">fillcolor</span><span class="o">=</span><span class="s1">&#39;turquoise3&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">null_state</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;plaintext&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">)</span>
    <span class="c1"># First phosphorylation: P -&gt; P1</span>
    <span class="k">if</span> <span class="n">num_sites</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;P1&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;S1&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
    <span class="c1"># Subsequent phosphorylations: P{i-1} -&gt; P{i}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;S</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;mediumvioletred&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
    <span class="c1"># Dephosphorylation (reverse reactions)</span>
    <span class="k">if</span> <span class="n">num_sites</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="s1">&#39;P1&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;slateblue&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>
    <span class="c1"># Degradation for phosphorylated states</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_sites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">dot</span><span class="o">.</span><span class="n">edge</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;NULL_P</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;D</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span> <span class="n">fontcolor</span><span class="o">=</span><span class="s1">&#39;dimgray&#39;</span><span class="p">,</span>
                 <span class="n">style</span><span class="o">=</span><span class="s1">&#39;dotted&#39;</span><span class="p">,</span> <span class="n">penwidth</span><span class="o">=</span><span class="s1">&#39;1.5&#39;</span><span class="p">)</span>

    <span class="n">dot</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">cleanup</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.diagram.helpers.powerset" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">powerset</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Return the list of all subsets (as frozensets) of the given iterable.
For example, for [1,2] returns:
  [frozenset(), frozenset({1}), frozenset({2}), frozenset({1,2})]</p>


            <details class="quote">
              <summary>Source code in <code>models/diagram/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">powerset</span><span class="p">(</span><span class="n">iterable</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return the list of all subsets (as frozensets) of the given iterable.</span>
<span class="sd">    For example, for [1,2] returns:</span>
<span class="sd">      [frozenset(), frozenset({1}), frozenset({2}), frozenset({1,2})]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">s</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span>
    <span class="n">all_subsets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">r</span><span class="p">):</span>
            <span class="n">all_subsets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">combo</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">all_subsets</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.diagram.helpers.state_label" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">state_label</span><span class="p">(</span><span class="n">state</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Convert a set of phosphorylation sites into a node label.
The unphosphorylated state (empty set) is labeled "P".
For nonempty states the label is "P" concatenated with the sorted site numbers.</p>


            <details class="quote">
              <summary>Source code in <code>models/diagram/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">state_label</span><span class="p">(</span><span class="n">state</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert a set of phosphorylation sites into a node label.</span>
<span class="sd">    The unphosphorylated state (empty set) is labeled &quot;P&quot;.</span>
<span class="sd">    For nonempty states the label is &quot;P&quot; concatenated with the sorted site numbers.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">state</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;P&quot;</span>
    <span class="n">sorted_sites</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">state</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
    <span class="k">return</span> <span class="s2">&quot;P&quot;</span> <span class="o">+</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sorted_sites</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.distmod"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="models.distmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>The core ODE system for the distributive phosphorylation model.</p>
<p>The system is defined by the following equations:</p>
<p>dR/dt = A - B * R
dP/dt = C * R - (D + sum(S_rates)) * P + sum(P_sites)
dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</p>
<p>where:</p>
<p>R: the concentration of the mRNA
P: the concentration of the protein
P_sites: the concentration of the phosphorylated sites
A: the rate of production of the mRNA
B: the rate of degradation of the mRNA
C: the rate of production of the protein
D: the rate of degradation of the protein
S_rates: the rates of phosphorylation of each site
D_rates: the rates of dephosphorylation of each site</p>
<p>:param y:
:param A:
:param B:
:param C:
:param D:
:param S_rates:
:param D_rates:
:return: Derivative of y</p>


            <details class="quote">
              <summary>Source code in <code>models/distmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The core ODE system for the distributive phosphorylation model.</span>

<span class="sd">    The system is defined by the following equations:</span>

<span class="sd">    dR/dt = A - B * R</span>
<span class="sd">    dP/dt = C * R - (D + sum(S_rates)) * P + sum(P_sites)</span>
<span class="sd">    dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</span>

<span class="sd">    where:</span>

<span class="sd">    R: the concentration of the mRNA</span>
<span class="sd">    P: the concentration of the protein</span>
<span class="sd">    P_sites: the concentration of the phosphorylated sites</span>
<span class="sd">    A: the rate of production of the mRNA</span>
<span class="sd">    B: the rate of degradation of the mRNA</span>
<span class="sd">    C: the rate of production of the protein</span>
<span class="sd">    D: the rate of degradation of the protein</span>
<span class="sd">    S_rates: the rates of phosphorylation of each site</span>
<span class="sd">    D_rates: the rates of dephosphorylation of each site</span>

<span class="sd">    :param y:</span>
<span class="sd">    :param A:</span>
<span class="sd">    :param B:</span>
<span class="sd">    :param C:</span>
<span class="sd">    :param D:</span>
<span class="sd">    :param S_rates:</span>
<span class="sd">    :param D_rates:</span>
<span class="sd">    :return: Derivative of y</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># y[0] is the concentration of the mRNA</span>
    <span class="n">R</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># y[1] is the concentration of the protein</span>
    <span class="n">P</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># Number of phosphorylation sites</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">S_rates</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Derivative of y</span>
    <span class="n">dydt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="c1"># dydt[0] is the rate of change of R</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
    <span class="c1"># S_rates</span>
    <span class="n">sum_S</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="c1"># sum_S is the sum of S_rates</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="c1"># S_rates[i] is the rate of phosphorylation of site i</span>
        <span class="n">sum_S</span> <span class="o">+=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># sum_P_sites is the sum of P sites</span>
    <span class="n">sum_P_sites</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="c1"># Loop over the number of phosphorylation sites</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="c1"># y[2:] are the concentrations of the phosphorylated sites</span>
        <span class="n">sum_P_sites</span> <span class="o">+=</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
    <span class="c1"># dydt[1] is the rate of change of P</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="p">(</span><span class="n">D</span> <span class="o">+</span> <span class="n">sum_S</span><span class="p">)</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="n">sum_P_sites</span>
    <span class="c1"># Loop over the number of phosphorylation sites</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span> 
        <span class="c1"># dydt[2 + i] is the rate of change of each P site</span>
        <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">dydt</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.distmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>The ODE system for the distributive phosphorylation model which calls the core ODE system.</p>
<p>:param y:
:param t:
:param params:
:param num_psites:
:return: ode_core(y, A, B, C, D, S_rates, D_rates)</p>


            <details class="quote">
              <summary>Source code in <code>models/distmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The ODE system for the distributive phosphorylation model which calls the core ODE system.</span>

<span class="sd">    :param y:</span>
<span class="sd">    :param t:</span>
<span class="sd">    :param params:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :return: ode_core(y, A, B, C, D, S_rates, D_rates)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Unpack the parameters</span>
    <span class="c1"># params[0] is A</span>
    <span class="c1"># params[1] is B</span>
    <span class="c1"># params[2] is C</span>
    <span class="c1"># params[3] is D</span>
    <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
    <span class="c1"># params[4 + i] is the S_rate for site i</span>
    <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">params</span><span class="p">[</span><span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
    <span class="c1"># params[4 + num_psites + i] is the D_rate for site i</span>
    <span class="n">D_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">params</span><span class="p">[</span><span class="mi">4</span> <span class="o">+</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
    <span class="k">return</span> <span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.distmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Solve the ODE system for the distributive phosphorylation model.</p>
<p>:param params:
:param init_cond:
:param num_psites:
:param t:
:return: solution of the ODE system, solution of phosphorylated sites</p>


            <details class="quote">
              <summary>Source code in <code>models/distmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Solve the ODE system for the distributive phosphorylation model.</span>

<span class="sd">    :param params:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param t:</span>
<span class="sd">    :return: solution of the ODE system, solution of phosphorylated sites</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Call the odeint function to solve the ODE system</span>
    <span class="n">sol</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">odeint</span><span class="p">(</span><span class="n">ode_system</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)))</span>
    <span class="c1"># Clip the solution to be non-negative</span>
    <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">sol</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">sol</span><span class="p">)</span>
    <span class="c1"># Normalize the solution if NORMALIZE_MODEL_OUTPUT is True</span>
    <span class="k">if</span> <span class="n">NORMALIZE_MODEL_OUTPUT</span><span class="p">:</span>
        <span class="c1"># Normalize the solution to the initial condition</span>
        <span class="n">norm_init</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">init_cond</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">sol</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
        <span class="c1"># Calculate the reciprocal of the norm_init</span>
        <span class="n">recip</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">norm_init</span>
        <span class="c1"># Normalize the solution by multiplying by the reciprocal of the norm_init</span>
        <span class="n">sol</span> <span class="o">*=</span> <span class="n">recip</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
    <span class="c1"># Extract the phosphorylated sites from the solution</span>
    <span class="n">P_fitted</span> <span class="o">=</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">T</span>
    <span class="c1"># Return the solution and the phosphorylated sites</span>
    <span class="k">return</span> <span class="n">sol</span><span class="p">,</span> <span class="n">P_fitted</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.randmod"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="models.randmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span><span class="p">,</span> <span class="o">*</span><span class="n">params</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>The ODE system for the random phosphorylation model. This function computes the derivatives of the
variables R, P, and X at each time step.</p>
<ul>
<li>R is the concentration of the receptor</li>
<li>P is the concentration of the protein</li>
<li>X is the concentration of the phosphorylated states</li>
<li>S is the phosphorylation rate for each site</li>
<li>D_deg is the degradation rate for each state</li>
<li>A, B, C, D are parameters for the system</li>
<li>num_sites is the number of phosphorylation sites</li>
<li>binary_states is a 2D array representing the binary states of the system</li>
<li>PHOSPHO_TARGET is a 2D array representing the phosphorylation targets</li>
<li>DEPHOSPHO_TARGET is a 2D array representing the dephosphorylation targets</li>
</ul>
<p>:param y:
    y[0] = R
    y[1] = P
    y[2:] = X
:param t:
:param A:
:param B:
:param C:
:param D:
:param num_sites:
:param binary_states:
:param PHOSPHO_TARGET:
:param DEPHOSPHO_TARGET:
:param params:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>models/randmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">,</span> <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span><span class="p">,</span> <span class="o">*</span><span class="n">params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The ODE system for the random phosphorylation model. This function computes the derivatives of the</span>
<span class="sd">    variables R, P, and X at each time step.</span>

<span class="sd">    - R is the concentration of the receptor</span>
<span class="sd">    - P is the concentration of the protein</span>
<span class="sd">    - X is the concentration of the phosphorylated states</span>
<span class="sd">    - S is the phosphorylation rate for each site</span>
<span class="sd">    - D_deg is the degradation rate for each state</span>
<span class="sd">    - A, B, C, D are parameters for the system</span>
<span class="sd">    - num_sites is the number of phosphorylation sites</span>
<span class="sd">    - binary_states is a 2D array representing the binary states of the system</span>
<span class="sd">    - PHOSPHO_TARGET is a 2D array representing the phosphorylation targets</span>
<span class="sd">    - DEPHOSPHO_TARGET is a 2D array representing the dephosphorylation targets</span>

<span class="sd">    :param y:</span>
<span class="sd">        y[0] = R</span>
<span class="sd">        y[1] = P</span>
<span class="sd">        y[2:] = X</span>
<span class="sd">    :param t:</span>
<span class="sd">    :param A:</span>
<span class="sd">    :param B:</span>
<span class="sd">    :param C:</span>
<span class="sd">    :param D:</span>
<span class="sd">    :param num_sites:</span>
<span class="sd">    :param binary_states:</span>
<span class="sd">    :param PHOSPHO_TARGET:</span>
<span class="sd">    :param DEPHOSPHO_TARGET:</span>
<span class="sd">    :param params:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Number of states</span>
    <span class="n">num_states</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">num_sites</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="c1"># Phosphorylation rates</span>
    <span class="n">S</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">num_sites</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
        <span class="c1"># Convert the j-th parameter to a phosphorylation rate</span>
        <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
    <span class="c1"># Dephosphorylation rates</span>
    <span class="n">D_deg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">num_states</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Convert the i-th parameter to a dephosphorylation rate</span>
        <span class="n">D_deg</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="n">num_sites</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
    <span class="c1"># Initialize the derivatives</span>
    <span class="n">R</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">P</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
    <span class="n">dR_dt</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
    <span class="n">sum_S</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
        <span class="n">sum_S</span> <span class="o">+=</span> <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
    <span class="c1"># Gain from phosphorylation</span>
    <span class="n">gain_1site</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="c1"># Loop over all states</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="n">cnt</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Count the number of phosphorylated sites in the state</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th site is phosphorylated in the state</span>
            <span class="n">cnt</span> <span class="o">+=</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
        <span class="c1"># If the state has only one phosphorylated site, add the gain to the total</span>
        <span class="k">if</span> <span class="n">cnt</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">gain_1site</span> <span class="o">+=</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># Protein dynamics</span>
    <span class="n">dP_dt</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">sum_S</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="n">gain_1site</span>
    <span class="c1"># Initialize the derivative of X</span>
    <span class="n">dX_dt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_states</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th site is not phosphorylated in the i-th state</span>
            <span class="k">if</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Get the phosphorylation target for the i-th state and j-th site</span>
                <span class="n">target</span> <span class="o">=</span> <span class="n">PHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
                <span class="c1"># If the target state is valid, add the phosphorylation rate to the derivative</span>
                <span class="k">if</span> <span class="n">target</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">dX_dt</span><span class="p">[</span><span class="n">target</span><span class="p">]</span> <span class="o">+=</span> <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Check if the i-th state is phosphorylated</span>
        <span class="n">loss</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th site is phosphorylated in the i-th state</span>
            <span class="k">if</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Add loss to the total</span>
                <span class="n">loss</span> <span class="o">+=</span> <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
        <span class="c1"># If the state has only one phosphorylated site, add the loss to the derivative</span>
        <span class="n">dX_dt</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-=</span> <span class="n">loss</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="n">cnt</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Count the number of phosphorylated sites in the state</span>
            <span class="n">cnt</span> <span class="o">+=</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
        <span class="c1"># Add the dephosphorylation rate to the derivative</span>
        <span class="n">dX_dt</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-=</span> <span class="n">cnt</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Check if the i-th state is phosphorylated</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th site is phosphorylated in the i-th state</span>
            <span class="k">if</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Get the dephosphorylation target for the i-th state and j-th site</span>
                <span class="n">lower</span> <span class="o">=</span> <span class="n">DEPHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
                <span class="c1"># If the target state is valid, add the dephosphorylation rate to the derivative of P</span>
                <span class="k">if</span> <span class="n">lower</span> <span class="o">==</span> <span class="o">-</span><span class="mi">2</span><span class="p">:</span>
                    <span class="n">dP_dt</span> <span class="o">+=</span> <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="c1"># If the target state is valid, add the dephosphorylation rate to the derivative of X</span>
                <span class="k">elif</span> <span class="n">lower</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">dX_dt</span><span class="p">[</span><span class="n">lower</span><span class="p">]</span> <span class="o">+=</span> <span class="n">S</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Add the degradation rates to the derivative of X</span>
        <span class="n">dX_dt</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-=</span> <span class="n">D_deg</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># Pack the derivatives into a single array</span>
    <span class="n">dydt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="mi">2</span> <span class="o">+</span> <span class="n">num_states</span><span class="p">)</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">dR_dt</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">dP_dt</span>
    <span class="c1"># Pack the derivatives of X into the array</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dX_dt</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="c1"># Return the derivatives</span>
    <span class="k">return</span> <span class="n">dydt</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.randmod.prepare_vectorized_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_vectorized_arrays</span><span class="p">(</span><span class="n">num_sites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prepare vectorized arrays for the Random ODE system.
It creates binary states, phosphorylation targets, and dephosphorylation targets.</p>
<p>Binary states are represented as a 2D array where each row corresponds to a state
and each column corresponds to a phosphorylation site.</p>
<p>Phosphorylation and Dephosphorylation targets are represented as 2D arrays where
each row corresponds to a state and each column corresponds to a phosphorylation site.</p>
<p>Target values indicate the resulting state after phosphorylation or dephosphorylation.</p>
<p>:param num_sites: Number of phosphorylation sites
:return: binary_states, PHOSPHO_TARGET, DEPHOSPHO_TARGET</p>


            <details class="quote">
              <summary>Source code in <code>models/randmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prepare_vectorized_arrays</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare vectorized arrays for the Random ODE system.</span>
<span class="sd">    It creates binary states, phosphorylation targets, and dephosphorylation targets.</span>

<span class="sd">    Binary states are represented as a 2D array where each row corresponds to a state</span>
<span class="sd">    and each column corresponds to a phosphorylation site.</span>

<span class="sd">    Phosphorylation and Dephosphorylation targets are represented as 2D arrays where</span>
<span class="sd">    each row corresponds to a state and each column corresponds to a phosphorylation site.</span>

<span class="sd">    Target values indicate the resulting state after phosphorylation or dephosphorylation.</span>

<span class="sd">    :param num_sites: Number of phosphorylation sites</span>
<span class="sd">    :return: binary_states, PHOSPHO_TARGET, DEPHOSPHO_TARGET</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the number of states</span>
    <span class="n">num_states</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">num_sites</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="c1"># Create binary states</span>
    <span class="n">binary_states</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">num_states</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="c1"># Fill the binary states array</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Convert the state number to binary representation</span>
        <span class="n">state</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="c1"># Fill the binary states array</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th bit is set in the state number</span>
            <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="p">(</span><span class="n">state</span> <span class="o">&amp;</span> <span class="p">(</span><span class="mi">1</span> <span class="o">&lt;&lt;</span> <span class="n">j</span><span class="p">))</span> <span class="o">!=</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="c1"># Create phosphorylation target arrays</span>
    <span class="n">PHOSPHO_TARGET</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">num_states</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="c1"># Fill the phosphorylation target array</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Convert the state number to binary representation</span>
        <span class="n">state</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="c1"># Fill the phosphorylation target array</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th bit is not set in the state number</span>
            <span class="k">if</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Set the j-th bit in the state number</span>
                <span class="n">target_state</span> <span class="o">=</span> <span class="n">state</span> <span class="o">|</span> <span class="p">(</span><span class="mi">1</span> <span class="o">&lt;&lt;</span> <span class="n">j</span><span class="p">)</span>
                <span class="c1"># Check if the target state is within the valid range</span>
                <span class="k">if</span> <span class="n">target_state</span> <span class="o">&lt;=</span> <span class="n">num_states</span><span class="p">:</span>
                    <span class="c1"># Set the target state in the phosphorylation target array</span>
                    <span class="n">PHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">target_state</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Set the target state to -1 if out of range</span>
                    <span class="n">PHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If the j-th bit is set, set the target state to -1</span>
                <span class="n">PHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="c1"># Create dephosphorylation target array</span>
    <span class="n">DEPHOSPHO_TARGET</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">num_states</span><span class="p">,</span> <span class="n">num_sites</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="c1"># Fill the dephosphorylation target array</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_states</span><span class="p">):</span>
        <span class="c1"># Convert the state number to binary representation</span>
        <span class="n">state</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="c1"># Fill the dephosphorylation target array</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_sites</span><span class="p">):</span>
            <span class="c1"># Check if the j-th bit is set in the state number</span>
            <span class="k">if</span> <span class="n">binary_states</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Set the j-th bit to 0 in the state number</span>
                <span class="n">lower_state</span> <span class="o">=</span> <span class="n">state</span> <span class="o">&amp;</span> <span class="o">~</span><span class="p">(</span><span class="mi">1</span> <span class="o">&lt;&lt;</span> <span class="n">j</span><span class="p">)</span>
                <span class="c1"># Check if the lower state is within the valid range</span>
                <span class="k">if</span> <span class="n">lower_state</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># Set the target state to -2 if the lower state is 0</span>
                    <span class="n">DEPHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Set the target state in the dephosphorylation target array</span>
                    <span class="n">DEPHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">lower_state</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If the j-th bit is not set, set the target state to -1</span>
                <span class="n">DEPHOSPHO_TARGET</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="k">return</span> <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.randmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">initial_conditions</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Solve the Random ODE system using the provided parameters and initial conditions.
The function integrates the ODE system over the specified time points and returns
the solution.</p>
<p>:param popt:
:param initial_conditions:
:param num_psites:
:param time_points:
:return: solution, solution of phosphorylated states for each site</p>


            <details class="quote">
              <summary>Source code in <code>models/randmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">solve_ode</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">initial_conditions</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Solve the Random ODE system using the provided parameters and initial conditions.</span>
<span class="sd">    The function integrates the ODE system over the specified time points and returns</span>
<span class="sd">    the solution.</span>

<span class="sd">    :param popt:</span>
<span class="sd">    :param initial_conditions:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :return: solution, solution of phosphorylated states for each site</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span> <span class="o">=</span> <span class="n">prepare_vectorized_arrays</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">ode_params</span> <span class="o">=</span> <span class="n">popt</span>
    <span class="n">A_val</span><span class="p">,</span> <span class="n">B_val</span><span class="p">,</span> <span class="n">C_val</span><span class="p">,</span> <span class="n">D_val</span> <span class="o">=</span> <span class="n">ode_params</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
    <span class="n">remaining</span> <span class="o">=</span> <span class="n">ode_params</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
    <span class="n">sol</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">odeint</span><span class="p">(</span><span class="n">ode_system</span><span class="p">,</span> <span class="n">initial_conditions</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span>
                 <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">A_val</span><span class="p">,</span> <span class="n">B_val</span><span class="p">,</span> <span class="n">C_val</span><span class="p">,</span> <span class="n">D_val</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span>
                       <span class="n">binary_states</span><span class="p">,</span> <span class="n">PHOSPHO_TARGET</span><span class="p">,</span> <span class="n">DEPHOSPHO_TARGET</span><span class="p">,</span> <span class="o">*</span><span class="n">remaining</span><span class="p">)))</span>
    <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">sol</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">sol</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">NORMALIZE_MODEL_OUTPUT</span><span class="p">:</span>
        <span class="n">norm_ic</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">initial_conditions</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">sol</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
        <span class="n">recip</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">norm_ic</span>
        <span class="n">sol</span> <span class="o">*=</span> <span class="n">recip</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
    <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">P_fitted</span> <span class="o">=</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:</span><span class="mi">2</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">P_fitted</span> <span class="o">=</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">sol</span><span class="p">,</span> <span class="n">P_fitted</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.succmod"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="models.succmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>The core of the ODE system for the successive ODE model.</p>
<p>The system is defined by the following equations:
dR/dt = A - B * R
dP/dt = C * R - D * P - S_rates[0] * P + sum(P_sites)
dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</p>
<p>where:
R: the concentration of the mRNA
P: the concentration of the protein
P_sites: the concentration of the phosphorylated sites
A: the rate of production of the mRNA
B: the rate of degradation of the mRNA
C: the rate of production of the protein
D: the rate of degradation of the protein
S_rates: the rates of phosphorylation of each site
D_rates: the rates of dephosphorylation of each site</p>
<p>:param y:
:param A:
:param B:
:param C:
:param D:
:param S_rates:
:param D_rates:
:return: derivative of y</p>


            <details class="quote">
              <summary>Source code in <code>models/succmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The core of the ODE system for the successive ODE model.</span>

<span class="sd">    The system is defined by the following equations:</span>
<span class="sd">    dR/dt = A - B * R</span>
<span class="sd">    dP/dt = C * R - D * P - S_rates[0] * P + sum(P_sites)</span>
<span class="sd">    dP_sites[i]/dt = S_rates[i] * P - (1.0 + D_rates[i]) * P_sites[i]</span>

<span class="sd">    where:</span>
<span class="sd">    R: the concentration of the mRNA</span>
<span class="sd">    P: the concentration of the protein</span>
<span class="sd">    P_sites: the concentration of the phosphorylated sites</span>
<span class="sd">    A: the rate of production of the mRNA</span>
<span class="sd">    B: the rate of degradation of the mRNA</span>
<span class="sd">    C: the rate of production of the protein</span>
<span class="sd">    D: the rate of degradation of the protein</span>
<span class="sd">    S_rates: the rates of phosphorylation of each site</span>
<span class="sd">    D_rates: the rates of dephosphorylation of each site</span>

<span class="sd">    :param y:</span>
<span class="sd">    :param A:</span>
<span class="sd">    :param B:</span>
<span class="sd">    :param C:</span>
<span class="sd">    :param D:</span>
<span class="sd">    :param S_rates:</span>
<span class="sd">    :param D_rates:</span>
<span class="sd">    :return: derivative of y</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># mRNA</span>
    <span class="n">R</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Protein</span>
    <span class="n">P</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># Number of phosphorylated sites</span>
    <span class="n">num_psites</span> <span class="o">=</span> <span class="n">S_rates</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># mRNA dynamics</span>
    <span class="n">dR_dt</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
    <span class="c1"># Protein dynamics</span>
    <span class="n">dP_dt</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span>

    <span class="c1"># Adjust protein dynamics by phosphorylation/dephosphorylation of the first site if exists</span>
    <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Subtract phosphorylation contribution from the protein at site 0</span>
        <span class="n">dP_dt</span> <span class="o">-=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span>
        <span class="c1"># Add dephosphorylation feedback from the first phosphorylated site</span>
        <span class="n">dP_dt</span> <span class="o">+=</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

    <span class="c1"># Prepare output array for derivatives</span>
    <span class="n">dydt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty_like</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">dR_dt</span>
    <span class="n">dydt</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">dP_dt</span>

    <span class="c1"># Phosphorylated sites dynamics loop</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="c1"># When there is only one site, handle it separately</span>
        <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># For one phosphorylated site:</span>
            <span class="c1"># Calculate the site&#39;s rate: phosphorylation from the protein minus its degradation (combined rate).</span>
            <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># For the first site:</span>
                <span class="c1"># Phosphorylation of the protein contributes to the site dynamics.</span>
                <span class="c1"># Site feedback: rate from the second site affects the current site.</span>
                <span class="c1"># The term y[3] provides dephosphorylation feedback.</span>
                <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">y</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># For intermediate sites:</span>
                <span class="c1"># Phosphorylation from the preceding phosphorylated species (y[1+i]) drives the site.</span>
                <span class="c1"># The degradation rate is increased by the phosphorylation rate of the next site (S_rates[i+1]).</span>
                <span class="c1"># The term y[3+i] provides dephosphorylation feedback from the next site.</span>
                <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">1</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">y</span><span class="p">[</span><span class="mi">3</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># For the last site:</span>
                <span class="c1"># Phosphorylation from the preceding site (y[1+i]) drives the site.</span>
                <span class="c1"># There is no next phosphorylation term, so only include the dephosphorylation degradation.</span>
                <span class="n">dydt</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">1</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">y</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">dydt</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.succmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>The ODE system for the successive ODE model.</p>
<p>:param y:
:param t:
:param params:
:param num_psites:
:return: ode_core(y, A, B, C, D, S_rates, D_rates)</p>


            <details class="quote">
              <summary>Source code in <code>models/succmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">ode_system</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The ODE system for the successive ODE model.</span>

<span class="sd">    :param y:</span>
<span class="sd">    :param t:</span>
<span class="sd">    :param params:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :return: ode_core(y, A, B, C, D, S_rates, D_rates)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">params</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">params</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
    <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">params</span><span class="p">[</span><span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
    <span class="n">D_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">params</span><span class="p">[</span><span class="mi">4</span> <span class="o">+</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
    <span class="k">return</span> <span class="n">ode_core</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="n">S_rates</span><span class="p">,</span> <span class="n">D_rates</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.succmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Solve the ODE system using the given parameters and initial conditions.
The function integrates the ODE system over time and returns the solution.</p>
<p>:param params:
:param init_cond:
:param num_psites:
:param t:
:return: solution, solution of phosphorylated sites</p>


            <details class="quote">
              <summary>Source code in <code>models/succmod.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Solve the ODE system using the given parameters and initial conditions.</span>
<span class="sd">    The function integrates the ODE system over time and returns the solution.</span>

<span class="sd">    :param params:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param t:</span>
<span class="sd">    :return: solution, solution of phosphorylated sites</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sol</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">odeint</span><span class="p">(</span><span class="n">ode_system</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)))</span>
    <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">sol</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">sol</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">NORMALIZE_MODEL_OUTPUT</span><span class="p">:</span>
        <span class="n">norm_init</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">init_cond</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">sol</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
        <span class="n">recip</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">norm_init</span>
        <span class="n">sol</span> <span class="o">*=</span> <span class="n">recip</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">P_fitted</span> <span class="o">=</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">T</span>
    <span class="k">return</span> <span class="n">sol</span><span class="p">,</span> <span class="n">P_fitted</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.testmod"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="models.weights"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="models.weights.early_emphasis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">early_emphasis</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function that calculates custom weights for early time points in a dataset.
The weights are based on the data values and the time differences between points.</p>
<p>The weights are calculated in a way that emphasizes early time points,
while also considering the data values and time intervals.</p>
<p>:param p_data:
:param time_points:
:param num_psites:
:return: flattened array of custom weights</p>


            <details class="quote">
              <summary>Source code in <code>models/weights.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span>
<span class="k">def</span><span class="w"> </span><span class="nf">early_emphasis</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that calculates custom weights for early time points in a dataset.</span>
<span class="sd">    The weights are based on the data values and the time differences between points.</span>

<span class="sd">    The weights are calculated in a way that emphasizes early time points,</span>
<span class="sd">    while also considering the data values and time intervals.</span>

<span class="sd">    :param p_data:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :return: flattened array of custom weights</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">p_data</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">p_data</span> <span class="o">=</span> <span class="n">p_data</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">p_data</span><span class="o">.</span><span class="n">size</span><span class="p">)</span>

    <span class="n">n_times</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">time_points</span><span class="p">)</span>
    <span class="n">custom_weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">n_times</span><span class="p">))</span>

    <span class="n">time_diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">n_times</span><span class="p">)</span>
    <span class="n">time_diffs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>

    <span class="c1"># Calculate time differences</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_times</span><span class="p">):</span>
        <span class="c1"># Subtract the previous time point from the current one</span>
        <span class="n">time_diffs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">time_points</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">-</span> <span class="n">time_points</span><span class="p">[</span><span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="c1"># Emphasize early time points - first five</span>
        <span class="n">limit</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">n_times</span><span class="p">)</span>
        <span class="c1"># Compute weights for early time points</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">limit</span><span class="p">):</span>
            <span class="c1"># Calculate the data-based and time-based weights</span>
            <span class="n">data_based_weight</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">p_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">])</span> <span class="o">+</span> <span class="mf">1e-5</span><span class="p">)</span>
            <span class="n">time_based_weight</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="n">time_diffs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-5</span><span class="p">)</span>
            <span class="c1"># Combine the weights</span>
            <span class="n">custom_weights</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_based_weight</span> <span class="o">*</span> <span class="n">time_based_weight</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">n_times</span><span class="p">):</span>
            <span class="c1"># For later time points, use a fixed weight</span>
            <span class="n">custom_weights</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">return</span> <span class="n">custom_weights</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="models.weights.get_weight_options" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_weight_options</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">use_regularization</span><span class="p">,</span> <span class="n">reg_len</span><span class="p">,</span> <span class="n">early_weights</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Function to calculate weights for parameter estimation based on the target data and time points.
The weights are designed to emphasize early time points and account for noise in the data.
The function also includes options for regularization and custom early point emphasis.</p>
<p>The following are the weighting schemes:
- Inverse data: 1 / abs(target)
- Exponential decay: exp(-0.5 * target)
- Log scale: 1 / log(1 + abs(target))
- Time difference: 1 / abs(time_diff)
- Moving average: 1 / abs(target - moving_avg)
- Sigmoid time decay: 1 / (1 + exp(time_indices - 5))
- Exponential early emphasis: exp(-0.5 * time_indices)
- Polynomial decay: 1 / (1 + 0.5 * time_indices)
- MS SNR model: 1 / sqrt(signal)
- MS inverse variance: 1 / (abs(target) ** 0.7)
- Flat region penalty: 1 / abs(grad)
- Steady state decay: exp(-0.1 * time_indices)
- Combined data time: 1 / (abs(target) * (1 + 0.5 * time_indices))
- Inverse sqrt data: 1 / sqrt(abs(target))
- Early emphasis moderate: ones
- Early emphasis steep decay: ones
- Custom early points emphasis: early_weights</p>
<p>:param target:
:param t_target:
:param num_psites:
:param use_regularization:
:param reg_len:
:param early_weights:
:return: dictionary of weights for parameter estimation</p>


            <details class="quote">
              <summary>Source code in <code>models/weights.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_weight_options</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">use_regularization</span><span class="p">,</span> <span class="n">reg_len</span><span class="p">,</span> <span class="n">early_weights</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to calculate weights for parameter estimation based on the target data and time points.</span>
<span class="sd">    The weights are designed to emphasize early time points and account for noise in the data.</span>
<span class="sd">    The function also includes options for regularization and custom early point emphasis.</span>

<span class="sd">    The following are the weighting schemes:</span>
<span class="sd">    - Inverse data: 1 / abs(target)</span>
<span class="sd">    - Exponential decay: exp(-0.5 * target)</span>
<span class="sd">    - Log scale: 1 / log(1 + abs(target))</span>
<span class="sd">    - Time difference: 1 / abs(time_diff)</span>
<span class="sd">    - Moving average: 1 / abs(target - moving_avg)</span>
<span class="sd">    - Sigmoid time decay: 1 / (1 + exp(time_indices - 5))</span>
<span class="sd">    - Exponential early emphasis: exp(-0.5 * time_indices)</span>
<span class="sd">    - Polynomial decay: 1 / (1 + 0.5 * time_indices)</span>
<span class="sd">    - MS SNR model: 1 / sqrt(signal)</span>
<span class="sd">    - MS inverse variance: 1 / (abs(target) ** 0.7)</span>
<span class="sd">    - Flat region penalty: 1 / abs(grad)</span>
<span class="sd">    - Steady state decay: exp(-0.1 * time_indices)</span>
<span class="sd">    - Combined data time: 1 / (abs(target) * (1 + 0.5 * time_indices))</span>
<span class="sd">    - Inverse sqrt data: 1 / sqrt(abs(target))</span>
<span class="sd">    - Early emphasis moderate: ones</span>
<span class="sd">    - Early emphasis steep decay: ones</span>
<span class="sd">    - Custom early points emphasis: early_weights</span>

<span class="sd">    :param target:</span>
<span class="sd">    :param t_target:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :param reg_len:</span>
<span class="sd">    :param early_weights:</span>
<span class="sd">    :return: dictionary of weights for parameter estimation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">time_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">t_target</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">num_psites</span><span class="p">)</span>
    <span class="n">log_scale</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">))</span>
    <span class="n">sqrt_signal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">))</span>

    <span class="c1"># Noise-aware weights</span>
    <span class="n">signal_noise_model</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">sqrt_signal</span>  <span class="c1"># MS-like error model</span>
    <span class="n">inverse_variance_model</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span> <span class="o">**</span> <span class="mf">0.7</span><span class="p">)</span>  <span class="c1"># empirical fit</span>

    <span class="c1"># Biological modeling</span>
    <span class="n">early_sigmoid</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">time_indices</span> <span class="o">-</span> <span class="mi">5</span><span class="p">)))</span>
    <span class="n">steady_state_decay</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span> <span class="o">*</span> <span class="n">time_indices</span><span class="p">)</span>  <span class="c1"># Less weight to late points</span>
    <span class="c1"># fallback for flat region penalty</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">target</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">grad</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">gradient</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="n">flat_region_penalty</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">grad</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Early time pointsbiologically inspired</span>
        <span class="n">flat_region_penalty</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span>  <span class="c1"># inverse signal intensity</span>

    <span class="n">base_weights</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;inverse_data&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">),</span>
        <span class="s2">&quot;exp_decay&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">target</span><span class="p">),</span>
        <span class="s2">&quot;log_scale&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">log_scale</span><span class="p">,</span> <span class="mf">1e-5</span><span class="p">),</span>
        <span class="s2">&quot;time_diff&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">prepend</span><span class="o">=</span><span class="n">target</span><span class="p">[</span><span class="mi">0</span><span class="p">])),</span> <span class="mf">1e-5</span><span class="p">),</span>
        <span class="s2">&quot;moving_avg&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span> <span class="o">-</span> <span class="n">uniform_filter1d</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span> <span class="mf">1e-5</span><span class="p">),</span>

        <span class="s2">&quot;sigmoid_time_decay&quot;</span><span class="p">:</span> <span class="n">early_sigmoid</span><span class="p">,</span>
        <span class="s2">&quot;exponential_early_emphasis&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">time_indices</span><span class="p">),</span>
        <span class="s2">&quot;polynomial_decay&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">time_indices</span><span class="p">),</span>

        <span class="s2">&quot;ms_snr_model&quot;</span><span class="p">:</span> <span class="n">signal_noise_model</span><span class="p">,</span>
        <span class="s2">&quot;ms_inverse_variance&quot;</span><span class="p">:</span> <span class="n">inverse_variance_model</span><span class="p">,</span>
        <span class="s2">&quot;flat_region_penalty&quot;</span><span class="p">:</span> <span class="n">flat_region_penalty</span><span class="p">,</span>
        <span class="s2">&quot;steady_state_decay&quot;</span><span class="p">:</span> <span class="n">steady_state_decay</span><span class="p">,</span>

        <span class="s2">&quot;combined_data_time&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">time_indices</span><span class="p">)),</span>
        <span class="s2">&quot;inverse_sqrt_data&quot;</span><span class="p">:</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">sqrt_signal</span><span class="p">,</span>

        <span class="s2">&quot;early_emphasis_moderate&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">target</span><span class="p">),</span>
        <span class="s2">&quot;early_emphasis_steep_decay&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span>
                <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">t_target</span><span class="p">)</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">10</span><span class="p">))]),</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">t_target</span><span class="p">)</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">10</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">target</span><span class="p">))</span>
        <span class="p">),</span>
        <span class="s2">&quot;custom_early_points_emphasis&quot;</span><span class="p">:</span> <span class="n">early_weights</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
        <span class="n">base_weights</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">k</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">v</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">reg_len</span><span class="p">)])</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">base_weights</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>

    <span class="k">return</span> <span class="n">base_weights</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.adapest"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.adapest.adaptive_estimation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">adaptive_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">extra_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Adaptive estimation of parameters using curve fitting.
This function estimates the parameters of the model using curve fitting.
Parameter will be fixed at certain time points.</p>
<p>:param p_data:
:param init_cond:
:param num_psites:
:param time_points:
:param t:
:param bounds:
:param fixed_params:
:param gene:
:param bootstraps:
:param extra_fixed:
:param use_regularization:
:param lambda_reg:
:return: estimated parameters</p>


            <details class="quote">
              <summary>Source code in <code>paramest/adapest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">adaptive_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span>
                        <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">extra_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adaptive estimation of parameters using curve fitting.</span>
<span class="sd">    This function estimates the parameters of the model using curve fitting.</span>
<span class="sd">    Parameter will be fixed at certain time points.</span>

<span class="sd">    :param p_data:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param t:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param gene:</span>
<span class="sd">    :param bootstraps:</span>
<span class="sd">    :param extra_fixed:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :param lambda_reg:</span>
<span class="sd">    :return: estimated parameters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the index of the time points that are less than or equal to t.</span>
    <span class="n">t_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">time_points</span> <span class="o">&lt;=</span> <span class="n">t</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">t_target</span> <span class="o">=</span> <span class="n">time_points</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>
    <span class="n">p_interp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_psites</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">t_target</span><span class="p">)))</span>

    <span class="c1"># Interpolate the data to get the values at the target time points.</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="c1"># Get the values at the target time points.</span>
        <span class="n">y_vals</span> <span class="o">=</span> <span class="n">p_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="k">else</span> <span class="n">p_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">:]</span>
        <span class="c1"># Interpolate the values at the target time points.</span>
        <span class="n">interp_func</span> <span class="o">=</span> <span class="n">PchipInterpolator</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">y_vals</span><span class="p">)</span>
        <span class="c1"># Get the values at the target time points.</span>
        <span class="n">p_interp</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">interp_func</span><span class="p">(</span><span class="n">t_target</span><span class="p">)</span>

    <span class="n">full_fixed</span> <span class="o">=</span> <span class="n">fixed_params</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">extra_fixed</span><span class="p">:</span>
        <span class="n">full_fixed</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">extra_fixed</span><span class="p">)</span>

    <span class="c1"># Get model function and bounds.</span>
    <span class="n">model_func</span><span class="p">,</span> <span class="n">free_idx</span><span class="p">,</span> <span class="n">free_bounds</span><span class="p">,</span> <span class="n">fixed_vals</span><span class="p">,</span> <span class="n">total_params</span> <span class="o">=</span> <span class="n">prepare_model_func</span><span class="p">(</span>
        <span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span>
        <span class="n">full_fixed</span><span class="p">,</span> <span class="n">use_regularization</span><span class="p">,</span> <span class="n">lambda_reg</span>
    <span class="p">)</span>

    <span class="c1"># Initial values for the free parameters.</span>
    <span class="n">p0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="n">l</span> <span class="o">+</span> <span class="n">u</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">u</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">free_bounds</span><span class="p">)])</span>
    <span class="n">target</span> <span class="o">=</span> <span class="n">p_interp</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="c1"># Make the fitting vector the same size as the model.</span>
    <span class="n">target_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p0</span><span class="p">))])</span> <span class="k">if</span> <span class="n">use_regularization</span> <span class="k">else</span> <span class="n">target</span>
    <span class="n">default_sigma</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target_fit</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Perform initial fit to get starting values.</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                      <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">target_fit</span><span class="p">,</span> <span class="n">p0</span><span class="o">=</span><span class="n">p0</span><span class="p">,</span>
                                <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">default_sigma</span><span class="p">,</span>
                                <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
        <span class="n">popt_init</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] T=</span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> init fit failed: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">popt_init</span> <span class="o">=</span> <span class="n">p0</span>

    <span class="c1"># Get weight options for the fit.</span>
    <span class="n">early_weights</span> <span class="o">=</span> <span class="n">early_emphasis</span><span class="p">(</span><span class="n">p_interp</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span>
    <span class="n">weight_options</span> <span class="o">=</span> <span class="n">get_weight_options</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">use_regularization</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">p0</span><span class="p">),</span> <span class="n">early_weights</span><span class="p">)</span>

    <span class="c1"># Make a dictionary of the weights, and parameters.</span>
    <span class="n">scores</span><span class="p">,</span> <span class="n">popts</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">wname</span><span class="p">,</span> <span class="n">sigma</span> <span class="ow">in</span> <span class="n">weight_options</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Perform the fit with the weights.</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                          <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">target_fit</span><span class="p">,</span> <span class="n">p0</span><span class="o">=</span><span class="n">popt_init</span><span class="p">,</span>
                                    <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span>
                                    <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
            <span class="n">popt_weight</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] T=</span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> fit failed for </span><span class="si">{</span><span class="n">wname</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">popt_weight</span> <span class="o">=</span> <span class="n">popt_init</span>
        <span class="n">popts</span><span class="p">[</span><span class="n">wname</span><span class="p">]</span> <span class="o">=</span> <span class="n">popt_weight</span>
        <span class="n">pred</span> <span class="o">=</span> <span class="n">model_func</span><span class="p">(</span><span class="n">t_target</span><span class="p">,</span> <span class="o">*</span><span class="n">popt_weight</span><span class="p">)</span>
        <span class="c1"># Calculate the score for the fit.</span>
        <span class="n">scores</span><span class="p">[</span><span class="n">wname</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_fit</span><span class="p">(</span><span class="n">target_fit</span><span class="p">,</span> <span class="n">pred</span><span class="p">,</span> <span class="n">popt_weight</span><span class="p">)</span>

    <span class="c1"># Get the best weight and score.</span>
    <span class="n">best_weight</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">scores</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>
    <span class="n">best_score</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">best_weight</span><span class="p">]</span>
    <span class="c1"># Get the best parameters.</span>
    <span class="n">popt_best</span> <span class="o">=</span> <span class="n">popts</span><span class="p">[</span><span class="n">best_weight</span><span class="p">]</span>
    <span class="c1"># Display the best weight and score for time points.</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Time = </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> Best Weight = </span><span class="si">{</span><span class="n">best_weight</span><span class="si">}</span><span class="s2"> Score: </span><span class="si">{</span><span class="n">best_score</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bootstraps</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Bootstrapping </span><span class="si">{</span><span class="n">bootstraps</span><span class="si">}</span><span class="s2">x at T = </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Perform bootstrapping to get the final parameters.</span>
        <span class="n">est_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bootstraps</span><span class="p">):</span>
            <span class="c1"># Generate bootstrap samples.</span>
            <span class="n">noise</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">target_fit</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">noisy_target</span> <span class="o">=</span> <span class="n">target_fit</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">noise</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Perform the fit with the bootstrap samples.</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                              <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">t_target</span><span class="p">,</span> <span class="n">noisy_target</span><span class="p">,</span> <span class="n">p0</span><span class="o">=</span><span class="n">popt_best</span><span class="p">,</span>
                                        <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">weight_options</span><span class="p">[</span><span class="n">best_weight</span><span class="p">],</span>
                                        <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
                <span class="n">popt_weight_bootstrap</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">RuntimeError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Bootstrap fit failed: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">popt_weight_bootstrap</span> <span class="o">=</span> <span class="n">popt_best</span>
            <span class="n">est_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">popt_weight_bootstrap</span><span class="p">)</span>

        <span class="c1"># Get the best parameters from bootstrap samples.</span>
        <span class="n">popt_best</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">est_list</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">p_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">total_params</span><span class="p">)</span>
    <span class="n">free_iter</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">popt_best</span><span class="p">)</span>

    <span class="c1"># Fill in the fixed values.</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">total_params</span><span class="p">):</span>
        <span class="c1"># If the index is in the fixed values, use the fixed value.</span>
        <span class="n">p_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_vals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fixed_vals</span> <span class="k">else</span> <span class="nb">next</span><span class="p">(</span><span class="n">free_iter</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">p_full</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="paramest.adapest.estimate_profiles" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">estimate_profiles</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">time_fixed_dict</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Estimate adaptive profiles for a given gene at specified time points.
This function estimates the parameters of the model using adaptive estimation.
The parameters are estimated at the specified time points and stored in a DataFrame.
The function also plots the profiles.</p>
<p>:param gene:
:param p_data:
:param init_cond:
:param num_psites:
:param time_points:
:param desired_times:
:param bounds:
:param fixed_params:
:param bootstraps:
:param time_fixed_dict:
:return: dataframe, profiles</p>


            <details class="quote">
              <summary>Source code in <code>paramest/adapest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">estimate_profiles</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span>
                      <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">time_fixed_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Estimate adaptive profiles for a given gene at specified time points.</span>
<span class="sd">    This function estimates the parameters of the model using adaptive estimation.</span>
<span class="sd">    The parameters are estimated at the specified time points and stored in a DataFrame.</span>
<span class="sd">    The function also plots the profiles.</span>

<span class="sd">    :param gene:</span>
<span class="sd">    :param p_data:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param desired_times:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param bootstraps:</span>
<span class="sd">    :param time_fixed_dict:</span>
<span class="sd">    :return: dataframe, profiles</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">profiles</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">T</span> <span class="ow">in</span> <span class="n">desired_times</span><span class="p">:</span>
        <span class="c1"># logger.info(f&quot;[{gene}] Estimating Adaptive Profile at T = {T}&quot;)</span>
        <span class="n">extra_fixed</span> <span class="o">=</span> <span class="n">time_fixed_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">T</span><span class="p">),</span> <span class="p">{})</span>
        <span class="n">profiled_params</span> <span class="o">=</span> <span class="n">adaptive_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">T</span><span class="p">,</span>
                                              <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">extra_fixed</span><span class="p">)</span>
        <span class="n">profiles</span><span class="p">[</span><span class="n">T</span><span class="p">]</span> <span class="o">=</span> <span class="n">profiled_params</span>

    <span class="n">param_names</span> <span class="o">=</span> <span class="n">get_param_names</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Time&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">profiles</span><span class="o">.</span><span class="n">keys</span><span class="p">())}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">param_names</span><span class="p">):</span>
        <span class="n">data</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">profiles</span><span class="p">[</span><span class="n">T</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">T</span> <span class="ow">in</span> <span class="n">desired_times</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="n">plotter</span> <span class="o">=</span> <span class="n">Plotter</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">plotter</span><span class="o">.</span><span class="n">plot_profiles</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span><span class="p">,</span> <span class="n">profiles</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="paramest.adapest.prepare_model_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prepares the model function for curve fitting.
Adaptive Estimation is used to estimate the parameters of the model.</p>
<p>:param num_psites:
:param init_cond:
:param bounds:
:param fixed_params:
:param use_regularization:
:param lambda_reg:
:return: model function, free indices, free bounds, fixed values, total params</p>


            <details class="quote">
              <summary>Source code in <code>paramest/adapest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span>
                       <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepares the model function for curve fitting.</span>
<span class="sd">    Adaptive Estimation is used to estimate the parameters of the model.</span>

<span class="sd">    :param num_psites:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :param lambda_reg:</span>
<span class="sd">    :return: model function, free indices, free bounds, fixed values, total params</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
        <span class="c1"># Build lower and upper bounds from config.</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="c1"># For phosphorylation parameters: use Ssite bounds.</span>
        <span class="n">lower_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="n">upper_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="c1"># For dephosphorylation parameters: for each combination, use Dsite bounds.</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">r</span><span class="p">):</span>
                <span class="n">lower_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">upper_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">num_total_params</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lower_bounds_full</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Existing approach for distributive or successive models.</span>
        <span class="n">num_total_params</span> <span class="o">=</span> <span class="mi">4</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>
    <span class="c1"># Create a dictionary of fixed parameters.</span>
    <span class="n">fixed_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Create a dictionary of fixed values.</span>
    <span class="n">free_indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Create a list of free indices.</span>
    <span class="n">param_names</span> <span class="o">=</span> <span class="n">get_param_names</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">param_names</span><span class="p">):</span>
        <span class="n">val</span> <span class="o">=</span> <span class="n">fixed_params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">val</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fixed_values</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">free_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">model_func</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="o">*</span><span class="n">p_free</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Model function for curve fitting.</span>

<span class="sd">        :param t:</span>
<span class="sd">        :param p_free:</span>
<span class="sd">        :return: model</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">p_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">)</span>
        <span class="c1"># Fill in the fixed values.</span>
        <span class="n">free_iter</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">p_free</span><span class="p">)</span>
        <span class="c1"># Fill in the free values.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">):</span>
            <span class="c1"># If the index is in the fixed values, use the fixed value.</span>
            <span class="n">p_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fixed_values</span> <span class="k">else</span> <span class="nb">next</span><span class="p">(</span><span class="n">free_iter</span><span class="p">)</span>

        <span class="n">_</span><span class="p">,</span> <span class="n">p_fitted</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">p_full</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">atleast_1d</span><span class="p">(</span><span class="n">t</span><span class="p">))</span>
        <span class="n">y_model</span> <span class="o">=</span> <span class="n">p_fitted</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="c1"># If using regularization, add the regularization term.</span>
        <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
            <span class="c1"># Add the regularization term.</span>
            <span class="n">reg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">lambda_reg</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">p_free</span><span class="p">)</span>
            <span class="c1"># Add the regularization term to the model.</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">y_model</span><span class="p">,</span> <span class="n">reg</span><span class="p">])</span>
        <span class="c1"># Return the model.</span>
        <span class="k">return</span> <span class="n">y_model</span>

    <span class="c1"># Create the free bounds.</span>
    <span class="n">free_bounds</span> <span class="o">=</span> <span class="p">([</span><span class="n">lower_bounds_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">free_indices</span><span class="p">],</span>
                   <span class="p">[</span><span class="n">upper_bounds_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">free_indices</span><span class="p">])</span>

    <span class="c1"># logger.info(&quot;Model built&quot;)</span>
    <span class="k">return</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">free_indices</span><span class="p">,</span> <span class="n">free_bounds</span><span class="p">,</span> <span class="n">fixed_values</span><span class="p">,</span> <span class="n">num_total_params</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.core"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.core.process_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">desired_times</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">time_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Process a single gene by estimating its parameters and generating plots.
This function extracts gene-specific data, estimates parameters using the
specified estimation mode, and generates plots for the estimated parameters
and the model fits. It also calculates error metrics and saves the results
to Excel files.</p>
<p>:param gene:
:param measurement_data:
:param time_points:
:param bounds:
:param fixed_params:
:param desired_times:
:param time_fixed:
:param bootstraps:
:param out_dir:
:return:
    - gene: The gene being processed.
    - estimated_params: Estimated parameters for the gene.
    - model_fits: Model fits for the gene.
    - seq_model_fit: Sequential model fit for the gene.
    - errors: Error metrics (MSE, MAE).
    - final_params: Final estimated parameters.
    - profiles: Adaptive profile estimates (if applicable).
    - profiles_df: DataFrame of adaptive profile estimates (if applicable).
    - param_df: DataFrame of estimated parameters.
    - gene_psite_data: Dictionary of gene-specific data.</p>


            <details class="quote">
              <summary>Source code in <code>paramest/core.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_gene</span><span class="p">(</span>
    <span class="n">gene</span><span class="p">,</span>
    <span class="n">measurement_data</span><span class="p">,</span>
    <span class="n">time_points</span><span class="p">,</span>
    <span class="n">bounds</span><span class="p">,</span>
    <span class="n">fixed_params</span><span class="p">,</span>
    <span class="n">desired_times</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">time_fixed</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process a single gene by estimating its parameters and generating plots.</span>
<span class="sd">    This function extracts gene-specific data, estimates parameters using the</span>
<span class="sd">    specified estimation mode, and generates plots for the estimated parameters</span>
<span class="sd">    and the model fits. It also calculates error metrics and saves the results</span>
<span class="sd">    to Excel files.</span>

<span class="sd">    :param gene:</span>
<span class="sd">    :param measurement_data:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param desired_times:</span>
<span class="sd">    :param time_fixed:</span>
<span class="sd">    :param bootstraps:</span>
<span class="sd">    :param out_dir:</span>
<span class="sd">    :return:</span>
<span class="sd">        - gene: The gene being processed.</span>
<span class="sd">        - estimated_params: Estimated parameters for the gene.</span>
<span class="sd">        - model_fits: Model fits for the gene.</span>
<span class="sd">        - seq_model_fit: Sequential model fit for the gene.</span>
<span class="sd">        - errors: Error metrics (MSE, MAE).</span>
<span class="sd">        - final_params: Final estimated parameters.</span>
<span class="sd">        - profiles: Adaptive profile estimates (if applicable).</span>
<span class="sd">        - profiles_df: DataFrame of adaptive profile estimates (if applicable).</span>
<span class="sd">        - param_df: DataFrame of estimated parameters.</span>
<span class="sd">        - gene_psite_data: Dictionary of gene-specific data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 1. Extract Gene-specific Data</span>
    <span class="n">gene_data</span> <span class="o">=</span> <span class="n">measurement_data</span><span class="p">[</span><span class="n">measurement_data</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">]</span>
    <span class="n">num_psites</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">psite_values</span> <span class="o">=</span> <span class="n">gene_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">init_cond</span> <span class="o">=</span> <span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">P_data</span> <span class="o">=</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># 2. Choose estimation mode</span>
    <span class="n">estimation_mode</span> <span class="o">=</span> <span class="n">ESTIMATION_MODE</span>

    <span class="n">model_fits</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">,</span> <span class="n">errors</span> <span class="o">=</span> <span class="n">estimate_parameters</span><span class="p">(</span>
        <span class="n">estimation_mode</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">P_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span>
    <span class="p">)</span>

    <span class="c1"># 7. Error Metrics</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">P_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">seq_model_fit</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">P_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">seq_model_fit</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">  MSE: </span><span class="si">{</span><span class="n">mse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, MAE: </span><span class="si">{</span><span class="n">mae</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># 3. Adaptive Profile Estimation (Optional)</span>
    <span class="n">profiles_df</span><span class="p">,</span> <span class="n">profiles_dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">desired_times</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">time_fixed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">profiles_df</span><span class="p">,</span> <span class="n">profiles_dict</span> <span class="o">=</span> <span class="n">estimate_profiles</span><span class="p">(</span>
            <span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span>
            <span class="n">time_points</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span>
            <span class="n">bootstraps</span><span class="p">,</span> <span class="n">time_fixed</span>
        <span class="p">)</span>
        <span class="c1"># Save profile Excel</span>
        <span class="n">profile_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_profiles.xlsx&quot;</span><span class="p">)</span>
        <span class="n">profiles_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">profile_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># logger.info(f&quot;Profiled Estimates: {profile_path}&quot;)</span>

    <span class="c1"># 4. Solve Full ODE with Final Params</span>
    <span class="n">final_params</span> <span class="o">=</span> <span class="n">estimated_params</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">gene_psite_dict_local</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Protein&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">get_param_names</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)):</span>
        <span class="n">gene_psite_dict_local</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">final_params</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>

    <span class="n">sol_full</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">final_params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span>

    <span class="c1"># 5. Plotting Outputs</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="n">generate_labels</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">illustrate</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span>
    <span class="c1"># Create a single Plotter instance.</span>
    <span class="n">plotter</span> <span class="o">=</span> <span class="n">Plotter</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">)</span>
    <span class="c1"># Call plot_all with all necessary data.</span>
    <span class="n">plotter</span><span class="o">.</span><span class="n">plot_all</span><span class="p">(</span><span class="n">solution</span><span class="o">=</span><span class="n">sol_full</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
                     <span class="n">estimated_params</span><span class="o">=</span><span class="n">estimated_params</span><span class="p">,</span> <span class="n">time_points</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
                     <span class="n">P_data</span><span class="o">=</span><span class="n">P_data</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="o">=</span><span class="n">seq_model_fit</span><span class="p">,</span>
                     <span class="n">psite_labels</span><span class="o">=</span><span class="n">psite_values</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">target_variance</span><span class="o">=</span><span class="mf">0.99</span><span class="p">)</span>

    <span class="c1"># 6. Save Sequential Parameters to Excel</span>
    <span class="n">df_params</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">get_param_names</span><span class="p">(</span><span class="n">num_psites</span><span class="p">))</span>
    <span class="n">df_params</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Time&quot;</span><span class="p">,</span> <span class="n">time_points</span><span class="p">[:</span><span class="nb">len</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">)])</span>
    <span class="n">param_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_parameters.xlsx&quot;</span><span class="p">)</span>
    <span class="n">df_params</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">param_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># logger.info(f&quot;Estimated Parameters: {param_path}&quot;)</span>

    <span class="c1"># 8. Return Results</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span>
        <span class="s2">&quot;estimated_params&quot;</span><span class="p">:</span> <span class="n">estimated_params</span><span class="p">,</span>
        <span class="s2">&quot;model_fits&quot;</span><span class="p">:</span> <span class="n">model_fits</span><span class="p">,</span>
        <span class="s2">&quot;seq_model_fit&quot;</span><span class="p">:</span> <span class="n">seq_model_fit</span><span class="p">,</span>
        <span class="s2">&quot;errors&quot;</span><span class="p">:</span> <span class="n">errors</span><span class="p">,</span>
        <span class="s2">&quot;final_params&quot;</span><span class="p">:</span> <span class="n">final_params</span><span class="p">,</span>
        <span class="s2">&quot;profiles&quot;</span><span class="p">:</span> <span class="n">profiles_dict</span><span class="p">,</span>
        <span class="s2">&quot;profiles_df&quot;</span><span class="p">:</span> <span class="n">profiles_df</span><span class="p">,</span>
        <span class="s2">&quot;param_df&quot;</span><span class="p">:</span> <span class="n">df_params</span><span class="p">,</span>
        <span class="s2">&quot;gene_psite_data&quot;</span><span class="p">:</span> <span class="n">gene_psite_dict_local</span><span class="p">,</span>
        <span class="s2">&quot;mse&quot;</span><span class="p">:</span> <span class="n">mse</span><span class="p">,</span>
        <span class="s2">&quot;mae&quot;</span><span class="p">:</span> <span class="n">mae</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="paramest.core.process_gene_wrapper" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_gene_wrapper</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">desired_times</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Wrapper function to process a gene. This function is a placeholder for
any additional processing or modifications needed before calling the
main processing function.</p>
<p>:param gene:
:param measurement_data:
:param time_points:
:param bounds:
:param fixed_params:
:param desired_times:
:param time_fixed:
:param bootstraps:
:param out_dir:
:return:
    - result: Dictionary containing the results of the gene processing.</p>


            <details class="quote">
              <summary>Source code in <code>paramest/core.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_gene_wrapper</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">measurement_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span>
                         <span class="n">desired_times</span><span class="p">,</span> <span class="n">time_fixed</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper function to process a gene. This function is a placeholder for</span>
<span class="sd">    any additional processing or modifications needed before calling the</span>
<span class="sd">    main processing function.</span>

<span class="sd">    :param gene:</span>
<span class="sd">    :param measurement_data:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param desired_times:</span>
<span class="sd">    :param time_fixed:</span>
<span class="sd">    :param bootstraps:</span>
<span class="sd">    :param out_dir:</span>
<span class="sd">    :return:</span>
<span class="sd">        - result: Dictionary containing the results of the gene processing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">process_gene</span><span class="p">(</span>
        <span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span>
        <span class="n">measurement_data</span><span class="o">=</span><span class="n">measurement_data</span><span class="p">,</span>
        <span class="n">time_points</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
        <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span>
        <span class="n">fixed_params</span><span class="o">=</span><span class="n">fixed_params</span><span class="p">,</span>
        <span class="n">desired_times</span><span class="o">=</span><span class="n">desired_times</span><span class="p">,</span>
        <span class="n">time_fixed</span><span class="o">=</span><span class="n">time_fixed</span><span class="p">,</span>
        <span class="n">bootstraps</span><span class="o">=</span><span class="n">bootstraps</span><span class="p">,</span>
        <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.identifiability.ci"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.identifiability.ci.confidence_intervals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">confidence_intervals</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">pcov</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">alpha_val</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Computes the confidence intervals for parameter estimates using a linearization approach.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="popt">popt</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>1D numpy array of best-fit parameter estimates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="pcov">pcov</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Square covariance matrix (numpy array) corresponding to popt.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="target">target</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>1D numpy array of observed data (used to compute degrees of freedom).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="alpha_val">alpha_val</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Significance level (default 0.05 for a 95% confidence interval).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary with the following keys:
'beta_hat': Best-fit parameter estimates.
'se_lin': Standard errors (sqrt of diagonal of pcov).
'df_lin': Degrees of freedom (n_obs - n_params).
't_stat': t-statistics for each parameter.
'pval': Two-sided p-values for each parameter.
'qt_lin': t critical value for the given alpha and degrees of freedom.
'lwr_ci': Lower 95% confidence intervals.
'upr_ci': Upper 95% confidence intervals.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>paramest/identifiability/ci.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">confidence_intervals</span><span class="p">(</span><span class="n">popt</span><span class="p">,</span> <span class="n">pcov</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">alpha_val</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the confidence intervals for parameter estimates using a linearization approach.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      - popt: 1D numpy array of best-fit parameter estimates.</span>
<span class="sd">      - pcov: Square covariance matrix (numpy array) corresponding to popt.</span>
<span class="sd">      - target: 1D numpy array of observed data (used to compute degrees of freedom).</span>
<span class="sd">      - alpha_val: Significance level (default 0.05 for a 95% confidence interval).</span>

<span class="sd">    Returns:</span>
<span class="sd">      A dictionary with the following keys:</span>
<span class="sd">        &#39;beta_hat&#39;: Best-fit parameter estimates.</span>
<span class="sd">        &#39;se_lin&#39;: Standard errors (sqrt of diagonal of pcov).</span>
<span class="sd">        &#39;df_lin&#39;: Degrees of freedom (n_obs - n_params).</span>
<span class="sd">        &#39;t_stat&#39;: t-statistics for each parameter.</span>
<span class="sd">        &#39;pval&#39;: Two-sided p-values for each parameter.</span>
<span class="sd">        &#39;qt_lin&#39;: t critical value for the given alpha and degrees of freedom.</span>
<span class="sd">        &#39;lwr_ci&#39;: Lower 95% confidence intervals.</span>
<span class="sd">        &#39;upr_ci&#39;: Upper 95% confidence intervals.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">pcov</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;No covariance matrix available; cannot compute confidence intervals using linearization.&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">beta_hat</span> <span class="o">=</span> <span class="n">popt</span>
    <span class="n">se_lin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">pcov</span><span class="p">))</span>
    <span class="c1"># Degrees of freedom: number of observations (target should be 1D) minus number of parameters.</span>
    <span class="n">df_lin</span> <span class="o">=</span> <span class="n">target</span><span class="o">.</span><span class="n">size</span> <span class="o">-</span> <span class="n">beta_hat</span><span class="o">.</span><span class="n">size</span>
    <span class="c1"># Compute t-statistics for each parameter.</span>
    <span class="n">t_stat</span> <span class="o">=</span> <span class="n">beta_hat</span> <span class="o">/</span> <span class="n">se_lin</span>
    <span class="c1"># Two-sided p-values.</span>
    <span class="n">pval</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">t</span><span class="o">.</span><span class="n">sf</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">t_stat</span><span class="p">),</span> <span class="n">df_lin</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span>
    <span class="c1"># t critical value for a two-tailed confidence interval.</span>
    <span class="n">qt_lin</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">t</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">alpha_val</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">df_lin</span><span class="p">)</span>
    <span class="c1"># Calculate confidence intervals.</span>
    <span class="n">lwr_ci</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">beta_hat</span> <span class="o">-</span> <span class="n">qt_lin</span> <span class="o">*</span> <span class="n">se_lin</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">upr_ci</span> <span class="o">=</span> <span class="n">beta_hat</span> <span class="o">+</span> <span class="n">qt_lin</span> <span class="o">*</span> <span class="n">se_lin</span>

    <span class="c1"># Log the summary.</span>
    <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;Parameter</span><span class="se">\t</span><span class="s2">Estimate</span><span class="se">\t</span><span class="s2">Std. Error</span><span class="se">\t</span><span class="s2">Pr(&gt;|t|)</span><span class="se">\t\t</span><span class="s2">95% CI&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Confidence Intervals:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">b</span><span class="p">,</span> <span class="n">se</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">lwr</span><span class="p">,</span> <span class="n">upr</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">beta_hat</span><span class="p">,</span> <span class="n">se_lin</span><span class="p">,</span> <span class="n">pval</span><span class="p">,</span> <span class="n">lwr_ci</span><span class="p">,</span> <span class="n">upr_ci</span><span class="p">)):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Param </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="s2"> </span><span class="si">{</span><span class="n">b</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="se">\t\t</span><span class="s2"> </span><span class="si">{</span><span class="n">se</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="se">\t\t</span><span class="s2"> </span><span class="si">{</span><span class="n">p</span><span class="si">:</span><span class="s2">.1e</span><span class="si">}</span><span class="se">\t\t</span><span class="s2"> (</span><span class="si">{</span><span class="n">lwr</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">upr</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;beta_hat&#39;</span><span class="p">:</span> <span class="n">beta_hat</span><span class="p">,</span>
        <span class="s1">&#39;se_lin&#39;</span><span class="p">:</span> <span class="n">se_lin</span><span class="p">,</span>
        <span class="s1">&#39;df_lin&#39;</span><span class="p">:</span> <span class="n">df_lin</span><span class="p">,</span>
        <span class="s1">&#39;t_stat&#39;</span><span class="p">:</span> <span class="n">t_stat</span><span class="p">,</span>
        <span class="s1">&#39;pval&#39;</span><span class="p">:</span> <span class="n">pval</span><span class="p">,</span>
        <span class="s1">&#39;qt_lin&#39;</span><span class="p">:</span> <span class="n">qt_lin</span><span class="p">,</span>
        <span class="s1">&#39;lwr_ci&#39;</span><span class="p">:</span> <span class="n">lwr_ci</span><span class="p">,</span>
        <span class="s1">&#39;upr_ci&#39;</span><span class="p">:</span> <span class="n">upr_ci</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.normest"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.normest.normest" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">normest</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Perform normal parameter estimation using all provided time points at once.
Uses the provided bounds (ignores fixed_params so that all parameters are estimated)
and supports bootstrapping if specified.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="p_data">p_data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Measurement data (DataFrame or numpy array). Assumes data starts at column index 2.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="init_cond">init_cond</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial condition for the ODE solver.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="num_psites">num_psites</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="time_points">time_points</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of time points to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bounds">bounds</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of parameter bounds.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="fixed_params">fixed_params</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(Ignored in normest) Provided for interface consistency.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bootstraps">bootstraps</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of bootstrapping iterations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="use_regularization">use_regularization</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flag to apply Tikhonov regularization.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="lambda_reg">lambda_reg</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization strength.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>est_params: List with the full estimated parameter vector.</li>
</ul>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>model_fits: List with the ODE solution and model predictions.</li>
</ul>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>error_vals: List with the squared error (data vs. model prediction).</li>
</ul>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>paramest/normest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">normest</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span>
            <span class="n">bootstraps</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform normal parameter estimation using all provided time points at once.</span>
<span class="sd">    Uses the provided bounds (ignores fixed_params so that all parameters are estimated)</span>
<span class="sd">    and supports bootstrapping if specified.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      - p_data: Measurement data (DataFrame or numpy array). Assumes data starts at column index 2.</span>
<span class="sd">      - init_cond: Initial condition for the ODE solver.</span>
<span class="sd">      - num_psites: Number of phosphorylation sites.</span>
<span class="sd">      - time_points: Array of time points to use.</span>
<span class="sd">      - bounds: Dictionary of parameter bounds.</span>
<span class="sd">      - fixed_params: (Ignored in normest) Provided for interface consistency.</span>
<span class="sd">      - bootstraps: Number of bootstrapping iterations.</span>
<span class="sd">      - use_regularization: Flag to apply Tikhonov regularization.</span>
<span class="sd">      - lambda_reg: Regularization strength.</span>

<span class="sd">    Returns:</span>
<span class="sd">      - est_params: List with the full estimated parameter vector.</span>
<span class="sd">      - model_fits: List with the ODE solution and model predictions.</span>
<span class="sd">      - error_vals: List with the squared error (data vs. model prediction).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">est_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">error_vals</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
        <span class="c1"># Build lower and upper bounds from config.</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="c1"># For phosphorylation parameters: use Ssite bounds.</span>
        <span class="n">lower_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="n">upper_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="c1"># For dephosphorylation parameters: for each combination, use Dsite bounds.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span><span class="p">):</span>
                <span class="n">lower_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">upper_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># If using log scale, transform bounds (ensure lower bounds &gt; 0)</span>
        <span class="n">eps</span> <span class="o">=</span> <span class="mf">1e-8</span>  <span class="c1"># small epsilon to avoid log(0)</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">b</span><span class="p">,</span> <span class="n">eps</span><span class="p">))</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">lower_bounds_full</span><span class="p">]</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">upper_bounds_full</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Existing approach for distributive or successive models.</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">model_func</span><span class="p">(</span><span class="n">tpts</span><span class="p">,</span> <span class="o">*</span><span class="n">params</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Define the model function for curve fitting.</span>

<span class="sd">        :param tpts:</span>
<span class="sd">        :param params:</span>
<span class="sd">        :return: model predictions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
            <span class="n">param_vec</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">params</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">param_vec</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">p_fitted</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">param_vec</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">atleast_1d</span><span class="p">(</span><span class="n">tpts</span><span class="p">))</span>
        <span class="n">y_model</span> <span class="o">=</span> <span class="n">p_fitted</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
            <span class="n">reg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">lambda_reg</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">y_model</span><span class="p">,</span> <span class="n">reg</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">y_model</span>

    <span class="n">free_bounds</span> <span class="o">=</span> <span class="p">(</span><span class="n">lower_bounds_full</span><span class="p">,</span> <span class="n">upper_bounds_full</span><span class="p">)</span>

    <span class="c1"># Set initial guess for all parameters (midpoint of bounds).</span>
    <span class="n">p0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="n">l</span> <span class="o">+</span> <span class="n">u</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">u</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">free_bounds</span><span class="p">)])</span>

    <span class="c1"># Build the target vector from the measured data.</span>
    <span class="n">target</span> <span class="o">=</span> <span class="n">p_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">target_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p0</span><span class="p">))])</span> <span class="k">if</span> <span class="n">use_regularization</span> <span class="k">else</span> <span class="n">target</span>

    <span class="n">default_sigma</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target_fit</span><span class="p">),</span> <span class="mf">1e-5</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Attempt to get a good initial estimate using curve_fit.</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                      <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">target_fit</span><span class="p">,</span> <span class="n">x_scale</span><span class="o">=</span><span class="s1">&#39;jac&#39;</span><span class="p">,</span>
                      <span class="n">p0</span><span class="o">=</span><span class="n">p0</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">default_sigma</span><span class="p">,</span>
                      <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
        <span class="n">popt_init</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Normal initial estimation failed: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">popt_init</span> <span class="o">=</span> <span class="n">p0</span>

    <span class="c1"># Get weights for the model fitting.</span>
    <span class="n">early_weights</span> <span class="o">=</span> <span class="n">early_emphasis</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span>
    <span class="n">weight_options</span> <span class="o">=</span> <span class="n">get_weight_options</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span>
                                        <span class="n">use_regularization</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">p0</span><span class="p">),</span> <span class="n">early_weights</span><span class="p">)</span>

    <span class="n">scores</span><span class="p">,</span> <span class="n">popts</span><span class="p">,</span> <span class="n">pcovs</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{},</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">wname</span><span class="p">,</span> <span class="n">sigma</span> <span class="ow">in</span> <span class="n">weight_options</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Attempt to fit the model using the specified weights.</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                          <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">target_fit</span><span class="p">,</span> <span class="n">p0</span><span class="o">=</span><span class="n">popt_init</span><span class="p">,</span>
                          <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span> <span class="n">x_scale</span><span class="o">=</span><span class="s1">&#39;jac&#39;</span><span class="p">,</span>
                          <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
            <span class="n">popt</span><span class="p">,</span> <span class="n">pcov</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Fit failed for </span><span class="si">{</span><span class="n">wname</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">popt</span> <span class="o">=</span> <span class="n">popt_init</span>
            <span class="n">pcov</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">popts</span><span class="p">[</span><span class="n">wname</span><span class="p">]</span> <span class="o">=</span> <span class="n">popt</span>
        <span class="n">pcovs</span><span class="p">[</span><span class="n">wname</span><span class="p">]</span> <span class="o">=</span> <span class="n">pcov</span>
        <span class="n">pred</span> <span class="o">=</span> <span class="n">model_func</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="o">*</span><span class="n">popt</span><span class="p">)</span>
        <span class="c1"># Calculate the score for the fit.</span>
        <span class="n">scores</span><span class="p">[</span><span class="n">wname</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_fit</span><span class="p">(</span><span class="n">target_fit</span><span class="p">,</span> <span class="n">pred</span><span class="p">,</span> <span class="n">popt</span><span class="p">)</span>

    <span class="c1"># Select the best weight based on the score.</span>
    <span class="n">best_weight</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">scores</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>
    <span class="n">best_score</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">best_weight</span><span class="p">]</span>
    <span class="c1"># Get the best parameters and covariance matrix.</span>
    <span class="n">popt_best</span> <span class="o">=</span> <span class="n">popts</span><span class="p">[</span><span class="n">best_weight</span><span class="p">]</span>
    <span class="n">pcov_best</span> <span class="o">=</span> <span class="n">pcovs</span><span class="p">[</span><span class="n">best_weight</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Best weight: </span><span class="si">{</span><span class="n">best_weight</span><span class="si">}</span><span class="s2"> with score: </span><span class="si">{</span><span class="n">best_score</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Get confidence intervals for the best parameters.</span>
    <span class="n">ci_results</span> <span class="o">=</span> <span class="n">confidence_intervals</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">popt_best</span><span class="p">)</span> <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span> <span class="k">else</span> <span class="n">popt_best</span><span class="p">,</span>
        <span class="n">pcov_best</span><span class="p">,</span>
        <span class="n">target</span><span class="p">,</span>
        <span class="n">alpha_val</span><span class="o">=</span><span class="n">ALPHA_CI</span>
    <span class="p">)</span>

    <span class="c1"># Bootstrapping</span>
    <span class="n">boot_estimates</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">boot_covariances</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">bootstraps</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Performing bootstrapping with </span><span class="si">{</span><span class="n">bootstraps</span><span class="si">}</span><span class="s2"> iterations&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bootstraps</span><span class="p">):</span>
            <span class="n">noise</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">target_fit</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">noisy_target</span> <span class="o">=</span> <span class="n">target_fit</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">noise</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Attempt to fit the model using the noisy target.</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                              <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">noisy_target</span><span class="p">,</span>
                                        <span class="n">p0</span><span class="o">=</span><span class="n">popt_best</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">default_sigma</span><span class="p">,</span>
                                        <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
                <span class="n">popt_bs</span><span class="p">,</span> <span class="n">pcov_bs</span> <span class="o">=</span> <span class="n">result</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Bootstrapping iteration failed: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">popt_bs</span> <span class="o">=</span> <span class="n">popt_best</span>
                <span class="n">pcov_bs</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">boot_estimates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">popt_bs</span><span class="p">)</span>
            <span class="n">boot_covariances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pcov_bs</span><span class="p">)</span>

        <span class="c1"># Convert boot_estimates to an array and compute the mean parameter estimates.</span>
        <span class="n">popt_best</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">boot_estimates</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Process bootstrap covariance matrices:</span>
        <span class="c1"># Only include iterations where pcov_bs is not None.</span>
        <span class="n">valid_covs</span> <span class="o">=</span> <span class="p">[</span><span class="n">cov</span> <span class="k">for</span> <span class="n">cov</span> <span class="ow">in</span> <span class="n">boot_covariances</span> <span class="k">if</span> <span class="n">cov</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">valid_covs</span><span class="p">:</span>
            <span class="c1"># Compute an average covariance matrix from the valid ones.</span>
            <span class="n">pcov_best</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">valid_covs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pcov_best</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># Compute confidence intervals for the bootstrapped estimates.</span>
        <span class="n">ci_results</span> <span class="o">=</span> <span class="n">confidence_intervals</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">popt_best</span><span class="p">)</span> <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span> <span class="k">else</span> <span class="n">popt_best</span><span class="p">,</span>
            <span class="n">pcov_best</span><span class="p">,</span>
            <span class="n">target</span><span class="p">,</span>
            <span class="n">alpha_val</span><span class="o">=</span><span class="n">ALPHA_CI</span>
        <span class="p">)</span>
    <span class="c1"># Since all parameters are free, param_final is simply the best-fit vector.</span>
    <span class="c1"># If parameters were estimated in log-space, convert them back.</span>
    <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
        <span class="n">param_final</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">popt_best</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">param_final</span> <span class="o">=</span> <span class="n">popt_best</span>
    <span class="n">est_params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">param_final</span><span class="p">)</span>
    <span class="n">sol</span><span class="p">,</span> <span class="n">p_fit</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">param_final</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span>
    <span class="n">model_fits</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sol</span><span class="p">,</span> <span class="n">p_fit</span><span class="p">))</span>
    <span class="n">error_vals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">target</span> <span class="o">-</span> <span class="n">p_fit</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span> <span class="o">**</span> <span class="mi">2</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">est_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">error_vals</span> <span class="c1">#, ci_results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.seqest"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.seqest.fit_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fit_parameters</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">p0_free</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">weight_options</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Fit the parameters of the model using curve fitting with different
weighting options. This function iterates over the provided
weight options, performs the curve fitting, and evaluates the
goodness of fit using a scoring function. The best fitting
parameters and their corresponding score are returned.</p>
<p>:param time_points:
:param p_data:
:param model_func:
:param p0_free:
:param bounds:
:param weight_options:
:param use_regularization:
:return: best_fit_params, best_weight_key, scores</p>


            <details class="quote">
              <summary>Source code in <code>paramest/seqest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">fit_parameters</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">p0_free</span><span class="p">,</span>
                   <span class="n">bounds</span><span class="p">,</span> <span class="n">weight_options</span><span class="p">,</span>
                   <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fit the parameters of the model using curve fitting with different</span>
<span class="sd">    weighting options. This function iterates over the provided</span>
<span class="sd">    weight options, performs the curve fitting, and evaluates the</span>
<span class="sd">    goodness of fit using a scoring function. The best fitting</span>
<span class="sd">    parameters and their corresponding score are returned.</span>

<span class="sd">    :param time_points:</span>
<span class="sd">    :param p_data:</span>
<span class="sd">    :param model_func:</span>
<span class="sd">    :param p0_free:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param weight_options:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :return: best_fit_params, best_weight_key, scores</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">scores</span><span class="p">,</span> <span class="n">popts</span> <span class="o">=</span> <span class="p">{},</span> <span class="p">{}</span>
    <span class="n">target</span> <span class="o">=</span> <span class="n">p_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p0_free</span><span class="p">))])</span>

    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">sigma</span> <span class="ow">in</span> <span class="n">weight_options</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                          <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span>
                          <span class="n">p0</span><span class="o">=</span><span class="n">p0_free</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span>
                          <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span> <span class="n">absolute_sigma</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                          <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
            <span class="n">popt</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fit failed with </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">popt</span> <span class="o">=</span> <span class="n">p0_free</span>

        <span class="n">popts</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">popt</span>
        <span class="n">prediction</span> <span class="o">=</span> <span class="n">model_func</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="o">*</span><span class="n">popt</span><span class="p">)</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">score_fit</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">prediction</span><span class="p">,</span> <span class="n">popt</span><span class="p">)</span>
        <span class="n">scores</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">] Score: </span><span class="si">{</span><span class="n">score</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">best_key</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">scores</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>
    <span class="n">best_score</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="n">best_key</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Score: </span><span class="si">{</span><span class="n">best_score</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">popts</span><span class="p">[</span><span class="n">best_key</span><span class="p">],</span> <span class="n">best_key</span><span class="p">,</span> <span class="n">scores</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="paramest.seqest.prepare_model_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prepare the model function for sequential parameter estimation.</p>
<p>This function builds the model function based on the specified
ODE model type and the number of phosphorylation sites. It also
sets up the bounds for the free parameters and handles fixed
parameters. The model function is used for curve fitting to
estimate the parameters of the ODE model.</p>
<p>:param num_psites:
:param init_cond:
:param bounds:
:param fixed_params:
:param use_regularization:
:param lambda_reg:
:return: model_func, free_indices, free_bounds, fixed_values, num_total_params</p>


            <details class="quote">
              <summary>Source code in <code>paramest/seqest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span>
                       <span class="n">use_regularization</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare the model function for sequential parameter estimation.</span>

<span class="sd">    This function builds the model function based on the specified</span>
<span class="sd">    ODE model type and the number of phosphorylation sites. It also</span>
<span class="sd">    sets up the bounds for the free parameters and handles fixed</span>
<span class="sd">    parameters. The model function is used for curve fitting to</span>
<span class="sd">    estimate the parameters of the ODE model.</span>

<span class="sd">    :param num_psites:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :param lambda_reg:</span>
<span class="sd">    :return: model_func, free_indices, free_bounds, fixed_values, num_total_params</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
        <span class="c1"># Build lower and upper bounds from config.</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="c1"># For phosphorylation parameters: use Ssite bounds.</span>
        <span class="n">lower_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="n">upper_bounds_full</span> <span class="o">+=</span> <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="c1"># For dephosphorylation parameters: for each combination, use Dsite bounds.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span><span class="p">):</span>
                <span class="n">lower_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">upper_bounds_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">num_total_params</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lower_bounds_full</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Existing approach for distributive or successive models.</span>
        <span class="n">num_total_params</span> <span class="o">=</span> <span class="mi">4</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="n">lower_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>
        <span class="n">upper_bounds_full</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Ssite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span>
            <span class="p">[</span><span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;Dsite&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>
        <span class="p">)</span>

    <span class="n">fixed_values</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">free_indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">param_names</span> <span class="o">=</span> <span class="n">get_param_names</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">param_names</span><span class="p">):</span>
        <span class="n">val</span> <span class="o">=</span> <span class="n">fixed_params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">val</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fixed_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">free_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">model_func</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="o">*</span><span class="n">p_free</span><span class="p">):</span>
        <span class="n">p_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">)</span>
        <span class="n">free_iter</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">p_free</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">):</span>
            <span class="n">p_full</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fixed_values</span> <span class="k">else</span> <span class="nb">next</span><span class="p">(</span><span class="n">free_iter</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">p_fitted</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">p_full</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">atleast_1d</span><span class="p">(</span><span class="n">t</span><span class="p">))</span>
        <span class="n">y_model</span> <span class="o">=</span> <span class="n">p_fitted</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
            <span class="n">reg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">lambda_reg</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">p_free</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">y_model</span><span class="p">,</span> <span class="n">reg</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">y_model</span>

    <span class="n">free_bounds</span> <span class="o">=</span> <span class="p">([</span><span class="n">lower_bounds_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">free_indices</span><span class="p">],</span>
                   <span class="p">[</span><span class="n">upper_bounds_full</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">free_indices</span><span class="p">])</span>
    <span class="c1"># logger.info(f&quot;Model Built&quot;)</span>
    <span class="k">return</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">free_indices</span><span class="p">,</span> <span class="n">free_bounds</span><span class="p">,</span> <span class="n">fixed_values</span><span class="p">,</span> <span class="n">num_total_params</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="paramest.seqest.sequential_estimation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sequential_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Perform sequential parameter estimation for a given gene using
the specified model function. This function iteratively fits the
model to the data at each time point, updating the initial guess
for the parameters based on the previous fit. The estimated
parameters, model fits, and error values are returned.</p>
<p>:param p_data:
:param time_points:
:param init_cond:
:param bounds:
:param fixed_params:
:param num_psites:
:param gene:
:param use_regularization:
:param lambda_reg:
:return:</p>


            <details class="quote">
              <summary>Source code in <code>paramest/seqest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">sequential_estimation</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span>
                          <span class="n">fixed_params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span>
                          <span class="n">use_regularization</span><span class="o">=</span><span class="n">USE_REGULARIZATION</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="o">=</span><span class="n">LAMBDA_REG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform sequential parameter estimation for a given gene using</span>
<span class="sd">    the specified model function. This function iteratively fits the</span>
<span class="sd">    model to the data at each time point, updating the initial guess</span>
<span class="sd">    for the parameters based on the previous fit. The estimated</span>
<span class="sd">    parameters, model fits, and error values are returned.</span>

<span class="sd">    :param p_data:</span>
<span class="sd">    :param time_points:</span>
<span class="sd">    :param init_cond:</span>
<span class="sd">    :param bounds:</span>
<span class="sd">    :param fixed_params:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param gene:</span>
<span class="sd">    :param use_regularization:</span>
<span class="sd">    :param lambda_reg:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">est_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">error_vals</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>

    <span class="n">model_func</span><span class="p">,</span> <span class="n">free_indices</span><span class="p">,</span> <span class="n">free_bounds</span><span class="p">,</span> <span class="n">fixed_values</span><span class="p">,</span> <span class="n">num_total_params</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">prepare_model_func</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span>
                           <span class="n">use_regularization</span><span class="p">,</span> <span class="n">lambda_reg</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">p0_free</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="n">lb</span> <span class="o">+</span> <span class="n">ub</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">free_bounds</span><span class="p">)])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">time_points</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">t_now</span> <span class="o">=</span> <span class="n">time_points</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span>
        <span class="n">y_now</span> <span class="o">=</span> <span class="n">p_data</span><span class="p">[:,</span> <span class="p">:</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">p_data</span><span class="o">.</span><span class="n">ndim</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">p_data</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">y_flat</span> <span class="o">=</span> <span class="n">y_now</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">use_regularization</span><span class="p">:</span>
            <span class="n">target_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">y_flat</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p0_free</span><span class="p">))])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">target_fit</span> <span class="o">=</span> <span class="n">y_flat</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Attempt to get a good initial estimate using curve_fit.</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
                          <span class="n">curve_fit</span><span class="p">(</span><span class="n">model_func</span><span class="p">,</span> <span class="n">t_now</span><span class="p">,</span> <span class="n">target_fit</span><span class="p">,</span>
                          <span class="n">p0</span><span class="o">=</span><span class="n">p0_free</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">free_bounds</span><span class="p">,</span>
                          <span class="n">maxfev</span><span class="o">=</span><span class="mi">20000</span><span class="p">))</span>
            <span class="n">popt_init</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initial fit failed at time index </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> for gene </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">popt_init</span> <span class="o">=</span> <span class="n">p0_free</span>

        <span class="c1"># Get weights for the model fitting.</span>
        <span class="n">early_emphasis_weights</span> <span class="o">=</span> <span class="n">early_emphasis</span><span class="p">(</span><span class="n">y_now</span><span class="p">,</span> <span class="n">t_now</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="n">get_weight_options</span><span class="p">(</span><span class="n">y_flat</span><span class="p">,</span> <span class="n">t_now</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span>
                                     <span class="n">use_regularization</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">p0_free</span><span class="p">),</span> <span class="n">early_emphasis_weights</span><span class="p">)</span>

        <span class="c1"># Perform the fit with the best weights.</span>
        <span class="n">best_fit</span><span class="p">,</span> <span class="n">weight_key</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fit_parameters</span><span class="p">(</span><span class="n">t_now</span><span class="p">,</span> <span class="n">y_now</span><span class="p">,</span> <span class="n">model_func</span><span class="p">,</span> <span class="n">popt_init</span><span class="p">,</span>
                                                 <span class="n">free_bounds</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span>
                                                 <span class="n">use_regularization</span><span class="p">)</span>
        <span class="n">p_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">)</span>
        <span class="c1"># Fill in the fixed parameters and free parameters.</span>
        <span class="n">free_iter</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">best_fit</span><span class="p">)</span>
        <span class="c1"># Iterate over the total parameters and assign values.</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_total_params</span><span class="p">):</span>
            <span class="c1"># If the parameter is fixed, use the fixed value.</span>
            <span class="n">p_full</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">fixed_values</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="k">if</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">fixed_values</span> <span class="k">else</span> <span class="nb">next</span><span class="p">(</span><span class="n">free_iter</span><span class="p">)</span>
        <span class="c1"># Append the estimated parameters and model fit.</span>
        <span class="n">est_params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p_full</span><span class="p">)</span>
        <span class="c1"># Solve the ODE with the estimated parameters.</span>
        <span class="n">sol</span><span class="p">,</span> <span class="n">p_fit</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">p_full</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">t_now</span><span class="p">)</span>
        <span class="c1"># Flatten the model fit for error calculation.</span>
        <span class="n">model_fits</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sol</span><span class="p">,</span> <span class="n">p_fit</span><span class="p">))</span>
        <span class="c1"># Calculate the mean square error value.</span>
        <span class="n">error_vals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">y_flat</span> <span class="o">-</span> <span class="n">p_fit</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span> <span class="o">**</span> <span class="mi">2</span><span class="p">))</span>
        <span class="c1"># Update the initial guess for the next iteration.</span>
        <span class="n">p0_free</span> <span class="o">=</span> <span class="n">best_fit</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">] Time Index </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> Best Weight = </span><span class="si">{</span><span class="n">weight_key</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">est_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">error_vals</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="paramest.toggle"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="paramest.toggle.estimate_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">estimate_parameters</span><span class="p">(</span><span class="n">mode</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Toggle between sequential and normal (all timepoints) estimation.</p>
<p>This function allows for the selection of the estimation mode
and handles the parameter estimation process accordingly.</p>
<p>It uses the sequential estimation method for "sequential" mode
and the normal estimation method for "normal" mode.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="mode">mode</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The estimation mode, either "sequential" or "normal".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="gene">gene</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The gene name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="p_data">p_data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Measurement data (DataFrame or numpy array).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="init_cond">init_cond</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial condition for the ODE solver.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="num_psites">num_psites</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="time_points">time_points</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of time points to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bounds">bounds</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of parameter bounds.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="fixed_params">fixed_params</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of fixed parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>-</code>
            </td>
            <td>
                  <code><span title="bootstraps">bootstraps</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of bootstrapping iterations (only used in normal mode).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>:returns:
    - model_fits: List with the ODE solution and model predictions.
    - estimated_params: List with the full estimated parameter vector.
    - seq_model_fit: Sequential model fit for the gene.
    - errors: Error metrics (MSE, MAE).</p>


            <details class="quote">
              <summary>Source code in <code>paramest/toggle.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">estimate_parameters</span><span class="p">(</span><span class="n">mode</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">bootstraps</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Toggle between sequential and normal (all timepoints) estimation.</span>

<span class="sd">    This function allows for the selection of the estimation mode</span>
<span class="sd">    and handles the parameter estimation process accordingly.</span>

<span class="sd">    It uses the sequential estimation method for &quot;sequential&quot; mode</span>
<span class="sd">    and the normal estimation method for &quot;normal&quot; mode.</span>

<span class="sd">    Args:</span>
<span class="sd">        - mode: The estimation mode, either &quot;sequential&quot; or &quot;normal&quot;.</span>
<span class="sd">        - gene: The gene name.</span>
<span class="sd">        - p_data: Measurement data (DataFrame or numpy array).</span>
<span class="sd">        - init_cond: Initial condition for the ODE solver.</span>
<span class="sd">        - num_psites: Number of phosphorylation sites.</span>
<span class="sd">        - time_points: Array of time points to use.</span>
<span class="sd">        - bounds: Dictionary of parameter bounds.</span>
<span class="sd">        - fixed_params: Dictionary of fixed parameters.</span>
<span class="sd">        - bootstraps: Number of bootstrapping iterations (only used in normal mode).</span>
<span class="sd">    :returns:</span>
<span class="sd">        - model_fits: List with the ODE solution and model predictions.</span>
<span class="sd">        - estimated_params: List with the full estimated parameter vector.</span>
<span class="sd">        - seq_model_fit: Sequential model fit for the gene.</span>
<span class="sd">        - errors: Error metrics (MSE, MAE).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;sequential&quot;</span><span class="p">:</span>

        <span class="c1"># For sequential estimation, we need to set up the model function</span>
        <span class="n">estimated_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">errors</span> <span class="o">=</span> <span class="n">sequential_estimation</span><span class="p">(</span>
            <span class="n">p_data</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">fixed_params</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">gene</span>
        <span class="p">)</span>

        <span class="c1"># For sequential estimation, assemble the fitted predictions at each time point:</span>
        <span class="n">seq_model_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_psites</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">time_points</span><span class="p">)))</span>

        <span class="c1"># Iterate over the model fits and extract the last column (predictions)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">P_fitted</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">model_fits</span><span class="p">):</span>
            <span class="c1"># P_fitted is expected to be of shape (num_psites, len(time_points))</span>
            <span class="n">seq_model_fit</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">P_fitted</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;normal&quot;</span><span class="p">:</span>

        <span class="c1"># For normal estimation, we use the provided bounds and fixed parameters</span>
        <span class="n">estimated_params</span><span class="p">,</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">errors</span> <span class="o">=</span> <span class="n">normest</span><span class="p">(</span>
            <span class="n">gene</span><span class="p">,</span> <span class="n">p_data</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">bootstraps</span>
        <span class="p">)</span>

        <span class="c1"># For normal estimation, model_fits[0][1] is already an array of shape (num_psites, len(time_points))</span>
        <span class="n">seq_model_fit</span> <span class="o">=</span> <span class="n">model_fits</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid estimation mode. Choose &#39;sequential&#39; or &#39;normal&#39;.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">model_fits</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">,</span> <span class="n">errors</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="plotting.plotting"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="plotting.plotting.Plotter" class="doc doc-heading">
            <code>Plotter</code>


</h2>


    <div class="doc doc-contents ">


        <p>A class to encapsulate plotting functionalities for ODE model analysis.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.gene">gene</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The gene or experiment name.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.out_dir">out_dir</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where plots will be saved.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="plotting.plotting.Plotter.color_palette">color_palette</span></code></td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of color codes used for plotting.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>plotting/plotting.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Plotter</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to encapsulate plotting functionalities for ODE model analysis.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        gene (str): The gene or experiment name.</span>
<span class="sd">        out_dir (str): The directory where plots will be saved.</span>
<span class="sd">        color_palette (list): List of color codes used for plotting.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">OUT_DIR</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span> <span class="o">=</span> <span class="n">out_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span> <span class="o">=</span> <span class="n">COLOR_PALETTE</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_save_fig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fig</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dpi</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">300</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Saves and closes the given matplotlib figure.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_parallel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a parallel coordinates plot for the given solution.</span>

<span class="sd">        :param solution: 2D numpy array of shape (sampels, features)</span>
<span class="sd">        :param labels: list of labels</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">labels</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">parallel_coordinates</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">class_column</span><span class="o">=</span><span class="s1">&#39;Time&#39;</span><span class="p">,</span> <span class="n">colormap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;States&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Values&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Time Points&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper right&quot;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_parallel_coordinates_.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">pca_components</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a scree plot showing the explained variance ratio for PCA components.</span>

<span class="sd">        :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">        :param target_variance: The target cumulative explained variance to determine the required number of components.</span>
<span class="sd">        :return: A tuple containing the number of required components and the explained variance ratio.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">solution</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="n">pca</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
        <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
        <span class="n">cumulative_explained_variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span>
        <span class="n">required_components</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">cumulative_explained_variance</span> <span class="o">&gt;=</span> <span class="n">target_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">explained_variance</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
               <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Individual&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cumulative_explained_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cumulative_explained_variance</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
                <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Cumulative&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">required_components</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
                   <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">required_components</span><span class="si">}</span><span class="s1"> Components&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Principal Component&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Explained Variance (%)&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_scree_plot_.png&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">required_components</span><span class="p">,</span> <span class="n">explained_variance</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the PCA results for the given solution.</span>

<span class="sd">        :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">        :param components: Number of PCA components to plot. Defaults to 3.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">components</span><span class="p">)</span>
        <span class="n">pca_result</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
        <span class="n">ev</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">components</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span> <span class="o">=</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>
            <span class="n">cs_x</span><span class="p">,</span> <span class="n">cs_y</span><span class="p">,</span> <span class="n">cs_z</span> <span class="o">=</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">y</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">z</span><span class="p">)</span>
            <span class="n">si</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
            <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">indices</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Time Index&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cs_x</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_y</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_z</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Temporal Path&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="n">zi</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">)):</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="n">zi</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC1 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC2 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC3 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_pca_plot_.png&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Optionally handle non-3D cases here</span>
            <span class="k">pass</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_tsne</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a t-SNE visualization of the given solution.</span>

<span class="sd">        :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">        :param perplexity: Perplexity parameter for t-SNE. Defaults to 30.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">perplexity</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">perplexity</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">tsne_result</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="n">perplexity</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">tsne_result</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">tsne_result</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">))</span>
        <span class="n">cs_x</span><span class="p">,</span> <span class="n">cs_y</span> <span class="o">=</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
        <span class="n">si</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">indices</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cs_x</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_y</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Temporal Path&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;t-SNE 1&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;t-SNE 2&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_tsne_plot_.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_param_bar</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">s_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a bar chart of parameter values for the given gene.</span>

<span class="sd">        This method visualizes the estimated parameter values for a specific gene</span>
<span class="sd">        and its phosphorylation sites. It uses color coding to distinguish between</span>
<span class="sd">        different phosphorylation sites and other parameters.</span>

<span class="sd">        :param params_df: DataFrame containing parameter values.</span>
<span class="sd">        :param s_df: DataFrame containing phosphorylation site information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">unique_psites</span> <span class="o">=</span> <span class="n">s_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">s_df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">psite</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab20</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">)}</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">single_psite</span> <span class="o">=</span> <span class="n">unique_psites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">color_map</span><span class="p">[</span><span class="n">single_psite</span><span class="p">]</span>
            <span class="k">if</span> <span class="s1">&#39;S&#39;</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">single_psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">):</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">color_map</span><span class="p">[</span><span class="n">psite</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
                           <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">psite</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">h</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span>
                                                             <span class="n">ax</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">other_params</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span>
                        <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Protein&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>
                                                                                                             <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>
                                                                                                             <span class="nb">range</span><span class="p">(</span>
                                                                                                                 <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))]]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">other_params</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Paired</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">other_params</span><span class="p">)),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Parameters&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_bar_.png&quot;</span><span class="p">)</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">plot_param_series</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">param_names</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the time series of estimated parameters over the given time points.</span>

<span class="sd">        This method visualizes the evolution of kinetic rates or parameters</span>
<span class="sd">        over time for a specific gene.</span>

<span class="sd">        :param estimated_params: List of estimated parameter values at each time point.</span>
<span class="sd">        :param param_names: List of parameter names corresponding to the estimated parameters.</span>
<span class="sd">        :param time_points: 1D numpy array of time points.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">arr</span><span class="p">[:,</span> <span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">param_names</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Kinetic Rates&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;best&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_series_.png&quot;</span><span class="p">)</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">plot_profiles</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the profiles of estimated parameters over time.</span>

<span class="sd">        :param data: DataFrame containing the estimated parameters and time points.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s2">&quot;Time&quot;</span><span class="p">:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Time&quot;</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="n">col</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">col</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (min)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Kinetic Rates&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_profiles.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_model_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">sol</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                       <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the model fit for the given data.</span>

<span class="sd">        :param model_fit: Estimated model fit values.</span>
<span class="sd">        :param P_data: Observed data for phosphorylation levels.</span>
<span class="sd">        :param sol: ODE solution for mRNA and protein levels.</span>
<span class="sd">        :param num_psites: number of phosphorylation sites.</span>
<span class="sd">        :param psite_labels: labels for the phosphorylation sites.</span>
<span class="sd">        :param time_points: time points for the data.</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;mRNA (R)&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Protein (P)&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">P_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">model_fit</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1"> (model)&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation level (FC)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_fit_.png&quot;</span><span class="p">)</span>

        <span class="c1"># Plot using Plotly for an interactive version.</span>
        <span class="n">fig_plotly</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">()</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">sol</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mRNA(R)(model)&#39;</span><span class="p">,</span>
            <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="p">))</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">sol</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Protein(P)(model)&#39;</span><span class="p">,</span>
            <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
        <span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
            <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">P_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">P_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
                <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
                <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
                <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">dash</span><span class="o">=</span><span class="s1">&#39;dash&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="p">))</span>
            <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
                <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="n">model_fit</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
                <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
                <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1"> (model)&#39;</span><span class="p">,</span>
                <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="p">))</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span>
                                 <span class="n">xaxis_title</span><span class="o">=</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">,</span>
                                 <span class="n">yaxis_title</span><span class="o">=</span><span class="s2">&quot;Phosphorylation level (FC)&quot;</span><span class="p">,</span>
                                 <span class="n">template</span><span class="o">=</span><span class="s2">&quot;plotly_white&quot;</span><span class="p">,</span>
                                 <span class="n">width</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">900</span><span class="p">)</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_fit_.html&quot;</span><span class="p">))</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_A_S</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">est_arr</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">time_vals</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the scatter plot of A vs S and the density contour plot.</span>

<span class="sd">        :param est_arr: Estimated parameters array.</span>
<span class="sd">        :param num_psites: Number of phosphorylation sites.</span>
<span class="sd">        :param time_vals: Time values for the data.</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">est_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">est_arr</span><span class="p">)</span>
        <span class="n">A_vals</span> <span class="o">=</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;viridis&quot;</span><span class="p">)</span>
        <span class="n">norm</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">time_vals</span><span class="p">),</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">time_vals</span><span class="p">))</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">legend_handles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
            <span class="n">S_vals</span> <span class="o">=</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
            <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">S_vals</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">time_vals</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span>
                            <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">available_markers</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">slope</span><span class="p">,</span> <span class="n">intercept</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">S_vals</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">x_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">A_vals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">A_vals</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
            <span class="n">y_fit</span> <span class="o">=</span> <span class="n">slope</span> <span class="o">*</span> <span class="n">x_fit</span> <span class="o">+</span> <span class="n">intercept</span>
            <span class="n">line_color</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;C</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_fit</span><span class="p">,</span> <span class="n">y_fit</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">legend_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Line2D</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                         <span class="n">marker</span><span class="o">=</span><span class="n">available_markers</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                         <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">,</span>
                                         <span class="n">markerfacecolor</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span>
                                         <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span>
                                         <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                                         <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;A (mRNA production rate)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;S (Phosphorylation rate)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">cbar</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Time (min)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">legend_handles</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_scatter_A_S_.png&quot;</span><span class="p">)</span>

        <span class="c1"># Density contour plot for A and S.</span>
        <span class="n">all_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
        <span class="n">kde</span> <span class="o">=</span> <span class="n">gaussian_kde</span><span class="p">(</span><span class="n">all_points</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="n">A_lin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">A_vals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">A_vals</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
        <span class="n">all_S</span> <span class="o">=</span> <span class="n">all_points</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">S_lin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">all_S</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">all_S</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
        <span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">A_lin</span><span class="p">,</span> <span class="n">S_lin</span><span class="p">)</span>
        <span class="n">grid_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">A_grid</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">S_grid</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
        <span class="n">density</span> <span class="o">=</span> <span class="n">kde</span><span class="p">(</span><span class="n">grid_coords</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">A_grid</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">all_points</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">all_points</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="n">contourf</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">contourf</span><span class="p">(</span><span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span><span class="p">,</span> <span class="n">density</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;inferno&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">contour</span><span class="p">(</span><span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span><span class="p">,</span> <span class="n">density</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">CONTOUR_LEVELS</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">contourf</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">cbar</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Density&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_density_A_S_.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_all</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
                 <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                 <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Function that calls parallel, t-SNE, PCA, and model fit plots.</span>
<span class="sd">        If mode is sequential, it also calls parameter series and A-S plots.</span>

<span class="sd">        :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">        :param labels: List of labels for the solution.</span>
<span class="sd">        :param estimated_params: List of estimated parameter values.</span>
<span class="sd">        :param time_points: 1D numpy array of time points.</span>
<span class="sd">        :param P_data: Observed data for phosphorylation levels.</span>
<span class="sd">        :param seq_model_fit: Estimated model fit values.</span>
<span class="sd">        :param psite_labels: Labels for the phosphorylation sites.</span>
<span class="sd">        :param perplexity: Perplexity parameter for t-SNE.</span>
<span class="sd">        :param components: Number of PCA components to plot.</span>
<span class="sd">        :param target_variance: The target cumulative explained variance to determine the required number of components.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_parallel</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_tsne</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="n">perplexity</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="n">components</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pca_components</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">target_variance</span><span class="o">=</span><span class="n">target_variance</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_model_fit</span><span class="p">(</span><span class="n">seq_model_fit</span><span class="p">,</span> <span class="n">P_data</span><span class="p">,</span> <span class="n">solution</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">),</span> <span class="n">psite_labels</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ESTIMATION_MODE</span> <span class="o">==</span> <span class="s1">&#39;sequential&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_param_series</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="n">get_param_names</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">)),</span> <span class="n">time_points</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_A_S</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">),</span> <span class="n">time_points</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_clusters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s_values_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">cluster_labels</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the clusters of S values for the given gene.</span>
<span class="sd">        Expects s_values_df to have columns &#39;S_value&#39;, &#39;GeneID&#39;, and &#39;Psite&#39;.</span>

<span class="sd">        :param s_values_df: DataFrame containing S values and gene information.</span>
<span class="sd">        :param cluster_labels: Cluster labels for each S value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">s_values_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_labels</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;S_value&#39;</span><span class="p">],</span> <span class="n">hue</span><span class="o">=</span><span class="n">cluster_labels</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;S_value&#39;</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_protein_clusters.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">param_value_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Expects param_value_df to have a &#39;Protein&#39; column.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">param_value_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;Protein&#39;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Protein&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_heatmap_protein.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_error_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">error_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Expects error_df to have a &#39;MAE&#39; column.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">error_df</span><span class="p">[</span><span class="s1">&#39;MAE&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;MSE&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">error_df</span><span class="p">[</span><span class="s1">&#39;MAE&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;MAE&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Error&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_error.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_1</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Expects merged_data to contain &#39;GeneID&#39;, &#39;Psite&#39;, and columns &#39;x1_obs&#39; to &#39;x14_obs&#39; and &#39;x1_est&#39; to &#39;x14_est&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">gene_handles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">obs_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">est_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                   <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                   <span class="n">obs_array</span><span class="p">,</span> <span class="n">est_array</span><span class="p">):</span>
            <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
            <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                       <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">handle</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">gene_handles</span><span class="p">]:</span>
                <span class="n">handle</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                                    <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
        <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">gene_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_1.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_2</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">gene_handles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">obs_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">est_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                   <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                   <span class="n">obs_array</span><span class="p">,</span> <span class="n">est_array</span><span class="p">):</span>
            <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
            <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                       <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">handle</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">gene_handles</span><span class="p">]:</span>
                <span class="n">handle</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                                    <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
        <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Expand axis limits slightly</span>
        <span class="n">x_min</span> <span class="o">=</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">gene_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_2.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">text_annotations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">obs_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">est_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                   <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                   <span class="n">obs_array</span><span class="p">,</span> <span class="n">est_array</span><span class="p">):</span>
            <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
            <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                       <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                    <span class="n">txt</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                                  <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                                  <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                    <span class="n">text_annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">txt</span><span class="p">)</span>
                    <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">adjust_text</span><span class="p">(</span><span class="n">text_annotations</span><span class="p">,</span> <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;gof_3.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_4</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">text_annotations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">obs_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">est_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                                   <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                                   <span class="n">obs_array</span><span class="p">,</span> <span class="n">est_array</span><span class="p">):</span>
            <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
            <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                       <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                    <span class="n">txt</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                                  <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                                  <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                    <span class="n">text_annotations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">txt</span><span class="p">)</span>
                    <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
                <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Expand axis limits</span>
        <span class="n">x_min</span> <span class="o">=</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">adjust_text</span><span class="p">(</span><span class="n">text_annotations</span><span class="p">,</span> <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_4.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_5</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Uses the row means of observed (x1_obs:x14_obs) and estimated (x1_est:x14_est) values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;Observed_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;Estimated_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                        <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                        <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()),</span>
                  <span class="nb">max</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_5.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_6</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;Observed_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;Estimated_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">overall_std</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
        <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

        <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">sorted_indices</span><span class="p">:</span>
            <span class="n">obs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="n">est</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                        <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                        <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
                <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">x_min</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>
        <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
                <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_6.png&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_kld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Expects merged_data to have columns &#39;x1_obs&#39; to &#39;x14_obs&#39; and &#39;x1_est&#39; to &#39;x14_est&#39;,</span>
<span class="sd">        as well as &#39;GeneID&#39; and &#39;Psite&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">obs_data</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span>
        <span class="n">est_data</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span>
        <span class="n">normalized_obs</span> <span class="o">=</span> <span class="n">obs_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">obs_data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">normalized_est</span> <span class="o">=</span> <span class="n">est_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">est_data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">kl_div</span> <span class="o">=</span> <span class="n">normalized_obs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">entropy</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">normalized_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">kl_df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">kl_df</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kl_div</span><span class="o">.</span><span class="n">values</span>
        <span class="n">kl_by_gene</span> <span class="o">=</span> <span class="n">kl_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;GeneID&#39;</span><span class="p">)[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">kl_by_gene</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">kl_by_gene</span><span class="o">.</span><span class="n">values</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Mean Normalized&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Entropy&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Kullback-Liebler Divergence&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_kld.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.pca_components" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pca_components</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">target_variance</span><span class="o">=</span><span class="mf">0.99</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a scree plot showing the explained variance ratio for PCA components.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param target_variance: The target cumulative explained variance to determine the required number of components.
:return: A tuple containing the number of required components and the explained variance ratio.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">pca_components</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a scree plot showing the explained variance ratio for PCA components.</span>

<span class="sd">    :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">    :param target_variance: The target cumulative explained variance to determine the required number of components.</span>
<span class="sd">    :return: A tuple containing the number of required components and the explained variance ratio.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">solution</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
    <span class="n">pca</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
    <span class="n">explained_variance</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
    <span class="n">cumulative_explained_variance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span>
    <span class="n">required_components</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">cumulative_explained_variance</span> <span class="o">&gt;=</span> <span class="n">target_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">explained_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">explained_variance</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
           <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Individual&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cumulative_explained_variance</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cumulative_explained_variance</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Cumulative&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">required_components</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">required_components</span><span class="si">}</span><span class="s1"> Components&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Principal Component&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Explained Variance (%)&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_scree_plot_.png&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">required_components</span><span class="p">,</span> <span class="n">explained_variance</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_A_S" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_A_S</span><span class="p">(</span><span class="n">est_arr</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_vals</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the scatter plot of A vs S and the density contour plot.</p>
<p>:param est_arr: Estimated parameters array.
:param num_psites: Number of phosphorylation sites.
:param time_vals: Time values for the data.
:return:</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_A_S</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">est_arr</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">time_vals</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the scatter plot of A vs S and the density contour plot.</span>

<span class="sd">    :param est_arr: Estimated parameters array.</span>
<span class="sd">    :param num_psites: Number of phosphorylation sites.</span>
<span class="sd">    :param time_vals: Time values for the data.</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">est_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">est_arr</span><span class="p">)</span>
    <span class="n">A_vals</span> <span class="o">=</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;viridis&quot;</span><span class="p">)</span>
    <span class="n">norm</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">time_vals</span><span class="p">),</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">time_vals</span><span class="p">))</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">legend_handles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="n">S_vals</span> <span class="o">=</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]</span>
        <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">S_vals</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">time_vals</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span>
                        <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">available_markers</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">slope</span><span class="p">,</span> <span class="n">intercept</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">S_vals</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">x_fit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">A_vals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">A_vals</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
        <span class="n">y_fit</span> <span class="o">=</span> <span class="n">slope</span> <span class="o">*</span> <span class="n">x_fit</span> <span class="o">+</span> <span class="n">intercept</span>
        <span class="n">line_color</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;C</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_fit</span><span class="p">,</span> <span class="n">y_fit</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">legend_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Line2D</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                     <span class="n">marker</span><span class="o">=</span><span class="n">available_markers</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                                     <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">,</span>
                                     <span class="n">markerfacecolor</span><span class="o">=</span><span class="n">line_color</span><span class="p">,</span>
                                     <span class="n">markeredgecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span>
                                     <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
                                     <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;A (mRNA production rate)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;S (Phosphorylation rate)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">cbar</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Time (min)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">legend_handles</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_scatter_A_S_.png&quot;</span><span class="p">)</span>

    <span class="c1"># Density contour plot for A and S.</span>
    <span class="n">all_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">A_vals</span><span class="p">,</span> <span class="n">est_arr</span><span class="p">[:,</span> <span class="mi">4</span> <span class="o">+</span> <span class="n">i</span><span class="p">]))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)])</span>
    <span class="n">kde</span> <span class="o">=</span> <span class="n">gaussian_kde</span><span class="p">(</span><span class="n">all_points</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
    <span class="n">A_lin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">A_vals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">A_vals</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
    <span class="n">all_S</span> <span class="o">=</span> <span class="n">all_points</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">S_lin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">all_S</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">all_S</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="mi">100</span><span class="p">)</span>
    <span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">A_lin</span><span class="p">,</span> <span class="n">S_lin</span><span class="p">)</span>
    <span class="n">grid_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">A_grid</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">S_grid</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
    <span class="n">density</span> <span class="o">=</span> <span class="n">kde</span><span class="p">(</span><span class="n">grid_coords</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">A_grid</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">all_points</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">all_points</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">contourf</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">contourf</span><span class="p">(</span><span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span><span class="p">,</span> <span class="n">density</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;inferno&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">contour</span><span class="p">(</span><span class="n">A_grid</span><span class="p">,</span> <span class="n">S_grid</span><span class="p">,</span> <span class="n">density</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">CONTOUR_LEVELS</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">contourf</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">cbar</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s2">&quot;Density&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_density_A_S_.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_all" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_all</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">P_data</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">target_variance</span><span class="o">=</span><span class="mf">0.99</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Function that calls parallel, t-SNE, PCA, and model fit plots.
If mode is sequential, it also calls parameter series and A-S plots.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param labels: List of labels for the solution.
:param estimated_params: List of estimated parameter values.
:param time_points: 1D numpy array of time points.
:param P_data: Observed data for phosphorylation levels.
:param seq_model_fit: Estimated model fit values.
:param psite_labels: Labels for the phosphorylation sites.
:param perplexity: Perplexity parameter for t-SNE.
:param components: Number of PCA components to plot.
:param target_variance: The target cumulative explained variance to determine the required number of components.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_all</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
             <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">seq_model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
             <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">target_variance</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that calls parallel, t-SNE, PCA, and model fit plots.</span>
<span class="sd">    If mode is sequential, it also calls parameter series and A-S plots.</span>

<span class="sd">    :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">    :param labels: List of labels for the solution.</span>
<span class="sd">    :param estimated_params: List of estimated parameter values.</span>
<span class="sd">    :param time_points: 1D numpy array of time points.</span>
<span class="sd">    :param P_data: Observed data for phosphorylation levels.</span>
<span class="sd">    :param seq_model_fit: Estimated model fit values.</span>
<span class="sd">    :param psite_labels: Labels for the phosphorylation sites.</span>
<span class="sd">    :param perplexity: Perplexity parameter for t-SNE.</span>
<span class="sd">    :param components: Number of PCA components to plot.</span>
<span class="sd">    :param target_variance: The target cumulative explained variance to determine the required number of components.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">plot_parallel</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">plot_tsne</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="n">perplexity</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="n">components</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">pca_components</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">target_variance</span><span class="o">=</span><span class="n">target_variance</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">plot_model_fit</span><span class="p">(</span><span class="n">seq_model_fit</span><span class="p">,</span> <span class="n">P_data</span><span class="p">,</span> <span class="n">solution</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">),</span> <span class="n">psite_labels</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ESTIMATION_MODE</span> <span class="o">==</span> <span class="s1">&#39;sequential&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_param_series</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="n">get_param_names</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">)),</span> <span class="n">time_points</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_A_S</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">psite_labels</span><span class="p">),</span> <span class="n">time_points</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_clusters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_clusters</span><span class="p">(</span><span class="n">s_values_df</span><span class="p">,</span> <span class="n">cluster_labels</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the clusters of S values for the given gene.
Expects s_values_df to have columns 'S_value', 'GeneID', and 'Psite'.</p>
<p>:param s_values_df: DataFrame containing S values and gene information.
:param cluster_labels: Cluster labels for each S value.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_clusters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">s_values_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">cluster_labels</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the clusters of S values for the given gene.</span>
<span class="sd">    Expects s_values_df to have columns &#39;S_value&#39;, &#39;GeneID&#39;, and &#39;Psite&#39;.</span>

<span class="sd">    :param s_values_df: DataFrame containing S values and gene information.</span>
<span class="sd">    :param cluster_labels: Cluster labels for each S value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">s_values_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_labels</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;S_value&#39;</span><span class="p">],</span> <span class="n">hue</span><span class="o">=</span><span class="n">cluster_labels</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;S_value&#39;</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_protein_clusters.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_error_distribution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_error_distribution</span><span class="p">(</span><span class="n">error_df</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Expects error_df to have a 'MAE' column.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_error_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">error_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Expects error_df to have a &#39;MAE&#39; column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">error_df</span><span class="p">[</span><span class="s1">&#39;MAE&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;MSE&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">error_df</span><span class="p">[</span><span class="s1">&#39;MAE&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;MAE&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Error&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_error.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_gof_1" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_gof_1</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Expects merged_data to contain 'GeneID', 'Psite', and columns 'x1_obs' to 'x14_obs' and 'x1_est' to 'x14_est'.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_1</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Expects merged_data to contain &#39;GeneID&#39;, &#39;Psite&#39;, and columns &#39;x1_obs&#39; to &#39;x14_obs&#39; and &#39;x1_est&#39; to &#39;x14_est&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">overall_std</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
    <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
    <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

    <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
    <span class="n">gene_color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">gene_handles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">obs_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">est_array</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">psite</span><span class="p">,</span> <span class="n">obs_vals</span><span class="p">,</span> <span class="n">est_vals</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">],</span>
                                               <span class="n">merged_data</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
                                               <span class="n">obs_array</span><span class="p">,</span> <span class="n">est_array</span><span class="p">):</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">obs_vals</span><span class="p">)</span>
        <span class="n">obs_vals_sorted</span> <span class="o">=</span> <span class="n">obs_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">est_vals_sorted</span> <span class="o">=</span> <span class="n">est_vals</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs_vals_sorted</span><span class="p">,</span> <span class="n">est_vals_sorted</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                   <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">handle</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">gene_handles</span><span class="p">]:</span>
            <span class="n">handle</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color_map</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                                <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
    <span class="n">min_val</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">max_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">obs_array</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">est_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span> <span class="p">[</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
            <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_95</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
            <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
            <span class="p">[</span><span class="n">min_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">+</span> <span class="n">ci_offset_99</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">],</span>
            <span class="p">[</span><span class="n">min_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">,</span> <span class="n">max_val</span> <span class="o">-</span> <span class="n">ci_offset_99</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">gene_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_1.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_gof_5" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_gof_5</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Uses the row means of observed (x1_obs:x14_obs) and estimated (x1_est:x14_est) values.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_gof_5</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Uses the row means of observed (x1_obs:x14_obs) and estimated (x1_est:x14_est) values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="s1">&#39;Observed_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;Estimated_Mean&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">overall_std</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
    <span class="n">ci_offset_95</span> <span class="o">=</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">overall_std</span>
    <span class="n">ci_offset_99</span> <span class="o">=</span> <span class="mf">2.576</span> <span class="o">*</span> <span class="n">overall_std</span>

    <span class="n">unique_genes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">))</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="n">palette</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">)}</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">plotted_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_genes</span> <span class="ow">and</span> <span class="p">(</span><span class="n">est</span> <span class="o">&gt;</span> <span class="n">obs</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="ow">or</span> <span class="n">est</span> <span class="o">&lt;</span> <span class="n">obs</span> <span class="o">-</span> <span class="n">ci_offset_95</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">est</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span>
                    <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span>
                    <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">))</span>
            <span class="n">plotted_genes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()),</span>
              <span class="nb">max</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Observed_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Estimated_Mean&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_95</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;99% CI&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">ci_offset_99</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_vals</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Observed&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fitted&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2"> model&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_gof_5.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_heatmap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">param_value_df</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Expects param_value_df to have a 'Protein' column.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">param_value_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Expects param_value_df to have a &#39;Protein&#39; column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">param_value_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="s1">&#39;Protein&#39;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Protein&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_heatmap_protein.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_kld</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Expects merged_data to have columns 'x1_obs' to 'x14_obs' and 'x1_est' to 'x14_est',
as well as 'GeneID' and 'Psite'.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_kld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">merged_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Expects merged_data to have columns &#39;x1_obs&#39; to &#39;x14_obs&#39; and &#39;x1_est&#39; to &#39;x14_est&#39;,</span>
<span class="sd">    as well as &#39;GeneID&#39; and &#39;Psite&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">obs_data</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_obs&#39;</span><span class="p">:</span><span class="s1">&#39;x14_obs&#39;</span><span class="p">]</span>
    <span class="n">est_data</span> <span class="o">=</span> <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1_est&#39;</span><span class="p">:</span><span class="s1">&#39;x14_est&#39;</span><span class="p">]</span>
    <span class="n">normalized_obs</span> <span class="o">=</span> <span class="n">obs_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">obs_data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">normalized_est</span> <span class="o">=</span> <span class="n">est_data</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">est_data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">kl_div</span> <span class="o">=</span> <span class="n">normalized_obs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">entropy</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">normalized_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">kl_df</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">kl_df</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kl_div</span><span class="o">.</span><span class="n">values</span>
    <span class="n">kl_by_gene</span> <span class="o">=</span> <span class="n">kl_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;GeneID&#39;</span><span class="p">)[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">kl_by_gene</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">kl_by_gene</span><span class="o">.</span><span class="n">values</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Mean Normalized&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Entropy&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Kullback-Liebler Divergence&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_kld.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_model_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_model_fit</span><span class="p">(</span><span class="n">model_fit</span><span class="p">,</span> <span class="n">P_data</span><span class="p">,</span> <span class="n">sol</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the model fit for the given data.</p>
<p>:param model_fit: Estimated model fit values.
:param P_data: Observed data for phosphorylation levels.
:param sol: ODE solution for mRNA and protein levels.
:param num_psites: number of phosphorylation sites.
:param psite_labels: labels for the phosphorylation sites.
:param time_points: time points for the data.
:return:</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_model_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model_fit</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">P_data</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">sol</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                   <span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">psite_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the model fit for the given data.</span>

<span class="sd">    :param model_fit: Estimated model fit values.</span>
<span class="sd">    :param P_data: Observed data for phosphorylation levels.</span>
<span class="sd">    :param sol: ODE solution for mRNA and protein levels.</span>
<span class="sd">    :param num_psites: number of phosphorylation sites.</span>
<span class="sd">    :param psite_labels: labels for the phosphorylation sites.</span>
<span class="sd">    :param time_points: time points for the data.</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;mRNA (R)&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">sol</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Protein (P)&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">P_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">model_fit</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1"> (model)&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation level (FC)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_fit_.png&quot;</span><span class="p">)</span>

    <span class="c1"># Plot using Plotly for an interactive version.</span>
    <span class="n">fig_plotly</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">()</span>
    <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="n">sol</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mRNA(R)(model)&#39;</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="p">))</span>
    <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="n">sol</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Protein(P)(model)&#39;</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
    <span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">P_data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">P_data</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">dash</span><span class="o">=</span><span class="s1">&#39;dash&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="p">))</span>
        <span class="n">fig_plotly</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_points</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">model_fit</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;P+</span><span class="si">{</span><span class="n">psite_labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1"> (model)&#39;</span><span class="p">,</span>
            <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_palette</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="p">))</span>
    <span class="n">fig_plotly</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span>
                             <span class="n">xaxis_title</span><span class="o">=</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">,</span>
                             <span class="n">yaxis_title</span><span class="o">=</span><span class="s2">&quot;Phosphorylation level (FC)&quot;</span><span class="p">,</span>
                             <span class="n">template</span><span class="o">=</span><span class="s2">&quot;plotly_white&quot;</span><span class="p">,</span>
                             <span class="n">width</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">900</span><span class="p">)</span>
    <span class="n">fig_plotly</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_model_fit_.html&quot;</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_parallel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_parallel</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a parallel coordinates plot for the given solution.</p>
<p>:param solution: 2D numpy array of shape (sampels, features)
:param labels: list of labels</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_parallel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a parallel coordinates plot for the given solution.</span>

<span class="sd">    :param solution: 2D numpy array of shape (sampels, features)</span>
<span class="sd">    :param labels: list of labels</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">labels</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">parallel_coordinates</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">class_column</span><span class="o">=</span><span class="s1">&#39;Time&#39;</span><span class="p">,</span> <span class="n">colormap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;tab20&quot;</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;States&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Values&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Time Points&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper right&quot;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_parallel_coordinates_.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_param_bar" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_param_bar</span><span class="p">(</span><span class="n">params_df</span><span class="p">,</span> <span class="n">s_df</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a bar chart of parameter values for the given gene.</p>
<p>This method visualizes the estimated parameter values for a specific gene
and its phosphorylation sites. It uses color coding to distinguish between
different phosphorylation sites and other parameters.</p>
<p>:param params_df: DataFrame containing parameter values.
:param s_df: DataFrame containing phosphorylation site information.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_param_bar</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">s_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a bar chart of parameter values for the given gene.</span>

<span class="sd">    This method visualizes the estimated parameter values for a specific gene</span>
<span class="sd">    and its phosphorylation sites. It uses color coding to distinguish between</span>
<span class="sd">    different phosphorylation sites and other parameters.</span>

<span class="sd">    :param params_df: DataFrame containing parameter values.</span>
<span class="sd">    :param s_df: DataFrame containing phosphorylation site information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">unique_psites</span> <span class="o">=</span> <span class="n">s_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">s_df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">psite</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab20</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">)}</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">single_psite</span> <span class="o">=</span> <span class="n">unique_psites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">color_map</span><span class="p">[</span><span class="n">single_psite</span><span class="p">]</span>
        <span class="k">if</span> <span class="s1">&#39;S&#39;</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">single_psite</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">psite</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">):</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">color_map</span><span class="p">[</span><span class="n">psite</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
                       <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">psite</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">psite</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">h</span><span class="o">.</span><span class="n">get_label</span><span class="p">()</span> <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span>
                                                         <span class="n">ax</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>

    <span class="n">other_params</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span>
                    <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Protein&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span>
                                                                                                         <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>
                                                                                                         <span class="nb">range</span><span class="p">(</span>
                                                                                                             <span class="nb">len</span><span class="p">(</span><span class="n">unique_psites</span><span class="p">))]]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">other_params</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">params_df</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="n">params_df</span><span class="p">[</span><span class="n">param</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Paired</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">other_params</span><span class="p">)),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Parameters&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Residue_Position&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_bar_.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_param_series" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_param_series</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">,</span> <span class="n">param_names</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the time series of estimated parameters over the given time points.</p>
<p>This method visualizes the evolution of kinetic rates or parameters
over time for a specific gene.</p>
<p>:param estimated_params: List of estimated parameter values at each time point.
:param param_names: List of parameter names corresponding to the estimated parameters.
:param time_points: 1D numpy array of time points.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_param_series</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">estimated_params</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">param_names</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">time_points</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the time series of estimated parameters over the given time points.</span>

<span class="sd">    This method visualizes the evolution of kinetic rates or parameters</span>
<span class="sd">    over time for a specific gene.</span>

<span class="sd">    :param estimated_params: List of estimated parameter values at each time point.</span>
<span class="sd">    :param param_names: List of parameter names corresponding to the estimated parameters.</span>
<span class="sd">    :param time_points: 1D numpy array of time points.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">estimated_params</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">arr</span><span class="p">[:,</span> <span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">param_names</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Kinetic Rates&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;best&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_series_.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">components</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the PCA results for the given solution.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param components: Number of PCA components to plot. Defaults to 3.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the PCA results for the given solution.</span>

<span class="sd">    :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">    :param components: Number of PCA components to plot. Defaults to 3.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">components</span><span class="p">)</span>
    <span class="n">pca_result</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
    <span class="n">ev</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">components</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span> <span class="o">=</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">pca_result</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>
        <span class="n">cs_x</span><span class="p">,</span> <span class="n">cs_y</span><span class="p">,</span> <span class="n">cs_z</span> <span class="o">=</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">y</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">z</span><span class="p">)</span>
        <span class="n">si</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">,</span> <span class="n">projection</span><span class="o">=</span><span class="s1">&#39;3d&#39;</span><span class="p">)</span>
        <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">indices</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Time Index&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cs_x</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_y</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_z</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Temporal Path&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="n">zi</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">)):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="n">zi</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC1 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC2 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PC3 (</span><span class="si">{</span><span class="n">ev</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_pca_plot_.png&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Optionally handle non-3D cases here</span>
        <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_profiles" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_profiles</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the profiles of estimated parameters over time.</p>
<p>:param data: DataFrame containing the estimated parameters and time points.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_profiles</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the profiles of estimated parameters over time.</span>

<span class="sd">    :param data: DataFrame containing the estimated parameters and time points.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s2">&quot;Time&quot;</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Time&quot;</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="n">col</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">col</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (min)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Kinetic Rates&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_params_profiles.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="plotting.plotting.Plotter.plot_tsne" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_tsne</span><span class="p">(</span><span class="n">solution</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a t-SNE visualization of the given solution.</p>
<p>:param solution: 2D numpy array of shape (samples, features) representing the data.
:param perplexity: Perplexity parameter for t-SNE. Defaults to 30.</p>


            <details class="quote">
              <summary>Source code in <code>plotting/plotting.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_tsne</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">solution</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">perplexity</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a t-SNE visualization of the given solution.</span>

<span class="sd">    :param solution: 2D numpy array of shape (samples, features) representing the data.</span>
<span class="sd">    :param perplexity: Perplexity parameter for t-SNE. Defaults to 30.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">perplexity</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">perplexity</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">tsne_result</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="n">perplexity</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">tsne_result</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">tsne_result</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">))</span>
    <span class="n">cs_x</span><span class="p">,</span> <span class="n">cs_y</span> <span class="o">=</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">x</span><span class="p">),</span> <span class="n">CubicSpline</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
    <span class="n">si</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">solution</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">indices</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cs_x</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">cs_y</span><span class="p">(</span><span class="n">si</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Temporal Path&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">yi</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;t-SNE 1&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;t-SNE 2&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_save_fig</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">_tsne_plot_.png&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="processing.cleanup"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.format_site" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_site</span><span class="p">(</span><span class="n">site</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Formats a phosphorylation site string.</p>
<p>If the input is NaN or an empty string, returns an empty string.
If the input contains an underscore ('_'), splits the string into two parts,
converts the first part to uppercase, and appends the second part unchanged.
Otherwise, converts the entire string to uppercase.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>site</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The phosphorylation site string to format.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The formatted phosphorylation site string.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_site</span><span class="p">(</span><span class="n">site</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Formats a phosphorylation site string.</span>

<span class="sd">    If the input is NaN or an empty string, returns an empty string.</span>
<span class="sd">    If the input contains an underscore (&#39;_&#39;), splits the string into two parts,</span>
<span class="sd">    converts the first part to uppercase, and appends the second part unchanged.</span>
<span class="sd">    Otherwise, converts the entire string to uppercase.</span>

<span class="sd">    Args:</span>
<span class="sd">        site (str): The phosphorylation site string to format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The formatted phosphorylation site string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">site</span><span class="p">)</span> <span class="ow">or</span> <span class="n">site</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span>
    <span class="k">if</span> <span class="s1">&#39;_&#39;</span> <span class="ow">in</span> <span class="n">site</span><span class="p">:</span>
        <span class="n">before</span><span class="p">,</span> <span class="n">after</span> <span class="o">=</span> <span class="n">site</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">before</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">after</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">site</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.move_processed_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">move_processed_files</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Moves or copies processed files to their respective directories.</p>
<p>This function organizes processed files into specific directories for
transcription factor optimization (TFOpt) and kinase optimization (KinOpt).
It creates the target directories if they do not exist and moves or copies
the files based on whether they have already been moved.</p>


<details class="directories" open>
  <summary>Directories</summary>
  <ul>
<li><code>../tfopt/data</code>: Target directory for TFOpt files.</li>
<li><code>../kinopt/data</code>: Target directory for KinOpt files.</li>
</ul>
</details>

<details class="files" open>
  <summary>Files</summary>
  <ul>
<li>KinOpt files: ["input1.csv", "input2.csv"]</li>
<li>TFOpt files: ["input1.csv", "input3.csv", "input4.csv"]</li>
</ul>
</details>

<details class="behavior" open>
  <summary>Behavior</summary>
  <ul>
<li>If a file has already been moved, it is copied to the target directory.</li>
<li>If a file has not been moved, it is moved to the target directory.</li>
<li>If a file does not exist, a message is printed.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>Messages indicating whether files were moved, copied, or not found.</li>
</ul>
</details>

            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">move_processed_files</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Moves or copies processed files to their respective directories.</span>

<span class="sd">    This function organizes processed files into specific directories for</span>
<span class="sd">    transcription factor optimization (TFOpt) and kinase optimization (KinOpt).</span>
<span class="sd">    It creates the target directories if they do not exist and moves or copies</span>
<span class="sd">    the files based on whether they have already been moved.</span>

<span class="sd">    Directories:</span>
<span class="sd">        - `../tfopt/data`: Target directory for TFOpt files.</span>
<span class="sd">        - `../kinopt/data`: Target directory for KinOpt files.</span>

<span class="sd">    Files:</span>
<span class="sd">        - KinOpt files: [&quot;input1.csv&quot;, &quot;input2.csv&quot;]</span>
<span class="sd">        - TFOpt files: [&quot;input1.csv&quot;, &quot;input3.csv&quot;, &quot;input4.csv&quot;]</span>

<span class="sd">    Behavior:</span>
<span class="sd">        - If a file has already been moved, it is copied to the target directory.</span>
<span class="sd">        - If a file has not been moved, it is moved to the target directory.</span>
<span class="sd">        - If a file does not exist, a message is printed.</span>

<span class="sd">    Prints:</span>
<span class="sd">        - Messages indicating whether files were moved, copied, or not found.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Create a new directory if it doesn&#39;t exist</span>
    <span class="n">tf_data_dir</span> <span class="o">=</span> <span class="s2">&quot;../tfopt/data&quot;</span>
    <span class="n">kin_data_dir</span> <span class="o">=</span> <span class="s2">&quot;../kinopt/data&quot;</span>

    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">tf_data_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">kin_data_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># List of files to move</span>
    <span class="n">kinopt_files</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;input1.csv&quot;</span><span class="p">,</span>
        <span class="s2">&quot;input2.csv&quot;</span>
    <span class="p">]</span>
    <span class="n">tfopt_files</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;input1.csv&quot;</span><span class="p">,</span>
        <span class="c1"># &quot;input1_wstd.csv&quot;,</span>
        <span class="s2">&quot;input3.csv&quot;</span><span class="p">,</span>
        <span class="s2">&quot;input4.csv&quot;</span>
    <span class="p">]</span>

    <span class="c1"># Track files already moved so we copy them next time</span>
    <span class="n">moved_files</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># Move or copy files to their respective directories</span>
    <span class="k">for</span> <span class="n">file_list</span><span class="p">,</span> <span class="n">target_dir</span> <span class="ow">in</span> <span class="p">[(</span><span class="n">kinopt_files</span><span class="p">,</span> <span class="n">kin_data_dir</span><span class="p">),</span> <span class="p">(</span><span class="n">tfopt_files</span><span class="p">,</span> <span class="n">tf_data_dir</span><span class="p">)]:</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">file_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
                <span class="n">target_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_dir</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">moved_files</span><span class="p">:</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_dir</span><span class="p">,</span> <span class="n">f</span><span class="p">))</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Copied </span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">target_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">target_path</span><span class="p">)</span>
                    <span class="n">moved_files</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Moved </span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">target_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2"> does not exist in the current directory or has already been moved.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.process_collecttri" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_collecttri</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Processes the CollecTRI file to clean and filter mRNA-TF interactions.
Removes complex interactions, filters by target genes, and saves the result.</p>


            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_collecttri</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the CollecTRI file to clean and filter mRNA-TF interactions.</span>
<span class="sd">    Removes complex interactions, filters by target genes, and saves the result.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load CollecTRI.csv and keep only the source and target columns</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;CollecTRI.csv&quot;</span><span class="p">))</span>

    <span class="c1"># Remove thr rows that starts with &#39;COMPLEX&#39; in the source column</span>
    <span class="c1"># Model doesn&#39;t support complex interactions, ony single mRNA</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COMPLEX&#39;</span><span class="p">)]</span>

    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;source_genesymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;target_genesymbol&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;source_genesymbol&#39;</span><span class="p">:</span> <span class="s1">&#39;Source&#39;</span><span class="p">,</span> <span class="s1">&#39;target_genesymbol&#39;</span><span class="p">:</span> <span class="s1">&#39;Target&#39;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="c1"># Remove rows with NaN, empty strings or whitespace, and drop duplicates</span>
    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df_readable</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df_readable</span><span class="p">[</span><span class="n">df_readable</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df_readable</span><span class="p">[</span><span class="n">df_readable</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df_readable</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="c1"># Load phospho-kinase interaction data and search CollectTRI for same TFs</span>
    <span class="n">df_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;input2.csv&quot;</span><span class="p">))</span>
    <span class="n">df_genes</span> <span class="o">=</span> <span class="n">df_genes</span><span class="p">[[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="s1">&#39;Target&#39;</span><span class="p">})</span>
    <span class="n">df_genes</span> <span class="o">=</span> <span class="n">df_genes</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
    <span class="n">df_genes</span> <span class="o">=</span> <span class="n">df_genes</span><span class="p">[</span><span class="n">df_genes</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
    <span class="n">df_genes</span> <span class="o">=</span> <span class="n">df_genes</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="c1"># Keep only interactions where Target (TFs) is present in input2.csv</span>
    <span class="n">df_readable</span> <span class="o">=</span> <span class="n">df_readable</span><span class="p">[</span><span class="n">df_readable</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">df_genes</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">])]</span>

    <span class="c1"># Save the cleaned mRNA - TFs interactions to input4.csv</span>
    <span class="n">df_readable</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;input4.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Copy the input2.csv file to the current directory using shutil</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;input2.csv&quot;</span><span class="p">),</span> <span class="s2">&quot;input2.csv&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saved TF-mRNA interactions to input4.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.process_msgauss" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_msgauss</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to generate time series data.</p>
<p>This function performs the following steps:
1. Loads the <code>MS_Gaussian_updated_09032023.csv</code> file.
2. Computes the transformed values as 2^(predict_mean).
3. Pivots the data to create a time series for each (GeneID, Psite) pair.
4. Renames the time point columns to <code>x1</code> to <code>x14</code> and formats the <code>Psite</code> column.
5. Saves the cleaned time series to <code>input1.csv</code>.
6. Filters the data to keep only rows where <code>Psite</code> starts with <code>Y_</code>, <code>S_</code>, <code>T_</code>, or is empty.
7. Saves the filtered time series to <code>input1.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input1.csv</code>: Cleaned and filtered time series data.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_msgauss</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the MS Gaussian data file to generate time series data.</span>

<span class="sd">    This function performs the following steps:</span>
<span class="sd">    1. Loads the `MS_Gaussian_updated_09032023.csv` file.</span>
<span class="sd">    2. Computes the transformed values as 2^(predict_mean).</span>
<span class="sd">    3. Pivots the data to create a time series for each (GeneID, Psite) pair.</span>
<span class="sd">    4. Renames the time point columns to `x1` to `x14` and formats the `Psite` column.</span>
<span class="sd">    5. Saves the cleaned time series to `input1.csv`.</span>
<span class="sd">    6. Filters the data to keep only rows where `Psite` starts with `Y_`, `S_`, `T_`, or is empty.</span>
<span class="sd">    7. Saves the filtered time series to `input1.csv`.</span>

<span class="sd">    Outputs:</span>
<span class="sd">        - `input1.csv`: Cleaned and filtered time series data.</span>

<span class="sd">    Prints:</span>
<span class="sd">        - A message indicating that the time series data has been saved.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the input file is not found in the specified directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the MS_Gaussian_updated_09032023.csv file</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;MS_Gaussian_updated_09032023.csv&quot;</span><span class="p">))</span>

    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="c1"># Compute 2^(predict_mean)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_trans&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_mean&#39;</span><span class="p">]</span>

    <span class="c1"># Pivot so that each (GeneID, Psite) pair has time series values per unit_time (0-13)</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
        <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;unit_time&#39;</span><span class="p">,</span>
        <span class="n">values</span><span class="o">=</span><span class="s1">&#39;predict_trans&#39;</span><span class="p">,</span>
        <span class="n">aggfunc</span><span class="o">=</span><span class="s1">&#39;first&#39;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># Rename time point columns to x1 to x14 and format the Psite column</span>
    <span class="n">new_names</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">14</span><span class="p">)}</span>
    <span class="n">pivot_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">new_names</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">pivot_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">format_site</span><span class="p">)</span>

    <span class="c1"># Save the cleaned time series to input1.csv</span>
    <span class="n">pivot_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;input1.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Filter to keep only rows where &#39;Psite&#39; starts with Y_, S_, T_, or is empty</span>
    <span class="n">kinopt_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span>
        <span class="n">pivot_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s1">&#39;Y_&#39;</span><span class="p">,</span> <span class="s1">&#39;S_&#39;</span><span class="p">,</span> <span class="s1">&#39;T_&#39;</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span><span class="n">pivot_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span>

    <span class="c1"># Save the filtered time series to input1.csv</span>
    <span class="n">kinopt_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;input1.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saved MS Gaussian (predict_mean) time series to input1.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.process_msgauss_std" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_msgauss_std</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to compute transformed means and standard deviations.</p>
<p>This function performs the following steps:
1. Loads the <code>MS_Gaussian_updated_09032023.csv</code> file.
2. Computes transformed values as 2^(predict_mean) and propagates errors using the formula:
   sigma_y = 2^(x) * ln(2) * sigma_x.
3. Pivots the data to create time series for each (GeneID, Psite) pair for both means and standard deviations.
4. Merges the pivoted data for means and standard deviations.
5. Filters the data to keep only rows where <code>Psite</code> starts with <code>Y_</code>, <code>S_</code>, <code>T_</code>, or is empty.
6. Saves the resulting data to <code>input1_wstd.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input1_wstd.csv</code>: Cleaned time series data with transformed means and standard deviations.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data with standard deviations has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_msgauss_std</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the MS Gaussian data file to compute transformed means and standard deviations.</span>

<span class="sd">    This function performs the following steps:</span>
<span class="sd">    1. Loads the `MS_Gaussian_updated_09032023.csv` file.</span>
<span class="sd">    2. Computes transformed values as 2^(predict_mean) and propagates errors using the formula:</span>
<span class="sd">       sigma_y = 2^(x) * ln(2) * sigma_x.</span>
<span class="sd">    3. Pivots the data to create time series for each (GeneID, Psite) pair for both means and standard deviations.</span>
<span class="sd">    4. Merges the pivoted data for means and standard deviations.</span>
<span class="sd">    5. Filters the data to keep only rows where `Psite` starts with `Y_`, `S_`, `T_`, or is empty.</span>
<span class="sd">    6. Saves the resulting data to `input1_wstd.csv`.</span>

<span class="sd">    Outputs:</span>
<span class="sd">        - `input1_wstd.csv`: Cleaned time series data with transformed means and standard deviations.</span>

<span class="sd">    Prints:</span>
<span class="sd">        - A message indicating that the time series data with standard deviations has been saved.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the input file is not found in the specified directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;MS_Gaussian_updated_09032023.csv&quot;</span><span class="p">))</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_trans&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_mean&#39;</span><span class="p">]</span>

    <span class="c1"># Error propagation: sigma_y = 2^(x) * ln(2) * sigma_x</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_trans_std&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_trans&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;predict_std&#39;</span><span class="p">]</span>

    <span class="c1"># Pivot for the transformed means</span>
    <span class="n">pivot_trans</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
        <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;unit_time&#39;</span><span class="p">,</span>
        <span class="n">values</span><span class="o">=</span><span class="s1">&#39;predict_trans&#39;</span><span class="p">,</span>
        <span class="n">aggfunc</span><span class="o">=</span><span class="s1">&#39;first&#39;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="n">pivot_trans</span> <span class="o">=</span> <span class="n">pivot_trans</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">14</span><span class="p">)})</span>

    <span class="c1"># Pivot for the transformed standard deviations</span>
    <span class="n">pivot_std</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">],</span>
        <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;unit_time&#39;</span><span class="p">,</span>
        <span class="n">values</span><span class="o">=</span><span class="s1">&#39;predict_trans_std&#39;</span><span class="p">,</span>
        <span class="n">aggfunc</span><span class="o">=</span><span class="s1">&#39;first&#39;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">pivot_std</span> <span class="o">=</span> <span class="n">pivot_std</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;x</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">_std&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">14</span><span class="p">)})</span>

    <span class="c1"># Merge the two pivot tables and format Psite</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pivot_trans</span><span class="p">,</span> <span class="n">pivot_std</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">,</span> <span class="s1">&#39;Psite&#39;</span><span class="p">])</span>
    <span class="n">result</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">format_site</span><span class="p">)</span>

    <span class="c1"># Filter to keep only rows where &#39;Psite&#39; starts with Y_, S_, T_, or is empty</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s1">&#39;Y_&#39;</span><span class="p">,</span> <span class="s1">&#39;S_&#39;</span><span class="p">,</span> <span class="s1">&#39;T_&#39;</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">)]</span>

    <span class="c1"># Save to input1_wstd.csv</span>
    <span class="n">result</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;input1_wstd.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saved MS Gaussian time-series with standard deviations to input1_wstd.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.process_routlimma" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_routlimma</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Processes the Rout Limma table to generate time series data for mRNA.</p>
<p>This function performs the following steps:
1. Loads the <code>Rout_LimmaTable.csv</code> file.
2. Selects specific columns related to time points and conditions.
3. Renames the selected columns to a standardized format (<code>x1</code> to <code>x9</code>).
4. Converts the values in the renamed columns using the formula <code>2^(value)</code>.
5. Saves the resulting time series data to <code>input3.csv</code>.</p>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li><code>input3.csv</code>: Cleaned and transformed time series data for mRNA.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A message indicating that the time series data has been saved.</li>
</ul>
</details>

<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the input file <code>Rout_LimmaTable.csv</code> is not found in the specified directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_routlimma</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the Rout Limma table to generate time series data for mRNA.</span>

<span class="sd">    This function performs the following steps:</span>
<span class="sd">    1. Loads the `Rout_LimmaTable.csv` file.</span>
<span class="sd">    2. Selects specific columns related to time points and conditions.</span>
<span class="sd">    3. Renames the selected columns to a standardized format (`x1` to `x9`).</span>
<span class="sd">    4. Converts the values in the renamed columns using the formula `2^(value)`.</span>
<span class="sd">    5. Saves the resulting time series data to `input3.csv`.</span>

<span class="sd">    Outputs:</span>
<span class="sd">        - `input3.csv`: Cleaned and transformed time series data for mRNA.</span>

<span class="sd">    Prints:</span>
<span class="sd">        - A message indicating that the time series data has been saved.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the input file `Rout_LimmaTable.csv` is not found in the specified directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load Rout_LimmaTable.csv and select desired columns</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="s2">&quot;Rout_LimmaTable.csv&quot;</span><span class="p">))</span>
    <span class="n">selected_cols</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;GeneID&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Min4vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Min8vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Min15vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Min30vsCtrl&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr1vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Hr2vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Hr4vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Hr8vsCtrl&#39;</span><span class="p">,</span> <span class="s1">&#39;Hr16vsCtrl&#39;</span>
    <span class="p">]</span>
    <span class="n">df_new</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">selected_cols</span><span class="p">]</span>

    <span class="c1"># Rename the condition columns to x1 to x9</span>
    <span class="n">rename_mapping</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Min4vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x1&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Min8vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x2&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Min15vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x3&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Min30vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x4&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr1vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x5&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr2vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x6&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr4vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x7&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr8vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x8&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Hr16vsCtrl&#39;</span><span class="p">:</span> <span class="s1">&#39;x9&#39;</span>
    <span class="p">}</span>
    <span class="n">df_new</span> <span class="o">=</span> <span class="n">df_new</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">rename_mapping</span><span class="p">)</span>

    <span class="c1"># Convert each value in x1 to x9 as 2^(value)</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">rename_mapping</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">df_new</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">**</span> <span class="n">df_new</span><span class="p">[</span><span class="n">col</span><span class="p">]</span>

    <span class="c1"># Save the resulting time series to input3.csv</span>
    <span class="n">df_new</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;input3.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saved Rout Limma time series - mRNA to input3.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="processing.cleanup.update_gene_symbols" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_gene_symbols</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Updates the GeneID column in a CSV file by mapping GeneIDs to gene/protein symbols.</p>
<p>This function reads a CSV file, queries the MyGeneInfo database to map GeneIDs to their
corresponding gene/protein symbols, and writes the updated DataFrame back to the same file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the CSV file to be updated. The file must contain a 'GeneID' column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the specified file does not exist.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the 'GeneID' column is missing in the file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="outputs" open>
  <summary>Outputs</summary>
  <ul>
<li>The input file is updated in place with the 'GeneID' column replaced by gene/protein symbols.</li>
</ul>
</details>

<details class="prints" open>
  <summary>Prints</summary>
  <ul>
<li>A progress bar indicating the mapping process.</li>
<li>A message confirming the update of gene symbols in the file.</li>
</ul>
</details>

            <details class="quote">
              <summary>Source code in <code>processing/cleanup.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">update_gene_symbols</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Updates the GeneID column in a CSV file by mapping GeneIDs to gene/protein symbols.</span>

<span class="sd">    This function reads a CSV file, queries the MyGeneInfo database to map GeneIDs to their</span>
<span class="sd">    corresponding gene/protein symbols, and writes the updated DataFrame back to the same file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str): The path to the CSV file to be updated. The file must contain a &#39;GeneID&#39; column.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the specified file does not exist.</span>
<span class="sd">        ValueError: If the &#39;GeneID&#39; column is missing in the file.</span>

<span class="sd">    Outputs:</span>
<span class="sd">        - The input file is updated in place with the &#39;GeneID&#39; column replaced by gene/protein symbols.</span>

<span class="sd">    Prints:</span>
<span class="sd">        - A progress bar indicating the mapping process.</span>
<span class="sd">        - A message confirming the update of gene symbols in the file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;GeneID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">unique_gene_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Initialize MyGeneInfo client and query in bulk.</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>
    <span class="n">query_results</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span><span class="n">unique_gene_ids</span><span class="p">,</span>
                                 <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;ensembl.gene,entrezgene,symbol&#39;</span><span class="p">,</span>
                                 <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span>
                                 <span class="n">as_dataframe</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Filter out not found results if available.</span>
    <span class="k">if</span> <span class="s1">&#39;notfound&#39;</span> <span class="ow">in</span> <span class="n">query_results</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">query_results</span> <span class="o">=</span> <span class="n">query_results</span><span class="p">[</span><span class="n">query_results</span><span class="p">[</span><span class="s1">&#39;notfound&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="kc">True</span><span class="p">]</span>

    <span class="c1"># Build the mapping dictionary from queried GeneIDs to gene symbols.</span>
    <span class="n">mapping</span> <span class="o">=</span> <span class="n">query_results</span><span class="p">[</span><span class="s1">&#39;symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">map_geneid</span><span class="p">(</span><span class="n">geneid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Maps a single GeneID to its corresponding gene/protein symbol.</span>

<span class="sd">        Args:</span>
<span class="sd">            geneid (str): The GeneID to be mapped.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The corresponding gene/protein symbol if found, otherwise the original GeneID.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">mapping</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">geneid</span><span class="p">),</span> <span class="n">geneid</span><span class="p">)</span>

    <span class="c1"># Use ThreadPoolExecutor with tqdm to parallelize mapping.</span>
    <span class="k">with</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">ThreadPoolExecutor</span><span class="p">()</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">tqdm</span><span class="p">(</span><span class="n">executor</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">map_geneid</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]),</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]),</span>
                            <span class="n">desc</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">))</span>

    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updated gene symbols in </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="processing.postoptmerge"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="processing.postoptmerge.map_optimization_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">map_optimization_results</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Reads the TF-mRNA optimization results from an Excel file and maps mRNA to each TF.</p>
<p>This function processes the 'Alpha Values' sheet of the provided Excel file to extract
non-zero optimization results, groups mRNA by TF, and merges the results with additional
data from a CSV file containing TF, Psite, and Kinase information. The final DataFrame
is cleaned and formatted for further analysis.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the Excel file containing TF-mRNA optimization results.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the mapped TF, mRNA, Psite, and Kinase information.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>processing/postoptmerge.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">map_optimization_results</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads the TF-mRNA optimization results from an Excel file and maps mRNA to each TF.</span>

<span class="sd">    This function processes the &#39;Alpha Values&#39; sheet of the provided Excel file to extract</span>
<span class="sd">    non-zero optimization results, groups mRNA by TF, and merges the results with additional</span>
<span class="sd">    data from a CSV file containing TF, Psite, and Kinase information. The final DataFrame</span>
<span class="sd">    is cleaned and formatted for further analysis.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path (str): The path to the Excel file containing TF-mRNA optimization results.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the mapped TF, mRNA, Psite, and Kinase information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Read the Excel file</span>
    <span class="n">tfopt_file</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelFile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>

    <span class="c1"># Read the &#39;Alpha Values&#39; sheet</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">tfopt_file</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Alpha Values&#39;</span><span class="p">)</span>

    <span class="c1"># Filter the DataFrame for non-zero values</span>
    <span class="n">non_zero_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Filter for positive values</span>
    <span class="c1"># non_zero_df = df[df[&#39;Value&#39;] &gt; 0]</span>

    <span class="c1"># Take all values without filtering</span>
    <span class="c1"># non_zero_df = df</span>

    <span class="c1"># Filter very small values out</span>
    <span class="c1"># non_zero_df = non_zero_df[non_zero_df[&#39;Value&#39;] &gt; 0.0001]</span>

    <span class="c1"># Extract the mRNA for each TF where Values was not zero</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">non_zero_df</span><span class="p">[[</span><span class="s1">&#39;TF&#39;</span><span class="p">,</span> <span class="s1">&#39;mRNA&#39;</span><span class="p">]]</span>

    <span class="c1"># Display value along with TF and mRNA</span>
    <span class="c1"># result[&#39;Value&#39;] = str(non_zero_df[&#39;Value&#39;])</span>

    <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;TF&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">))</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># List number of mRNA for each TF</span>
    <span class="c1"># result[&#39;mRNA_Count&#39;] = result[&#39;mRNA&#39;].apply(lambda x: len(x.split(&#39;,&#39;)))</span>

    <span class="c1"># Read another csv file which has TF aka GeneID and Psites and Kinases to merge with the result</span>
    <span class="n">kinopt_file</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;raw/input2.csv&#39;</span><span class="p">)</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">kinopt_file</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;GeneID&#39;</span><span class="p">:</span> <span class="s1">&#39;TF&#39;</span><span class="p">})</span>
    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;TF&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>

    <span class="c1"># Remove {} from the Kinases column with format {kinase1, kinase2, kinase3}</span>
    <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;{&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;}&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Remove any extra spaces</span>
    <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Delete the rows where Psite doesn&#39;t match Psite in input2</span>
    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">df2</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">])]</span>

    <span class="c1"># Empty cells in mRNA column if they are repeateed in the next row</span>
    <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shift</span><span class="p">(),</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">merged_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">merged_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">merged_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="sensitivity.analysis"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="sensitivity.analysis.define_sensitivity_problem_ds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">define_sensitivity_problem_ds</span><span class="p">(</span><span class="n">ub</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Problem definition for sensitivity analysis.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>sensitivity/analysis.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">define_sensitivity_problem_ds</span><span class="p">(</span><span class="n">ub</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Problem definition for sensitivity analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">num_vars</span> <span class="o">=</span> <span class="mi">4</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">num_psites</span>  <span class="c1"># A, B, C, D, and S1, S2, ..., Sn, D1, D2, ..., Dn</span>
    <span class="n">param_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">]</span> <span class="o">+</span> \
                  <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;S</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span> <span class="o">+</span> \
                  <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;D</span><span class="si">{</span><span class="n">i</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="p">[</span>
                 <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">],</span>  <span class="c1"># A</span>
                 <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">],</span>  <span class="c1"># B</span>
                 <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">],</span>  <span class="c1"># C</span>
                 <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">],</span>  <span class="c1"># D</span>
             <span class="p">]</span> <span class="o">+</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="n">ub</span><span class="p">]]</span> <span class="o">*</span> <span class="n">num_psites</span>  <span class="c1"># S and D parameters</span>
    <span class="n">problem</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;num_vars&#39;</span><span class="p">:</span> <span class="n">num_vars</span><span class="p">,</span>
        <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="n">param_names</span><span class="p">,</span>
        <span class="s1">&#39;bounds&#39;</span><span class="p">:</span> <span class="n">bounds</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">problem</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sensitivity.analysis.define_sensitivity_problem_rand" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">define_sensitivity_problem_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Problem definition for sensitivity analysis.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>sensitivity/analysis.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">define_sensitivity_problem_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Defines the Morris sensitivity analysis problem for a dynamic number of parameters.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Problem definition for sensitivity analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">num_vars</span> <span class="o">=</span> <span class="n">get_number_of_params_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">param_names</span> <span class="o">=</span> <span class="n">get_param_names_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="n">get_bounds_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">problem</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;num_vars&#39;</span><span class="p">:</span> <span class="n">num_vars</span><span class="p">,</span>
        <span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="n">param_names</span><span class="p">,</span>
        <span class="s1">&#39;bounds&#39;</span><span class="p">:</span> <span class="n">bounds</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">problem</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sensitivity.analysis.sensitivity_analysis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sensitivity_analysis</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Performs sensitivity analysis using the Morris method for a given ODE model.</p>
<p>This function defines the sensitivity problem based on the ODE model type,
generates parameter samples, evaluates the model for each sample, and computes
sensitivity indices. It also generates various plots to visualize the results.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>time_points</code>
            </td>
            <td>
                  <code><span title="list">list</span> or <span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time points for the ODE simulation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>init_cond</code>
            </td>
            <td>
                  <code><span title="list">list</span> or <span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial conditions for the ODE model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene or protein being analyzed.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>None</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function saves sensitivity analysis results and plots to the output directory.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>sensitivity/analysis.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">sensitivity_analysis</span><span class="p">(</span><span class="n">time_points</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">gene</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs sensitivity analysis using the Morris method for a given ODE model.</span>

<span class="sd">    This function defines the sensitivity problem based on the ODE model type,</span>
<span class="sd">    generates parameter samples, evaluates the model for each sample, and computes</span>
<span class="sd">    sensitivity indices. It also generates various plots to visualize the results.</span>

<span class="sd">    Args:</span>
<span class="sd">        time_points (list or np.ndarray): Time points for the ODE simulation.</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites in the model.</span>
<span class="sd">        init_cond (list or np.ndarray): Initial conditions for the ODE model.</span>
<span class="sd">        gene (str): Name of the gene or protein being analyzed.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: The function saves sensitivity analysis results and plots to the output directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span><span class="p">:</span>
        <span class="n">problem</span> <span class="o">=</span> <span class="n">define_sensitivity_problem_rand</span><span class="p">(</span><span class="n">num_psites</span><span class="o">=</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">problem</span> <span class="o">=</span> <span class="n">define_sensitivity_problem_ds</span><span class="p">(</span><span class="n">num_psites</span><span class="o">=</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">N</span> <span class="o">=</span> <span class="mi">10000</span>
    <span class="n">num_levels</span> <span class="o">=</span> <span class="mi">400</span>
    <span class="n">param_values</span> <span class="o">=</span> <span class="n">morris</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">,</span> <span class="n">num_levels</span><span class="o">=</span><span class="n">num_levels</span><span class="p">,</span> <span class="n">local_optimization</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">param_values</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">X</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">param_values</span><span class="p">):</span>
        <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="o">*</span><span class="n">rest</span> <span class="o">=</span> <span class="n">X</span>
        <span class="n">S_list</span> <span class="o">=</span> <span class="n">rest</span><span class="p">[:</span><span class="n">num_psites</span><span class="p">]</span>
        <span class="n">D_list</span> <span class="o">=</span> <span class="n">rest</span><span class="p">[</span><span class="n">num_psites</span><span class="p">:]</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span><span class="p">,</span> <span class="o">*</span><span class="n">S_list</span><span class="p">,</span> <span class="o">*</span><span class="n">D_list</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sol</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">solve_ode</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">init_cond</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">time_points</span><span class="p">)</span>
            <span class="c1"># Sum last time point P1, P2, ..., Pn</span>
            <span class="n">Y</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sol</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">))])</span> <span class="k">if</span> <span class="n">ODE_MODEL</span> <span class="o">==</span> <span class="s1">&#39;randmod&#39;</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sol</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="mi">2</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="n">Y</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">posinf</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">neginf</span><span class="o">=</span><span class="mf">0.0</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">Sensitivity Analysis for Protein: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">Si</span> <span class="o">=</span> <span class="n">analyze</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">param_values</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">num_levels</span><span class="o">=</span><span class="n">num_levels</span><span class="p">,</span> <span class="n">conf_level</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">scaled</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">print_to_console</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Absolute Mean of Elementary Effects : represents the overall importance</span>
    <span class="c1"># of each parameter, reflecting its sensitivity</span>
    <span class="c1">## Bar Plot of mu* ##</span>
    <span class="c1"># Standard Deviation of Elementary Effects: High standard deviation suggests</span>
    <span class="c1"># that the parameter has nonlinear effects or is involved in interactions</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">],</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">],</span> <span class="n">yerr</span><span class="o">=</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star_conf&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;skyblue&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;mu* (Importance)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/bar_plot_mu_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">],</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;sigma&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39; (Standard Deviation)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/bar_plot_sigma_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1">## Bar Plot of sigma ##</span>
    <span class="c1"># Distinguish between parameters with purely linear effects (low sigma) and</span>
    <span class="c1"># those with nonlinear or interaction effects (high sigma).</span>
    <span class="c1"># **--- Parameters with high mu* and high sigma ---**</span>
    <span class="c1">#           &lt;particularly important to watch&gt;</span>
    <span class="c1">## Scatter Plot of mu* vs sigma ##</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">],</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;sigma&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;sigma&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">param</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;mu* (Mean Absolute Effect)&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39; (Standard Deviation)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/scatter_plot_musigma_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># A radial plot (also known as a spider or radar plot) can give a visual</span>
    <span class="c1"># overview of multiple sensitivity metrics (e.g., mu*, sigma, etc.) for</span>
    <span class="c1"># each parameter in a circular format.</span>

    <span class="c1"># Each parameter gets a spoke, and the distance from the center represents</span>
    <span class="c1"># the sensitivity for a given metric.</span>
    <span class="c1">## Radial Plot (Spider Plot) of Sensitivity Metrics ##</span>
    <span class="n">categories</span> <span class="o">=</span> <span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span>
    <span class="n">N_cat</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span>
    <span class="n">mu_star</span> <span class="o">=</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">]</span>
    <span class="n">sigma</span> <span class="o">=</span> <span class="n">Si</span><span class="p">[</span><span class="s1">&#39;sigma&#39;</span><span class="p">]</span>
    <span class="n">angles</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span> <span class="n">N_cat</span><span class="p">,</span> <span class="n">endpoint</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">mu_star</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">mu_star</span><span class="p">,</span> <span class="p">[</span><span class="n">mu_star</span><span class="p">[</span><span class="mi">0</span><span class="p">]]))</span>
    <span class="n">sigma</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">sigma</span><span class="p">,</span> <span class="p">[</span><span class="n">sigma</span><span class="p">[</span><span class="mi">0</span><span class="p">]]))</span>
    <span class="n">angles</span> <span class="o">+=</span> <span class="n">angles</span><span class="p">[:</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">subplot_kw</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">polar</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">angles</span><span class="p">,</span> <span class="n">mu_star</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;skyblue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Mu*&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">angles</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Sigma&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">angles</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/radial_plot_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># CDF can show how often the effects of certain parameters are strong or</span>
    <span class="c1"># weak across the model outputs.</span>
    <span class="c1"># Visualizing how many times a parameter has a strong effect across</span>
    <span class="c1"># different sample runs.</span>
    <span class="c1">## Cumulative Distribution Function (CDF) of Sensitivity Indices ##</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">])),</span> <span class="n">label</span><span class="o">=</span><span class="n">param</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Sensitivity Index&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Cumulative Probability&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/cdf_plot_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Visualize the proportion of total sensitivity contributed by each</span>
    <span class="c1"># parameter using a pie chart, showing the relative importance of each</span>
    <span class="c1"># parameter&#39;s contribution to sensitivity.</span>
    <span class="c1">## Pie Chart for Sensitivity Contribution ##</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">Si</span><span class="p">[</span><span class="s1">&#39;mu_star&#39;</span><span class="p">],</span> <span class="n">labels</span><span class="o">=</span><span class="n">problem</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">],</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">startangle</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Paired</span><span class="o">.</span><span class="n">colors</span><span class="p">,</span>
           <span class="n">textprops</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;fontsize&#39;</span><span class="p">:</span> <span class="mi">8</span><span class="p">})</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">OUT_DIR</span><span class="si">}</span><span class="s2">/pie_chart_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="steady.initdist"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="steady.initdist.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for distributive phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>steady/initdist.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the initial steady-state conditions for a given number of phosphorylation sites</span>
<span class="sd">    for distributive phosphorylation model.</span>

<span class="sd">    This function defines a system of equations representing the steady-state conditions</span>
<span class="sd">    of an ODE model and solves it using numerical optimization. The steady-state conditions</span>
<span class="sd">    are used as initial conditions for further simulations.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites in the model.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: A list of steady-state values for the variables [R, P, P_sites].</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the optimization fails to find a solution for the steady-state conditions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>
    <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">D_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Defines the system of equations for the steady-state conditions.</span>

<span class="sd">        Args:</span>
<span class="sd">            y (list or np.ndarray): Current values of the variables [R, P, P_sites].</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: Residuals of the steady-state equations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
        <span class="n">dR_dt</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
        <span class="n">dP_dt</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="p">(</span><span class="n">D</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">S_rates</span><span class="p">))</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">P_sites</span><span class="p">)</span>
        <span class="n">dP_sites_dt</span> <span class="o">=</span> <span class="p">[</span><span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">dR_dt</span><span class="p">,</span> <span class="n">dP_dt</span><span class="p">]</span> <span class="o">+</span> <span class="n">dP_sites_dt</span>

    <span class="n">y0_guess</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">bounds_local</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">1e-6</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="n">y0_guess</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;SLSQP&#39;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">bounds_local</span><span class="p">,</span>
                      <span class="n">constraints</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;eq&#39;</span><span class="p">,</span> <span class="s1">&#39;fun&#39;</span><span class="p">:</span> <span class="n">steady_state_equations</span><span class="p">})</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Steady-State conditions calculated&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to find steady-state conditions&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="steady.initrand"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="steady.initrand.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for random phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>steady/initrand.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the initial steady-state conditions for a given number of phosphorylation sites</span>
<span class="sd">    for random phosphorylation model.</span>

<span class="sd">    This function defines a system of equations representing the steady-state conditions</span>
<span class="sd">    of an ODE model and solves it using numerical optimization. The steady-state conditions</span>
<span class="sd">    are used as initial conditions for further simulations.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites in the model.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: A list of steady-state values for the variables [R, P, P_sites].</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the optimization fails to find a solution for the steady-state conditions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">subsets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">comb</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">k</span><span class="p">):</span>
            <span class="n">subsets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">comb</span><span class="p">)</span>
    <span class="n">subset_to_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">subset</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">subset</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subsets</span><span class="p">)}</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Defines the system of equations for the steady-state conditions.</span>

<span class="sd">        Args:</span>
<span class="sd">            y (list or np.ndarray): Current values of the variables [R, P, P_sites].</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: Residuals of the steady-state equations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">phos</span> <span class="o">=</span> <span class="n">y</span>
        <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>
        <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
        <span class="n">D_params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subsets</span><span class="p">))</span>
        <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
        <span class="n">gain_from_dephos</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">phos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">subset</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subsets</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">S_rates</span><span class="p">)</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="n">gain_from_dephos</span>
        <span class="n">eqs_phos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">subset</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subsets</span><span class="p">):</span>
            <span class="n">P_state</span> <span class="o">=</span> <span class="n">phos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">gain_phos</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">subset</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">gain_phos</span> <span class="o">+=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">reduced</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span> <span class="o">-</span> <span class="p">{</span><span class="n">site</span><span class="p">}))</span>
                    <span class="k">if</span> <span class="n">reduced</span> <span class="ow">in</span> <span class="n">subset_to_index</span><span class="p">:</span>
                        <span class="n">gain_phos</span> <span class="o">+=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">phos</span><span class="p">[</span><span class="n">subset_to_index</span><span class="p">[</span><span class="n">reduced</span><span class="p">]</span> <span class="o">-</span> <span class="mi">2</span><span class="p">]</span>
            <span class="n">loss_phos</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">S_rates</span><span class="p">[</span><span class="n">site</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">site</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">subset</span><span class="p">)</span> <span class="o">*</span> <span class="n">P_state</span>
            <span class="n">basal_loss</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">)</span> <span class="o">*</span> <span class="n">P_state</span>
            <span class="n">additional_loss</span> <span class="o">=</span> <span class="n">D_params</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_state</span>
            <span class="n">gain_from_further</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">site</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">subset</span><span class="p">:</span>
                    <span class="n">new_state</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">subset</span> <span class="o">+</span> <span class="p">(</span><span class="n">site</span><span class="p">,)))</span>
                    <span class="k">if</span> <span class="n">new_state</span> <span class="ow">in</span> <span class="n">subset_to_index</span><span class="p">:</span>
                        <span class="n">gain_from_further</span> <span class="o">+=</span> <span class="n">phos</span><span class="p">[</span><span class="n">subset_to_index</span><span class="p">[</span><span class="n">new_state</span><span class="p">]</span> <span class="o">-</span> <span class="mi">2</span><span class="p">]</span>
            <span class="n">eqs_phos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gain_phos</span> <span class="o">-</span> <span class="n">loss_phos</span> <span class="o">-</span> <span class="n">basal_loss</span> <span class="o">-</span> <span class="n">additional_loss</span> <span class="o">+</span> <span class="n">gain_from_further</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs_phos</span>

    <span class="n">y0_guess</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subsets</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">bounds_local</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">1e-6</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subsets</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="n">y0_guess</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;SLSQP&#39;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">bounds_local</span><span class="p">,</span>
                      <span class="n">constraints</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;eq&#39;</span><span class="p">,</span> <span class="s1">&#39;fun&#39;</span><span class="p">:</span> <span class="n">steady_state_equations</span><span class="p">})</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Steady-State conditions calculated&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to find steady-state conditions&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="steady.initsucc"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="steady.initsucc.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for successive phosphorylation model.</p>
<p>This function defines a system of equations representing the steady-state conditions
of an ODE model and solves it using numerical optimization. The steady-state conditions
are used as initial conditions for further simulations.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>steady/initsucc.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates the initial steady-state conditions for a given number of phosphorylation sites</span>
<span class="sd">    for successive phosphorylation model.</span>

<span class="sd">    This function defines a system of equations representing the steady-state conditions</span>
<span class="sd">    of an ODE model and solves it using numerical optimization. The steady-state conditions</span>
<span class="sd">    are used as initial conditions for further simulations.</span>

<span class="sd">    Args:</span>
<span class="sd">        num_psites (int): Number of phosphorylation sites in the model.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: A list of steady-state values for the variables [R, P, P_sites].</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the optimization fails to find a solution for the steady-state conditions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>
    <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">D_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Defines the system of equations for the steady-state conditions.</span>

<span class="sd">        Args:</span>
<span class="sd">            y (list or np.ndarray): Current values of the variables [R, P, P_sites].</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: Residuals of the steady-state equations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
        <span class="n">dR_dt</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
        <span class="n">dP_dt</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="p">(</span><span class="n">D</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">S_rates</span><span class="p">))</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">P_sites</span><span class="p">)</span>
        <span class="n">dP_sites_dt</span> <span class="o">=</span> <span class="p">[</span><span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)]</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">dR_dt</span><span class="p">,</span> <span class="n">dP_dt</span><span class="p">]</span> <span class="o">+</span> <span class="n">dP_sites_dt</span>

    <span class="n">y0_guess</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">bounds_local</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">1e-6</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">y0_guess</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s1">&#39;SLSQP&#39;</span><span class="p">,</span>
        <span class="n">bounds</span><span class="o">=</span><span class="n">bounds_local</span><span class="p">,</span>
        <span class="n">constraints</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;eq&#39;</span><span class="p">,</span> <span class="s1">&#39;fun&#39;</span><span class="p">:</span> <span class="n">steady_state_equations</span><span class="p">}</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Steady-State conditions calculated&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to find steady-state conditions&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="steady.inittest"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="steady.inittest.initial_condition" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Compute steady-state conditions for the given model.</p>


<details class="common-parameters" open>
  <summary>Common parameters</summary>
  <p>A, B, C, D = 1, 1, 1, 1
S_rates = ones(num_psites)
D_rates = ones(num_psites)</p>
</details>

<details class="the-parameter-vector-for-the-ode-models-is" open>
  <summary>The parameter vector for the ODE models is</summary>
  <p>[A, B, C, D, S1, ..., S_n, D1, ..., D_n]</p>
</details>        <p>Extra (hardcoded) parameters per model:
  - semi_processive: p_proc = 0.8
  - oscillating:      A_osc = 1.0, omega = 0.5 (steady state computed at t=0)
  - noise:            noise_std = 0.01 (set to 0 for steady state)
  - cooperative:      hill_n = 2, K_half = 0.5</p>


            <details class="quote">
              <summary>Source code in <code>steady/inittest.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">initial_condition</span><span class="p">(</span><span class="n">num_psites</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute steady-state conditions for the given model.</span>

<span class="sd">    Common parameters:</span>
<span class="sd">      A, B, C, D = 1, 1, 1, 1</span>
<span class="sd">      S_rates = ones(num_psites)</span>
<span class="sd">      D_rates = ones(num_psites)</span>

<span class="sd">    The parameter vector for the ODE models is:</span>
<span class="sd">      [A, B, C, D, S1, ..., S_n, D1, ..., D_n]</span>

<span class="sd">    Extra (hardcoded) parameters per model:</span>
<span class="sd">      - semi_processive: p_proc = 0.8</span>
<span class="sd">      - oscillating:      A_osc = 1.0, omega = 0.5 (steady state computed at t=0)</span>
<span class="sd">      - noise:            noise_std = 0.01 (set to 0 for steady state)</span>
<span class="sd">      - cooperative:      hill_n = 2, K_half = 0.5</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>
    <span class="n">S_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>
    <span class="n">D_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">TEST_MODEL</span> <span class="o">==</span> <span class="s1">&#39;semi_processive&#39;</span><span class="p">:</span>
        <span class="c1"># Differential Equations (Semi-processive):</span>
        <span class="c1">#   dR/dt = A - BR</span>
        <span class="c1">#   dP/dt = CR - DP - p_procS1P + P1</span>
        <span class="c1">#   For n = 1: dP1/dt = p_procS1P - (1 + D1)P1</span>
        <span class="c1">#   For n &gt; 1:</span>
        <span class="c1">#     dP1/dt = p_procS1P - (1 + p_procS2 + D1)P1 + P2</span>
        <span class="c1">#     dP/dt = p_procSP - (1 + p_procS + D)P + P  (i=2..n-1)</span>
        <span class="c1">#     dP/dt = p_procSP - (1 + D)P</span>
        <span class="n">p_proc</span> <span class="o">=</span> <span class="n">PROCESSIVITY</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
            <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
            <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="p">(</span><span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">,</span> <span class="n">eq_P1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_P1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">eq_pi</span> <span class="o">=</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span>
                        <span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_pi</span><span class="p">)</span>
                <span class="n">eq_last</span> <span class="o">=</span> <span class="n">p_proc</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_last</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs</span>

    <span class="k">elif</span> <span class="n">TEST_MODEL</span> <span class="o">==</span> <span class="s1">&#39;neg_feedback&#39;</span><span class="p">:</span>
        <span class="n">k_fb</span> <span class="o">=</span> <span class="n">NEGATIVE_FEEDBACK_CONSTANT</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
            <span class="c1"># Compute the feedback factor: f = 1/(1 + k_fb * (P1 + P2 + ... + Pn))</span>
            <span class="n">f</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">k_fb</span> <span class="o">*</span> <span class="nb">sum</span><span class="p">(</span><span class="n">P_sites</span><span class="p">))</span>
            <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
            <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="p">(</span><span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">,</span> <span class="n">eq_P1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_P1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">eq_pi</span> <span class="o">=</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span>
                        <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_pi</span><span class="p">)</span>
                <span class="n">eq_last</span> <span class="o">=</span> <span class="n">f</span> <span class="o">*</span> <span class="n">S_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_last</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs</span>

    <span class="k">elif</span> <span class="n">TEST_MODEL</span> <span class="o">==</span> <span class="s1">&#39;site_affinity&#39;</span><span class="p">:</span>
        <span class="c1"># Differential Equations (Site-specific Binding Affinity):</span>
        <span class="c1">#   dR/dt = A - BR</span>
        <span class="c1">#   dP/dt = CR - DP - S1P + P1</span>
        <span class="c1">#   For n = 1: dP1/dt = S1P - (1 + D1)P1</span>
        <span class="c1">#   For n &gt; 1:</span>
        <span class="c1">#     dP1/dt = S1P - (1 + S2 + D1)P1 + P2</span>
        <span class="c1">#     dP/dt = SP - (1 + S + D)P + P, i = 2..n-1</span>
        <span class="c1">#     dP/dt = SP - (1 + D)P</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
            <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
            <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">+</span> <span class="p">(</span><span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">,</span> <span class="n">eq_P1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_P1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">eq_pi</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span>
                        <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_pi</span><span class="p">)</span>
                <span class="n">eq_last</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_last</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs</span>

    <span class="k">elif</span> <span class="n">TEST_MODEL</span> <span class="o">==</span> <span class="s1">&#39;crosstalk&#39;</span><span class="p">:</span>
        <span class="c1"># Differential Equations (Crosstalk):</span>
        <span class="c1">#   Let  = (P1 + ... + Pn)/n.</span>
        <span class="c1">#   dR/dt = A - BR</span>
        <span class="c1">#   dP/dt = CR - DP - S1P(1 + ) + P1</span>
        <span class="c1">#   For n = 1: dP1/dt = S1P(1 + ) - (1 + D1)P1</span>
        <span class="c1">#   For n &gt; 1:</span>
        <span class="c1">#     dP1/dt = S1P(1 + ) - (1 + S2(1 + ) + D1)P1 + P2</span>
        <span class="c1">#     dP/dt = SP(1 + ) - (1 + S(1 + ) + D)P + P</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
            <span class="n">mod_factor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">P_sites</span><span class="p">)</span> <span class="o">/</span> <span class="n">num_psites</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
            <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
            <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">,</span> <span class="n">eq_P1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span>
                    <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_P1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">eq_pi</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span>
                                <span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_pi</span><span class="p">)</span>
                <span class="n">eq_last</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mod_factor</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_last</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs</span>

    <span class="k">elif</span> <span class="n">TEST_MODEL</span> <span class="o">==</span> <span class="s1">&#39;cooperative&#39;</span><span class="p">:</span>
        <span class="c1"># Differential Equations (Cooperative, Hill-type):</span>
        <span class="c1">#   Let activation  = (P_i)^hill_n / (K_half^hill_n + (P_i)^hill_n + ), with hill_n = 2, K_half = 0.5.</span>
        <span class="c1">#   dR/dt = A - BR</span>
        <span class="c1">#   dP/dt = CR - DP - S1P + P1</span>
        <span class="c1">#   For n = 1: dP1/dt = S1P - (1 + D1)P1</span>
        <span class="c1">#   For n &gt; 1:</span>
        <span class="c1">#     dP1/dt = S1P - (1 + S2 + D1)P1 + P2</span>
        <span class="c1">#     dP/dt = SP - (1 + S + D)P + P, for i = 2,...,n-1</span>
        <span class="c1">#     dP/dt = SP - (1 + D)P</span>

        <span class="c1"># Works well for hill_n = 0.5 and K_half = 2</span>
        <span class="n">hill_n</span> <span class="o">=</span> <span class="n">HILL_N</span>
        <span class="n">K_half</span> <span class="o">=</span> <span class="n">K_HALF</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">steady_state_equations</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="n">R</span><span class="p">,</span> <span class="n">P</span><span class="p">,</span> <span class="o">*</span><span class="n">P_sites</span> <span class="o">=</span> <span class="n">y</span>
            <span class="n">sum_P</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">P_sites</span><span class="p">)</span>
            <span class="n">activation</span> <span class="o">=</span> <span class="p">(</span><span class="n">sum_P</span> <span class="o">**</span> <span class="n">hill_n</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">K_half</span> <span class="o">**</span> <span class="n">hill_n</span> <span class="o">+</span> <span class="n">sum_P</span> <span class="o">**</span> <span class="n">hill_n</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">)</span>
            <span class="n">eq_R</span> <span class="o">=</span> <span class="n">A</span> <span class="o">-</span> <span class="n">B</span> <span class="o">*</span> <span class="n">R</span>
            <span class="n">eq_P</span> <span class="o">=</span> <span class="n">C</span> <span class="o">*</span> <span class="n">R</span> <span class="o">-</span> <span class="n">D</span> <span class="o">*</span> <span class="n">P</span> <span class="o">-</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">+</span> <span class="p">(</span><span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">num_psites</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">num_psites</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">,</span> <span class="n">eq_P1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">eq_P1</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">P</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span>
                    <span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_P1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_psites</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">eq_pi</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">S_rates</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">*</span> \
                            <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">P_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_pi</span><span class="p">)</span>
                <span class="n">eq_last</span> <span class="o">=</span> <span class="n">S_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="n">activation</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">D_rates</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">P_sites</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">eqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">eq_last</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">eq_R</span><span class="p">,</span> <span class="n">eq_P</span><span class="p">]</span> <span class="o">+</span> <span class="n">eqs</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown model type: &quot;</span> <span class="o">+</span> <span class="n">TEST_MODEL</span><span class="p">)</span>

    <span class="n">y0_guess</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">bounds_local</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">1e-8</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="n">num_psites</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span>

    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="n">y0_guess</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;SLSQP&#39;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="n">bounds_local</span><span class="p">,</span>
                      <span class="n">constraints</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;eq&#39;</span><span class="p">,</span> <span class="s1">&#39;fun&#39;</span><span class="p">:</span> <span class="n">steady_state_equations</span><span class="p">},</span>
                      <span class="n">options</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;maxiter&#39;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">})</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Steady-State conditions calculated for TEST model: &quot;</span> <span class="o">+</span> <span class="n">TEST_MODEL</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;result: </span><span class="si">{</span><span class="n">result</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to find steady-state conditions for model: &quot;</span> <span class="o">+</span> <span class="n">TEST_MODEL</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.config.constants"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov
- optimizer: Global Evolutionary Optimization method (default: 0).
    Options:
    0: NGSA2
    1: SMSEMOA
    2: AGEMOEA</p>
<p>:returns
- lower_bound: Lower bound for the optimization variables.
- upper_bound: Upper bound for the optimization variables.
- loss_type: Type of loss function to use.
- optimizer: Global Evolutionary Optimization method.
:rtype: tuple
:raises argparse.ArgumentError: If an invalid argument is provided.
:raises SystemExit: If the script is run with invalid arguments.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/config/constants.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse command line arguments for the PhosKinTime optimization problem.</span>
<span class="sd">    This function uses argparse to handle input parameters for the optimization process.</span>
<span class="sd">    The parameters include:</span>
<span class="sd">    - lower_bound: Lower bound for the optimization variables (default: -2).</span>
<span class="sd">    - upper_bound: Upper bound for the optimization variables (default: 2).</span>
<span class="sd">    - loss_type: Type of loss function to use (default: 0).</span>
<span class="sd">        Options:</span>
<span class="sd">        0: MSE</span>
<span class="sd">        1: MAE</span>
<span class="sd">        2: soft L1</span>
<span class="sd">        3: Cauchy</span>
<span class="sd">        4: Arctan</span>
<span class="sd">        5: Elastic Net</span>
<span class="sd">        6: Tikhonov</span>
<span class="sd">    - optimizer: Global Evolutionary Optimization method (default: 0).</span>
<span class="sd">        Options:</span>
<span class="sd">        0: NGSA2</span>
<span class="sd">        1: SMSEMOA</span>
<span class="sd">        2: AGEMOEA</span>

<span class="sd">    :returns</span>
<span class="sd">    - lower_bound: Lower bound for the optimization variables.</span>
<span class="sd">    - upper_bound: Upper bound for the optimization variables.</span>
<span class="sd">    - loss_type: Type of loss function to use.</span>
<span class="sd">    - optimizer: Global Evolutionary Optimization method.</span>
<span class="sd">    :rtype: tuple</span>
<span class="sd">    :raises argparse.ArgumentError: If an invalid argument is provided.</span>
<span class="sd">    :raises SystemExit: If the script is run with invalid arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;PhosKinTime - Global Optimization mRNA-TF Optimization Problem.&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Adding command line arguments for lower and upper bounds, loss type, and optimizer</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--lower_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Lower Beta bound.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--upper_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Upper Beta bound.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--loss_type&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Loss function to use:  &quot;</span>
                             <span class="s2">&quot;0: MSE, 1: MAE, 2: soft L1, 3: Cauchy,&quot;</span>
                             <span class="s2">&quot;4: Arctan, 5: Elastic Net, 6: Tikhonov.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--optimizer&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Global Evolutionary Optimization method:  &quot;</span>
                             <span class="s2">&quot;0: NGSA2, 1: SMSEMOA , 2: AGEMOEA&quot;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">args</span><span class="o">.</span><span class="n">lower_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">upper_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">loss_type</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">optimizer</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.config.logconf"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.exporter.plotout"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.exporter.plotout.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Compute the predicted expression levels based on the optimization variables.
This function calculates the predicted expression levels for each mRNA based on the optimization variables,
the regulators, and the protein and phosphorylation site tensors.</p>
<p>:param x:
:param regulators:
:param protein_mat:
:param psite_tensor:
:param n_reg:
:param T_use:
:param n_mRNA:
:param beta_start_indices:
:param num_psites:
:return: predictions</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute the predicted expression levels based on the optimization variables.</span>
<span class="sd">    This function calculates the predicted expression levels for each mRNA based on the optimization variables,</span>
<span class="sd">    the regulators, and the protein and phosphorylation site tensors.</span>

<span class="sd">    :param x:</span>
<span class="sd">    :param regulators:</span>
<span class="sd">    :param protein_mat:</span>
<span class="sd">    :param psite_tensor:</span>
<span class="sd">    :param n_reg:</span>
<span class="sd">    :param T_use:</span>
<span class="sd">    :param n_mRNA:</span>
<span class="sd">    :param beta_start_indices:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :return: predictions</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_mRNA</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">T_use</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_mRNA</span><span class="p">):</span>
        <span class="n">R_pred</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T_use</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">tf_idx</span> <span class="o">=</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_idx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="c1"># No valid TF for this regulator</span>
                <span class="k">continue</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="n">protein_mat</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span> <span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>
            <span class="n">tf_effect</span> <span class="o">=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">protein</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]):</span>
                <span class="n">tf_effect</span> <span class="o">+=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">psite_tensor</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">R_pred</span> <span class="o">+=</span> <span class="n">a</span> <span class="o">*</span> <span class="n">tf_effect</span>
        <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">R_pred</span>
    <span class="k">return</span> <span class="n">predictions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Plot the estimated vs observed expression levels for a set of genes.
This function generates two plots for each gene:
1. A full time series plot showing the observed and estimated expression levels.
2. A plot showing the observed and estimated expression levels for the first 5 time points.
Additionally, it plots the protein signals for the regulators of each gene.
The plots are saved as PNG files in the specified save path.</p>
<p>:param predictions: Predicted expression levels.
:param expression_matrix: Observed expression levels.
:param gene_ids: List of gene identifiers.
:param time_points: Time points for the expression data.
:param regulators: Matrix of regulators for each gene.
:param tf_protein_matrix: Matrix of protein signals for each transcription factor.
:param tf_ids: List of transcription factor identifiers.
:param num_targets: Number of target genes to plot.
:param save_path: Path to save the plots.
:type predictions: np.ndarray
:type expression_matrix: np.ndarray
:type gene_ids: list
:type time_points: np.ndarray
:type regulators: np.ndarray
:type tf_protein_matrix: np.ndarray
:type tf_ids: list
:type num_targets: int
:type save_path: str</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span>
                               <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot the estimated vs observed expression levels for a set of genes.</span>
<span class="sd">    This function generates two plots for each gene:</span>
<span class="sd">    1. A full time series plot showing the observed and estimated expression levels.</span>
<span class="sd">    2. A plot showing the observed and estimated expression levels for the first 5 time points.</span>
<span class="sd">    Additionally, it plots the protein signals for the regulators of each gene.</span>
<span class="sd">    The plots are saved as PNG files in the specified save path.</span>

<span class="sd">    :param predictions: Predicted expression levels.</span>
<span class="sd">    :param expression_matrix: Observed expression levels.</span>
<span class="sd">    :param gene_ids: List of gene identifiers.</span>
<span class="sd">    :param time_points: Time points for the expression data.</span>
<span class="sd">    :param regulators: Matrix of regulators for each gene.</span>
<span class="sd">    :param tf_protein_matrix: Matrix of protein signals for each transcription factor.</span>
<span class="sd">    :param tf_ids: List of transcription factor identifiers.</span>
<span class="sd">    :param num_targets: Number of target genes to plot.</span>
<span class="sd">    :param save_path: Path to save the plots.</span>
<span class="sd">    :type predictions: np.ndarray</span>
<span class="sd">    :type expression_matrix: np.ndarray</span>
<span class="sd">    :type gene_ids: list</span>
<span class="sd">    :type time_points: np.ndarray</span>
<span class="sd">    :type regulators: np.ndarray</span>
<span class="sd">    :type tf_protein_matrix: np.ndarray</span>
<span class="sd">    :type tf_ids: list</span>
<span class="sd">    :type num_targets: int</span>
<span class="sd">    :type save_path: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">T</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">time_points</span><span class="p">)</span>
    <span class="n">time_vals_expr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">240</span><span class="p">,</span> <span class="mi">480</span><span class="p">,</span> <span class="mi">960</span><span class="p">])</span>
    <span class="n">time_vals_tf</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">240</span><span class="p">,</span> <span class="mi">480</span><span class="p">,</span> <span class="mi">960</span><span class="p">])</span>
    <span class="n">combined_ticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">time_vals_tf</span><span class="p">)))</span>
    <span class="n">num_targets</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_targets</span><span class="p">,</span> <span class="n">predictions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_targets</span><span class="p">):</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

        <span class="c1"># --- Full time series plot ---</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;s-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="n">T</span><span class="p">]</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">,</span> <span class="n">protein_signal</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;mRNA: </span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fold Changes&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">combined_ticks</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">combined_ticks</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

        <span class="c1"># --- First 5 time points plot ---</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;s-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">protein_signal</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;TFs&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">_model_fit_.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># This block is for saving two plots for one TF</span>
        <span class="c1"># One for full time series and one for first 5 time points</span>
        <span class="c1"># To see clearly the dynamics early on</span>

        <span class="c1"># plt.figure(figsize=(8, 8))</span>
        <span class="c1"># plt.plot(time_vals_expr, expression_matrix[i, :], &#39;s-&#39;, label=&#39;Observed&#39;)</span>
        <span class="c1"># plt.plot(time_vals_expr, predictions[i, :], &#39;-&#39;, label=&#39;Estimated&#39;)</span>
        <span class="c1"># # Plot protein time series for each regulator of mRNA i (only unique TFs)</span>
        <span class="c1"># plotted_tfs = set()</span>
        <span class="c1"># for r in regulators[i, :]:</span>
        <span class="c1">#     if r == -1:  # Skip invalid TF</span>
        <span class="c1">#         continue</span>
        <span class="c1">#     tf_name = tf_ids[r]</span>
        <span class="c1">#     if tf_name not in plotted_tfs:</span>
        <span class="c1">#         protein_signal = tf_protein_matrix[r, :T]</span>
        <span class="c1">#         plt.plot(time_vals_tf, protein_signal, &#39;:&#39;, label=f&quot;mRNA: {tf_name}&quot;)</span>
        <span class="c1">#         plotted_tfs.add(tf_name)</span>
        <span class="c1"># plt.title(f&quot;mRNA: {gene_ids[i]}&quot;)</span>
        <span class="c1"># plt.xlabel(&quot;Time (minutes)&quot;)</span>
        <span class="c1"># plt.ylabel(&quot;Fold Changes&quot;)</span>
        <span class="c1"># plt.xticks(combined_ticks, combined_ticks, rotation=45)</span>
        <span class="c1"># plt.legend()</span>
        <span class="c1"># plt.tight_layout()</span>
        <span class="c1"># plt.grid(True, alpha=0.3)</span>
        <span class="c1"># plt.show()</span>
        <span class="c1">#</span>
        <span class="c1"># plt.figure(figsize=(8, 8))</span>
        <span class="c1"># plt.plot(time_vals_expr[:5], expression_matrix[i, :5], &#39;s-&#39;, label=&#39;Observed&#39;)</span>
        <span class="c1"># plt.plot(time_vals_expr[:5], predictions[i, :5], &#39;-&#39;, label=&#39;Estimated&#39;)</span>
        <span class="c1"># # Plot protein time series for each regulator of mRNA i (only unique TFs)</span>
        <span class="c1"># plotted_tfs = set()</span>
        <span class="c1"># for r in regulators[i, :]:</span>
        <span class="c1">#     if r == -1:  # Skip invalid TF</span>
        <span class="c1">#         continue</span>
        <span class="c1">#     tf_name = tf_ids[r]</span>
        <span class="c1">#     if tf_name not in plotted_tfs:</span>
        <span class="c1">#         protein_signal = tf_protein_matrix[r, :5]</span>
        <span class="c1">#         plt.plot(time_vals_tf[:5], protein_signal, &#39;:&#39;, label=f&quot;mRNA: {tf_name}&quot;)</span>
        <span class="c1">#         plotted_tfs.add(tf_name)</span>
        <span class="c1"># plt.title(f&quot;mRNA: {gene_ids[i]}&quot;)</span>
        <span class="c1"># plt.xlabel(&quot;Time (minutes)&quot;)</span>
        <span class="c1"># plt.ylabel(&quot;Fold Changes&quot;)</span>
        <span class="c1"># plt.xticks(time_vals_expr[:5], time_vals_expr[:5], rotation=45)</span>
        <span class="c1"># plt.legend()</span>
        <span class="c1"># plt.tight_layout()</span>
        <span class="c1"># plt.grid(True, alpha=0.3)</span>
        <span class="c1"># plt.show()</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">()</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_vals_expr</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers+lines&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">symbol</span><span class="o">=</span><span class="s1">&#39;square&#39;</span><span class="p">)</span>
        <span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_vals_expr</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span>
        <span class="p">))</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">)]</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
                    <span class="n">x</span><span class="o">=</span><span class="n">time_vals_tf</span><span class="p">,</span>
                    <span class="n">y</span><span class="o">=</span><span class="n">protein_signal</span><span class="p">,</span>
                    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines&#39;</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;mRNA: </span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">dash</span><span class="o">=</span><span class="s1">&#39;dot&#39;</span><span class="p">)</span>
                <span class="p">))</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;mRNA: </span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">xaxis_title</span><span class="o">=</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">,</span>
            <span class="n">yaxis_title</span><span class="o">=</span><span class="s2">&quot;Fold Changes&quot;</span><span class="p">,</span>
            <span class="n">xaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">tickmode</span><span class="o">=</span><span class="s1">&#39;array&#39;</span><span class="p">,</span>
                <span class="n">tickvals</span><span class="o">=</span><span class="n">combined_ticks</span><span class="p">,</span>
                <span class="n">ticktext</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">combined_ticks</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">legend</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;h&quot;</span><span class="p">,</span> <span class="n">yanchor</span><span class="o">=</span><span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mf">1.02</span><span class="p">,</span> <span class="n">xanchor</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">_model_fit_.html&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.exporter.sheetutils"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_results_to_excel</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.
This function creates multiple sheets in the Excel file to store different types of data,
including alpha values, beta values, residuals, observed values, estimated values,
and optimization results.
Each sheet is formatted with appropriate column names and data types.
The function also calculates various metrics (MSE, MAE, MAPE, R^2)
to evaluate the performance of the optimization.</p>
<p>:param gene_ids:
:param tf_ids:
:param final_alpha:
:param final_beta:
:param psite_labels_arr:
:param expression_matrix:
:param predictions:
:param objective_value:
:param reg_map:
:param filename:</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/exporter/sheetutils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_results_to_excel</span><span class="p">(</span>
        <span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span>
        <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span>
        <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span>
        <span class="n">objective_value</span><span class="p">,</span>
        <span class="n">reg_map</span><span class="p">,</span>
        <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the optimization results to an Excel file.</span>
<span class="sd">    This function creates multiple sheets in the Excel file to store different types of data,</span>
<span class="sd">    including alpha values, beta values, residuals, observed values, estimated values,</span>
<span class="sd">    and optimization results.</span>
<span class="sd">    Each sheet is formatted with appropriate column names and data types.</span>
<span class="sd">    The function also calculates various metrics (MSE, MAE, MAPE, R^2)</span>
<span class="sd">    to evaluate the performance of the optimization.</span>

<span class="sd">    :param gene_ids:</span>
<span class="sd">    :param tf_ids:</span>
<span class="sd">    :param final_alpha:</span>
<span class="sd">    :param final_beta:</span>
<span class="sd">    :param psite_labels_arr:</span>
<span class="sd">    :param expression_matrix:</span>
<span class="sd">    :param predictions:</span>
<span class="sd">    :param objective_value:</span>
<span class="sd">    :param reg_map:</span>
<span class="sd">    :param filename:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># --- Alpha Values ---</span>
    <span class="n">alpha_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span> <span class="o">=</span> <span class="n">final_alpha</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">actual_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf_name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">actual_tfs</span><span class="p">):</span>
            <span class="n">alpha_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">,</span> <span class="n">tf_name</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]])</span>
    <span class="n">df_alpha</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">alpha_rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;TF&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="c1"># --- Beta Values ---</span>
    <span class="n">beta_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">):</span>
        <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">final_beta</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">beta_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">tf</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>  <span class="c1"># Protein beta</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)):</span>
            <span class="n">beta_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">tf</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">j</span><span class="p">]])</span>
    <span class="n">df_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">beta_rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;TF&quot;</span><span class="p">,</span> <span class="s2">&quot;PSite&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="c1"># --- Residuals ---</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="n">expression_matrix</span> <span class="o">-</span> <span class="n">predictions</span>
    <span class="n">df_residuals</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">residuals</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_residuals</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Observed ---</span>
    <span class="n">df_observed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">expression_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_observed</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Estimated ---</span>
    <span class="n">df_estimated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predictions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_estimated</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Optimization Results ---</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">expression_matrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">y_pred</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">mape</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">((</span><span class="n">y_true</span> <span class="o">-</span> <span class="n">y_pred</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">y_true</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">r2</span> <span class="o">=</span> <span class="n">r2_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">df_metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
        <span class="p">[</span><span class="s2">&quot;Objective Value&quot;</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MSE&quot;</span><span class="p">,</span> <span class="n">mse</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MAE&quot;</span><span class="p">,</span> <span class="n">mae</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MAPE&quot;</span><span class="p">,</span> <span class="n">mape</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;R^2&quot;</span><span class="p">,</span> <span class="n">r2</span><span class="p">],</span>
    <span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Metric&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;xlsxwriter&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
        <span class="n">df_alpha</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Alpha Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_beta</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Beta Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_residuals</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Residuals&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_observed</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Observed&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_estimated</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Estimated&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_metrics</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Optimization Results&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.objfn.minfn"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="doc doc-heading">
            <code>TFOptimizationMultiObjectiveProblem</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.Problem">Problem</span></code></p>


        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Multi-objective optimization problem for TF optimization.
This class defines a multi-objective optimization problem for the
transcription factor (TF) optimization problem. It inherits from the
<code>Problem</code> class in the pymoo library. The problem is defined with three
objectives: f1 (error), f2 (alpha violation), and f3 (beta violation).</p>
<p>The problem is initialized with the following parameters:
- n_var: Number of decision variables.
- n_mRNA: Number of mRNA genes.
- n_TF: Number of transcription factors.
- n_reg: Number of regulators.
- n_psite_max: Maximum number of phosphorylation sites.
- n_alpha: Number of alpha parameters.
- mRNA_mat: Matrix of measured mRNA values.
- regulators: Matrix of TF regulators for each mRNA.
- protein_mat: Matrix of TF protein time series.
- psite_tensor: Tensor of phosphorylation site signals.
- T_use: Number of time points used.
- beta_start_indices: Integer array giving the starting index (in the segment)
                      for each TF.
- num_psites: Integer array with the actual number of phosphorylation sites for each TF.
- no_psite_tf: Boolean array indicating if the TF has no phosphorylation sites.
- xl: Lower bounds for the decision variables.
- xu: Upper bounds for the decision variables.
- kwargs: Additional keyword arguments.</p>







              <details class="quote">
                <summary>Source code in <code>tfopt/evol/objfn/minfn.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">TFOptimizationMultiObjectiveProblem</span><span class="p">(</span><span class="n">Problem</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Originally implemented by Julius Normann.</span>

<span class="sd">    This version has been modified and optimized</span>
<span class="sd">    for consistency &amp; speed in submodules by Abhinav Mishra.</span>

<span class="sd">    Multi-objective optimization problem for TF optimization.</span>
<span class="sd">    This class defines a multi-objective optimization problem for the</span>
<span class="sd">    transcription factor (TF) optimization problem. It inherits from the</span>
<span class="sd">    `Problem` class in the pymoo library. The problem is defined with three</span>
<span class="sd">    objectives: f1 (error), f2 (alpha violation), and f3 (beta violation).</span>

<span class="sd">    The problem is initialized with the following parameters:</span>
<span class="sd">    - n_var: Number of decision variables.</span>
<span class="sd">    - n_mRNA: Number of mRNA genes.</span>
<span class="sd">    - n_TF: Number of transcription factors.</span>
<span class="sd">    - n_reg: Number of regulators.</span>
<span class="sd">    - n_psite_max: Maximum number of phosphorylation sites.</span>
<span class="sd">    - n_alpha: Number of alpha parameters.</span>
<span class="sd">    - mRNA_mat: Matrix of measured mRNA values.</span>
<span class="sd">    - regulators: Matrix of TF regulators for each mRNA.</span>
<span class="sd">    - protein_mat: Matrix of TF protein time series.</span>
<span class="sd">    - psite_tensor: Tensor of phosphorylation site signals.</span>
<span class="sd">    - T_use: Number of time points used.</span>
<span class="sd">    - beta_start_indices: Integer array giving the starting index (in the segment)</span>
<span class="sd">                          for each TF.</span>
<span class="sd">    - num_psites: Integer array with the actual number of phosphorylation sites for each TF.</span>
<span class="sd">    - no_psite_tf: Boolean array indicating if the TF has no phosphorylation sites.</span>
<span class="sd">    - xl: Lower bounds for the decision variables.</span>
<span class="sd">    - xu: Upper bounds for the decision variables.</span>
<span class="sd">    - kwargs: Additional keyword arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_var</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span>
                 <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span>
                 <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize the multi-objective optimization problem.</span>

<span class="sd">        :param n_var:</span>
<span class="sd">        :param n_mRNA:</span>
<span class="sd">        :param n_TF:</span>
<span class="sd">        :param n_reg:</span>
<span class="sd">        :param n_psite_max:</span>
<span class="sd">        :param n_alpha:</span>
<span class="sd">        :param mRNA_mat:</span>
<span class="sd">        :param regulators:</span>
<span class="sd">        :param protein_mat:</span>
<span class="sd">        :param psite_tensor:</span>
<span class="sd">        :param T_use:</span>
<span class="sd">        :param beta_start_indices:</span>
<span class="sd">        :param num_psites:</span>
<span class="sd">        :param no_psite_tf:</span>
<span class="sd">        :param xl:</span>
<span class="sd">        :param xu:</span>
<span class="sd">        :param kwargs:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">n_var</span><span class="o">=</span><span class="n">n_var</span><span class="p">,</span> <span class="n">n_obj</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_constr</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="n">xl</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="n">xu</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_mRNA</span> <span class="o">=</span> <span class="n">n_mRNA</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_TF</span> <span class="o">=</span> <span class="n">n_TF</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_reg</span> <span class="o">=</span> <span class="n">n_reg</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_psite_max</span> <span class="o">=</span> <span class="n">n_psite_max</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_alpha</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mRNA_mat</span> <span class="o">=</span> <span class="n">mRNA_mat</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">regulators</span> <span class="o">=</span> <span class="n">regulators</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_mat</span> <span class="o">=</span> <span class="n">protein_mat</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psite_tensor</span> <span class="o">=</span> <span class="n">psite_tensor</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">T_use</span> <span class="o">=</span> <span class="n">T_use</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">beta_start_indices</span> <span class="o">=</span> <span class="n">beta_start_indices</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_psites</span> <span class="o">=</span> <span class="n">num_psites</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">no_psite_tf</span> <span class="o">=</span> <span class="n">no_psite_tf</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">loss_type</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;loss_type&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_evaluate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Evaluate the objectives for the given decision variables.</span>
<span class="sd">        This function computes the objectives for each individual in the population.</span>
<span class="sd">        The objectives are defined as follows:</span>
<span class="sd">        - f1: Error (objective function value).</span>
<span class="sd">        - f2: Alpha violation (sum of squares of alpha values).</span>
<span class="sd">        - f3: Beta violation (sum of squares of beta values).</span>

<span class="sd">        The function computes the objectives for each individual in the population</span>
<span class="sd">        and stores the results in the output dictionary.</span>
<span class="sd">        :param X: Decision variables (population).</span>
<span class="sd">        :param out: Output dictionary to store the results.</span>
<span class="sd">        :param args: Additional arguments.</span>
<span class="sd">        :param kwargs: Additional keyword arguments.</span>
<span class="sd">        :return: Final Objective values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n_pop</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">F</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">n_pop</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
        <span class="n">n_alpha</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_alpha</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_pop</span><span class="p">):</span>
            <span class="n">xi</span> <span class="o">=</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">f1</span> <span class="o">=</span> <span class="n">objective_</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mRNA_mat</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">regulators</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_mat</span><span class="p">,</span>
                                       <span class="bp">self</span><span class="o">.</span><span class="n">psite_tensor</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_reg</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">T_use</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_start_indices</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_psites</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">loss_type</span><span class="p">)</span>
            <span class="n">f2</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_mRNA</span><span class="p">):</span>
                <span class="n">s</span> <span class="o">=</span> <span class="mf">0.0</span>
                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_reg</span><span class="p">):</span>
                    <span class="n">s</span> <span class="o">+=</span> <span class="n">xi</span><span class="p">[</span><span class="n">m</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
                <span class="n">f2</span> <span class="o">+=</span> <span class="p">(</span><span class="n">s</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
            <span class="n">f3</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_TF</span><span class="p">):</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
                <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_psites</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
                <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">xi</span><span class="p">[</span><span class="n">start</span> <span class="p">:</span> <span class="n">start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>
                <span class="n">f3</span> <span class="o">+=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">no_psite_tf</span><span class="p">[</span><span class="n">tf</span><span class="p">]:</span>
                    <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
                        <span class="n">f3</span> <span class="o">+=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">q</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span>
            <span class="c1"># Three objectives:</span>
            <span class="c1"># f1 (error)</span>
            <span class="n">F</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">f1</span>
            <span class="c1"># f2 (alpha violation)</span>
            <span class="n">F</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">f2</span>
            <span class="c1"># f3 (beta violation)</span>
            <span class="n">F</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">f3</span>
        <span class="n">out</span><span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">F</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">n_var</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initialize the multi-objective optimization problem.</p>
<p>:param n_var:
:param n_mRNA:
:param n_TF:
:param n_reg:
:param n_psite_max:
:param n_alpha:
:param mRNA_mat:
:param regulators:
:param protein_mat:
:param psite_tensor:
:param T_use:
:param beta_start_indices:
:param num_psites:
:param no_psite_tf:
:param xl:
:param xu:
:param kwargs:</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/objfn/minfn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_var</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span>
             <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span>
             <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
             <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize the multi-objective optimization problem.</span>

<span class="sd">    :param n_var:</span>
<span class="sd">    :param n_mRNA:</span>
<span class="sd">    :param n_TF:</span>
<span class="sd">    :param n_reg:</span>
<span class="sd">    :param n_psite_max:</span>
<span class="sd">    :param n_alpha:</span>
<span class="sd">    :param mRNA_mat:</span>
<span class="sd">    :param regulators:</span>
<span class="sd">    :param protein_mat:</span>
<span class="sd">    :param psite_tensor:</span>
<span class="sd">    :param T_use:</span>
<span class="sd">    :param beta_start_indices:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param no_psite_tf:</span>
<span class="sd">    :param xl:</span>
<span class="sd">    :param xu:</span>
<span class="sd">    :param kwargs:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">n_var</span><span class="o">=</span><span class="n">n_var</span><span class="p">,</span> <span class="n">n_obj</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_constr</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">xl</span><span class="o">=</span><span class="n">xl</span><span class="p">,</span> <span class="n">xu</span><span class="o">=</span><span class="n">xu</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_mRNA</span> <span class="o">=</span> <span class="n">n_mRNA</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_TF</span> <span class="o">=</span> <span class="n">n_TF</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_reg</span> <span class="o">=</span> <span class="n">n_reg</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_psite_max</span> <span class="o">=</span> <span class="n">n_psite_max</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_alpha</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">mRNA_mat</span> <span class="o">=</span> <span class="n">mRNA_mat</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">regulators</span> <span class="o">=</span> <span class="n">regulators</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_mat</span> <span class="o">=</span> <span class="n">protein_mat</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psite_tensor</span> <span class="o">=</span> <span class="n">psite_tensor</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">T_use</span> <span class="o">=</span> <span class="n">T_use</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">beta_start_indices</span> <span class="o">=</span> <span class="n">beta_start_indices</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_psites</span> <span class="o">=</span> <span class="n">num_psites</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">no_psite_tf</span> <span class="o">=</span> <span class="n">no_psite_tf</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">loss_type</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;loss_type&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Computes a loss value using one of several loss functions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mRNA_mat</code>
            </td>
            <td>
                  <code>       </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_mRNA x T_use) measured mRNA values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>     </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_mRNA x n_reg) indices of TF regulators for each mRNA.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>protein_mat</code>
            </td>
            <td>
                  <code>    </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>   </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per mRNA.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_mRNA,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of mRNA and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the segment)
                 for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>     </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>      </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1,
                 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lam1,</code>
            </td>
            <td>
                  <code><span title="lam2">lam2</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization parameters (used for loss_type 5 and 6).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/objfn/minfn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span><span class="p">(</span><span class="n">cache</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fastmath</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">parallel</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">nogil</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span>
                   <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes a loss value using one of several loss functions.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      x               : Decision vector.</span>
<span class="sd">      mRNA_mat        : (n_mRNA x T_use) measured mRNA values.</span>
<span class="sd">      regulators      : (n_mRNA x n_reg) indices of TF regulators for each mRNA.</span>
<span class="sd">      protein_mat     : (n_TF x T_use) TF protein time series.</span>
<span class="sd">      psite_tensor    : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</span>
<span class="sd">      n_reg           : Maximum number of regulators per mRNA.</span>
<span class="sd">      T_use           : Number of time points used.</span>
<span class="sd">      n_mRNA, n_TF    : Number of mRNA and TF respectively.</span>
<span class="sd">      beta_start_indices: Integer array giving the starting index (in the segment)</span>
<span class="sd">                         for each TF.</span>
<span class="sd">      num_psites      : Integer array with the actual number of PSites for each TF.</span>
<span class="sd">      loss_type       : Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1,</span>
<span class="sd">                         3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</span>
<span class="sd">      lam1, lam2      : Regularization parameters (used for loss_type 5 and 6).</span>

<span class="sd">    Returns:</span>
<span class="sd">      The computed loss (a scalar).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_loss</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_mRNA</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">prange</span><span class="p">(</span><span class="n">n_mRNA</span><span class="p">):</span>
        <span class="n">R_meas</span> <span class="o">=</span> <span class="n">mRNA_mat</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
        <span class="n">R_pred</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T_use</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">tf_idx</span> <span class="o">=</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_idx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="c1"># No valid TF for this regulator</span>
                <span class="k">continue</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="n">protein_mat</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>  <span class="c1"># actual length of beta vector for TF</span>
            <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span><span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>
            <span class="n">tf_effect</span> <span class="o">=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">protein</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]):</span>
                <span class="n">tf_effect</span> <span class="o">+=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">psite_tensor</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">R_pred</span> <span class="o">+=</span> <span class="n">a</span> <span class="o">*</span> <span class="n">tf_effect</span>
        <span class="c1"># For each time point, add loss according to loss_type.</span>
        <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">T_use</span><span class="p">):</span>
            <span class="n">e</span> <span class="o">=</span> <span class="n">R_meas</span><span class="p">[</span><span class="n">t</span><span class="p">]</span> <span class="o">-</span> <span class="n">R_pred</span><span class="p">[</span><span class="n">t</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># MSE</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">e</span> <span class="o">*</span> <span class="n">e</span>
            <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># MAE</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>  <span class="c1"># Soft L1 (pseudo-Huber)</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">e</span> <span class="o">*</span> <span class="n">e</span><span class="p">)</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>  <span class="c1"># Cauchy</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">e</span> <span class="o">*</span> <span class="n">e</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># Arctan</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">arctan</span><span class="p">(</span><span class="n">e</span> <span class="o">*</span> <span class="n">e</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Default to MSE if unknown.</span>
                <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">e</span> <span class="o">*</span> <span class="n">e</span>
    <span class="n">loss</span> <span class="o">=</span> <span class="n">total_loss</span> <span class="o">/</span> <span class="p">(</span><span class="n">n_mRNA</span> <span class="o">*</span> <span class="n">T_use</span><span class="p">)</span>

    <span class="c1"># For elastic net (loss_type 5), add L1 and L2 penalties on the beta portion.</span>
    <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>
        <span class="n">l1</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="n">l2</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="c1"># Compute over beta parameters only.</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">l1</span> <span class="o">+=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
            <span class="n">l2</span> <span class="o">+=</span> <span class="n">v</span> <span class="o">*</span> <span class="n">v</span>
        <span class="n">loss</span> <span class="o">+=</span> <span class="n">lam1</span> <span class="o">*</span> <span class="n">l1</span> <span class="o">+</span> <span class="n">lam2</span> <span class="o">*</span> <span class="n">l2</span>

    <span class="c1"># For Tikhonov (loss_type 6), add L2 penalty on the beta portion.</span>
    <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">6</span><span class="p">:</span>
        <span class="n">l2</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">l2</span> <span class="o">+=</span> <span class="n">v</span> <span class="o">*</span> <span class="n">v</span>
        <span class="n">loss</span> <span class="o">+=</span> <span class="n">lam1</span> <span class="o">*</span> <span class="n">l2</span>

    <span class="k">return</span> <span class="n">loss</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.opt.optrun"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimization</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">total_dim</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Run the optimization using the specified algorithm and problem.
This function sets up the algorithm parameters, initializes the optimizer,
and runs the optimization process.</p>
<p>:param problem:
:param total_dim:
:param optimizer:
:return: result</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_optimization</span><span class="p">(</span><span class="n">problem</span><span class="p">,</span> <span class="n">total_dim</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run the optimization using the specified algorithm and problem.</span>
<span class="sd">    This function sets up the algorithm parameters, initializes the optimizer,</span>
<span class="sd">    and runs the optimization process.</span>

<span class="sd">    :param problem:</span>
<span class="sd">    :param total_dim:</span>
<span class="sd">    :param optimizer:</span>
<span class="sd">    :return: result</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define algorithm settings.</span>
    <span class="k">global</span> <span class="n">algo</span>
    <span class="n">pop_size</span> <span class="o">=</span> <span class="mi">2000</span>
    <span class="n">crossover</span> <span class="o">=</span> <span class="n">TwoPointCrossover</span><span class="p">(</span><span class="n">prob</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span>
    <span class="n">mutation</span> <span class="o">=</span> <span class="n">PolynomialMutation</span><span class="p">(</span><span class="n">prob</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">total_dim</span><span class="p">,</span> <span class="n">eta</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
    <span class="n">eliminate_duplicates</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="c1"># Choose the optimizer based on the input parameter.</span>
    <span class="k">if</span> <span class="n">optimizer</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># NSGA2 settings.</span>
        <span class="n">algo</span> <span class="o">=</span> <span class="n">NSGA2</span><span class="p">(</span>
            <span class="n">pop_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">,</span>
            <span class="n">crossover</span><span class="o">=</span><span class="n">crossover</span><span class="p">,</span>
            <span class="n">mutation</span><span class="o">=</span><span class="n">mutation</span><span class="p">,</span>
            <span class="n">eliminate_duplicates</span><span class="o">=</span><span class="n">eliminate_duplicates</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">optimizer</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="c1"># SMSEMOA settings.</span>
        <span class="n">algo</span><span class="o">=</span> <span class="n">SMSEMOA</span><span class="p">(</span>
            <span class="n">pop_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">,</span>
            <span class="n">crossover</span><span class="o">=</span><span class="n">crossover</span><span class="p">,</span>
            <span class="n">mutation</span><span class="o">=</span><span class="n">mutation</span><span class="p">,</span>
            <span class="n">eliminate_duplicates</span><span class="o">=</span><span class="n">eliminate_duplicates</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">optimizer</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="c1"># AGEMOEA settings.</span>
        <span class="n">algo</span> <span class="o">=</span> <span class="n">AGEMOEA</span><span class="p">(</span>
            <span class="n">pop_size</span><span class="o">=</span><span class="n">pop_size</span><span class="p">,</span>
            <span class="n">crossover</span><span class="o">=</span><span class="n">crossover</span><span class="p">,</span>
            <span class="n">mutation</span><span class="o">=</span><span class="n">mutation</span><span class="p">,</span>
            <span class="n">eliminate_duplicates</span><span class="o">=</span><span class="n">eliminate_duplicates</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Unknown optimizer type. Please choose 0 (NSGA2), 1 (SMSEMOA), or 2 (AGEMOEA).&quot;</span><span class="p">)</span>

    <span class="n">termination</span> <span class="o">=</span> <span class="n">DefaultMultiObjectiveTermination</span><span class="p">()</span>

    <span class="c1"># Run the optimization</span>
    <span class="c1"># buf = io.StringIO()</span>
    <span class="c1"># with contextlib.redirect_stdout(buf):</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">pymoo_minimize</span><span class="p">(</span><span class="n">problem</span><span class="o">=</span><span class="n">problem</span><span class="p">,</span>
                         <span class="n">algorithm</span><span class="o">=</span><span class="n">algo</span><span class="p">,</span>
                         <span class="n">termination</span><span class="o">=</span><span class="n">termination</span><span class="p">,</span>
                         <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                         <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Log the captured pymoo progress</span>
    <span class="c1"># pymoo_progress = buf.getvalue()</span>
    <span class="c1"># if pymoo_progress.strip():  # only log if there&#39;s actual text</span>
    <span class="c1">#     logger.info(&quot;--- Progress Output ---\n&quot; + pymoo_progress)</span>

    <span class="k">return</span> <span class="n">res</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.optcon.construct"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_fixed_arrays</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.
Returns:
  - mRNA_mat: array of shape (n_mRNA, T)
  - regulators: array of shape (n_mRNA, n_reg) with indices into TF_ids.
  - protein_mat: array of shape (n_TF, T)
  - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
  - n_reg: maximum number of regulators per mRNA.
  - n_psite_max: maximum number of PSites among TFs.
  - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
  - num_psites: array of length n_TF with the actual number of PSites for each TF.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_fixed_arrays</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Builds fixed-shape arrays from the input data.</span>
<span class="sd">    Returns:</span>
<span class="sd">      - mRNA_mat: array of shape (n_mRNA, T)</span>
<span class="sd">      - regulators: array of shape (n_mRNA, n_reg) with indices into TF_ids.</span>
<span class="sd">      - protein_mat: array of shape (n_TF, T)</span>
<span class="sd">      - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.</span>
<span class="sd">      - n_reg: maximum number of regulators per mRNA.</span>
<span class="sd">      - n_psite_max: maximum number of PSites among TFs.</span>
<span class="sd">      - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).</span>
<span class="sd">      - num_psites: array of length n_TF with the actual number of PSites for each TF.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">T</span> <span class="o">=</span> <span class="n">mRNA_mat</span><span class="o">.</span><span class="n">shape</span>

    <span class="c1"># Map TF_id to index.</span>
    <span class="n">TF_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">)}</span>
    <span class="n">n_TF</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">)</span>

    <span class="c1"># # Determine maximum number of regulators per mRNA.</span>
    <span class="c1"># max_reg = 0</span>
    <span class="c1"># reg_list = []</span>
    <span class="c1"># for gene in mRNA_ids:</span>
    <span class="c1">#     regs = reg_map.get(gene, [])</span>
    <span class="c1">#     max_reg = max(max_reg, len(regs))</span>
    <span class="c1">#     reg_list.append(regs)</span>
    <span class="c1"># n_reg = max_reg if max_reg &gt; 0 else 1</span>
    <span class="c1">#</span>
    <span class="c1"># # Build regulators array (n_mRNA x n_reg), padded with index 0.</span>
    <span class="c1"># regulators = np.zeros((n_mRNA, n_reg), dtype=np.int32)</span>
    <span class="c1"># for i, regs in enumerate(reg_list):</span>
    <span class="c1">#     for j in range(n_reg):</span>
    <span class="c1">#         if j &lt; len(regs):</span>
    <span class="c1">#             regulators[i, j] = TF_index.get(regs[j], 0)</span>
    <span class="c1">#         else:</span>
    <span class="c1">#             regulators[i, j] = 0</span>

    <span class="c1"># Determine max number of valid regulators across all mRNA, and keep valid indices only.</span>
    <span class="n">reg_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">mRNA_ids</span><span class="p">:</span>
        <span class="n">regs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[])</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids</span><span class="p">]</span>
        <span class="n">reg_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">regs</span><span class="p">)</span>
    <span class="n">n_reg</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">regs</span><span class="p">)</span> <span class="k">for</span> <span class="n">regs</span> <span class="ow">in</span> <span class="n">reg_list</span><span class="p">)</span> <span class="k">if</span> <span class="n">reg_list</span> <span class="k">else</span> <span class="mi">1</span>

    <span class="c1"># Build regulators array (n_mRNA x n_reg), padded with -1 to mark invalid.</span>
    <span class="n">regulators</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">((</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">regs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reg_list</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regs</span><span class="p">):</span>
            <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">TF_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Build protein_mat.</span>
    <span class="n">protein_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">T</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">TF_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">protein_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">protein_mat</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">protein_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">][:</span><span class="n">T</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_mat</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># For each TF, record the actual number of PSites.</span>
    <span class="n">num_psites</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">):</span>
        <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">psite_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[]))</span>
    <span class="c1"># Maximum number of PSites across all TFs.</span>
    <span class="n">n_psite_max</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">num_psites</span><span class="p">))</span> <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="c1"># Build psite_tensor and psite_labels_arr.</span>
    <span class="c1"># psite_tensor will have shape (n_TF, n_psite_max, T) and we pad shorter vectors with zeros.</span>
    <span class="n">psite_tensor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">T</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="n">psite_labels_arr</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">TF_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">psites</span> <span class="o">=</span> <span class="n">psite_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[])</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">psite_labels_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_psite_max</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">psites</span><span class="p">):</span>
                <span class="n">psite_tensor</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">psites</span><span class="p">[</span><span class="n">j</span><span class="p">][:</span><span class="n">T</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">psite_tensor</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>
        <span class="n">padded_labels</span> <span class="o">=</span> <span class="n">labels</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">n_psite_max</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
        <span class="n">psite_labels_arr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">padded_labels</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">protein_mat</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">num_psites</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.optcon.filter"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.filter.determine_T_use" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">determine_T_use</span><span class="p">(</span><span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_time_cols</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Determine the number of time points to use for the analysis.
This function takes the mRNA matrix and TF time columns as input
and returns the minimum number of time points available across both datasets.
This is important for ensuring that the analysis is consistent and
comparable across different datasets.
The function checks the shape of the mRNA matrix and the length of the TF time columns
to determine the minimum number of time points.
If the mRNA matrix has fewer time points than the TF time columns,
it uses the number of time points in the mRNA matrix.
If the TF time columns have fewer time points, it uses that value instead.
This ensures that the analysis is based on the same number of time points
for both mRNA and TF data.</p>
<p>:param mRNA_mat: Matrix of mRNA expression data.
:param TF_time_cols: Time points for TF data.
:return: T_use: Number of time points to use for the analysis.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">determine_T_use</span><span class="p">(</span><span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">TF_time_cols</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine the number of time points to use for the analysis.</span>
<span class="sd">    This function takes the mRNA matrix and TF time columns as input</span>
<span class="sd">    and returns the minimum number of time points available across both datasets.</span>
<span class="sd">    This is important for ensuring that the analysis is consistent and</span>
<span class="sd">    comparable across different datasets.</span>
<span class="sd">    The function checks the shape of the mRNA matrix and the length of the TF time columns</span>
<span class="sd">    to determine the minimum number of time points.</span>
<span class="sd">    If the mRNA matrix has fewer time points than the TF time columns,</span>
<span class="sd">    it uses the number of time points in the mRNA matrix.</span>
<span class="sd">    If the TF time columns have fewer time points, it uses that value instead.</span>
<span class="sd">    This ensures that the analysis is based on the same number of time points</span>
<span class="sd">    for both mRNA and TF data.</span>

<span class="sd">    :param mRNA_mat: Matrix of mRNA expression data.</span>
<span class="sd">    :param TF_time_cols: Time points for TF data.</span>
<span class="sd">    :return: T_use: Number of time points to use for the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">T_use</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">mRNA_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">TF_time_cols</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">T_use</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.filter.filter_TF" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_TF</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">relevant_TFs</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Filter transcription factors to only those present in the relevant_TFs set.
This function returns the filtered TF_ids and their corresponding protein and phosphorylation site data.
This is important for ensuring that only relevant transcription factors are included in the analysis.
:param TF_ids: List of transcription factor identifiers.
:param protein_dict: Dictionary mapping TF_ids to their protein data.
:param psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.
:param psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.
:param relevant_TFs: Set of relevant transcription factors.
:return: filtered TF_ids, protein_dict, psite_dict, psite_labels_dict</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_TF</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">relevant_TFs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter transcription factors to only those present in the relevant_TFs set.</span>
<span class="sd">    This function returns the filtered TF_ids and their corresponding protein and phosphorylation site data.</span>
<span class="sd">    This is important for ensuring that only relevant transcription factors are included in the analysis.</span>
<span class="sd">    :param TF_ids: List of transcription factor identifiers.</span>
<span class="sd">    :param protein_dict: Dictionary mapping TF_ids to their protein data.</span>
<span class="sd">    :param psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.</span>
<span class="sd">    :param psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.</span>
<span class="sd">    :param relevant_TFs: Set of relevant transcription factors.</span>
<span class="sd">    :return: filtered TF_ids, protein_dict, psite_dict, psite_labels_dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">TF_ids_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">relevant_TFs</span><span class="p">]</span>
    <span class="n">protein_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">protein_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids_filtered</span><span class="p">}</span>
    <span class="n">psite_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">psite_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids_filtered</span><span class="p">}</span>
    <span class="n">psite_labels_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">psite_labels_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids_filtered</span><span class="p">}</span>
    <span class="k">return</span> <span class="n">TF_ids_filtered</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.filter.filter_mrna" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_mrna</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Filter mRNA genes to only those with regulators present in the regulation map.
This function returns the filtered mRNA_ids and their corresponding expression matrix.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param mRNA_mat: Matrix of mRNA expression data.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:return: filtered_mRNA_ids, filtered_mRNA_mat</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_mrna</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter mRNA genes to only those with regulators present in the regulation map.</span>
<span class="sd">    This function returns the filtered mRNA_ids and their corresponding expression matrix.</span>

<span class="sd">    :param mRNA_ids: List of mRNA gene identifiers.</span>
<span class="sd">    :param mRNA_mat: Matrix of mRNA expression data.</span>
<span class="sd">    :param reg_map: Regulation map, mapping mRNA genes to their regulators.</span>
<span class="sd">    :return: filtered_mRNA_ids, filtered_mRNA_mat</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filtered_indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">)</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">reg_map</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">filtered_indices</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No mRNA with regulators found.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">mRNA_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">filtered_indices</span><span class="p">],</span> <span class="n">mRNA_mat</span><span class="p">[</span><span class="n">filtered_indices</span><span class="p">,</span> <span class="p">:]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.filter.load_raw_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_raw_data</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load raw data from files.
This includes mRNA data, TF data, and regulation maps.
The function returns the following:
- mRNA_ids: List of mRNA gene identifiers.
- mRNA_mat: Matrix of mRNA expression data.
- mRNA_time_cols: Time points for mRNA data.
- TF_ids: List of transcription factor identifiers.
- protein_dict: Dictionary mapping TF_ids to their protein data.
- psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.
- psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.
- TF_time_cols: Time points for TF data.
- reg_map: Regulation map, mapping mRNA genes to their regulators.</p>
<p>:return: mRNA_ids, mRNA_mat, mRNA_time_cols, TF_ids, protein_dict, psite_dict, psite_labels_dict, TF_time_cols, reg_map</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_raw_data</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load raw data from files.</span>
<span class="sd">    This includes mRNA data, TF data, and regulation maps.</span>
<span class="sd">    The function returns the following:</span>
<span class="sd">    - mRNA_ids: List of mRNA gene identifiers.</span>
<span class="sd">    - mRNA_mat: Matrix of mRNA expression data.</span>
<span class="sd">    - mRNA_time_cols: Time points for mRNA data.</span>
<span class="sd">    - TF_ids: List of transcription factor identifiers.</span>
<span class="sd">    - protein_dict: Dictionary mapping TF_ids to their protein data.</span>
<span class="sd">    - psite_dict: Dictionary mapping TF_ids to their phosphorylation site data.</span>
<span class="sd">    - psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels.</span>
<span class="sd">    - TF_time_cols: Time points for TF data.</span>
<span class="sd">    - reg_map: Regulation map, mapping mRNA genes to their regulators.</span>

<span class="sd">    :return: mRNA_ids, mRNA_mat, mRNA_time_cols, TF_ids, protein_dict, psite_dict, psite_labels_dict, TF_time_cols, reg_map</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">mRNA_time_cols</span> <span class="o">=</span> <span class="n">load_mRNA_data</span><span class="p">()</span>
    <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">TF_time_cols</span> <span class="o">=</span> <span class="n">load_TF_data</span><span class="p">()</span>
    <span class="n">reg_map</span> <span class="o">=</span> <span class="n">load_regulation</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">mRNA_time_cols</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">TF_time_cols</span><span class="p">,</span> <span class="n">reg_map</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.optcon.filter.update_regulations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_regulations</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Update the regulation map to only include relevant transcription factors.
This function modifies the reg_map in place and returns a set of relevant transcription factors.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:param TF_ids: List of transcription factor identifiers.
:return: relevant_TFs: Set of relevant transcription factors.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">update_regulations</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update the regulation map to only include relevant transcription factors.</span>
<span class="sd">    This function modifies the reg_map in place and returns a set of relevant transcription factors.</span>

<span class="sd">    :param mRNA_ids: List of mRNA gene identifiers.</span>
<span class="sd">    :param reg_map: Regulation map, mapping mRNA genes to their regulators.</span>
<span class="sd">    :param TF_ids: List of transcription factor identifiers.</span>
<span class="sd">    :return: relevant_TFs: Set of relevant transcription factors.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">relevant_TFs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">mRNA_ids</span><span class="p">:</span>
        <span class="n">regs</span> <span class="o">=</span> <span class="n">reg_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[])</span>
        <span class="n">regs_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">regs</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids</span><span class="p">]</span>
        <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">regs_filtered</span>
        <span class="n">relevant_TFs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">regs_filtered</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">relevant_TFs</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.utils.iodata"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;report.html&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a single global report HTML file from all gene folders inside the results directory.</span>

<span class="sd">    For each gene folder (e.g. &quot;ABL2&quot;), the report will include:</span>
<span class="sd">      - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.</span>
<span class="sd">      - Each plot is confined to a fixed size of 900px by 900px.</span>
<span class="sd">      - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</span>

<span class="sd">    Args:</span>
<span class="sd">        results_dir (str): Path to the root results directory.</span>
<span class="sd">        output_file (str): Name of the generated global report file (placed inside results_dir).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Gather gene folders (skip &quot;General&quot; and &quot;logs&quot;)</span>
    <span class="n">gene_folders</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">results_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="ow">and</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;General&quot;</span><span class="p">,</span> <span class="s2">&quot;logs&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Build HTML content with updated CSS for spacing.</span>
    <span class="n">html_parts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;meta charset=&#39;UTF-8&#39;&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;title&gt;Estimation Report&lt;/title&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;body { font-family: Arial, sans-serif; margin: 20px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h1 { color: #333; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h2 { color: #555; font-size: 1.8em; border-bottom: 1px solid #ccc; padding-bottom: 5px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h3 { color: #666; font-size: 1.4em; margin-top: 10px; margin-bottom: 10px; }&quot;</span><span class="p">,</span>
        <span class="c1"># /* CSS grid for plots: two per row, fixed size 500px x 500px, extra space between rows */</span>
        <span class="s2">&quot;.plot-container {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  display: grid;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  grid-template-columns: repeat(2, 500px);&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  column-gap: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  row-gap: 40px;&quot;</span><span class="p">,</span> <span class="c1"># /* extra vertical gap */</span>
        <span class="s2">&quot;  justify-content: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;.plot-item {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;img, iframe {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  object-fit: contain;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: none;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="c1"># /* Data tables: full width, one per row */</span>
        <span class="s2">&quot;.data-table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 50%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border-collapse: collapse;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-top: 10px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;th, td {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: 1px solid #ccc;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  padding: 8px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  text-align: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;body&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;h1&gt;mRNA-TF Optimization Report&lt;/h1&gt;&quot;</span>
    <span class="p">]</span>

    <span class="c1"># For each gene folder, create a section in the report.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_folders</span><span class="p">):</span>
        <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;mRNA: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>

        <span class="c1"># Create grid container for fixed-size plots.</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;div class=&quot;plot-container&quot;&gt;&#39;</span><span class="p">)</span>
        <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                    <span class="n">rel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;plot-item&quot;&gt;&lt;h3&gt;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;img src=&quot;</span><span class="si">{</span><span class="n">rel_path</span><span class="si">}</span><span class="s1">&quot; alt=&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&quot;&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;/div&gt;&#39;</span><span class="p">)</span>  <span class="c1"># End of plot container</span>

        <span class="c1"># Data tables: display XLSX or CSV files from the gene folder, one per row.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.xlsx&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                    <span class="n">table_html</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">border</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;</span><span class="si">{</span><span class="n">table_html</span><span class="si">}</span><span class="s1">&lt;/div&gt;&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;p&gt;Error reading </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&lt;/p&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>

    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span>

    <span class="c1"># Write the report into the results directory.</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">html_parts</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.
The function converts the duration into hours, minutes, or seconds,
depending on the length of the duration.
Args:
    seconds (float): Duration in seconds.
Returns:
    str: Formatted duration string.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format a duration in seconds into a human-readable string.</span>
<span class="sd">    The function converts the duration into hours, minutes, or seconds,</span>
<span class="sd">    depending on the length of the duration.</span>
<span class="sd">    Args:</span>
<span class="sd">        seconds (float): Duration in seconds.</span>
<span class="sd">    Returns:</span>
<span class="sd">        str: Formatted duration string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">60</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> sec&quot;</span>
    <span class="k">elif</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">3600</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> min&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hr&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.load_TF_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_TF_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load TF data from a CSV file.
The file is expected to have a "GeneID" column, a "Psite" column, and time columns for protein data.
The function returns a list of TF gene identifiers, a dictionary mapping TF_ids to their protein data,
a dictionary mapping TF_ids to their phosphorylation site data, a dictionary mapping TF_ids to their phosphorylation site labels,
and the time columns.
The TF_ids are converted to strings, and the protein data is stored in a dictionary with TF_ids as keys.
The phosphorylation site data is stored in a list for each TF_id, and the phosphorylation site labels are also stored in a list.
The time columns are extracted from the DataFrame, excluding the "GeneID" and "Psite" columns.
The function returns:
    - TF_ids: List of TF gene identifiers (strings).
    - protein_dict: Dictionary mapping TF_ids to their protein data (numpy arrays).
    - psite_dict: Dictionary mapping TF_ids to their phosphorylation site data (lists of numpy arrays).
    - psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels (lists of strings).
    - time_cols: List of time columns (excluding "GeneID" and "Psite").</p>
<p>:param filename: str
:return: TF_ids, protein_dict, psite_dict, psite_labels_dict, time_cols</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_TF_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load TF data from a CSV file.</span>
<span class="sd">    The file is expected to have a &quot;GeneID&quot; column, a &quot;Psite&quot; column, and time columns for protein data.</span>
<span class="sd">    The function returns a list of TF gene identifiers, a dictionary mapping TF_ids to their protein data,</span>
<span class="sd">    a dictionary mapping TF_ids to their phosphorylation site data, a dictionary mapping TF_ids to their phosphorylation site labels,</span>
<span class="sd">    and the time columns.</span>
<span class="sd">    The TF_ids are converted to strings, and the protein data is stored in a dictionary with TF_ids as keys.</span>
<span class="sd">    The phosphorylation site data is stored in a list for each TF_id, and the phosphorylation site labels are also stored in a list.</span>
<span class="sd">    The time columns are extracted from the DataFrame, excluding the &quot;GeneID&quot; and &quot;Psite&quot; columns.</span>
<span class="sd">    The function returns:</span>
<span class="sd">        - TF_ids: List of TF gene identifiers (strings).</span>
<span class="sd">        - protein_dict: Dictionary mapping TF_ids to their protein data (numpy arrays).</span>
<span class="sd">        - psite_dict: Dictionary mapping TF_ids to their phosphorylation site data (lists of numpy arrays).</span>
<span class="sd">        - psite_labels_dict: Dictionary mapping TF_ids to their phosphorylation site labels (lists of strings).</span>
<span class="sd">        - time_cols: List of time columns (excluding &quot;GeneID&quot; and &quot;Psite&quot;).</span>

<span class="sd">    :param filename: str</span>
<span class="sd">    :return: TF_ids, protein_dict, psite_dict, psite_labels_dict, time_cols</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">protein_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">psite_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">psite_labels_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">tf</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">psite</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Psite&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">]]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">tf</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">protein_dict</span><span class="p">:</span>
            <span class="n">protein_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span>
            <span class="n">psite_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">psite_labels_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">psite_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
            <span class="n">psite_labels_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psite</span><span class="p">)</span>
    <span class="n">TF_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">protein_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">protein_dict</span><span class="p">,</span> <span class="n">psite_dict</span><span class="p">,</span> <span class="n">psite_labels_dict</span><span class="p">,</span> <span class="n">time_cols</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.load_mRNA_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_mRNA_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load mRNA data from a CSV file.
The file is expected to have a "GeneID" column and time columns for expression data.
The function returns a list of mRNA gene identifiers, a matrix of expression data,
and the time columns.
The mRNA_ids are converted to strings, and the expression data is converted to a numpy array of floats.
The time columns are extracted from the DataFrame, excluding the "GeneID" column.
The function returns:
    - mRNA_ids: List of mRNA gene identifiers (strings).
    - mRNA_mat: Numpy array of expression data (floats).
    - time_cols: List of time columns (excluding "GeneID").</p>
<p>:param filename: str
:return: mRNA_ids, mRNA_mat, time_cols</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_mRNA_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load mRNA data from a CSV file.</span>
<span class="sd">    The file is expected to have a &quot;GeneID&quot; column and time columns for expression data.</span>
<span class="sd">    The function returns a list of mRNA gene identifiers, a matrix of expression data,</span>
<span class="sd">    and the time columns.</span>
<span class="sd">    The mRNA_ids are converted to strings, and the expression data is converted to a numpy array of floats.</span>
<span class="sd">    The time columns are extracted from the DataFrame, excluding the &quot;GeneID&quot; column.</span>
<span class="sd">    The function returns:</span>
<span class="sd">        - mRNA_ids: List of mRNA gene identifiers (strings).</span>
<span class="sd">        - mRNA_mat: Numpy array of expression data (floats).</span>
<span class="sd">        - time_cols: List of time columns (excluding &quot;GeneID&quot;).</span>

<span class="sd">    :param filename: str</span>
<span class="sd">    :return: mRNA_ids, mRNA_mat, time_cols</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">mRNA_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">]</span>
    <span class="n">mRNA_mat</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">mRNA_mat</span><span class="p">,</span> <span class="n">time_cols</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load regulation data from a CSV file.
The file is expected to have "Source" and "Target" columns.
The function returns a dictionary mapping mRNA gene identifiers to their regulators (TFs).
The mRNA gene identifiers and TF identifiers are converted to strings and stripped of whitespace.
The function returns:
    - reg_map: Dictionary mapping mRNA gene identifiers to their regulators (TFs).</p>
<p>:param filename: str
:return: reg_map</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load regulation data from a CSV file.</span>
<span class="sd">    The file is expected to have &quot;Source&quot; and &quot;Target&quot; columns.</span>
<span class="sd">    The function returns a dictionary mapping mRNA gene identifiers to their regulators (TFs).</span>
<span class="sd">    The mRNA gene identifiers and TF identifiers are converted to strings and stripped of whitespace.</span>
<span class="sd">    The function returns:</span>
<span class="sd">        - reg_map: Dictionary mapping mRNA gene identifiers to their regulators (TFs).</span>

<span class="sd">    :param filename: str</span>
<span class="sd">    :return: reg_map</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">reg_map</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">mrna</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Source&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">tf</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Target&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">mrna</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">:</span>
            <span class="n">reg_map</span><span class="p">[</span><span class="n">mrna</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">tf</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">mrna</span><span class="p">]:</span>
            <span class="n">reg_map</span><span class="p">[</span><span class="n">mrna</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tf</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">reg_map</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Organizes output files from multiple directories into separate folders for each protein.
Files are moved into folders named after the protein identifier extracted from the filename.
Remaining files are moved to a "General" folder.
Args:
    directories (str): List of directories to organize.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Organizes output files from multiple directories into separate folders for each protein.</span>
<span class="sd">    Files are moved into folders named after the protein identifier extracted from the filename.</span>
<span class="sd">    Remaining files are moved to a &quot;General&quot; folder.</span>
<span class="sd">    Args:</span>
<span class="sd">        directories (str): List of directories to organize.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Za-z0-9]+)_.*\.(json|svg|png|html|csv|xlsx)$&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">directory</span> <span class="ow">in</span> <span class="n">directories</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">directory</span><span class="si">}</span><span class="s2">&#39; is not a valid directory. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Move files matching the protein pattern.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">protein_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">protein_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">protein</span><span class="p">)</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>

        <span class="c1"># After protein files have been moved, move remaining files to a &quot;General&quot; folder.</span>
        <span class="n">general_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="s2">&quot;General&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.evol.utils.params"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.compute_beta_indices" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_beta_indices</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Compute the starting indices for the beta parameters for each TF.
The beta parameters are stored in a flat array, and this function computes
the starting index for each TF based on the number of phosphorylation sites.
The starting index for the beta parameters of TF i is given by:
beta_start_indices[i] = sum(1 + num_psites[j] for j in range(i))
This function is used to extract the beta parameters from the flat array
during the optimization process.</p>
<p>:param num_psites: array of integers indicating the number of phosphorylation sites for each TF
:param n_TF: number of TFs
:return: beta_start_indices: array of integers indicating the starting index for each TF
         cum: total number of beta parameters</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_beta_indices</span><span class="p">(</span><span class="n">num_psites</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute the starting indices for the beta parameters for each TF.</span>
<span class="sd">    The beta parameters are stored in a flat array, and this function computes</span>
<span class="sd">    the starting index for each TF based on the number of phosphorylation sites.</span>
<span class="sd">    The starting index for the beta parameters of TF i is given by:</span>
<span class="sd">    beta_start_indices[i] = sum(1 + num_psites[j] for j in range(i))</span>
<span class="sd">    This function is used to extract the beta parameters from the flat array</span>
<span class="sd">    during the optimization process.</span>

<span class="sd">    :param num_psites: array of integers indicating the number of phosphorylation sites for each TF</span>
<span class="sd">    :param n_TF: number of TFs</span>
<span class="sd">    :return: beta_start_indices: array of integers indicating the starting index for each TF</span>
<span class="sd">             cum: total number of beta parameters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">beta_start_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="n">cum</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">cum</span>
        <span class="n">cum</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">cum</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.create_bounds" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_bounds</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_beta_total</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create the lower and upper bounds for the optimization variables.
The lower bounds are set to 0 for the alpha parameters and lb for the beta parameters.
The upper bounds are set to 1 for the alpha parameters and ub for the beta parameters.
The bounds are used to constrain the optimization process and ensure that the parameters
are within a reasonable range.
The bounds are stored in two separate arrays: xl and xu.
The xl array contains the lower bounds for each parameter,
and the xu array contains the upper bounds.</p>
<p>:param n_alpha: number of alpha parameters
:param n_beta_total: total number of beta parameters
:param lb: lower bound for beta parameters
:param ub: upper bound for beta parameters
:return: xl: lower bounds array
         xu: upper bounds array</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_bounds</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_beta_total</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the lower and upper bounds for the optimization variables.</span>
<span class="sd">    The lower bounds are set to 0 for the alpha parameters and lb for the beta parameters.</span>
<span class="sd">    The upper bounds are set to 1 for the alpha parameters and ub for the beta parameters.</span>
<span class="sd">    The bounds are used to constrain the optimization process and ensure that the parameters</span>
<span class="sd">    are within a reasonable range.</span>
<span class="sd">    The bounds are stored in two separate arrays: xl and xu.</span>
<span class="sd">    The xl array contains the lower bounds for each parameter,</span>
<span class="sd">    and the xu array contains the upper bounds.</span>

<span class="sd">    :param n_alpha: number of alpha parameters</span>
<span class="sd">    :param n_beta_total: total number of beta parameters</span>
<span class="sd">    :param lb: lower bound for beta parameters</span>
<span class="sd">    :param ub: upper bound for beta parameters</span>
<span class="sd">    :return: xl: lower bounds array</span>
<span class="sd">             xu: upper bounds array</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">xl</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">),</span> <span class="n">lb</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_beta_total</span><span class="p">)])</span>
    <span class="n">xu</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">),</span> <span class="n">ub</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_beta_total</span><span class="p">)])</span>
    <span class="k">return</span> <span class="n">xl</span><span class="p">,</span> <span class="n">xu</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.create_initial_guess" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_initial_guess</span><span class="p">(</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create the initial guess for the optimization variables.
The initial guess is a flat array containing the alpha and beta parameters.
The alpha parameters are initialized to 1.0 / n_reg for each mRNA-regulator pair.
The beta parameters are initialized to 1.0 for the protein and 1.0 / (1 + num_psites[i])
for each phosphorylation site of TF i.
The beta parameters for TFs with no phosphorylation sites are initialized to 1.0.
The initial guess is used as the starting point for the optimization process.
The initial guess is a flat array of length n_alpha + n_beta_total,
where n_alpha is the number of alpha parameters and n_beta_total is the total number of beta parameters.
The alpha parameters are stored in the first n_alpha elements of the array,
and the beta parameters are stored in the remaining elements.
The beta parameters are stored in the order of the TFs, with the protein parameter first
followed by the phosphorylation site parameters.
The beta parameters for TFs with no phosphorylation sites are stored as a single value.
The beta parameters for TFs with phosphorylation sites are stored as a vector of length 1 + num_psites[i].</p>
<p>:param n_mRNA:
:param n_reg:
:param n_TF:
:param num_psites:
:param no_psite_tf:
:return: initial guess array, n_alpha</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_initial_guess</span><span class="p">(</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the initial guess for the optimization variables.</span>
<span class="sd">    The initial guess is a flat array containing the alpha and beta parameters.</span>
<span class="sd">    The alpha parameters are initialized to 1.0 / n_reg for each mRNA-regulator pair.</span>
<span class="sd">    The beta parameters are initialized to 1.0 for the protein and 1.0 / (1 + num_psites[i])</span>
<span class="sd">    for each phosphorylation site of TF i.</span>
<span class="sd">    The beta parameters for TFs with no phosphorylation sites are initialized to 1.0.</span>
<span class="sd">    The initial guess is used as the starting point for the optimization process.</span>
<span class="sd">    The initial guess is a flat array of length n_alpha + n_beta_total,</span>
<span class="sd">    where n_alpha is the number of alpha parameters and n_beta_total is the total number of beta parameters.</span>
<span class="sd">    The alpha parameters are stored in the first n_alpha elements of the array,</span>
<span class="sd">    and the beta parameters are stored in the remaining elements.</span>
<span class="sd">    The beta parameters are stored in the order of the TFs, with the protein parameter first</span>
<span class="sd">    followed by the phosphorylation site parameters.</span>
<span class="sd">    The beta parameters for TFs with no phosphorylation sites are stored as a single value.</span>
<span class="sd">    The beta parameters for TFs with phosphorylation sites are stored as a vector of length 1 + num_psites[i].</span>

<span class="sd">    :param n_mRNA:</span>
<span class="sd">    :param n_reg:</span>
<span class="sd">    :param n_TF:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param no_psite_tf:</span>
<span class="sd">    :return: initial guess array, n_alpha</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_mRNA</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="n">x0_alpha</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">n_reg</span><span class="p">)</span>
    <span class="n">x0_beta_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">no_psite_tf</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
            <span class="n">x0_beta_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="mf">1.0</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">x0_beta_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">length</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span><span class="p">)</span>
    <span class="n">x0_beta</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x0_beta_list</span><span class="p">)</span>
    <span class="n">x0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">x0_alpha</span><span class="p">,</span> <span class="n">x0_beta</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">x0</span><span class="p">,</span> <span class="n">n_alpha</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.create_no_psite_array" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_no_psite_array</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create an array indicating whether each TF has no phosphorylation sites.
A TF is considered to have no phosphorylation sites if:
1. The number of phosphorylation sites is zero.
2. All labels for the phosphorylation sites are empty strings.
This function is used to determine the initial guess for the beta parameters
in the optimization process.</p>
<p>:param n_TF:
:param num_psites:
:param psite_labels_arr:
:return: array of booleans indicating no phosphorylation sites for each TF</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_no_psite_array</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create an array indicating whether each TF has no phosphorylation sites.</span>
<span class="sd">    A TF is considered to have no phosphorylation sites if:</span>
<span class="sd">    1. The number of phosphorylation sites is zero.</span>
<span class="sd">    2. All labels for the phosphorylation sites are empty strings.</span>
<span class="sd">    This function is used to determine the initial guess for the beta parameters</span>
<span class="sd">    in the optimization process.</span>

<span class="sd">    :param n_TF:</span>
<span class="sd">    :param num_psites:</span>
<span class="sd">    :param psite_labels_arr:</span>
<span class="sd">    :return: array of booleans indicating no phosphorylation sites for each TF</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">all</span><span class="p">(</span><span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                     <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">)])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.extract_best_solution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_best_solution</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Extract the best solution from the optimization results.
This function finds the best solution based on the Pareto front,
which is a set of non-dominated solutions in the objective space.
The best solution is determined by minimizing the weighted sum of the objectives.
The alpha parameters are reshaped into a matrix of shape (n_mRNA, n_reg),
and the beta parameters are extracted based on the starting indices and
the number of phosphorylation sites for each TF.
The beta parameters are stored in a list of arrays, where each array corresponds to a TF
and contains the protein parameter and the phosphorylation site parameters.
The function returns the final alpha and beta parameters, the best objectives,
and the final decision variables.</p>
<p>:param res: optimization results
:param n_alpha: number of alpha parameters
:param n_mRNA: number of mRNAs
:param n_reg: number of regulators
:param n_TF: number of transcription factors
:param num_psites: number of phosphorylation sites for each TF
:param beta_start_indices: starting indices for the beta parameters</p>
<p>:return: final_alpha: final alpha parameters (n_mRNA x n_reg)
            final_beta: final beta parameters (n_TF x (1 + num_psites))
            best_objectives: best objectives (3 objectives)
            final_x: final decision variables (flat array)</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">extract_best_solution</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the best solution from the optimization results.</span>
<span class="sd">    This function finds the best solution based on the Pareto front,</span>
<span class="sd">    which is a set of non-dominated solutions in the objective space.</span>
<span class="sd">    The best solution is determined by minimizing the weighted sum of the objectives.</span>
<span class="sd">    The alpha parameters are reshaped into a matrix of shape (n_mRNA, n_reg),</span>
<span class="sd">    and the beta parameters are extracted based on the starting indices and</span>
<span class="sd">    the number of phosphorylation sites for each TF.</span>
<span class="sd">    The beta parameters are stored in a list of arrays, where each array corresponds to a TF</span>
<span class="sd">    and contains the protein parameter and the phosphorylation site parameters.</span>
<span class="sd">    The function returns the final alpha and beta parameters, the best objectives,</span>
<span class="sd">    and the final decision variables.</span>

<span class="sd">    :param res: optimization results</span>
<span class="sd">    :param n_alpha: number of alpha parameters</span>
<span class="sd">    :param n_mRNA: number of mRNAs</span>
<span class="sd">    :param n_reg: number of regulators</span>
<span class="sd">    :param n_TF: number of transcription factors</span>
<span class="sd">    :param num_psites: number of phosphorylation sites for each TF</span>
<span class="sd">    :param beta_start_indices: starting indices for the beta parameters</span>

<span class="sd">    :return: final_alpha: final alpha parameters (n_mRNA x n_reg)</span>
<span class="sd">                final_beta: final beta parameters (n_TF x (1 + num_psites))</span>
<span class="sd">                best_objectives: best objectives (3 objectives)</span>
<span class="sd">                final_x: final decision variables (flat array)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pareto_front</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">ind</span><span class="o">.</span><span class="n">F</span> <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">res</span><span class="o">.</span><span class="n">pop</span><span class="p">])</span>
    <span class="c1"># Scoring the Pareto front</span>
    <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">weights</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">weights</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">pareto_front</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">])</span>
    <span class="n">best_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="n">best_solution</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">pop</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>
    <span class="n">best_objectives</span> <span class="o">=</span> <span class="n">pareto_front</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>
    <span class="n">final_x</span> <span class="o">=</span> <span class="n">best_solution</span><span class="o">.</span><span class="n">X</span>
    <span class="n">final_alpha</span> <span class="o">=</span> <span class="n">final_x</span><span class="p">[:</span><span class="n">n_alpha</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">n_mRNA</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">))</span>
    <span class="n">final_beta</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">final_beta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">final_x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">start</span> <span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">start</span> <span class="o">+</span> <span class="n">length</span><span class="p">])</span>
    <span class="n">final_beta</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">final_beta</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">best_objectives</span><span class="p">,</span> <span class="n">final_x</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.get_parallel_runner" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_parallel_runner</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Get a parallel runner for multi-threading.
This function uses the lscpu command to determine the number of threads available
on the system and creates a ThreadPool with that number of threads.
The ThreadPool is used to parallelize the optimization process and speed up the computation.</p>
<p>:return: runner: StarmapParallelization object for parallel execution
            pool: ThreadPool object for managing threads</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_parallel_runner</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a parallel runner for multi-threading.</span>
<span class="sd">    This function uses the lscpu command to determine the number of threads available</span>
<span class="sd">    on the system and creates a ThreadPool with that number of threads.</span>
<span class="sd">    The ThreadPool is used to parallelize the optimization process and speed up the computation.</span>

<span class="sd">    :return: runner: StarmapParallelization object for parallel execution</span>
<span class="sd">                pool: ThreadPool object for managing threads</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_threads_cmd</span> <span class="o">=</span> <span class="s2">&quot;lscpu -p | grep -v &#39;^#&#39; | wc -l&quot;</span>
    <span class="n">n_threads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">n_threads_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">pool</span> <span class="o">=</span> <span class="n">ThreadPool</span><span class="p">(</span><span class="n">n_threads</span><span class="p">)</span>
    <span class="n">runner</span> <span class="o">=</span> <span class="n">StarmapParallelization</span><span class="p">(</span><span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">runner</span><span class="p">,</span> <span class="n">pool</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.print_alpha_mapping" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_alpha_mapping</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to mRNAs with their corresponding alpha values.
This function iterates through the mRNA IDs and their corresponding regulators,
and logs the TFs that are present in the final alpha matrix.
The alpha values are printed for each TF that regulates the mRNA.</p>
<p>:param mRNA_ids: List of mRNA gene identifiers.
:param reg_map: Regulation map, mapping mRNA genes to their regulators.
:param TF_ids: List of transcription factor identifiers.
:param final_alpha: Final alpha parameters (n_mRNA x n_reg).</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">print_alpha_mapping</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print the mapping of transcription factors (TFs) to mRNAs with their corresponding alpha values.</span>
<span class="sd">    This function iterates through the mRNA IDs and their corresponding regulators,</span>
<span class="sd">    and logs the TFs that are present in the final alpha matrix.</span>
<span class="sd">    The alpha values are printed for each TF that regulates the mRNA.</span>

<span class="sd">    :param mRNA_ids: List of mRNA gene identifiers.</span>
<span class="sd">    :param reg_map: Regulation map, mapping mRNA genes to their regulators.</span>
<span class="sd">    :param TF_ids: List of transcription factor identifiers.</span>
<span class="sd">    :param final_alpha: Final alpha parameters (n_mRNA x n_reg).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping of TFs to mRNAs ( values):&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mrna</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mRNA_ids</span><span class="p">):</span>
        <span class="n">actual_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">mrna</span><span class="p">]</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">TF_ids</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;mRNA </span><span class="si">{</span><span class="n">mrna</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">actual_tfs</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TF   </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">final_alpha</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.evol.utils.params.print_beta_mapping" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">print_beta_mapping</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to their beta parameters.
This function iterates through the TF IDs and their corresponding beta values,
and logs the beta values for each TF.
The beta values are printed for the protein and each phosphorylation site,
with appropriate labels.</p>
<p>:param TF_ids: List of transcription factor identifiers.
:param final_beta: Final beta parameters (n_TF x (1 + num_psites)).
:param psite_labels_arr: Array of phosphorylation site labels for each TF.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/evol/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">print_beta_mapping</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print the mapping of transcription factors (TFs) to their beta parameters.</span>
<span class="sd">    This function iterates through the TF IDs and their corresponding beta values,</span>
<span class="sd">    and logs the beta values for each TF.</span>
<span class="sd">    The beta values are printed for the protein and each phosphorylation site,</span>
<span class="sd">    with appropriate labels.</span>

<span class="sd">    :param TF_ids: List of transcription factor identifiers.</span>
<span class="sd">    :param final_beta: Final beta parameters (n_TF x (1 + num_psites)).</span>
<span class="sd">    :param psite_labels_arr: Array of phosphorylation site labels for each TF.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping of TFs to  parameters:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">TF_ids</span><span class="p">):</span>
        <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">final_beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   TF </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)):</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="n">q</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">label</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PSite</span><span class="si">{</span><span class="n">q</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">beta_vec</span><span class="p">[</span><span class="n">q</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.fitanalysis.helper"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="tfopt.fitanalysis.helper.Plotter" class="doc doc-heading">
            <code>Plotter</code>


</h2>


    <div class="doc doc-contents ">


        <p>A class to plot various analysis results from an Excel file.
The class provides methods to visualize the alpha and beta values,
residuals, observed and estimated values, and other metrics.</p>
<p>The plots include:
- Alpha distribution
- Beta bar plots
- Heatmap of absolute residuals
- Goodness of fit
- Kullback-Leibler divergence
- PCA
- Box plots for alpha and beta values
- CDF for alpha and beta values
- Time-wise residuals</p>







              <details class="quote">
                <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Plotter</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to plot various analysis results from an Excel file.</span>
<span class="sd">    The class provides methods to visualize the alpha and beta values,</span>
<span class="sd">    residuals, observed and estimated values, and other metrics.</span>

<span class="sd">    The plots include:</span>
<span class="sd">    - Alpha distribution</span>
<span class="sd">    - Beta bar plots</span>
<span class="sd">    - Heatmap of absolute residuals</span>
<span class="sd">    - Goodness of fit</span>
<span class="sd">    - Kullback-Leibler divergence</span>
<span class="sd">    - PCA</span>
<span class="sd">    - Box plots for alpha and beta values</span>
<span class="sd">    - CDF for alpha and beta values</span>
<span class="sd">    - Time-wise residuals</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">savepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes the Plotter instance by loading data from the Excel file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">savepath</span> <span class="o">=</span> <span class="n">savepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_data</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">load_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads data from the specified Excel file.</span>
<span class="sd">        The data includes residuals, observed values, estimated values,</span>
<span class="sd">        alpha values, and beta values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Residuals&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Alpha Values&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Beta Values&#39;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_alpha_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the distribution of alpha parameter values grouped by transcription factors (TFs)</span>
<span class="sd">        using a strip plot.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
        <span class="n">all_markers</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;&lt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;*&#39;</span><span class="p">,</span> <span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;x&#39;</span><span class="p">]</span>
        <span class="n">unique_tfs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">marker_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">all_markers</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_markers</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_tfs</span><span class="p">)}</span>

        <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">marker_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">tf</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span>
                <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">],</span>     <span class="c1"># mRNA names</span>
                <span class="n">x</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">][</span><span class="n">mask</span><span class="p">],</span>                     <span class="c1"># Alpha values</span>
                <span class="n">hue</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">],</span>    <span class="c1"># mRNA used for hue</span>
                <span class="n">label</span><span class="o">=</span><span class="n">tf</span><span class="p">,</span>
                <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">,</span>
                <span class="n">dodge</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">size</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
                <span class="n">jitter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span>
                <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span>
            <span class="p">)</span>

        <span class="c1"># Create a custom legend for each TF</span>
        <span class="n">legend_handles</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">mlines</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker_map</span><span class="p">[</span><span class="n">tf</span><span class="p">],</span>
                          <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;None&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">tf</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">unique_tfs</span>
        <span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
            <span class="n">handles</span><span class="o">=</span><span class="n">legend_handles</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="s1">&#39;TFs&#39;</span><span class="p">,</span>
            <span class="n">frameon</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span>
            <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
            <span class="n">title_fontsize</span><span class="o">=</span><span class="s1">&#39;10&#39;</span><span class="p">,</span>
            <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;7&#39;</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Effect of Transcription Factors (TFs) on mRNA Expression&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s2">/mRNA_alpha_group.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_beta_barplots</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Processes the beta values DataFrame and creates a separate bar plot</span>
<span class="sd">        for each unique transcription factor (TF).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Data cleaning for df_beta</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">])</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="s1">&#39;&#39;</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>

        <span class="n">unique_tfs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="c1"># Plot bar plot for each TF</span>
        <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">unique_tfs</span><span class="p">:</span>
            <span class="n">tf_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tf</span><span class="p">]</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
                <span class="n">data</span><span class="o">=</span><span class="n">tf_data</span><span class="p">,</span>
                <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PSite&#39;</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Value&#39;</span><span class="p">,</span>
                <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">,</span>
                <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
                <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span>
            <span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation - Residue Position&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Transcription Factor&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Effect of Phosphorylation on Transcription Factor </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2"> Activity&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/TF_</span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s1">_beta_group.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmap_abs_residuals</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a heatmap of the absolute values of the residuals.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="n">abs_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
        <span class="c1"># Use fixed x tick labels as given in the original code</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
            <span class="n">abs_df</span><span class="p">,</span>
            <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x1&#39;</span><span class="p">,</span> <span class="s1">&#39;x2&#39;</span><span class="p">,</span> <span class="s1">&#39;x3&#39;</span><span class="p">,</span> <span class="s1">&#39;x4&#39;</span><span class="p">,</span> <span class="s1">&#39;x5&#39;</span><span class="p">,</span> <span class="s1">&#39;x6&#39;</span><span class="p">,</span> <span class="s1">&#39;x7&#39;</span><span class="p">,</span> <span class="s1">&#39;x8&#39;</span><span class="p">,</span> <span class="s1">&#39;x9&#39;</span><span class="p">],</span>
            <span class="n">yticklabels</span><span class="o">=</span><span class="n">abs_df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
            <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Absolute Residuals&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time Points&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Residual_Heatmap.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_goodness_of_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a scatter plot comparing observed vs. estimated values,</span>
<span class="sd">        fits a linear regression model, plots the 95% confidence interval,</span>
<span class="sd">        and labels points outside the confidence interval.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

        <span class="n">mRNAs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mRNAs</span><span class="p">):</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="n">label</span><span class="o">=</span><span class="n">mRNA</span><span class="p">,</span>
                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span>
            <span class="p">)</span>

        <span class="c1"># Linear Regression and Confidence Interval Calculation</span>
        <span class="n">X</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span>
        <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
        <span class="n">y_pred</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
        <span class="n">std_error</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>

        <span class="c1"># Flatten arrays for plotting</span>
        <span class="n">obs_flat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">est_flat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
            <span class="n">obs_flat</span><span class="p">,</span>
            <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="o">-</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">,</span>
            <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="o">+</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span>
        <span class="p">)</span>

        <span class="c1"># Label any points that lie outside the 95% confidence interval</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">y_val</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">obs_flat</span><span class="p">,</span> <span class="n">est_flat</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">mRNAs</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">y_val</span> <span class="o">&lt;</span> <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span> <span class="ow">or</span> <span class="n">y_val</span> <span class="o">&gt;</span> <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">:</span>
                    <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span>
                        <span class="n">x_val</span><span class="p">,</span> <span class="n">y_val</span><span class="p">,</span> <span class="n">mRNAs</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                        <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span>
                        <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;light&#39;</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span>
                    <span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">obs_flat</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Observed Values&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Goodness of Fit&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Goodness_of_Fit.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_kld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the Kullback-Leibler Divergence (KLD) for each mRNA.</span>
<span class="sd">        The KLD is calculated between the observed and estimated distributions</span>
<span class="sd">        of the mRNA expression levels.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Normalize observed and estimated values</span>
        <span class="n">normalized_obs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">normalized_est</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="c1"># Calculate KLD for each mRNA</span>
        <span class="n">kld</span> <span class="o">=</span> <span class="n">normalized_obs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">entropy</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">normalized_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Create a DataFrame for KLD values</span>
        <span class="n">kld_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mRNA&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;KL&#39;</span><span class="p">:</span> <span class="n">kld</span><span class="o">.</span><span class="n">values</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
        <span class="c1"># Sort KLD values by mRNA</span>
        <span class="n">kld_by_gene</span> <span class="o">=</span> <span class="n">kld_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;KL&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># Plot the KLD values</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="c1"># Add horizontal bar plot for KLD values with different color for bars above 0.03</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">kld_by_gene</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.03</span><span class="p">],</span> <span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">][</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.03</span><span class="p">],</span>
                 <span class="n">color</span><span class="o">=</span><span class="s1">&#39;coral&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">kld_by_gene</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.03</span><span class="p">],</span> <span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">][</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.03</span><span class="p">],</span>
                 <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cornflowerblue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Kullback-Leibler Divergence&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/KLD.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a PCA (Principal Component Analysis) of the observed and estimated values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Standardize the data</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
        <span class="n">scaled_data</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="p">)</span>
        <span class="c1"># Perform PCA</span>
        <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">pca_result</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">)</span>
        <span class="c1"># Create a DataFrame for PCA results</span>
        <span class="n">pca_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pca_result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="c1"># Perform KMeans clustering</span>
        <span class="n">kmeans</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">kmeans</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">pca_df</span><span class="p">)</span>
        <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kmeans</span><span class="o">.</span><span class="n">labels_</span>
        <span class="c1"># Add clusters to the PCA plot</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">],</span> <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC2&#39;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="c1"># Plot clusters</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pca_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;PC2&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;Cluster&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Set1&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pca_df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">mRNA</span><span class="p">,</span> <span class="p">(</span><span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
                         <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">annotation_clip</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Principal Component Analysis&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;PC1&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;PC2&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/PCA.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_boxplot_alpha</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a boxplot of the alpha values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Alpha Values&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Box_Plot_Alpha.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_boxplot_beta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots a boxplot of the beta values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Beta Values&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Beta Value&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Phosphorylation - Residue Position&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Box_Plot_Beta.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_cdf_alpha</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the cumulative distribution function (CDF) of the alpha values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">stat</span><span class="o">=</span><span class="s1">&#39;proportion&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Cumulative Probability&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/CDF_Alpha.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_cdf_beta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the cumulative distribution function (CDF) of the beta values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">stat</span><span class="o">=</span><span class="s1">&#39;proportion&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Beta Value&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Cumulative Probability&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/CDF_Beta.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_time_wise_residuals</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots the residuals over time for each mRNA.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="c1"># Generate a colormap for unique mRNAs with residuals &gt; 0.5</span>
        <span class="n">unique_colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab10</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)))</span>
        <span class="n">default_color</span> <span class="o">=</span> <span class="s1">&#39;lightgray&#39;</span>
        <span class="c1"># Calculate a single mean absolute value across all time points for all mRNAs</span>
        <span class="n">mean_residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">mean_residuals</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]):</span>
                <span class="c1"># Assign a unique color for mRNAs with residuals &gt; 0.5</span>
                <span class="n">color</span> <span class="o">=</span> <span class="n">unique_colors</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_colors</span><span class="p">)]</span>
                <span class="n">linestyle</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Use the default color for other mRNAs</span>
                <span class="n">color</span> <span class="o">=</span> <span class="n">default_color</span>
                <span class="n">linestyle</span> <span class="o">=</span> <span class="s1">&#39;:&#39;</span>
                <span class="n">mRNA</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># Hide label for mRNAs with no residuals &gt; 0.5</span>

            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">mRNA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="n">linestyle</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time Points&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Residuals&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Time_Wise_Residuals.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initializes the Plotter instance by loading data from the Excel file.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">savepath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the Plotter instance by loading data from the Excel file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">filepath</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">savepath</span> <span class="o">=</span> <span class="n">savepath</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">load_data</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.load_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_data</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Loads data from the specified Excel file.
The data includes residuals, observed values, estimated values,
alpha values, and beta values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads data from the specified Excel file.</span>
<span class="sd">    The data includes residuals, observed values, estimated values,</span>
<span class="sd">    alpha values, and beta values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Residuals&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Alpha Values&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;Beta Values&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_alpha_distribution</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the distribution of alpha parameter values grouped by transcription factors (TFs)
using a strip plot.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_alpha_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the distribution of alpha parameter values grouped by transcription factors (TFs)</span>
<span class="sd">    using a strip plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="n">all_markers</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;&lt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;*&#39;</span><span class="p">,</span> <span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;x&#39;</span><span class="p">]</span>
    <span class="n">unique_tfs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">marker_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">all_markers</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_markers</span><span class="p">)]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_tfs</span><span class="p">)}</span>

    <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">marker_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">tf</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span>
            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">],</span>     <span class="c1"># mRNA names</span>
            <span class="n">x</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">][</span><span class="n">mask</span><span class="p">],</span>                     <span class="c1"># Alpha values</span>
            <span class="n">hue</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">],</span>    <span class="c1"># mRNA used for hue</span>
            <span class="n">label</span><span class="o">=</span><span class="n">tf</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">,</span>
            <span class="n">dodge</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
            <span class="n">jitter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span>
        <span class="p">)</span>

    <span class="c1"># Create a custom legend for each TF</span>
    <span class="n">legend_handles</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">mlines</span><span class="o">.</span><span class="n">Line2D</span><span class="p">([],</span> <span class="p">[],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker_map</span><span class="p">[</span><span class="n">tf</span><span class="p">],</span>
                      <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;None&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">tf</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">unique_tfs</span>
    <span class="p">]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
        <span class="n">handles</span><span class="o">=</span><span class="n">legend_handles</span><span class="p">,</span>
        <span class="n">title</span><span class="o">=</span><span class="s1">&#39;TFs&#39;</span><span class="p">,</span>
        <span class="n">frameon</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span>
        <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="n">title_fontsize</span><span class="o">=</span><span class="s1">&#39;10&#39;</span><span class="p">,</span>
        <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;7&#39;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Effect of Transcription Factors (TFs) on mRNA Expression&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s2">/mRNA_alpha_group.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_beta_barplots</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Processes the beta values DataFrame and creates a separate bar plot
for each unique transcription factor (TF).</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_beta_barplots</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the beta values DataFrame and creates a separate bar plot</span>
<span class="sd">    for each unique transcription factor (TF).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Data cleaning for df_beta</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">])</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">])</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="s1">&#39;&#39;</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSite&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>

    <span class="n">unique_tfs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="c1"># Plot bar plot for each TF</span>
    <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">unique_tfs</span><span class="p">:</span>
        <span class="n">tf_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tf</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">tf_data</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PSite&#39;</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Value&#39;</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">,</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Phosphorylation - Residue Position&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Transcription Factor&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Effect of Phosphorylation on Transcription Factor </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2"> Activity&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/TF_</span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s1">_beta_group.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_boxplot_alpha</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the alpha values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_boxplot_alpha</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a boxplot of the alpha values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Alpha Values&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Box_Plot_Alpha.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_boxplot_beta</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the beta values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_boxplot_beta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a boxplot of the beta values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Dark2&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Beta Values&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Beta Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Phosphorylation - Residue Position&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Box_Plot_Beta.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cdf_alpha</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the alpha values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cdf_alpha</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the cumulative distribution function (CDF) of the alpha values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_alpha</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">stat</span><span class="o">=</span><span class="s1">&#39;proportion&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Alpha Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Cumulative Probability&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/CDF_Alpha.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_cdf_beta</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the beta values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cdf_beta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the cumulative distribution function (CDF) of the beta values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">ecdfplot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_beta</span><span class="p">[</span><span class="s1">&#39;Value&#39;</span><span class="p">],</span> <span class="n">stat</span><span class="o">=</span><span class="s1">&#39;proportion&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Beta Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Cumulative Probability&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/CDF_Beta.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_goodness_of_fit</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a scatter plot comparing observed vs. estimated values,
fits a linear regression model, plots the 95% confidence interval,
and labels points outside the confidence interval.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_goodness_of_fit</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a scatter plot comparing observed vs. estimated values,</span>
<span class="sd">    fits a linear regression model, plots the 95% confidence interval,</span>
<span class="sd">    and labels points outside the confidence interval.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="n">mRNAs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mRNAs</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">label</span><span class="o">=</span><span class="n">mRNA</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span>
        <span class="p">)</span>

    <span class="c1"># Linear Regression and Confidence Interval Calculation</span>
    <span class="n">X</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">LinearRegression</span><span class="p">()</span>
    <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
    <span class="n">y_pred</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">std_error</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>

    <span class="c1"># Flatten arrays for plotting</span>
    <span class="n">obs_flat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">est_flat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span>
        <span class="n">obs_flat</span><span class="p">,</span>
        <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="o">-</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">,</span>
        <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="o">+</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;95% CI&#39;</span>
    <span class="p">)</span>

    <span class="c1"># Label any points that lie outside the 95% confidence interval</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">y_val</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">obs_flat</span><span class="p">,</span> <span class="n">est_flat</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">mRNAs</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">y_val</span> <span class="o">&lt;</span> <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span> <span class="ow">or</span> <span class="n">y_val</span> <span class="o">&gt;</span> <span class="n">y_pred</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1.96</span> <span class="o">*</span> <span class="n">std_error</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span>
                    <span class="n">x_val</span><span class="p">,</span> <span class="n">y_val</span><span class="p">,</span> <span class="n">mRNAs</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span>
                    <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;light&#39;</span><span class="p">,</span> <span class="n">fontstyle</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span>
                <span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">obs_flat</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Observed Values&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Goodness of Fit&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Goodness_of_Fit.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_heatmap_abs_residuals</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a heatmap of the absolute values of the residuals.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_heatmap_abs_residuals</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a heatmap of the absolute values of the residuals.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
    <span class="n">abs_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
    <span class="c1"># Use fixed x tick labels as given in the original code</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
        <span class="n">abs_df</span><span class="p">,</span>
        <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x1&#39;</span><span class="p">,</span> <span class="s1">&#39;x2&#39;</span><span class="p">,</span> <span class="s1">&#39;x3&#39;</span><span class="p">,</span> <span class="s1">&#39;x4&#39;</span><span class="p">,</span> <span class="s1">&#39;x5&#39;</span><span class="p">,</span> <span class="s1">&#39;x6&#39;</span><span class="p">,</span> <span class="s1">&#39;x7&#39;</span><span class="p">,</span> <span class="s1">&#39;x8&#39;</span><span class="p">,</span> <span class="s1">&#39;x9&#39;</span><span class="p">],</span>
        <span class="n">yticklabels</span><span class="o">=</span><span class="n">abs_df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Absolute Residuals&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time Points&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Residual_Heatmap.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_kld</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the Kullback-Leibler Divergence (KLD) for each mRNA.
The KLD is calculated between the observed and estimated distributions
of the mRNA expression levels.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_kld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the Kullback-Leibler Divergence (KLD) for each mRNA.</span>
<span class="sd">    The KLD is calculated between the observed and estimated distributions</span>
<span class="sd">    of the mRNA expression levels.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Normalize observed and estimated values</span>
    <span class="n">normalized_obs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">normalized_est</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;x1&#39;</span><span class="p">:</span><span class="s1">&#39;x9&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Calculate KLD for each mRNA</span>
    <span class="n">kld</span> <span class="o">=</span> <span class="n">normalized_obs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">entropy</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">normalized_est</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># Create a DataFrame for KLD values</span>
    <span class="n">kld_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mRNA&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">df_obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;KL&#39;</span><span class="p">:</span> <span class="n">kld</span><span class="o">.</span><span class="n">values</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;mRNA&#39;</span><span class="p">)</span>
    <span class="c1"># Sort KLD values by mRNA</span>
    <span class="n">kld_by_gene</span> <span class="o">=</span> <span class="n">kld_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;KL&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Plot the KLD values</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
    <span class="c1"># Add horizontal bar plot for KLD values with different color for bars above 0.03</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">kld_by_gene</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.03</span><span class="p">],</span> <span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">][</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.03</span><span class="p">],</span>
             <span class="n">color</span><span class="o">=</span><span class="s1">&#39;coral&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">kld_by_gene</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.03</span><span class="p">],</span> <span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">][</span><span class="n">kld_by_gene</span><span class="p">[</span><span class="s1">&#39;KL&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.03</span><span class="p">],</span>
             <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cornflowerblue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Kullback-Leibler Divergence&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/KLD.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_pca</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots a PCA (Principal Component Analysis) of the observed and estimated values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a PCA (Principal Component Analysis) of the observed and estimated values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Standardize the data</span>
    <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
    <span class="n">scaled_data</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="p">)</span>
    <span class="c1"># Perform PCA</span>
    <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">pca_result</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">scaled_data</span><span class="p">)</span>
    <span class="c1"># Create a DataFrame for PCA results</span>
    <span class="n">pca_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pca_result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC2&#39;</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df_est</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="c1"># Perform KMeans clustering</span>
    <span class="n">kmeans</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    <span class="n">kmeans</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">pca_df</span><span class="p">)</span>
    <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kmeans</span><span class="o">.</span><span class="n">labels_</span>
    <span class="c1"># Add clusters to the PCA plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">],</span> <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC2&#39;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="c1"># Plot clusters</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pca_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;PC1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;PC2&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;Cluster&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;Set1&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pca_df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">mRNA</span><span class="p">,</span> <span class="p">(</span><span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">pca_df</span><span class="p">[</span><span class="s1">&#39;PC2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
                     <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">annotation_clip</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Principal Component Analysis&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;PC1&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;PC2&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/PCA.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_time_wise_residuals</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the residuals over time for each mRNA.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/fitanalysis/helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_time_wise_residuals</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the residuals over time for each mRNA.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="c1"># Generate a colormap for unique mRNAs with residuals &gt; 0.5</span>
    <span class="n">unique_colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab10</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">)))</span>
    <span class="n">default_color</span> <span class="o">=</span> <span class="s1">&#39;lightgray&#39;</span>
    <span class="c1"># Calculate a single mean absolute value across all time points for all mRNAs</span>
    <span class="n">mean_residuals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">mean_residuals</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]):</span>
            <span class="c1"># Assign a unique color for mRNAs with residuals &gt; 0.5</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">unique_colors</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_colors</span><span class="p">)]</span>
            <span class="n">linestyle</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Use the default color for other mRNAs</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">default_color</span>
            <span class="n">linestyle</span> <span class="o">=</span> <span class="s1">&#39;:&#39;</span>
            <span class="n">mRNA</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># Hide label for mRNAs with no residuals &gt; 0.5</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">mRNA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="n">linestyle</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time Points&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Residuals&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">savepath</span><span class="si">}</span><span class="s1">/Time_Wise_Residuals.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.config.constants"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov</p>
<p>:return: lower_bound, upper_bound, loss_type</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/config/constants.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse command line arguments for the PhosKinTime optimization problem.</span>
<span class="sd">    This function uses argparse to handle input parameters for the optimization process.</span>
<span class="sd">    The parameters include:</span>
<span class="sd">    - lower_bound: Lower bound for the optimization variables (default: -2).</span>
<span class="sd">    - upper_bound: Upper bound for the optimization variables (default: 2).</span>
<span class="sd">    - loss_type: Type of loss function to use (default: 0).</span>
<span class="sd">        Options:</span>
<span class="sd">        0: MSE</span>
<span class="sd">        1: MAE</span>
<span class="sd">        2: soft L1</span>
<span class="sd">        3: Cauchy</span>
<span class="sd">        4: Arctan</span>
<span class="sd">        5: Elastic Net</span>
<span class="sd">        6: Tikhonov</span>

<span class="sd">    :return: lower_bound, upper_bound, loss_type</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;PhosKinTime - SLSQP mRNA-TF Optimization Problem.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--lower_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Lower Beta bound.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--upper_bound&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Upper Beta bound.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--loss_type&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Loss function to use:  &quot;</span>
                             <span class="s2">&quot;0: MSE, 1: MAE, 2: soft L1, 3: Cauchy,&quot;</span>
                             <span class="s2">&quot;4: Arctan, 5: Elastic Net, 6: Tikhonov.&quot;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">args</span><span class="o">.</span><span class="n">lower_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">upper_bound</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">loss_type</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.config.logconf"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.exporter.plotout"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Plots the estimated vs observed values for a given set of genes and their corresponding TFs.
The function generates two types of plots:
1. A full time series plot showing the observed and estimated values for each gene.
2. A plot showing the first 5 time points for each gene.
Additionally, it generates an interactive plot using Plotly for each gene.</p>
<p>:param predictions: 2D numpy array of predicted values.
:param expression_matrix: 2D numpy array of observed values.
:param gene_ids: List of gene identifiers.
:param time_points: 1D numpy array of time points.
:param regulators: 2D numpy array of regulators for each gene.
:param tf_protein_matrix: 2D numpy array of TF protein signals.
:param tf_ids: List of TF identifiers.
:param num_targets: Number of target genes to plot.
:param save_path: Directory path to save the plots.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/exporter/plotout.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_estimated_vs_observed</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">time_points</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span>
                               <span class="n">tf_ids</span><span class="p">,</span> <span class="n">num_targets</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="n">OUT_DIR</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the estimated vs observed values for a given set of genes and their corresponding TFs.</span>
<span class="sd">    The function generates two types of plots:</span>
<span class="sd">    1. A full time series plot showing the observed and estimated values for each gene.</span>
<span class="sd">    2. A plot showing the first 5 time points for each gene.</span>
<span class="sd">    Additionally, it generates an interactive plot using Plotly for each gene.</span>

<span class="sd">    :param predictions: 2D numpy array of predicted values.</span>
<span class="sd">    :param expression_matrix: 2D numpy array of observed values.</span>
<span class="sd">    :param gene_ids: List of gene identifiers.</span>
<span class="sd">    :param time_points: 1D numpy array of time points.</span>
<span class="sd">    :param regulators: 2D numpy array of regulators for each gene.</span>
<span class="sd">    :param tf_protein_matrix: 2D numpy array of TF protein signals.</span>
<span class="sd">    :param tf_ids: List of TF identifiers.</span>
<span class="sd">    :param num_targets: Number of target genes to plot.</span>
<span class="sd">    :param save_path: Directory path to save the plots.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">T</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">time_points</span><span class="p">)</span>
    <span class="n">time_vals_expr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">240</span><span class="p">,</span> <span class="mi">480</span><span class="p">,</span> <span class="mi">960</span><span class="p">])</span>
    <span class="n">time_vals_tf</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">240</span><span class="p">,</span> <span class="mi">480</span><span class="p">,</span> <span class="mi">960</span><span class="p">])</span>
    <span class="n">combined_ticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">time_vals_tf</span><span class="p">)))</span>
    <span class="n">num_targets</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_targets</span><span class="p">,</span> <span class="n">predictions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_targets</span><span class="p">):</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>

        <span class="c1"># --- Full time series plot ---</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;s-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">,</span> <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="n">T</span><span class="p">]</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">,</span> <span class="n">protein_signal</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;mRNA: </span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Fold Changes&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">combined_ticks</span><span class="p">[</span><span class="mi">4</span><span class="p">:])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">combined_ticks</span><span class="p">[</span><span class="mi">4</span><span class="p">:])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

        <span class="c1"># --- First 5 time points plot ---</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;s-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">)</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">protein_signal</span><span class="p">,</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">time_vals_expr</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;TFs&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">_model_fit_.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">()</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_vals_expr</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers+lines&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Observed&#39;</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">symbol</span><span class="o">=</span><span class="s1">&#39;square&#39;</span><span class="p">)</span>
        <span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
            <span class="n">x</span><span class="o">=</span><span class="n">time_vals_expr</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Estimated&#39;</span>
        <span class="p">))</span>
        <span class="n">plotted_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]:</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">tf_name</span> <span class="o">=</span> <span class="n">tf_ids</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">plotted_tfs</span><span class="p">:</span>
                <span class="n">protein_signal</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">time_vals_tf</span><span class="p">)]</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">add_trace</span><span class="p">(</span><span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
                    <span class="n">x</span><span class="o">=</span><span class="n">time_vals_tf</span><span class="p">,</span>
                    <span class="n">y</span><span class="o">=</span><span class="n">protein_signal</span><span class="p">,</span>
                    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines&#39;</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;TF: </span><span class="si">{</span><span class="n">tf_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">dash</span><span class="o">=</span><span class="s1">&#39;dot&#39;</span><span class="p">)</span>
                <span class="p">))</span>
                <span class="n">plotted_tfs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">tf_name</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">update_layout</span><span class="p">(</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;mRNA: </span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">xaxis_title</span><span class="o">=</span><span class="s2">&quot;Time (minutes)&quot;</span><span class="p">,</span>
            <span class="n">yaxis_title</span><span class="o">=</span><span class="s2">&quot;Fold Changes&quot;</span><span class="p">,</span>
            <span class="n">xaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">tickmode</span><span class="o">=</span><span class="s1">&#39;array&#39;</span><span class="p">,</span>
                <span class="n">tickvals</span><span class="o">=</span><span class="n">combined_ticks</span><span class="p">,</span>
                <span class="n">ticktext</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">combined_ticks</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">legend</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;h&quot;</span><span class="p">,</span> <span class="n">yanchor</span><span class="o">=</span><span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mf">1.02</span><span class="p">,</span> <span class="n">xanchor</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
            <span class="n">template</span><span class="o">=</span><span class="s2">&quot;plotly_white&quot;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">900</span>
        <span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s2">_model_fit_.html&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.exporter.sheetutils"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_results_to_excel</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.
This function creates multiple sheets in the Excel file, each containing different
aspects of the optimization results, including alpha values, beta values,
residuals, observed values, estimated values, and optimization metrics.
Each sheet is formatted with appropriate column names and data types.
The function also computes various metrics such as Mean Squared Error (MSE),
Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared
score to evaluate the performance of the optimization.</p>
<p>:param gene_ids: List of gene identifiers.
:param tf_ids: List of TF identifiers.
:param final_alpha: 2D numpy array of alpha values.
:param final_beta: 2D numpy array of beta values.
:param psite_labels_arr: List of post-translational modification labels.
:param expression_matrix: 2D numpy array of observed values.
:param predictions: 2D numpy array of predicted values.
:param objective_value: Objective value of the optimization.
:param reg_map: Dictionary mapping genes to their regulators.
:param filename: Path to the output Excel file.
:type filename: str</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/exporter/sheetutils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_results_to_excel</span><span class="p">(</span>
        <span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span>
        <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span>
        <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span>
        <span class="n">objective_value</span><span class="p">,</span>
        <span class="n">reg_map</span><span class="p">,</span>
        <span class="n">filename</span><span class="o">=</span><span class="n">OUT_FILE</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the optimization results to an Excel file.</span>
<span class="sd">    This function creates multiple sheets in the Excel file, each containing different</span>
<span class="sd">    aspects of the optimization results, including alpha values, beta values,</span>
<span class="sd">    residuals, observed values, estimated values, and optimization metrics.</span>
<span class="sd">    Each sheet is formatted with appropriate column names and data types.</span>
<span class="sd">    The function also computes various metrics such as Mean Squared Error (MSE),</span>
<span class="sd">    Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared</span>
<span class="sd">    score to evaluate the performance of the optimization.</span>

<span class="sd">    :param gene_ids: List of gene identifiers.</span>
<span class="sd">    :param tf_ids: List of TF identifiers.</span>
<span class="sd">    :param final_alpha: 2D numpy array of alpha values.</span>
<span class="sd">    :param final_beta: 2D numpy array of beta values.</span>
<span class="sd">    :param psite_labels_arr: List of post-translational modification labels.</span>
<span class="sd">    :param expression_matrix: 2D numpy array of observed values.</span>
<span class="sd">    :param predictions: 2D numpy array of predicted values.</span>
<span class="sd">    :param objective_value: Objective value of the optimization.</span>
<span class="sd">    :param reg_map: Dictionary mapping genes to their regulators.</span>
<span class="sd">    :param filename: Path to the output Excel file.</span>
<span class="sd">    :type filename: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># --- Alpha Values ---</span>
    <span class="n">alpha_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span> <span class="o">=</span> <span class="n">final_alpha</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">actual_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf_name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">actual_tfs</span><span class="p">):</span>
            <span class="n">alpha_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">,</span> <span class="n">tf_name</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]])</span>
    <span class="n">df_alpha</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">alpha_rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;TF&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="c1"># --- Beta Values ---</span>
    <span class="n">beta_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">):</span>
        <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">final_beta</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">beta_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">tf</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>  <span class="c1"># Protein beta</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)):</span>
            <span class="n">beta_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">tf</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">j</span><span class="p">]])</span>
    <span class="n">df_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">beta_rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;TF&quot;</span><span class="p">,</span> <span class="s2">&quot;PSite&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="c1"># --- Residuals ---</span>
    <span class="n">residuals</span> <span class="o">=</span> <span class="n">expression_matrix</span> <span class="o">-</span> <span class="n">predictions</span>
    <span class="n">df_residuals</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">residuals</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_residuals</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Observed ---</span>
    <span class="n">df_observed</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">expression_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_observed</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Estimated ---</span>
    <span class="n">df_estimated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;x</span><span class="si">{</span><span class="n">j</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">predictions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
    <span class="n">df_estimated</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>

    <span class="c1"># --- Optimization Results ---</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">expression_matrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">y_pred</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">mape</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">((</span><span class="n">y_true</span> <span class="o">-</span> <span class="n">y_pred</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">y_true</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">)))</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">r2</span> <span class="o">=</span> <span class="n">r2_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
    <span class="n">df_metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
        <span class="p">[</span><span class="s2">&quot;Objective Value&quot;</span><span class="p">,</span> <span class="n">objective_value</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MSE&quot;</span><span class="p">,</span> <span class="n">mse</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MAE&quot;</span><span class="p">,</span> <span class="n">mae</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;MAPE&quot;</span><span class="p">,</span> <span class="n">mape</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;R^2&quot;</span><span class="p">,</span> <span class="n">r2</span><span class="p">],</span>
    <span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Metric&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>

    <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;xlsxwriter&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
        <span class="n">df_alpha</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Alpha Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_beta</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Beta Values&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_residuals</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Residuals&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_observed</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Observed&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_estimated</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Estimated&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_metrics</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Optimization Results&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.objfn.minfn"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.objfn.minfn.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Computes the predicted expression matrix based on the decision vector x.
This function uses the regulators, TF protein matrix, and post-translational modification tensor to generate
predictions for each gene at each time point.
Parameters:
  x                  : Decision vector.
  regulators         : (n_genes x n_reg) indices of TF regulators for each gene.
  tf_protein_matrix  : (n_TF x T_use) TF protein time series.
  psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).
  n_reg              : Maximum number of regulators per gene.
  T_use              : Number of time points used.
  n_genes, n_TF     : Number of genes and TF respectively.
  beta_start_indices : Integer array giving the starting index (in the segment) for each TF.
  num_psites         : Integer array with the actual number of PSites for each TF.
Returns:
    predictions        : (n_genes x T_use) matrix of predicted gene expression values.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/objfn/minfn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_predictions</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span>
                        <span class="n">num_psites</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the predicted expression matrix based on the decision vector x.</span>
<span class="sd">    This function uses the regulators, TF protein matrix, and post-translational modification tensor to generate</span>
<span class="sd">    predictions for each gene at each time point.</span>
<span class="sd">    Parameters:</span>
<span class="sd">      x                  : Decision vector.</span>
<span class="sd">      regulators         : (n_genes x n_reg) indices of TF regulators for each gene.</span>
<span class="sd">      tf_protein_matrix  : (n_TF x T_use) TF protein time series.</span>
<span class="sd">      psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</span>
<span class="sd">      n_reg              : Maximum number of regulators per gene.</span>
<span class="sd">      T_use              : Number of time points used.</span>
<span class="sd">      n_genes, n_TF     : Number of genes and TF respectively.</span>
<span class="sd">      beta_start_indices : Integer array giving the starting index (in the segment) for each TF.</span>
<span class="sd">      num_psites         : Integer array with the actual number of PSites for each TF.</span>
<span class="sd">    Returns:</span>
<span class="sd">        predictions        : (n_genes x T_use) matrix of predicted gene expression values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_genes</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">T_use</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">R_pred</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T_use</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">tf_idx</span> <span class="o">=</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_idx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span><span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>
            <span class="n">tf_effect</span> <span class="o">=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">protein</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]):</span>
                <span class="n">tf_effect</span> <span class="o">+=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">psite_tensor</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">R_pred</span> <span class="o">+=</span> <span class="n">a</span> <span class="o">*</span> <span class="n">tf_effect</span>
        <span class="n">predictions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">R_pred</span>
    <span class="k">return</span> <span class="n">predictions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">1e-06</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Computes a loss value using one of several loss functions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x</code>
            </td>
            <td>
                  <code>                 </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expression_matrix</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>        </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein_matrix</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>      </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the segment) for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>        </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lam1,</code>
            </td>
            <td>
                  <code><span title="lam2">lam2</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regularization parameters (used for loss_type 5 and 6).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/objfn/minfn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@njit</span><span class="p">(</span><span class="n">cache</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fastmath</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">parallel</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">nogil</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span>
               <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">,</span> <span class="n">lam1</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">lam2</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Originally implemented by Julius Normann.</span>

<span class="sd">    This version has been modified and optimized</span>
<span class="sd">    for consistency &amp; speed in submodules by Abhinav Mishra.</span>

<span class="sd">    Computes a loss value using one of several loss functions.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      x                  : Decision vector.</span>
<span class="sd">      expression_matrix  : (n_genes x T_use) measured gene expression values.</span>
<span class="sd">      regulators         : (n_genes x n_reg) indices of TF regulators for each gene.</span>
<span class="sd">      tf_protein_matrix  : (n_TF x T_use) TF protein time series.</span>
<span class="sd">      psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</span>
<span class="sd">      n_reg              : Maximum number of regulators per gene.</span>
<span class="sd">      T_use              : Number of time points used.</span>
<span class="sd">      n_genes, n_TF     : Number of genes and TF respectively.</span>
<span class="sd">      beta_start_indices : Integer array giving the starting index (in the segment) for each TF.</span>
<span class="sd">      num_psites         : Integer array with the actual number of PSites for each TF.</span>
<span class="sd">      loss_type          : Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</span>
<span class="sd">      lam1, lam2         : Regularization parameters (used for loss_type 5 and 6).</span>

<span class="sd">    Returns:</span>
<span class="sd">      The computed loss (a scalar).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_loss</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_genes</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="n">nT</span> <span class="o">=</span> <span class="n">n_genes</span> <span class="o">*</span> <span class="n">T_use</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">prange</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">R_meas</span> <span class="o">=</span> <span class="n">expression_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
        <span class="n">R_pred</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T_use</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">tf_idx</span> <span class="o">=</span> <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">r</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tf_idx</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>  <span class="c1"># No valid TF for this regulator</span>
                <span class="k">continue</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
            <span class="n">beta_start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>
            <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]</span>  <span class="c1"># actual length of beta vector for TF</span>
            <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span><span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>

            <span class="n">tf_effect</span> <span class="o">=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">protein</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">]):</span>
                <span class="n">tf_effect</span> <span class="o">+=</span> <span class="n">beta_vec</span><span class="p">[</span><span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">psite_tensor</span><span class="p">[</span><span class="n">tf_idx</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>

            <span class="n">R_pred</span> <span class="o">+=</span> <span class="n">a</span> <span class="o">*</span> <span class="n">tf_effect</span>

            <span class="c1"># VECTORIZED RESIDUAL - use when mRNAs are &gt; 500</span>
        <span class="c1">#     diff = R_meas - R_pred</span>
        <span class="c1">#     if loss_type == 0:  # MSE</span>
        <span class="c1">#         total_loss += np.dot(diff, diff)</span>
        <span class="c1">#     elif loss_type == 1:  # MAE</span>
        <span class="c1">#         total_loss += np.sum(np.abs(diff))</span>
        <span class="c1">#     elif loss_type == 2:  # Soft L1 (pseudo-Huber)</span>
        <span class="c1">#         total_loss += 2.0 * np.sum(np.sqrt(1.0 + diff * diff) - 1.0)</span>
        <span class="c1">#     elif loss_type == 3:  # Cauchy</span>
        <span class="c1">#         total_loss += np.sum(np.log(1.0 + diff * diff))</span>
        <span class="c1">#     elif loss_type == 4:  # Arctan</span>
        <span class="c1">#         total_loss += np.sum(np.arctan(diff * diff))</span>
        <span class="c1">#     else:</span>
        <span class="c1">#         total_loss += np.dot(diff, diff)</span>
        <span class="c1">#</span>
        <span class="c1"># loss = total_loss / nT</span>
        <span class="c1">#</span>
        <span class="c1"># # For elastic net penalty (loss_type 5) using vectorized operations.</span>
        <span class="c1"># if loss_type == 5:</span>
        <span class="c1">#     beta = x[n_alpha:]</span>
        <span class="c1">#     loss += lam1 * np.sum(np.abs(beta)) + lam2 * np.dot(beta, beta)</span>
        <span class="c1">#</span>
        <span class="c1"># # For Tikhonov regularization (loss_type 6).</span>
        <span class="c1"># if loss_type == 6:</span>
        <span class="c1">#     beta = x[n_alpha:]</span>
        <span class="c1">#     loss += lam1 * np.dot(beta, beta)</span>

            <span class="c1"># Residuals computed timepoint-by-timepoint</span>
            <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">T_use</span><span class="p">):</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">R_meas</span><span class="p">[</span><span class="n">t</span><span class="p">]</span> <span class="o">-</span> <span class="n">R_pred</span><span class="p">[</span><span class="n">t</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># MSE</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">diff</span> <span class="o">*</span> <span class="n">diff</span>
                <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># MAE</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>  <span class="c1"># Soft L1</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">diff</span> <span class="o">*</span> <span class="n">diff</span><span class="p">)</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>  <span class="c1"># Cauchy</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">diff</span> <span class="o">*</span> <span class="n">diff</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># Arctan</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">arctan</span><span class="p">(</span><span class="n">diff</span> <span class="o">*</span> <span class="n">diff</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># default to MSE</span>
                    <span class="n">total_loss</span> <span class="o">+=</span> <span class="n">diff</span> <span class="o">*</span> <span class="n">diff</span>

        <span class="n">loss</span> <span class="o">=</span> <span class="n">total_loss</span> <span class="o">/</span> <span class="n">nT</span>

        <span class="c1"># Regularization penalties</span>
        <span class="k">if</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>
            <span class="n">beta</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span><span class="p">:]</span>
            <span class="n">loss</span> <span class="o">+=</span> <span class="n">lam1</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">beta</span><span class="p">))</span> <span class="o">+</span> <span class="n">lam2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">loss_type</span> <span class="o">==</span> <span class="mi">6</span><span class="p">:</span>
            <span class="n">beta</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">n_alpha</span><span class="p">:]</span>
            <span class="n">loss</span> <span class="o">+=</span> <span class="n">lam1</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">loss</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.objfn.minfn.objective_wrapper" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">objective_wrapper</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Wrapper function for the objective function.
This function is used to call the objective function with the appropriate parameters.
Parameters:
  x                  : Decision vector.
  expression_matrix  : (n_genes x T_use) measured gene expression values.
  regulators         : (n_genes x n_reg) indices of TF regulators for each gene.
  tf_protein_matrix  : (n_TF x T_use) TF protein time series.
  psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).
  n_reg              : Maximum number of regulators per gene.
  T_use              : Number of time points used.
  n_genes, n_TF     : Number of genes and TF respectively.
  beta_start_indices : Integer array giving the starting index (in the segment) for each TF.
  num_psites         : Integer array with the actual number of PSites for each TF.
  loss_type          : Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).
Returns:
    loss               : The computed loss (a scalar).</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/objfn/minfn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">objective_wrapper</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span>
                      <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper function for the objective function.</span>
<span class="sd">    This function is used to call the objective function with the appropriate parameters.</span>
<span class="sd">    Parameters:</span>
<span class="sd">      x                  : Decision vector.</span>
<span class="sd">      expression_matrix  : (n_genes x T_use) measured gene expression values.</span>
<span class="sd">      regulators         : (n_genes x n_reg) indices of TF regulators for each gene.</span>
<span class="sd">      tf_protein_matrix  : (n_TF x T_use) TF protein time series.</span>
<span class="sd">      psite_tensor       : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</span>
<span class="sd">      n_reg              : Maximum number of regulators per gene.</span>
<span class="sd">      T_use              : Number of time points used.</span>
<span class="sd">      n_genes, n_TF     : Number of genes and TF respectively.</span>
<span class="sd">      beta_start_indices : Integer array giving the starting index (in the segment) for each TF.</span>
<span class="sd">      num_psites         : Integer array with the actual number of PSites for each TF.</span>
<span class="sd">      loss_type          : Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</span>
<span class="sd">    Returns:</span>
<span class="sd">        loss               : The computed loss (a scalar).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">objective_</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span>
                      <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.opt.optrun"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.opt.optrun.run_optimizer" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_optimizer</span><span class="p">(</span><span class="n">x0</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">lin_cons</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Runs the optimization algorithm to minimize the objective function.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>x0</code>
            </td>
            <td>
                  <code>                 </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initial guess for the optimization variables.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>bounds</code>
            </td>
            <td>
                  <code>             </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Bounds for the optimization variables.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lin_cons</code>
            </td>
            <td>
                  <code>           </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Linear constraints for the optimization problem.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expression_matrix</code>
            </td>
            <td>
                  <code>  </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>regulators</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein_matrix</code>
            </td>
            <td>
                  <code>  </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_tensor</code>
            </td>
            <td>
                  <code>       </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>T_use</code>
            </td>
            <td>
                  <code>              </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes,</code>
            </td>
            <td>
                  <code><span title="n_TF">n_TF</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
                  <code> </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the segment) for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code>         </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loss_type</code>
            </td>
            <td>
                  <code>          </code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of loss function to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    result             : Result of the optimization process, including the optimized parameters and objective value.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/opt/optrun.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_optimizer</span><span class="p">(</span><span class="n">x0</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">lin_cons</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs the optimization algorithm to minimize the objective function.</span>

<span class="sd">    Parameters:</span>
<span class="sd">      x0                  : Initial guess for the optimization variables.</span>
<span class="sd">      bounds              : Bounds for the optimization variables.</span>
<span class="sd">      lin_cons            : Linear constraints for the optimization problem.</span>
<span class="sd">      expression_matrix   : (n_genes x T_use) measured gene expression values.</span>
<span class="sd">      regulators          : (n_genes x n_reg) indices of TF regulators for each gene.</span>
<span class="sd">      tf_protein_matrix   : (n_TF x T_use) TF protein time series.</span>
<span class="sd">      psite_tensor        : (n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</span>
<span class="sd">      n_reg               : Maximum number of regulators per gene.</span>
<span class="sd">      T_use               : Number of time points used.</span>
<span class="sd">      n_genes, n_TF     : Number of genes and TF respectively.</span>
<span class="sd">      beta_start_indices  : Integer array giving the starting index (in the segment) for each TF.</span>
<span class="sd">      num_psites          : Integer array with the actual number of PSites for each TF.</span>
<span class="sd">      loss_type           : Type of loss function to use.</span>
<span class="sd">    Returns:</span>
<span class="sd">        result             : Result of the optimization process, including the optimized parameters and objective value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">m</span> <span class="o">=</span> <span class="s2">&quot;SLSQP&quot;</span> <span class="c1"># or trust-constr or SLSQP</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">minimize</span><span class="p">(</span>
        <span class="n">fun</span><span class="o">=</span><span class="n">objective_wrapper</span><span class="p">,</span>
        <span class="n">x0</span><span class="o">=</span><span class="n">x0</span><span class="p">,</span>
        <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">loss_type</span><span class="p">),</span>
        <span class="n">method</span><span class="o">=</span><span class="n">m</span><span class="p">,</span>
        <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span>
        <span class="n">constraints</span><span class="o">=</span><span class="n">lin_cons</span><span class="p">,</span>
        <span class="n">options</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;disp&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;maxiter&quot;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">}</span> <span class="k">if</span> <span class="n">m</span> <span class="o">==</span> <span class="s2">&quot;SLSQP&quot;</span> <span class="k">else</span> <span class="p">{</span><span class="s2">&quot;disp&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;maxiter&quot;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">,</span> <span class="s2">&quot;verbose&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.optcon.construct"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_fixed_arrays</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.
Returns:
  - expression_matrix: array of shape (n_genes, T)
  - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.
  - tf_protein_matrix: array of shape (n_TF, T)
  - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
  - n_reg: maximum number of regulators per gene.
  - n_psite_max: maximum number of PSites among TFs.
  - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
  - num_psites: array of length n_TF with the actual number of PSites for each TF.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_fixed_arrays</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Builds fixed-shape arrays from the input data.</span>
<span class="sd">    Returns:</span>
<span class="sd">      - expression_matrix: array of shape (n_genes, T)</span>
<span class="sd">      - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.</span>
<span class="sd">      - tf_protein_matrix: array of shape (n_TF, T)</span>
<span class="sd">      - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.</span>
<span class="sd">      - n_reg: maximum number of regulators per gene.</span>
<span class="sd">      - n_psite_max: maximum number of PSites among TFs.</span>
<span class="sd">      - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).</span>
<span class="sd">      - num_psites: array of length n_TF with the actual number of PSites for each TF.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_genes</span><span class="p">,</span> <span class="n">T</span> <span class="o">=</span> <span class="n">expression_matrix</span><span class="o">.</span><span class="n">shape</span>

    <span class="c1"># Map TF id to index.</span>
    <span class="n">tf_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">)}</span>
    <span class="n">n_TF</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">)</span>

    <span class="c1"># Determine max number of valid regulators per gene, filtering out TFs not present in tf_ids.</span>
    <span class="n">reg_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">:</span>
        <span class="n">regs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[])</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">]</span>
        <span class="n">reg_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">regs</span><span class="p">)</span>
    <span class="n">n_reg</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">regs</span><span class="p">)</span> <span class="k">for</span> <span class="n">regs</span> <span class="ow">in</span> <span class="n">reg_list</span><span class="p">)</span> <span class="k">if</span> <span class="n">reg_list</span> <span class="k">else</span> <span class="mi">1</span>

    <span class="c1"># Build regulators array (n_genes x n_reg), padded with -1 to mark invalid.</span>
    <span class="n">regulators</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">((</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">regs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reg_list</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regs</span><span class="p">):</span>
            <span class="n">regulators</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">tf_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">tf_protein_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">T</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">tf_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">tf_protein</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">tf_protein</span><span class="p">[</span><span class="n">tf</span><span class="p">][:</span><span class="n">T</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tf_protein_matrix</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># For each TF, record the actual number of PSites.</span>
    <span class="n">num_psites</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">):</span>
        <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tf_psite_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[]))</span>
    <span class="c1"># Maximum number of PSites across all TFs.</span>
    <span class="n">n_psite_max</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">num_psites</span><span class="p">))</span> <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">num_psites</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="c1"># Build psite_tensor and psite_labels_arr.</span>
    <span class="n">psite_tensor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">T</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="n">psite_labels_arr</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">tf_index</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">psites</span> <span class="o">=</span> <span class="n">tf_psite_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[])</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">tf_psite_labels</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_psite_max</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">psites</span><span class="p">):</span>
                <span class="n">psite_tensor</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">psites</span><span class="p">[</span><span class="n">j</span><span class="p">][:</span><span class="n">T</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">psite_tensor</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>
        <span class="n">padded_labels</span> <span class="o">=</span> <span class="n">labels</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">n_psite_max</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
        <span class="n">psite_labels_arr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">padded_labels</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">regulators</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">psite_tensor</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_psite_max</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">num_psites</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.construct.build_linear_constraints" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_linear_constraints</span><span class="p">(</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Build linear constraints for the optimization problem.</p>
<p>The constraints are:
1. For each mRNA, the sum of its alpha parameters must equal 1.
2. For each TF, the sum of its beta parameters must equal 1.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_linear_constraints</span><span class="p">(</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Build linear constraints for the optimization problem.</span>

<span class="sd">    The constraints are:</span>
<span class="sd">    1. For each mRNA, the sum of its alpha parameters must equal 1.</span>
<span class="sd">    2. For each TF, the sum of its beta parameters must equal 1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_vars</span> <span class="o">=</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">))</span>

    <span class="c1"># --- Alpha constraints ---</span>
    <span class="n">alpha_constraints_matrix</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">total_vars</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">row</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="n">alpha_constraints_matrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">alpha_constraints_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">alpha_constraints_matrix</span><span class="p">)</span>
    <span class="n">alpha_constraint</span> <span class="o">=</span> <span class="n">LinearConstraint</span><span class="p">(</span><span class="n">alpha_constraints_matrix</span><span class="p">,</span> <span class="n">lb</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">ub</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>

    <span class="c1"># --- Beta constraints ---</span>
    <span class="n">beta_constraint_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">lb_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">ub_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">total_vars</span><span class="p">)</span>
        <span class="n">row</span><span class="p">[</span><span class="n">start</span><span class="p">:</span> <span class="n">start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="n">beta_constraint_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
        <span class="n">lb_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span>
        <span class="n">ub_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">no_psite_tf</span><span class="p">[</span><span class="n">tf</span><span class="p">]:</span>
            <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">total_vars</span><span class="p">)</span>
                <span class="n">row</span><span class="p">[</span><span class="n">start</span> <span class="o">+</span> <span class="n">q</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
                <span class="n">beta_constraint_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">lb_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
                <span class="n">ub_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>

    <span class="n">beta_constraints_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">beta_constraint_rows</span><span class="p">)</span>
    <span class="n">beta_constraint</span> <span class="o">=</span> <span class="n">LinearConstraint</span><span class="p">(</span><span class="n">beta_constraints_matrix</span><span class="p">,</span> <span class="n">lb</span><span class="o">=</span><span class="n">lb_list</span><span class="p">,</span> <span class="n">ub</span><span class="o">=</span><span class="n">ub_list</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">alpha_constraint</span><span class="p">,</span> <span class="n">beta_constraint</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.construct.constraint_alpha_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">constraint_alpha_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>For each gene, the sum of its alpha parameters must equal 1.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">constraint_alpha_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For each gene, the sum of its alpha parameters must equal 1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_reg</span><span class="p">):</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span> <span class="o">+</span> <span class="n">r</span><span class="p">]</span>
        <span class="n">cons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.construct.constraint_beta_func" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">constraint_beta_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>For each TF, the sum of its beta parameters must equal 1.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/construct.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">constraint_beta_func</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For each TF, the sum of its beta parameters must equal 1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>  <span class="c1"># Total beta parameters for TF tf.</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
        <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">start</span><span class="p">:</span> <span class="n">start</span> <span class="o">+</span> <span class="n">length</span><span class="p">]</span>
        <span class="n">cons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">no_psite_tf</span><span class="p">[</span><span class="n">tf</span><span class="p">]:</span>
            <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
                <span class="n">cons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">[</span><span class="n">q</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cons</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.optcon.filter"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.filter.load_and_filter_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_and_filter_data</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load and filter data for the optimization problem.
This function loads gene expression data, transcription factor (TF) data,
and regulation data. It filters genes to only include those with at least
one regulator and filters regulators to only include those present in the
TF data. The function returns the filtered gene IDs, expression matrix,
TF IDs, TF protein data, TF phosphorylation site data, TF phosphorylation
site labels, and the regulation map.
It raises a ValueError if no genes with regulators are found.
The function also filters the TF data to only include those present in the
regulation map.</p>
<p>The function returns:
- gene_ids: List of filtered gene IDs.
- expr_matrix: Filtered expression matrix.
- expr_time_cols: Time columns for gene expression data.
- tf_ids: Filtered TF IDs.
- tf_protein: Dictionary mapping TF IDs to their protein data.
- tf_psite_data: Dictionary mapping TF IDs to their phosphorylation site data.
- tf_psite_labels: Dictionary mapping TF IDs to their phosphorylation site labels.
- tf_time_cols: Time columns for TF data.
- reg_map: Regulation map, mapping gene IDs to their regulators.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_and_filter_data</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load and filter data for the optimization problem.</span>
<span class="sd">    This function loads gene expression data, transcription factor (TF) data,</span>
<span class="sd">    and regulation data. It filters genes to only include those with at least</span>
<span class="sd">    one regulator and filters regulators to only include those present in the</span>
<span class="sd">    TF data. The function returns the filtered gene IDs, expression matrix,</span>
<span class="sd">    TF IDs, TF protein data, TF phosphorylation site data, TF phosphorylation</span>
<span class="sd">    site labels, and the regulation map.</span>
<span class="sd">    It raises a ValueError if no genes with regulators are found.</span>
<span class="sd">    The function also filters the TF data to only include those present in the</span>
<span class="sd">    regulation map.</span>

<span class="sd">    The function returns:</span>
<span class="sd">    - gene_ids: List of filtered gene IDs.</span>
<span class="sd">    - expr_matrix: Filtered expression matrix.</span>
<span class="sd">    - expr_time_cols: Time columns for gene expression data.</span>
<span class="sd">    - tf_ids: Filtered TF IDs.</span>
<span class="sd">    - tf_protein: Dictionary mapping TF IDs to their protein data.</span>
<span class="sd">    - tf_psite_data: Dictionary mapping TF IDs to their phosphorylation site data.</span>
<span class="sd">    - tf_psite_labels: Dictionary mapping TF IDs to their phosphorylation site labels.</span>
<span class="sd">    - tf_time_cols: Time columns for TF data.</span>
<span class="sd">    - reg_map: Regulation map, mapping gene IDs to their regulators.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">expr_time_cols</span> <span class="o">=</span> <span class="n">load_expression_data</span><span class="p">()</span>
    <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">tf_time_cols</span> <span class="o">=</span> <span class="n">load_tf_protein_data</span><span class="p">()</span>
    <span class="n">reg_map</span> <span class="o">=</span> <span class="n">load_regulation</span><span class="p">()</span>

    <span class="c1"># Filter genes: only keep those with at least one regulator.</span>
    <span class="n">filtered_indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">reg_map</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_indices</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No genes with regulators found. Exiting.&quot;</span><span class="p">)</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">filtered_indices</span><span class="p">]</span>
    <span class="n">expr_matrix</span> <span class="o">=</span> <span class="n">expr_matrix</span><span class="p">[</span><span class="n">filtered_indices</span><span class="p">,</span> <span class="p">:]</span>

    <span class="c1"># For each gene, filter regulators to those present in tf_ids.</span>
    <span class="n">relevant_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">:</span>
        <span class="n">regs</span> <span class="o">=</span> <span class="n">reg_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[])</span>
        <span class="n">regs_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">regs</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">]</span>
        <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">regs_filtered</span>
        <span class="n">relevant_tfs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">regs_filtered</span><span class="p">)</span>

    <span class="c1"># Filter TFs.</span>
    <span class="n">tf_ids_filtered</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">relevant_tfs</span><span class="p">]</span>
    <span class="n">tf_ids</span> <span class="o">=</span> <span class="n">tf_ids_filtered</span>
    <span class="n">tf_protein</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">tf_protein</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">}</span>
    <span class="n">tf_psite_data</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">tf_psite_data</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">}</span>
    <span class="n">tf_psite_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">tf</span><span class="p">:</span> <span class="n">tf_psite_labels</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">expr_time_cols</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">tf_time_cols</span><span class="p">,</span> <span class="n">reg_map</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.optcon.filter.prepare_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_data</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">tf_time_cols</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prepares the data for optimization by filtering the expression matrix
to match the number of time points and building fixed arrays.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expr_matrix</code>
            </td>
            <td>
                  <code><span title="np.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene expression matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_protein</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their protein data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_psite_data</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_psite_labels</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site labels.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_time_cols</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Time columns for TF data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reg_map</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    fixed_arrays (tuple): Tuple containing the fixed arrays:
        - expression_matrix: array of shape (n_genes, T)
        - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.
        - tf_protein_matrix: array of shape (n_TF, T)
        - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
        - n_reg: maximum number of regulators per gene.
        - n_psite_max: maximum number of PSites among TFs.
        - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
        - num_psites: array of length n_TF with the actual number of PSites for each TF.
    T_use (int): Number of time points used in the expression matrix.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/optcon/filter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prepare_data</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">tf_time_cols</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepares the data for optimization by filtering the expression matrix</span>
<span class="sd">    to match the number of time points and building fixed arrays.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_ids (list): List of gene IDs.</span>
<span class="sd">        expr_matrix (np.ndarray): Gene expression matrix.</span>
<span class="sd">        tf_ids (list): List of transcription factor IDs.</span>
<span class="sd">        tf_protein (dict): Dictionary mapping TF IDs to their protein data.</span>
<span class="sd">        tf_psite_data (dict): Dictionary mapping TF IDs to their phosphorylation site data.</span>
<span class="sd">        tf_psite_labels (dict): Dictionary mapping TF IDs to their phosphorylation site labels.</span>
<span class="sd">        tf_time_cols (list): Time columns for TF data.</span>
<span class="sd">        reg_map (dict): Regulation map, mapping gene IDs to their regulators.</span>
<span class="sd">    Returns:</span>
<span class="sd">        fixed_arrays (tuple): Tuple containing the fixed arrays:</span>
<span class="sd">            - expression_matrix: array of shape (n_genes, T)</span>
<span class="sd">            - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.</span>
<span class="sd">            - tf_protein_matrix: array of shape (n_TF, T)</span>
<span class="sd">            - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.</span>
<span class="sd">            - n_reg: maximum number of regulators per gene.</span>
<span class="sd">            - n_psite_max: maximum number of PSites among TFs.</span>
<span class="sd">            - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).</span>
<span class="sd">            - num_psites: array of length n_TF with the actual number of PSites for each TF.</span>
<span class="sd">        T_use (int): Number of time points used in the expression matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">T_use</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">expr_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">tf_time_cols</span><span class="p">))</span>
    <span class="n">expr_matrix</span> <span class="o">=</span> <span class="n">expr_matrix</span><span class="p">[:,</span> <span class="p">:</span><span class="n">T_use</span><span class="p">]</span>
    <span class="n">fixed_arrays</span> <span class="o">=</span> <span class="n">build_fixed_arrays</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">expr_matrix</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">fixed_arrays</span><span class="p">,</span> <span class="n">T_use</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.utils.iodata"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;report.html&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a single global report HTML file from all gene folders inside the results directory.</span>

<span class="sd">    For each gene folder (e.g. &quot;ABL2&quot;), the report will include:</span>
<span class="sd">      - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.</span>
<span class="sd">      - Each plot is confined to a fixed size of 900px by 900px.</span>
<span class="sd">      - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</span>

<span class="sd">    Args:</span>
<span class="sd">        results_dir (str): Path to the root results directory.</span>
<span class="sd">        output_file (str): Name of the generated global report file (placed inside results_dir).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Gather gene folders (skip &quot;General&quot; and &quot;logs&quot;)</span>
    <span class="n">gene_folders</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">results_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="ow">and</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;General&quot;</span><span class="p">,</span> <span class="s2">&quot;logs&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Build HTML content with updated CSS for spacing.</span>
    <span class="n">html_parts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;meta charset=&#39;UTF-8&#39;&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;title&gt;Estimation Report&lt;/title&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;body { font-family: Arial, sans-serif; margin: 20px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h1 { color: #333; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h2 { color: #555; font-size: 1.8em; border-bottom: 1px solid #ccc; padding-bottom: 5px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h3 { color: #666; font-size: 1.4em; margin-top: 10px; margin-bottom: 10px; }&quot;</span><span class="p">,</span>
        <span class="c1"># /* CSS grid for plots: two per row, fixed size 500px x 500px, extra space between rows */</span>
        <span class="s2">&quot;.plot-container {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  display: grid;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  grid-template-columns: repeat(2, 500px);&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  column-gap: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  row-gap: 40px;&quot;</span><span class="p">,</span> <span class="c1"># /* extra vertical gap */</span>
        <span class="s2">&quot;  justify-content: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;.plot-item {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;img, iframe {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  object-fit: contain;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: none;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="c1"># /* Data tables: full width, one per row */</span>
        <span class="s2">&quot;.data-table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 50%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border-collapse: collapse;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-top: 10px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;th, td {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: 1px solid #ccc;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  padding: 8px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  text-align: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;body&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;h1&gt;Global Report&lt;/h1&gt;&quot;</span>
    <span class="p">]</span>

    <span class="c1"># For each gene folder, create a section in the report.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_folders</span><span class="p">):</span>
        <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;mRNA: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>

        <span class="c1"># Create grid container for fixed-size plots.</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;div class=&quot;plot-container&quot;&gt;&#39;</span><span class="p">)</span>
        <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                    <span class="n">rel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;plot-item&quot;&gt;&lt;h3&gt;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;img src=&quot;</span><span class="si">{</span><span class="n">rel_path</span><span class="si">}</span><span class="s1">&quot; alt=&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&quot;&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;/div&gt;&#39;</span><span class="p">)</span>  <span class="c1"># End of plot container</span>

        <span class="c1"># Data tables: display XLSX or CSV files from the gene folder, one per row.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.xlsx&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                    <span class="n">table_html</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">border</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;</span><span class="si">{</span><span class="n">table_html</span><span class="si">}</span><span class="s1">&lt;/div&gt;&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;p&gt;Error reading </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&lt;/p&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>

    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span>

    <span class="c1"># Write the report into the results directory.</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">html_parts</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.load_expression_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_expression_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Loads gene expression (mRNA) data.
Expects a CSV with a 'GeneID' column and time-point columns.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_expression_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads gene expression (mRNA) data.</span>
<span class="sd">    Expects a CSV with a &#39;GeneID&#39; column and time-point columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="c1"># Normlaize for high unscaled variability</span>
    <span class="c1"># Exists often</span>
    <span class="c1"># df = min_max_normalize(df, 4)</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">]</span>
    <span class="n">expression_matrix</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">expression_matrix</span><span class="p">,</span> <span class="n">time_cols</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">



<details class="assumes-the-regulation-file-is-reversed" open>
  <summary>Assumes the regulation file is reversed</summary>
  <ul>
<li>The 'Source' column holds gene (mRNA) identifiers.</li>
<li>The 'Target' column holds TF identifiers.</li>
</ul>
</details>        <p>Returns a mapping from gene (source) to a list of TFs (targets).</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_regulation</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assumes the regulation file is reversed:</span>
<span class="sd">      - The &#39;Source&#39; column holds gene (mRNA) identifiers.</span>
<span class="sd">      - The &#39;Target&#39; column holds TF identifiers.</span>
<span class="sd">    Returns a mapping from gene (source) to a list of TFs (targets).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">reg_map</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Source&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">tf</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Target&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">:</span>
            <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">tf</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]:</span>
            <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tf</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">reg_map</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.load_tf_protein_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_tf_protein_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Loads TF protein data along with PSite information.
Expects a CSV with 'GeneID' and 'Psite' columns.
For rows without a valid PSite, the entire row is considered as the protein signal.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_tf_protein_data</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads TF protein data along with PSite information.</span>
<span class="sd">    Expects a CSV with &#39;GeneID&#39; and &#39;Psite&#39; columns.</span>
<span class="sd">    For rows without a valid PSite, the entire row is considered as the protein signal.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="c1"># Normlaize for high unscaled variability</span>
    <span class="c1"># Exists often</span>
    <span class="c1"># df = min_max_normalize(df, 4)</span>
    <span class="n">tf_protein</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tf_psite_data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tf_psite_labels</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Original time columns from TF data (should be 14 columns)</span>
    <span class="n">orig_time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">]]</span>
    <span class="c1"># Use only time points from index 5 onward (i.e. last 9 time points) if available.</span>
    <span class="c1"># To match the expression data, we need to ensure the same number of time points.</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">orig_time_cols</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">14</span><span class="p">:</span>
        <span class="n">time_cols</span> <span class="o">=</span> <span class="n">orig_time_cols</span><span class="p">[</span><span class="mi">5</span><span class="p">:]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">time_cols</span> <span class="o">=</span> <span class="n">orig_time_cols</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">tf</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">psite</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Psite&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="n">orig_time_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">vals</span><span class="p">[</span><span class="mi">5</span><span class="p">:]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">orig_time_cols</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">14</span> <span class="k">else</span> <span class="n">vals</span>
        <span class="k">if</span> <span class="n">tf</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">tf_protein</span><span class="p">:</span>
            <span class="n">tf_protein</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span>
            <span class="n">tf_psite_data</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">tf_psite_labels</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tf_psite_data</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
            <span class="n">tf_psite_labels</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psite</span><span class="p">)</span>
    <span class="n">tf_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">tf_protein</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">tf_protein</span><span class="p">,</span> <span class="n">tf_psite_data</span><span class="p">,</span> <span class="n">tf_psite_labels</span><span class="p">,</span> <span class="n">time_cols</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.min_max_normalize" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">min_max_normalize</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">custom_max</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Row-wise (per-sample) min-max normalize time-series columns starting with 'x'.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input DataFrame with time-series columns (x1-xN).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>custom_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If given, used as max for all rows.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: Normalized DataFrame with same shape.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">min_max_normalize</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">custom_max</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Row-wise (per-sample) min-max normalize time-series columns starting with &#39;x&#39;.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        df (pd.DataFrame): Input DataFrame with time-series columns (x1-xN).</span>
<span class="sd">        custom_max (float, optional): If given, used as max for all rows.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: Normalized DataFrame with same shape.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)]</span>
    <span class="n">other_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)]</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

    <span class="n">row_min</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">custom_max</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">row_max</span> <span class="o">=</span> <span class="n">custom_max</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="p">(</span><span class="n">custom_max</span> <span class="o">-</span> <span class="n">row_min</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row_max</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="p">(</span><span class="n">row_max</span> <span class="o">-</span> <span class="n">row_min</span><span class="p">)</span>

    <span class="n">normalized</span> <span class="o">=</span> <span class="p">(</span><span class="n">data</span> <span class="o">-</span> <span class="n">row_min</span><span class="p">)</span> <span class="o">/</span> <span class="n">denom</span>
    <span class="n">df</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span> <span class="o">=</span> <span class="n">normalized</span>

    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.iodata.summarize_stats" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">summarize_stats</span><span class="p">(</span><span class="n">input3</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">,</span> <span class="n">input1</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">,</span> <span class="n">input4</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Summarizes statistics for the expression data (input3) and TF protein data (input1).
It also summarizes the data after filtering based on the mapping file (input4).</p>


<details class="the-function-prints-the-following-statistics" open>
  <summary>The function prints the following statistics</summary>
  <ul>
<li>Global min, max, std, var for the full dataset.</li>
<li>Time-wise min, max, std, var for each time point.</li>
<li>Global min, max, std, var for the subset data (filtered by input4).</li>
<li>Time-wise min, max, std, var for the subset data.</li>
</ul>
</details>        <p>Args:
    input3 (str): Path to the expression data CSV file.
    input1 (str): Path to the TF protein data CSV file.
    input4 (str): Path to the mapping file CSV.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/iodata.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">summarize_stats</span><span class="p">(</span><span class="n">input3</span><span class="o">=</span><span class="n">INPUT3</span><span class="p">,</span> <span class="n">input1</span><span class="o">=</span><span class="n">INPUT1</span><span class="p">,</span> <span class="n">input4</span><span class="o">=</span><span class="n">INPUT4</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Summarizes statistics for the expression data (input3) and TF protein data (input1).</span>
<span class="sd">    It also summarizes the data after filtering based on the mapping file (input4).</span>

<span class="sd">    The function prints the following statistics:</span>
<span class="sd">        - Global min, max, std, var for the full dataset.</span>
<span class="sd">        - Time-wise min, max, std, var for each time point.</span>
<span class="sd">        - Global min, max, std, var for the subset data (filtered by input4).</span>
<span class="sd">        - Time-wise min, max, std, var for the subset data.</span>
<span class="sd">    Args:</span>
<span class="sd">        input3 (str): Path to the expression data CSV file.</span>
<span class="sd">        input1 (str): Path to the TF protein data CSV file.</span>
<span class="sd">        input4 (str): Path to the mapping file CSV.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load input3: expression data</span>
    <span class="n">expr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input3</span><span class="p">)</span>
    <span class="n">expr_data</span> <span class="o">=</span> <span class="n">expr_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">])</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=== Expression Data (input3)  Full Dataset ===&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global min: </span><span class="si">{</span><span class="n">expr_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global max: </span><span class="si">{</span><span class="n">expr_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global std: </span><span class="si">{</span><span class="n">expr_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global var: </span><span class="si">{</span><span class="n">expr_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">var</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Time-wise stats:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">expr_data</span><span class="o">.</span><span class="n">agg</span><span class="p">([</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;var&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># Load input1: TF protein data</span>
    <span class="n">prot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input1</span><span class="p">)</span>
    <span class="n">time_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">prot_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">,</span> <span class="s2">&quot;Psite&quot;</span><span class="p">]]</span>
    <span class="n">prot_data</span> <span class="o">=</span> <span class="n">prot_df</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== TF Protein Data (input1)  Full Dataset ===&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global min: </span><span class="si">{</span><span class="n">prot_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global max: </span><span class="si">{</span><span class="n">prot_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global std: </span><span class="si">{</span><span class="n">prot_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global var: </span><span class="si">{</span><span class="n">prot_data</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">var</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Time-wise stats:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">prot_data</span><span class="o">.</span><span class="n">agg</span><span class="p">([</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;var&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># Load mapping from input4</span>
    <span class="n">map_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input4</span><span class="p">)</span>
    <span class="n">expr_subset</span> <span class="o">=</span> <span class="n">expr_df</span><span class="p">[</span><span class="n">expr_df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">map_df</span><span class="p">[</span><span class="s2">&quot;Source&quot;</span><span class="p">])]</span>
    <span class="n">prot_subset</span> <span class="o">=</span> <span class="n">prot_df</span><span class="p">[</span><span class="n">prot_df</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">map_df</span><span class="p">[</span><span class="s2">&quot;Target&quot;</span><span class="p">])]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Expression Data  Subset from input4 ===&quot;</span><span class="p">)</span>
    <span class="n">expr_data_sub</span> <span class="o">=</span> <span class="n">expr_subset</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneID&quot;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global min: </span><span class="si">{</span><span class="n">expr_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global max: </span><span class="si">{</span><span class="n">expr_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global std: </span><span class="si">{</span><span class="n">expr_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global var: </span><span class="si">{</span><span class="n">expr_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">var</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Time-wise stats:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">expr_data_sub</span><span class="o">.</span><span class="n">agg</span><span class="p">([</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;var&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== TF Protein Data  Subset from input4 ===&quot;</span><span class="p">)</span>
    <span class="n">prot_data_sub</span> <span class="o">=</span> <span class="n">prot_subset</span><span class="p">[</span><span class="n">time_cols</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global min: </span><span class="si">{</span><span class="n">prot_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global max: </span><span class="si">{</span><span class="n">prot_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global std: </span><span class="si">{</span><span class="n">prot_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global var: </span><span class="si">{</span><span class="n">prot_data_sub</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">var</span><span class="p">()</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Time-wise stats:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">prot_data_sub</span><span class="o">.</span><span class="n">agg</span><span class="p">([</span><span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;var&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="tfopt.local.utils.params"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.params.get_optimization_parameters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_optimization_parameters</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prepare the optimization parameters for the optimization problem.
This function initializes the alpha and beta parameters, sets up the bounds,
and constructs the linear constraints for the optimization problem.
It returns the initial guess for the optimization variables, the number of
alpha parameters, the starting indices for beta parameters, the bounds,
the boolean array indicating TFs without phosphorylation sites, and the
number of genes and TFs.</p>
<p>Parameters:
expression_matrix (np.ndarray): Gene expression matrix.
tf_protein_matrix (np.ndarray): TF protein matrix.
n_reg (int): Number of regulators.
T_use (int): Number of time points used in the expression matrix.
psite_labels_arr (list): List of lists containing phosphorylation site labels.
num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.
lb (float): Lower bound for beta parameters.
ub (float): Upper bound for beta parameters.</p>
<p>Returns:
x0 (np.ndarray): Initial guess for the optimization variables.
n_alpha (int): Number of alpha parameters.
beta_start_indices (np.ndarray): Starting indices for beta parameters.
bounds (list): List of tuples specifying the bounds for each optimization variable.
no_psite_tf (np.ndarray): Boolean array indicating TFs without phosphorylation sites.
n_genes (int): Number of genes.
n_TF (int): Number of TFs.
num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.
lin_cons (LinearConstraint): Linear constraints for the optimization problem.
T_use (int): Number of time points used in the expression matrix.</p>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_optimization_parameters</span><span class="p">(</span><span class="n">expression_matrix</span><span class="p">,</span> <span class="n">tf_protein_matrix</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">T_use</span><span class="p">,</span>
                                <span class="n">psite_labels_arr</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare the optimization parameters for the optimization problem.</span>
<span class="sd">    This function initializes the alpha and beta parameters, sets up the bounds,</span>
<span class="sd">    and constructs the linear constraints for the optimization problem.</span>
<span class="sd">    It returns the initial guess for the optimization variables, the number of</span>
<span class="sd">    alpha parameters, the starting indices for beta parameters, the bounds,</span>
<span class="sd">    the boolean array indicating TFs without phosphorylation sites, and the</span>
<span class="sd">    number of genes and TFs.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    expression_matrix (np.ndarray): Gene expression matrix.</span>
<span class="sd">    tf_protein_matrix (np.ndarray): TF protein matrix.</span>
<span class="sd">    n_reg (int): Number of regulators.</span>
<span class="sd">    T_use (int): Number of time points used in the expression matrix.</span>
<span class="sd">    psite_labels_arr (list): List of lists containing phosphorylation site labels.</span>
<span class="sd">    num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.</span>
<span class="sd">    lb (float): Lower bound for beta parameters.</span>
<span class="sd">    ub (float): Upper bound for beta parameters.</span>

<span class="sd">    Returns:</span>
<span class="sd">    x0 (np.ndarray): Initial guess for the optimization variables.</span>
<span class="sd">    n_alpha (int): Number of alpha parameters.</span>
<span class="sd">    beta_start_indices (np.ndarray): Starting indices for beta parameters.</span>
<span class="sd">    bounds (list): List of tuples specifying the bounds for each optimization variable.</span>
<span class="sd">    no_psite_tf (np.ndarray): Boolean array indicating TFs without phosphorylation sites.</span>
<span class="sd">    n_genes (int): Number of genes.</span>
<span class="sd">    n_TF (int): Number of TFs.</span>
<span class="sd">    num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.</span>
<span class="sd">    lin_cons (LinearConstraint): Linear constraints for the optimization problem.</span>
<span class="sd">    T_use (int): Number of time points used in the expression matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_genes</span> <span class="o">=</span> <span class="n">expression_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">n_TF</span> <span class="o">=</span> <span class="n">tf_protein_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">no_psite_tf</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">all</span><span class="p">(</span><span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">)])</span>
    <span class="n">beta_start_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_TF</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="n">cum</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">cum</span>
        <span class="n">cum</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">n_alpha</span> <span class="o">=</span> <span class="n">n_genes</span> <span class="o">*</span> <span class="n">n_reg</span>
    <span class="c1"># x0_alpha = np.full(n_alpha, 1.0 / n_reg)</span>
    <span class="c1"># Initialize x0_alpha using uniform random numbers and normalize per gene.</span>
    <span class="n">x0_alpha</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">n_alpha</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">):</span>
        <span class="c1"># Sample n_reg values uniformly from [0,1)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">n_reg</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">/=</span> <span class="n">a</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># Normalize so that the regulators for gene i sum to 1.</span>
        <span class="n">x0_alpha</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">n_reg</span><span class="p">:(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">n_reg</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
    <span class="c1"># Initialize x0_beta by sampling uniformly from [lb, ub] and normalizing per TF.</span>
    <span class="n">x0_beta_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_TF</span><span class="p">):</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">no_psite_tf</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
            <span class="c1"># For TFs without any PSite, the beta vector has one element.</span>
            <span class="c1"># The constraint forces that element to be 1.0, so we simply use 1.0.</span>
            <span class="n">x0_beta_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="mf">1.0</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">length</span><span class="p">)</span>
            <span class="n">sample</span> <span class="o">/=</span> <span class="n">sample</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># Normalize so that this beta vector sums to 1.</span>
            <span class="n">x0_beta_list</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
            <span class="c1"># x0_beta_list.extend([1.0 / length] * length)</span>

    <span class="n">x0_beta</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x0_beta_list</span><span class="p">)</span>
    <span class="n">x0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">x0_alpha</span><span class="p">,</span> <span class="n">x0_beta</span><span class="p">])</span>

    <span class="n">bounds_alpha</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)]</span> <span class="o">*</span> <span class="n">n_alpha</span>
    <span class="n">bounds_beta</span> <span class="o">=</span> <span class="p">[(</span><span class="n">lb</span><span class="p">,</span> <span class="n">ub</span><span class="p">)]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">x0_beta</span><span class="p">)</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="n">bounds_alpha</span> <span class="o">+</span> <span class="n">bounds_beta</span>

    <span class="n">lin_cons</span> <span class="o">=</span> <span class="n">build_linear_constraints</span><span class="p">(</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">x0</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">bounds</span><span class="p">,</span> <span class="n">no_psite_tf</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_TF</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">lin_cons</span><span class="p">,</span> <span class="n">T_use</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="tfopt.local.utils.params.postprocess_results" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">postprocess_results</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span> <span class="n">psite_labels_arr</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Post-process the optimization results to extract the final alpha and beta parameters.
This function reshapes the optimization result into the final alpha and beta matrices,
and builds the mapping of TFs to mRNAs and phosphorylation sites.
It also logs the mappings for debugging purposes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
                  <code><span title="OptimizeResult">OptimizeResult</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The result of the optimization.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_alpha</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_reg</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beta_start_indices</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Starting indices for beta parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_psites</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array containing the number of phosphorylation sites for each TF.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reg_map</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tf_ids</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>psite_labels_arr</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of lists containing phosphorylation site labels.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>final_x</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final optimization result.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>final_alpha</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final alpha parameters reshaped into a matrix.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>final_beta</code></td>            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Final beta parameters reshaped into a matrix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>tfopt/local/utils/params.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">postprocess_results</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">n_alpha</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">,</span> <span class="n">beta_start_indices</span><span class="p">,</span> <span class="n">num_psites</span><span class="p">,</span> <span class="n">reg_map</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">tf_ids</span><span class="p">,</span>
                        <span class="n">psite_labels_arr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Post-process the optimization results to extract the final alpha and beta parameters.</span>
<span class="sd">    This function reshapes the optimization result into the final alpha and beta matrices,</span>
<span class="sd">    and builds the mapping of TFs to mRNAs and phosphorylation sites.</span>
<span class="sd">    It also logs the mappings for debugging purposes.</span>

<span class="sd">    Args:</span>
<span class="sd">        result (OptimizeResult): The result of the optimization.</span>
<span class="sd">        n_alpha (int): Number of alpha parameters.</span>
<span class="sd">        n_genes (int): Number of genes.</span>
<span class="sd">        n_reg (int): Number of regulators.</span>
<span class="sd">        beta_start_indices (np.ndarray): Starting indices for beta parameters.</span>
<span class="sd">        num_psites (np.ndarray): Array containing the number of phosphorylation sites for each TF.</span>
<span class="sd">        reg_map (dict): Regulation map, mapping gene IDs to their regulators.</span>
<span class="sd">        gene_ids (list): List of gene IDs.</span>
<span class="sd">        tf_ids (list): List of transcription factor IDs.</span>
<span class="sd">        psite_labels_arr (list): List of lists containing phosphorylation site labels.</span>

<span class="sd">    Returns:</span>
<span class="sd">        final_x (np.ndarray): Final optimization result.</span>
<span class="sd">        final_alpha (np.ndarray): Final alpha parameters reshaped into a matrix.</span>
<span class="sd">        final_beta (np.ndarray): Final beta parameters reshaped into a matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">final_x</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span>
    <span class="n">final_alpha</span> <span class="o">=</span> <span class="n">final_x</span><span class="p">[:</span><span class="n">n_alpha</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">n_reg</span><span class="p">))</span>
    <span class="n">final_beta</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">)):</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">beta_start_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">num_psites</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">final_beta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">final_x</span><span class="p">[</span><span class="n">n_alpha</span> <span class="o">+</span> <span class="n">start</span><span class="p">:</span> <span class="n">n_alpha</span> <span class="o">+</span> <span class="n">start</span> <span class="o">+</span> <span class="n">length</span><span class="p">])</span>
    <span class="n">final_beta</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">final_beta</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>

    <span class="c1"># Build and logger.info  mapping.</span>
    <span class="n">alpha_mapping</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">):</span>
        <span class="n">actual_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">tf</span> <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">reg_map</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="k">if</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_ids</span><span class="p">]</span>
        <span class="n">alpha_mapping</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">actual_tfs</span><span class="p">):</span>
            <span class="n">alpha_mapping</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="n">final_alpha</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping of TFs to mRNAs ( values):&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">mapping</span> <span class="ow">in</span> <span class="n">alpha_mapping</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;mRNA </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">tf</span><span class="p">,</span> <span class="n">a_val</span> <span class="ow">in</span> <span class="n">mapping</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;TF   </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">a_val</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping of TFs to  parameters:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_ids</span><span class="p">):</span>
        <span class="n">beta_vec</span> <span class="o">=</span> <span class="n">final_beta</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   TF </span><span class="si">{</span><span class="n">tf</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">beta_vec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">beta_vec</span><span class="p">)):</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">psite_labels_arr</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="n">q</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">label</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">label</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;PSite</span><span class="si">{</span><span class="n">q</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">beta_vec</span><span class="p">[</span><span class="n">q</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Build and logger.info  mapping.</span>
    <span class="c1"># beta_mapping = {}</span>
    <span class="c1"># for idx, tf in enumerate(tf_ids):</span>
    <span class="c1">#     beta_mapping[tf] = {}</span>
    <span class="c1">#     beta_vec = final_beta[idx]</span>
    <span class="c1">#     logger.info(f&quot;{tf}:&quot;)</span>
    <span class="c1">#     beta_mapping[tf][f&quot;mRNA {tf}&quot;] = beta_vec[0]</span>
    <span class="c1">#     for q in range(1, len(beta_vec)):</span>
    <span class="c1">#         label = psite_labels_arr[idx][q - 1]</span>
    <span class="c1">#         if label == &quot;&quot;:</span>
    <span class="c1">#             label = f&quot;PSite{q}&quot;</span>
    <span class="c1">#         beta_mapping[tf][label] = beta_vec[q]</span>
    <span class="c1"># logger.info(&quot;Mapping of phosphorylation sites to TFs ( parameters):&quot;)</span>
    <span class="c1"># for tf, mapping in beta_mapping.items():</span>
    <span class="c1">#     logger.info(f&quot;{tf}:&quot;)</span>
    <span class="c1">#     for label, b_val in mapping.items():</span>
    <span class="c1">#         logger.info(f&quot;   {label}: {b_val:.4f}&quot;)</span>
    <span class="k">return</span> <span class="n">final_x</span><span class="p">,</span> <span class="n">final_alpha</span><span class="p">,</span> <span class="n">final_beta</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="utils.display"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="utils.display.create_report" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;report.html&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>results_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </td>
            <td>
                  <code>&#39;report.html&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">results_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;report.html&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a single global report HTML file from all gene folders inside the results directory.</span>

<span class="sd">    For each gene folder (e.g. &quot;ABL2&quot;), the report will include:</span>
<span class="sd">      - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.</span>
<span class="sd">      - Each plot is confined to a fixed size of 900px by 900px.</span>
<span class="sd">      - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</span>

<span class="sd">    Args:</span>
<span class="sd">        results_dir (str): Path to the root results directory.</span>
<span class="sd">        output_file (str): Name of the generated global report file (placed inside results_dir).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Gather gene folders (skip &quot;General&quot; and &quot;logs&quot;)</span>
    <span class="n">gene_folders</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">results_dir</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span> <span class="ow">and</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;General&quot;</span><span class="p">,</span> <span class="s2">&quot;logs&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Build HTML content with updated CSS for spacing.</span>
    <span class="n">html_parts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;html&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;meta charset=&#39;UTF-8&#39;&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;title&gt;Estimation Report&lt;/title&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;body { font-family: Arial, sans-serif; margin: 20px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h1 { color: #333; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h2 { color: #555; font-size: 1.8em; border-bottom: 1px solid #ccc; padding-bottom: 5px; }&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h3 { color: #666; font-size: 1.4em; margin-top: 10px; margin-bottom: 10px; }&quot;</span><span class="p">,</span>
        <span class="c1"># /* CSS grid for plots: two per row, fixed size 500px x 500px, extra space between rows */</span>
        <span class="s2">&quot;.plot-container {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  display: grid;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  grid-template-columns: repeat(2, 500px);&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  column-gap: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  row-gap: 40px;&quot;</span><span class="p">,</span>  <span class="c1">#// extra vertical gap</span>
        <span class="s2">&quot;  justify-content: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 40px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;.plot-item {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 500px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;img, iframe {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  height: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  object-fit: contain;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: none;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="c1"># /* Data tables: full width, one per row */</span>
        <span class="s2">&quot;.data-table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 50%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-bottom: 20px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;table {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border-collapse: collapse;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  width: 100%;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  margin-top: 10px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;th, td {&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  border: 1px solid #ccc;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  padding: 8px;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;  text-align: left;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/style&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;/head&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;body&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;h1&gt;Modelling &amp; Parameter Estimation Report&lt;/h1&gt;&quot;</span>
    <span class="p">]</span>

    <span class="c1"># For each gene folder, create a section in the report.</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_folders</span><span class="p">):</span>
        <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&lt;h2&gt;Protein Group: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&lt;/h2&gt;&quot;</span><span class="p">)</span>

        <span class="c1"># Create grid container for fixed-size plots.</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;div class=&quot;plot-container&quot;&gt;&#39;</span><span class="p">)</span>
        <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                <span class="n">rel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                <span class="c1"># Remove the extension and split on &#39;_&#39;</span>
                <span class="n">base_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">tokens</span> <span class="o">=</span> <span class="p">[</span><span class="n">token</span> <span class="k">for</span> <span class="n">token</span> <span class="ow">in</span> <span class="n">base_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">token</span><span class="p">]</span>
                <span class="c1"># Remove the gene name if it matches (case-insensitive)</span>
                <span class="k">if</span> <span class="n">tokens</span> <span class="ow">and</span> <span class="n">tokens</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
                    <span class="n">tokens</span> <span class="o">=</span> <span class="n">tokens</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
                <span class="c1"># Join remaining tokens with space and convert to upper case</span>
                <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;plot-item&quot;&gt;&lt;h3&gt;</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;img src=&quot;</span><span class="si">{</span><span class="n">rel_path</span><span class="si">}</span><span class="s1">&quot; alt=&quot;</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&quot;&gt;&lt;/div&gt;&#39;</span>
                <span class="p">)</span>
        <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;/div&gt;&#39;</span><span class="p">)</span>  <span class="c1"># End of plot container</span>

        <span class="c1"># Data tables: display XLSX or CSV files from the gene folder, one per row.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.xlsx&quot;</span><span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                    <span class="n">table_html</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">border</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="c1"># Remove the extension and split on &#39;_&#39;</span>
                    <span class="n">base_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">tokens</span> <span class="o">=</span> <span class="p">[</span><span class="n">token</span> <span class="k">for</span> <span class="n">token</span> <span class="ow">in</span> <span class="n">base_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">token</span><span class="p">]</span>
                    <span class="c1"># Remove the gene name if it matches (case-insensitive)</span>
                    <span class="k">if</span> <span class="n">tokens</span> <span class="ow">and</span> <span class="n">tokens</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
                        <span class="n">tokens</span> <span class="o">=</span> <span class="n">tokens</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="c1"># Join remaining tokens with space and convert to upper case</span>
                    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s1">&lt;/h3&gt;</span><span class="si">{</span><span class="n">table_html</span><span class="si">}</span><span class="s1">&lt;/div&gt;&#39;</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s1">&#39;&lt;div class=&quot;data-table&quot;&gt;&lt;h3&gt;Data Table: </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">&lt;/h3&gt;&lt;p&gt;Error reading </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&lt;/p&gt;&lt;/div&gt;&#39;</span>
                    <span class="p">)</span>

    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/body&gt;&quot;</span><span class="p">)</span>
    <span class="n">html_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;/html&gt;&quot;</span><span class="p">)</span>

    <span class="c1"># Write the report into the results directory.</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">results_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">html_parts</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.display.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Ensure the output directory exists. If it doesn't, create it.</p>
<p>:param directory: str</p>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">ensure_output_directory</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Ensure the output directory exists. If it doesn&#39;t, create it.</span>

<span class="sd">    :param directory: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.display.format_duration" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.
The function converts seconds into a string representation in the format:
- "X sec" for seconds
- "X min" for minutes
- "X hr" for hours</p>
<p>:param seconds: float
:return: formatted string</p>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_duration</span><span class="p">(</span><span class="n">seconds</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format a duration in seconds into a human-readable string.</span>
<span class="sd">    The function converts seconds into a string representation in the format:</span>
<span class="sd">    - &quot;X sec&quot; for seconds</span>
<span class="sd">    - &quot;X min&quot; for minutes</span>
<span class="sd">    - &quot;X hr&quot; for hours</span>

<span class="sd">    :param seconds: float</span>
<span class="sd">    :return: formatted string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">60</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> sec&quot;</span>
    <span class="k">elif</span> <span class="n">seconds</span> <span class="o">&lt;</span> <span class="mi">3600</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> min&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seconds</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hr&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.display.load_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_data</span><span class="p">(</span><span class="n">excel_file</span><span class="p">,</span> <span class="n">sheet</span><span class="o">=</span><span class="s1">&#39;Estimated Values&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Load data from an Excel file. The default sheet is "Estimated Values".</p>
<p>:param excel_file: str
:param sheet: str
:return: DataFrame
:rtype: pd.DataFrame</p>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_data</span><span class="p">(</span><span class="n">excel_file</span><span class="p">,</span> <span class="n">sheet</span><span class="o">=</span><span class="s2">&quot;Estimated Values&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load data from an Excel file. The default sheet is &quot;Estimated Values&quot;.</span>

<span class="sd">    :param excel_file: str</span>
<span class="sd">    :param sheet: str</span>
<span class="sd">    :return: DataFrame</span>
<span class="sd">    :rtype: pd.DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">excel_file</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="n">sheet</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.display.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Organize output files into protein-specific folders and a general folder.
Files matching the pattern "protein_name_*.{json,svg,png,html,csv,xlsx}"
will be moved to a folder named after the protein.
Remaining files will be moved to a "General" folder within the same directory.</p>
<p>:param directories: List of directories to organize.
:type directories: list</p>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">organize_output_files</span><span class="p">(</span><span class="o">*</span><span class="n">directories</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Organize output files into protein-specific folders and a general folder.</span>
<span class="sd">    Files matching the pattern &quot;protein_name_*.{json,svg,png,html,csv,xlsx}&quot;</span>
<span class="sd">    will be moved to a folder named after the protein.</span>
<span class="sd">    Remaining files will be moved to a &quot;General&quot; folder within the same directory.</span>

<span class="sd">    :param directories: List of directories to organize.</span>
<span class="sd">    :type directories: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Za-z0-9]+)_.*\.(json|svg|png|html|csv|xlsx)$&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">directory</span> <span class="ow">in</span> <span class="n">directories</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">directory</span><span class="si">}</span><span class="s2">&#39; is not a valid directory. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Move files matching the protein pattern.</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">protein_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">protein</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">protein_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">protein</span><span class="p">)</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>

        <span class="c1"># After protein files have been moved, move remaining files to a &quot;General&quot; folder.</span>
        <span class="n">general_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="s2">&quot;General&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="n">destination_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">general_folder</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">destination_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.display.save_result" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_result</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">excel_filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Save the results to an Excel file with multiple sheets.
Each sheet corresponds to a different gene and contains:
- Sequential Parameter Estimates
- Profiled Estimates (if available)
- Errors summary
The sheet names are prefixed with the gene name, truncated to 25 characters.
Args:
    results (list): List of dictionaries containing results for each gene.
    excel_filename (str): Path to the output Excel file.</p>


            <details class="quote">
              <summary>Source code in <code>utils/display.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_result</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">excel_filename</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the results to an Excel file with multiple sheets.</span>
<span class="sd">    Each sheet corresponds to a different gene and contains:</span>
<span class="sd">    - Sequential Parameter Estimates</span>
<span class="sd">    - Profiled Estimates (if available)</span>
<span class="sd">    - Errors summary</span>
<span class="sd">    The sheet names are prefixed with the gene name, truncated to 25 characters.</span>
<span class="sd">    Args:</span>
<span class="sd">        results (list): List of dictionaries containing results for each gene.</span>
<span class="sd">        excel_filename (str): Path to the output Excel file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">excel_filename</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;xlsxwriter&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
            <span class="n">sheet_prefix</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[:</span><span class="mi">25</span><span class="p">]</span>  <span class="c1"># Excel sheet names must be 31 chars</span>

            <span class="c1"># 1. Save Sequential Parameter Estimates</span>
            <span class="n">param_df</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;param_df&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;errors&quot;</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                <span class="n">param_df</span><span class="p">[</span><span class="s2">&quot;MSE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;errors&quot;</span><span class="p">][:</span><span class="nb">len</span><span class="p">(</span><span class="n">param_df</span><span class="p">)])</span>
            <span class="n">param_df</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
            <span class="n">param_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sheet_prefix</span><span class="si">}</span><span class="s2">_params&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># 2. Save Profiled Estimates if available</span>
            <span class="k">if</span> <span class="s2">&quot;profiles_df&quot;</span> <span class="ow">in</span> <span class="n">res</span> <span class="ow">and</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;profiles_df&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prof_df</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;profiles_df&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                <span class="n">prof_df</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span>
                <span class="n">prof_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sheet_prefix</span><span class="si">}</span><span class="s2">_profiles&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># 4. Save errors summary</span>
            <span class="n">error_summary</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span>
                <span class="s2">&quot;MSE&quot;</span><span class="p">:</span> <span class="n">res</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mse&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
                <span class="s2">&quot;MAE&quot;</span><span class="p">:</span> <span class="n">res</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mae&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="p">}</span>
            <span class="n">err_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">error_summary</span><span class="p">])</span>
            <span class="n">err_df</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sheet_prefix</span><span class="si">}</span><span class="s2">_errors&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<a id="utils.tables"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="utils.tables.generate_tables" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">generate_tables</span><span class="p">(</span><span class="n">xlsx_file_path</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Generate hierarchical tables from the XLSX file containing alpha and beta values.
The function reads the alpha and beta values from the specified XLSX file,
processes them to create hierarchical tables, and returns a list of these tables.
Each table is a DataFrame with a MultiIndex for the columns, representing
the alpha and beta values for different kinases and phosphorylation sites.
The tables are structured to facilitate easy comparison and analysis of the
phosphorylation data.</p>


            <details class="quote">
              <summary>Source code in <code>utils/tables.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_tables</span><span class="p">(</span><span class="n">xlsx_file_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate hierarchical tables from the XLSX file containing alpha and beta values.</span>
<span class="sd">    The function reads the alpha and beta values from the specified XLSX file,</span>
<span class="sd">    processes them to create hierarchical tables, and returns a list of these tables.</span>
<span class="sd">    Each table is a DataFrame with a MultiIndex for the columns, representing</span>
<span class="sd">    the alpha and beta values for different kinases and phosphorylation sites.</span>
<span class="sd">    The tables are structured to facilitate easy comparison and analysis of the</span>
<span class="sd">    phosphorylation data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load alpha and beta values from the XLSX file</span>
    <span class="n">alpha_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">xlsx_file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Alpha Values&quot;</span><span class="p">)</span>
    <span class="n">beta_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">xlsx_file_path</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;Beta Values&quot;</span><span class="p">)</span>

    <span class="c1"># Prepare the tables</span>
    <span class="n">hierarchical_tables</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">format_float</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Custom formatter to remove trailing zeroes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">value</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">merge_kinase_columns</span><span class="p">(</span><span class="n">alpha_pivot</span><span class="p">,</span> <span class="n">beta_pivot</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge columns of kinases without duplicating names.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_kinases</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alpha_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">beta_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">unique_kinases</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">all_kinases</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">all_kinases</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

        <span class="n">merged_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">alpha_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">beta_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="n">unique_kinases</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">unique_kinases</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">alpha_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">alpha_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">kinase</span><span class="p">]</span> <span class="o">=</span> <span class="n">alpha_pivot</span><span class="p">[</span><span class="n">kinase</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">beta_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">merged_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">beta_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">kinase</span><span class="p">]</span> <span class="o">=</span> <span class="n">beta_pivot</span><span class="p">[</span><span class="n">kinase</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="k">return</span> <span class="n">merged_data</span>

    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Protein&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">protein_alpha</span> <span class="o">=</span> <span class="n">alpha_values</span><span class="p">[</span><span class="n">alpha_values</span><span class="p">[</span><span class="s1">&#39;Protein&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">protein</span><span class="p">]</span>
        <span class="n">protein_beta</span> <span class="o">=</span> <span class="n">beta_values</span><span class="p">[</span><span class="n">beta_values</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">protein_alpha</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())]</span>

        <span class="k">for</span> <span class="n">psite</span> <span class="ow">in</span> <span class="n">protein_alpha</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="c1"># Filter alpha and beta data for this specific psite</span>
            <span class="n">alpha_data</span> <span class="o">=</span> <span class="n">protein_alpha</span><span class="p">[</span><span class="n">protein_alpha</span><span class="p">[</span><span class="s1">&#39;Psite&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">psite</span><span class="p">]</span>
            <span class="n">beta_data</span> <span class="o">=</span> <span class="n">protein_beta</span><span class="p">[</span><span class="n">protein_beta</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">alpha_data</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())]</span>

            <span class="c1"># Prepare alpha and beta pivot data</span>
            <span class="n">alpha_pivot</span> <span class="o">=</span> <span class="n">alpha_data</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s1">&#39;Psite&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Alpha&#39;</span><span class="p">)</span>
            <span class="n">beta_pivot</span> <span class="o">=</span> <span class="n">beta_data</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;Psite&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Beta&#39;</span><span class="p">)</span>

            <span class="c1"># Round alpha and beta values to 3 decimal places and format</span>
            <span class="n">alpha_pivot</span> <span class="o">=</span> <span class="n">alpha_pivot</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">format_float</span><span class="p">)</span>
            <span class="n">beta_pivot</span> <span class="o">=</span> <span class="n">beta_pivot</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">format_float</span><span class="p">)</span>

            <span class="c1"># Combine alpha and beta with hierarchical levels: add latex symbol</span>
            <span class="n">alpha_pivot</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">([[</span><span class="s1">&#39;$</span><span class="se">\\</span><span class="s1">alpha$&#39;</span><span class="p">],</span> <span class="n">alpha_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">],</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">])</span>
            <span class="n">beta_pivot</span> <span class="o">=</span> <span class="n">beta_pivot</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># Transpose beta for matching structure</span>
            <span class="n">beta_pivot</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">([[</span><span class="s1">&#39;$</span><span class="se">\\</span><span class="s1">beta$&#39;</span><span class="p">],</span> <span class="n">beta_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">],</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">])</span>

            <span class="c1"># # Combine alpha and beta with hierarchical levels: DONT add latex symbol</span>
            <span class="c1"># alpha_pivot.columns = MultiIndex.from_product([[&#39;&#39;], alpha_pivot.columns], names=[&#39;&#39;, &#39;Kinase&#39;])</span>
            <span class="c1"># beta_pivot = beta_pivot.T  # Transpose beta for matching structure</span>
            <span class="c1"># beta_pivot.columns = MultiIndex.from_product([[&#39;&#39;], beta_pivot.columns], names=[&#39;&#39;, &#39;Kinase&#39;])</span>

            <span class="c1"># Concatenate alpha and beta tables</span>
            <span class="n">hierarchical_table</span> <span class="o">=</span> <span class="n">concat</span><span class="p">([</span><span class="n">alpha_pivot</span><span class="p">,</span> <span class="n">beta_pivot</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">hierarchical_table</span> <span class="o">=</span> <span class="n">hierarchical_table</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">hierarchical_table</span><span class="p">),</span>
                                                          <span class="s2">&quot;&quot;</span><span class="p">)</span>  <span class="c1"># Replace NaN with empty strings</span>
            <span class="c1"># Rename &#39;Psite&#39; to &#39;Site&#39; and replace underscores with \_</span>
            <span class="n">hierarchical_table</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">hierarchical_table</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\\</span><span class="s1">_&#39;</span><span class="p">))</span>
            <span class="n">hierarchical_table</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;Site&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">hierarchical_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(((</span><span class="n">protein</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">hierarchical_table</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">hierarchical_tables</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.tables.save_master_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_master_table</span><span class="p">(</span><span class="n">folder</span><span class="o">=</span><span class="s1">&#39;latex&#39;</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;latex/all_tables.tex&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Save a master LaTeX file that includes all individual LaTeX files from the specified folder.
This function generates a LaTeX file that includes all the individual LaTeX files
for each protein and phosphorylation site.</p>
<p>:param folder: Directory containing the individual LaTeX files.
:param output_file: Output LaTeX file name.
:type folder: str
:type output_file: str</p>


            <details class="quote">
              <summary>Source code in <code>utils/tables.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_master_table</span><span class="p">(</span><span class="n">folder</span><span class="o">=</span><span class="s2">&quot;latex&quot;</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s2">&quot;latex/all_tables.tex&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save a master LaTeX file that includes all individual LaTeX files from the specified folder.</span>
<span class="sd">    This function generates a LaTeX file that includes all the individual LaTeX files</span>
<span class="sd">    for each protein and phosphorylation site.</span>

<span class="sd">    :param folder: Directory containing the individual LaTeX files.</span>
<span class="sd">    :param output_file: Output LaTeX file name.</span>
<span class="sd">    :type folder: str</span>
<span class="sd">    :type output_file: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">folder</span><span class="p">)</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.tex&quot;</span><span class="p">)])</span>

    <span class="c1"># Write a LaTeX file that includes all these files</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out</span><span class="p">:</span>
        <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;% This file is auto-generated</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">input</span><span class="se">{{</span><span class="si">{</span><span class="n">folder</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="se">}}\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generated </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">files</span><span class="p">)</span><span class="si">}</span><span class="s2"> entries.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="utils.tables.save_tables" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">save_tables</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Save the generated tables as LaTeX and CSV files.
Each table is saved with a filename based on the protein and phosphorylation site.
The LaTeX files are formatted for easy inclusion in a larger document,
and the CSV files are saved for further analysis.</p>
<p>:param tables: List of tuples containing protein, psite, and the corresponding table.
:param output_dir: Directory where the tables will be saved.
:type output_dir: str</p>


            <details class="quote">
              <summary>Source code in <code>utils/tables.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_tables</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save the generated tables as LaTeX and CSV files.</span>
<span class="sd">    Each table is saved with a filename based on the protein and phosphorylation site.</span>
<span class="sd">    The LaTeX files are formatted for easy inclusion in a larger document,</span>
<span class="sd">    and the CSV files are saved for further analysis.</span>

<span class="sd">    :param tables: List of tuples containing protein, psite, and the corresponding table.</span>
<span class="sd">    :param output_dir: Directory where the tables will be saved.</span>
<span class="sd">    :type output_dir: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">psite</span><span class="p">),</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">tables</span><span class="p">:</span>
        <span class="n">base_filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">protein</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">psite</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="c1"># Save as LaTeX</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_filename</span><span class="si">}</span><span class="s2">.tex&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tex_file</span><span class="p">:</span>
            <span class="n">tex_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">table</span><span class="o">.</span><span class="n">to_latex</span><span class="p">(</span><span class="n">multicolumn</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">multirow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">escape</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
        <span class="c1"># Save as CSV</span>
        <span class="n">table</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_filename</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>